{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b4caa8e-d3eb-4840-a5c7-0c41314206fa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import rapids_singlecell as rsc\n",
    "import scvi\n",
    "import glob\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy import stats as sp_stats\n",
    "from datetime import datetime\n",
    "import torch\n",
    "import warnings\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sc.settings.n_jobs = 32\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17400223-3fc2-4573-8e9e-cc4671c128ad",
   "metadata": {},
   "source": [
    "### Define highly informative features from MTG reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c990b4e8-0874-4299-8c99-88b505952266",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the reference MTG dataset\n",
    "dataset = \"reference\"\n",
    "technology = \"singleomeCR6\"\n",
    "region = \"MTG\"\n",
    "date = \"2022-04-08\"\n",
    "adata_ref = sc.read_h5ad(filename=os.path.join(pwd, \"input\", region + \"_\" + dataset + \"_\" + technology + \".\" + date + \".h5ad\"))\n",
    "\n",
    "if os.path.exists(os.path.join(pwd, \"output\", \"de_genes.csv\")):\n",
    "    cells = []\n",
    "\n",
    "    for i in adata_ref.obs[\"cluster\"].cat.categories:\n",
    "        tmp = adata_ref[adata_ref.obs[\"cluster\"] == i].obs_names.to_list()\n",
    "        if len(tmp) > 100:\n",
    "            cells = cells + random.sample(tmp, k=100)\n",
    "        else:\n",
    "            cells = cells + tmp  \n",
    "    adata_ds = adata_ref[cells].copy()\n",
    "\n",
    "    marker_genes_a = {}\n",
    "    for i in adata_ref.obs[\"class\"].cat.categories:\n",
    "        try:\n",
    "            adata_tmp = adata_ds[adata_ds.obs[\"class\"] == i].copy()\n",
    "            sc.tl.rank_genes_groups(adata_tmp,\n",
    "                                    groupby=\"cluster\",\n",
    "                                    method=\"wilcoxon\",\n",
    "                                    tie_correct=True)\n",
    "\n",
    "            result = adata_tmp.uns['rank_genes_groups']\n",
    "            groups = result['names'].dtype.names\n",
    "            tmp = {group: pd.DataFrame({key: result[key][group] for key in ['names', 'pvals_adj', 'logfoldchanges']}) for group in groups}\n",
    "\n",
    "            marker_genes = {**marker_genes_a, **tmp}\n",
    "        except ZeroDivisionError:\n",
    "            pass\n",
    "\n",
    "    marker_genes_b = {}\n",
    "    for i in adata_ref.obs[\"subclass\"].cat.categories:\n",
    "        try:\n",
    "            adata_tmp = adata_ds[adata_ds.obs[\"subclass\"] == i].copy()\n",
    "            sc.tl.rank_genes_groups(adata_tmp,\n",
    "                                    groupby=\"cluster\",\n",
    "                                    method=\"wilcoxon\",\n",
    "                                    tie_correct=True)\n",
    "\n",
    "            result = adata_tmp.uns['rank_genes_groups']\n",
    "            groups = result['names'].dtype.names\n",
    "            tmp = {group: pd.DataFrame({key: result[key][group] for key in ['names', 'pvals_adj', 'logfoldchanges']}) for group in groups}\n",
    "\n",
    "            marker_genes_2 = {**marker_genes_b, **tmp}\n",
    "        except ZeroDivisionError:\n",
    "            pass\n",
    "\n",
    "    de_genes_a = []\n",
    "    for i,j in marker_genes_a.items():\n",
    "        de_genes_a.extend(j.iloc[:10, :].loc[:, \"names\"].to_list())\n",
    "    de_genes_a = np.unique(de_genes_a)\n",
    "\n",
    "    de_genes_b = []\n",
    "    for i,j in marker_genes_b.items():\n",
    "        de_genes_b.extend(j.iloc[:5, :].loc[:, \"names\"].to_list())\n",
    "    de_genes_b = np.unique(de_genes_b)\n",
    "\n",
    "    de_genes = np.union1d(de_genes_a, de_genes_b)\n",
    "    pd.DataFrame(index=de_genes).to_csv(os.path.join(pwd, \"output\", \"de_genes.csv\"), header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e20eeb0-5b81-4a7e-a6cb-ecd106dee519",
   "metadata": {},
   "source": [
    "### Build combined dataset on a subset of highly informative features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03011917-53b3-4155-8df3-36172a3e6a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = []\n",
    "external_datasets = glob.glob(os.path.join(pwd, \"input\", \"*_external_singleomeCR6.2023-10-12.h5ad\"))\n",
    "external_datasets.extend([os.path.join(pwd, \"input\", \"SEAAD_A9_RNAseq_final-nuclei.2023-07-19.h5ad\")])\n",
    "\n",
    "de_genes = pd.read_csv(os.path.join(pwd, \"output\", \"de_genes.csv\"), header=None)\n",
    "de_genes = de_genes[0].to_list()\n",
    "\n",
    "for k in external_datasets:\n",
    "    print(os.path.basename(k))\n",
    "    tmp = sc.read_h5ad(k)\n",
    "    tmp = tmp[:, de_genes].copy()\n",
    "    tmp.X = tmp.layers[\"UMIs\"].copy()\n",
    "    del tmp.layers[\"UMIs\"]\n",
    "    if \"external\" in k:\n",
    "        tmp.obs[\"Source\"] = os.path.basename(k).replace(\"PFC_\", \"\").replace(\"_external_singleomeCR6.2023-10-12.h5ad\", \"\")\n",
    "    else:\n",
    "        tmp.obs[\"Source\"] = \"Allen\" \n",
    "    \n",
    "    external_adata.append(tmp)\n",
    "\n",
    "first = external_adata.pop()\n",
    "external_adata = first.concatenate(external_adata, index_unique=None)\n",
    "\n",
    "for i in external_adata.obs.columns[external_adata.obs.isna().sum(axis=0) > 0]:\n",
    "    if any(external_adata.obs[i].notna()) == False:\n",
    "        print(\"Dropping no-value column \" + i)\n",
    "        external_adata.obs.drop([i], axis=1, inplace=True)\n",
    "            \n",
    "    else:            \n",
    "        replace_with = \"\"\n",
    "        \n",
    "        if isinstance(external_adata.obs.loc[external_adata.obs[i].notna(), i][0], np.float64) == True or isinstance(external_adata.obs.loc[external_adata.obs[i].notna(), i][0], np.float32) == True:\n",
    "            replace_with = 0.0\n",
    "\n",
    "        if isinstance(external_adata.obs[i].dtype, pd.core.dtypes.dtypes.CategoricalDtype) == True:\n",
    "            external_adata.obs[i] = external_adata.obs[i].astype(\"object\")\n",
    "        \n",
    "        print(\"Replacing NaNs with \" + str(replace_with) + \" for \" + i + \" with dtype \" + str(type(external_adata.obs.loc[external_adata.obs[i].notna(), i][0])))\n",
    "        external_adata.obs.loc[external_adata.obs[i].isna(), i] = replace_with\n",
    "        \n",
    "        if isinstance(external_adata.obs.loc[(external_adata.obs[i].notna()) & (external_adata.obs[i] != \"\"), i][0], bool) == True:\n",
    "            external_adata.obs[i] = external_adata.obs[i].astype(\"str\")\n",
    "            \n",
    "external_adata.obs[\"Donor ID\"] = external_adata.obs[\"Donor ID\"].astype(\"str\")\n",
    "external_adata.obs[\"Age at Death\"] = external_adata.obs[\"Age at Death\"].astype(\"str\")\n",
    "for i in external_adata.obs.columns[(external_adata.obs == \"Reference\").sum(axis=0) > 0]:\n",
    "    external_adata.obs[i] = external_adata.obs[i].astype(\"str\")\n",
    "\n",
    "external_adata.write(os.path.join(pwd, \"output\", \"PFC_external_SEAAD_singleomeCR6_UMIs_only_de_genes.2023-10-12.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c02992f-3248-4684-9863-5b6578671be2",
   "metadata": {},
   "source": [
    "### Load the dataset and map the original author study names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc446cd4-96e0-4253-bd58-31b2d8bb4838",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = sc.read_h5ad(os.path.join(pwd, \"output\", \"PFC_external_SEAAD_singleomeCR6_UMIs_only_de_genes.2023-10-12.h5ad\"))\n",
    "\n",
    "external_adata.obs[\"Original Cell ID\"] = external_adata.obs_names.copy()\n",
    "external_adata.obs[\"Original Cell ID Alternative\"] = external_adata.obs_names.copy()\n",
    "\n",
    "### Mathys_2023\n",
    "# ROSMAP_clinical.csv - https://www.synapse.org/#!Synapse:syn3191087\n",
    "# projid_and_subject_to_cellid_ending.tsv - Personal communication, Tain Luquez\n",
    "# Mathys_2023_FASTQ_batch_crosswalk.csv - Personal communication, Sudhagar Babu\n",
    "\n",
    "# Original study name format: [Batch]_[Barcode]-[Batch Number]\n",
    "# Original study name format (R): [Barcode]-[Batch Number]-[Dataset Number]\n",
    "# Example: SM_Last16_AAACCCATCCGTAGTA-1\n",
    "# Example (R): AAACCCATCCGTAGTA-1-1\n",
    "\n",
    "ROSMAP_metadata = pd.read_csv(os.path.join(pwd, \"external\", \"Mathys_2023\", \"ROSMAP_clinical.csv\"), index_col=0)\n",
    "\n",
    "tain_batch = pd.read_csv(os.path.join(pwd, \"external\", \"Mathys_2023\", \"projid_and_subject_to_cellid_ending.tsv\"), sep=\"\\t\", index_col=1)\n",
    "tain_batch = tain_batch.merge(ROSMAP_metadata.loc[:, [\"individualID\"]], how=\"left\", left_index=True, right_index=True)\n",
    "tain_batch[\"num_1\"] = [i.split(\"_\")[0] for i in tain_batch[\"cellid_ending\"]]\n",
    "tain_batch[\"num_2\"] = [i.split(\"_\")[1] for i in tain_batch[\"cellid_ending\"]]\n",
    "tain_batch = tain_batch.loc[tain_batch[\"num_2\"] != \"2\", :].copy()\n",
    "\n",
    "author_batch = pd.read_csv(os.path.join(pwd, \"external\", \"Mathys_2023\", \"Mathys_2023_FASTQ_batch_crosswalk.csv\"), index_col=0)\n",
    "author_batch[\"batch\"] = \"SM_\" + author_batch[\"batch\"]\n",
    "author_batch[\"batch\"] = author_batch[\"batch\"].replace(\n",
    "    {\"SM_Last_16\": \"SM_Last16\"},\n",
    ")\n",
    "author_batch[\"batch_annotation\"] = author_batch[\"batch_annotation\"].astype(\"str\")\n",
    "author_batch = author_batch.loc[:, [\"individualID\", \"library_id\", \"library_prep\", \"batch\", \"batch_annotation\"]].drop_duplicates().merge(tain_batch.loc[:, [\"individualID\", \"subject\", \"num_1\", \"num_2\"]], left_on=[\"individualID\", \"batch_annotation\"], right_on=[\"individualID\", \"num_2\"], how=\"left\")\n",
    "author_batch = author_batch.drop(\"batch_annotation\", axis=1)\n",
    "author_batch.index = author_batch[\"library_id\"].copy()\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2023\", \"Original Cell ID\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\2_\\\\1-\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2023\", \"Original Cell ID\"]]\n",
    "for i,j in author_batch[\"batch\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, j)\n",
    "for i,j in author_batch[\"num_1\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] + j\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2023\", \"Original Cell ID Alternative\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\1\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2023\", \"Original Cell ID Alternative\"]]\n",
    "for i,j in author_batch[\"num_1\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID Alternative\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID Alternative\"] + \"-\" + j\n",
    "for i,j in author_batch[\"num_2\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID Alternative\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID Alternative\"] + \"-\" + j\n",
    "    \n",
    "### Green_2023\n",
    "# cell-annotation.csv - https://www.synapse.org/#!Synapse:syn51218314\n",
    "# ROSMAP_snRNAseq_demultiplexed_ID_mapping.csv - https://www.synapse.org/#!Synapse:syn34572333\n",
    "# Original study name format: [Batch ID]_[Barcode]-1\n",
    "# Example: 190403-B4-A_AAACCCATCCGTAGTA-1\n",
    "\n",
    "author_annotation = pd.read_csv(os.path.join(pwd, \"external\", \"Green_2023\", \"cell-annotation.csv\"), index_col=0)\n",
    "author_annotation = author_annotation.loc[author_annotation.index.str.startswith(\"MAP\"), :]\n",
    "author_annotation[\"libraryBatch_fix\"] = [re.sub(\"_[ATGC]+-1\", \"\", i) for i in author_annotation.index]\n",
    "author_annotation = author_annotation.loc[:, [\"individualID\", \"libraryBatch_fix\"]].drop_duplicates()\n",
    "author_annotation[\"new_batch\"] = \"B6-\" + author_annotation[\"individualID\"] + \"-alone_Merged\"\n",
    "\n",
    "author_batch = pd.read_csv(os.path.join(pwd, \"external\", \"Green_2023\", \"ROSMAP_snRNAseq_demultiplexed_ID_mapping.csv\"))\n",
    "author_batch = author_batch.loc[:, [\"libraryBatch\", \"individualID\"]].drop_duplicates()\n",
    "author_batch[\"new_batch\"] = [re.sub(\"^[0-9]+-\", \"\", i) + \"_Merged\" for i in author_batch[\"libraryBatch\"]]\n",
    "author_batch[\"new_batch\"] = author_batch[\"new_batch\"].replace(\n",
    "    {\n",
    "        \"B6_Merged\": \"B6-A_Merged\"\n",
    "    }\n",
    ")\n",
    "author_batch = author_batch.reset_index(drop=True)\n",
    "\n",
    "author_batch = author_batch.loc[:, [\"libraryBatch\", \"new_batch\"]].drop_duplicates().merge(author_annotation.loc[:, [\"new_batch\", \"libraryBatch_fix\"]], left_on=\"new_batch\", right_on=\"new_batch\", how=\"left\")\n",
    "author_batch.loc[~author_batch[\"libraryBatch_fix\"].isna(), \"libraryBatch\"] = author_batch.loc[~author_batch[\"libraryBatch_fix\"].isna(), \"libraryBatch_fix\"]\n",
    "author_batch = author_batch.drop(\"libraryBatch_fix\", axis=1)\n",
    "author_batch.index = author_batch[\"new_batch\"].copy()\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Green_2023\", \"Original Cell ID\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\2_\\\\1-1\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Green_2023\", \"Original Cell ID\"]]\n",
    "for i,j in author_batch[\"libraryBatch\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, j)\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Green_2023\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Yang_2022\n",
    "# Original study name format: [Barcode]-[Donor ID]\n",
    "# Example: AAACCCATCCGTAGTA-AD 4\n",
    "\n",
    "author_batch = external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Yang_2022\", [\"library_prep\", \"Donor ID\"]].drop_duplicates().reset_index(drop=True)\n",
    "author_batch.index = author_batch[\"library_prep\"].copy()\n",
    "\n",
    "for i,j in author_batch[\"Donor ID\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, j)\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Yang_2022\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Cain_2022\n",
    "# Original study name format: [Batch ID]_[Sex]_[Barcode]\n",
    "# Example: MFC_B2_16_Cdx4_pAD1_M_GAACATCCACAGACTT\n",
    "author_batch = external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Cain_2022\", [\"library_prep\", \"Sex\"]].drop_duplicates().reset_index(drop=True)\n",
    "author_batch[\"new_batch\"] = author_batch[\"library_prep\"].str.replace(\"Cog\", \"Cdx\").str.replace(\"Path\", \"pAD\").str.replace(\"-\", \"_\")\n",
    "author_batch[\"Sex\"] = author_batch[\"Sex\"].replace(\n",
    "    {\n",
    "        \"Female\": \"F\",\n",
    "        \"Male\": \"M\"\n",
    "    }\n",
    ")\n",
    "author_batch[\"new_batch\"] = author_batch[\"new_batch\"] + \"_\" + author_batch[\"Sex\"].astype(\"str\")\n",
    "author_batch.index = author_batch[\"library_prep\"].copy()\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Cain_2022\", \"Original Cell ID\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\2_\\\\1\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Cain_2022\", \"Original Cell ID\"]]\n",
    "\n",
    "for i,j in author_batch[\"new_batch\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, j)\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Cain_2022\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Morabito_2021\n",
    "# donor_to_cellid_ending.tsv - Personal communication, Tain Luquez\n",
    "# Original study name format: [Barcode]-[Batch Number]\n",
    "# Example: AAACGCTAGGGTACGT-5\n",
    "\n",
    "author_batch = pd.read_csv(os.path.join(pwd, \"external\", \"Morabito_2021\", \"donor_to_cellid_ending.tsv\"), sep=\"\\t\")\n",
    "author_batch.index = author_batch[\"donor\"].copy()\n",
    "\n",
    "for i,j in author_batch[\"cellid_ending\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, str(j))\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Morabito_2021\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Leng_2021\n",
    "# Original study name format: [Donor ID]_[Barcode]\n",
    "# Example: SFG1_AAACGCTAGGGTACGT\n",
    "\n",
    "author_batch = external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Leng_2021\", [\"library_prep\", \"Donor ID\"]].drop_duplicates().reset_index(drop=True)\n",
    "author_batch[\"Donor ID\"] = \"SFG\" + author_batch[\"Donor ID\"].astype(\"str\")\n",
    "author_batch.index = author_batch[\"library_prep\"].copy()\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Leng_2021\", \"Original Cell ID\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\2_\\\\1\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Leng_2021\", \"Original Cell ID\"]]\n",
    "\n",
    "for i,j in author_batch[\"Donor ID\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, j)\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Leng_2021\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Lau_2020\n",
    "# Original study name format: [Barcode]-1_[Donor ID]\n",
    "# Example: AAACGCTAGCGATGGT-1_AD1\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Lau_2020\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Zhou_2020\n",
    "# Original study name format: [Barcode]-1_[Donor ID]\n",
    "# Example: AAACCTGGTCCGTGAC-1_AD11\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Zhou_2020\", \"Original Cell ID\"] = [re.sub(\"TWCC-([^-]+)-([^-]+)-lib1\", \"1_\\\\2\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Zhou_2020\", \"Original Cell ID\"]]\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Zhou_2020\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Olah_2020\n",
    "# Original study name format: [Barcode]-[Donor ID]\n",
    "# Example: AAACCTGAGCTAGTTC-Microglia_MO_AD8\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Olah_2020\", \"Original Cell ID Alternative\"] = \"\"\n",
    "\n",
    "### Mathys_2019\n",
    "# projid_to_cellid_ending.tsv - Personal communication, Tain Luquez\n",
    "# ROSMAP_clinical.csv - https://www.synapse.org/#!Synapse:syn3191087\n",
    "# Original study name format: [Barcode].[Batch Number]\n",
    "# Example: AAAGATGAGCCAGAAC.41\n",
    "\n",
    "ROSMAP_metadata = pd.read_csv(os.path.join(pwd, \"external\", \"Mathys_2019\", \"ROSMAP_clinical.csv\"), index_col=0)\n",
    "\n",
    "tain_batch = pd.read_csv(os.path.join(pwd, \"external\", \"Mathys_2019\", \"projid_to_cellid_ending.tsv\"), sep=\"\\t\", index_col=1)\n",
    "tain_batch = tain_batch.merge(ROSMAP_metadata.loc[:, [\"individualID\"]], how=\"left\", left_index=True, right_index=True)\n",
    "\n",
    "author_batch = external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2019\", [\"library_prep\", \"Donor ID\"]].drop_duplicates().reset_index(drop=True)\n",
    "author_batch = author_batch.merge(tain_batch, left_on=\"Donor ID\", right_on=\"individualID\", how=\"left\")\n",
    "author_batch.index = author_batch[\"library_prep\"].copy()\n",
    "author_batch\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2019\", \"Original Cell ID\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\1.\\\\2\", i) for i in external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2019\", \"Original Cell ID\"]]\n",
    "\n",
    "for i,j in author_batch[\"cellid_ending\"].to_dict().items():\n",
    "    external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"] = external_adata.obs.loc[external_adata.obs[\"library_prep\"] == i, \"Original Cell ID\"].str.replace(i, str(j))\n",
    "\n",
    "external_adata.obs.loc[external_adata.obs[\"Source\"] == \"Mathys_2019\", \"Original Cell ID Alternative\"] = \"\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a7ebb12-94dd-4294-96af-0e27690beb84",
   "metadata": {},
   "source": [
    "### Train the model to integrate it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ec3db1a-fa0b-4a51-9fc1-e1690eff67c1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "model_args = {\n",
    "    \"n_layers\": 2,\n",
    "    \"n_latent\": 20,\n",
    "    \"dispersion\": \"gene-batch\",\n",
    "}\n",
    "scvi.model.SCVI.setup_anndata(\n",
    "    external_adata,\n",
    "    layer=None,\n",
    "    batch_key=\"Source\",\n",
    "    categorical_covariate_keys=[\"library_prep\"]\n",
    ")\n",
    "if os.path.exists(os.path.join(pwd, \"output\", \"External_Data_Models\", \"Full\")) == False:\n",
    "    model = scvi.model.SCVI(external_adata, **model_args)\n",
    "    model.train(max_epochs=500, early_stopping=True, use_gpu=True)\n",
    "    model.save(os.path.join(pwd, \"output\", \"External_Data_Models\", \"Full\"))\n",
    "else:\n",
    "    model = scvi.model.SCVI.load(os.path.join(pwd, \"output\", \"External_Data_Models\", \"Full\"), external_adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f32a59d-31d1-48fe-bfcb-34817e9231b8",
   "metadata": {},
   "source": [
    "### Import results from iterative scANVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a186609a-1547-49c0-b05a-5827a202a610",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = glob.glob(os.path.join(pwd, \"external\", \"iterative_scANVI\", \"*\", \"iterative_scANVI_results.2023-11-06.csv\"))\n",
    "for i in results:\n",
    "    result = pd.read_csv(i, index_col=0)\n",
    "    result = result.loc[result[\"class\"] == \"Unknown\", :].copy()\n",
    "    external_adata.obs.loc[result.index, \"Class\"] = result.loc[:, \"class_scANVI\"].replace(\n",
    "        {\n",
    "            \"exc\": \"Neuronal: Glutamatergic\",\n",
    "            \"inh\": \"Neuronal: GABAergic\",\n",
    "            \"glia\": \"Non-neuronal and Non-neural\"\n",
    "        }\n",
    "    )\n",
    "    external_adata.obs.loc[result.index, \"Class confidence\"] = result.loc[:, \"class_conf_scANVI\"]\n",
    "    external_adata.obs.loc[result.index, \"Subclass\"] = result.loc[:, \"subclass_scANVI\"].replace(\n",
    "        {\n",
    "            \"Lamp5_Lhx6\": \"Lamp5 Lhx6\",\n",
    "            \"Astro\": \"Astrocyte\",\n",
    "            \"Oligo\": \"Oligodendrocyte\",\n",
    "            \"Endo\": \"Endothelial\",\n",
    "            \"Micro-PVM\": \"Microglia-PVM\"\n",
    "        }\n",
    "    )\n",
    "    external_adata.obs.loc[result.index, \"Subclass confidence\"] = result.loc[:, \"subclass_conf_scANVI\"]\n",
    "    try:\n",
    "        external_adata.obs[\"Supertype (non-expanded)\"] = external_adata.obs[\"Supertype (non-expanded)\"].cat.add_categories([\"L5 ET_1\", \"Endo_1\", \"Endo_3\"])\n",
    "    except:\n",
    "        pass\n",
    "    external_adata.obs.loc[result.index, \"Supertype (non-expanded)\"] = result.loc[:, \"supertype_scANVI\"]\n",
    "    external_adata.obs.loc[result.index, \"Supertype confidence\"] = result.loc[:, \"supertype_conf_scANVI\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42e4eb84-12ec-4785-80ad-8b84a8ae2ea3",
   "metadata": {},
   "source": [
    "### Compute a low dimensional representation and cluster the cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef468a00-636b-4ff5-b0a7-fe93f5bcb235",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "print(str(datetime.now()) + \" -- Computed latent representation\")\n",
    "rsc.pp.neighbors(external_adata, use_rep=\"X_scVI\")\n",
    "print(str(datetime.now()) + \" -- Computed neighbors\")\n",
    "rsc.tl.umap(external_adata, min_dist=0.3)\n",
    "print(str(datetime.now()) + \" -- Computed UMAP\")\n",
    "rsc.tl.leiden(external_adata, resolution=5)\n",
    "print(str(datetime.now()) + \" -- Computed clustering\")\n",
    "sc.pp.subsample(external_adata, fraction=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88fee420-be7d-43ad-ad33-8cfa3d1eb9c1",
   "metadata": {},
   "source": [
    "### Perform QC on the clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7493811d-3b62-469a-93bc-2a0888565252",
   "metadata": {},
   "outputs": [],
   "source": [
    "neuronal_clusters = external_adata.obs.loc[:, [\"leiden\", \"Class\"]].groupby(\"leiden\").value_counts(sort=False, normalize=True).reset_index()\n",
    "neuronal_clusters = neuronal_clusters.loc[neuronal_clusters[\"Class\"] == \"Non-neuronal and Non-neural\"]\n",
    "neuronal_clusters.columns = [\"leiden\", \"Class\", \"Fraction Non-neuronal\"]\n",
    "neuronal_clusters = neuronal_clusters.loc[neuronal_clusters[\"Fraction Non-neuronal\"] < 0.3, \"leiden\"]\n",
    "neuronal_clusters = neuronal_clusters.to_list()\n",
    "\n",
    "external_adata.obs[\"Neuronal Cluster\"] = external_adata.obs[\"leiden\"].isin(neuronal_clusters)\n",
    "external_adata.obs[\"Neuronal Cluster\"] = external_adata.obs[\"Neuronal Cluster\"].astype(\"category\")\n",
    "\n",
    "qc_glia_thresh = [0.20, 1000, 0.0325]\n",
    "qc_neuron_thresh = [0.20, 3000, 0.0325]\n",
    "qc_vars = [\"Doublet score\", \"Genes detected\", \"Fraction mitochondrial UMIs\"]\n",
    "qc_dir = [\"gt\", \"lt\", \"gt\"]\n",
    "for l,k in enumerate(qc_vars):\n",
    "    plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "    tmp = external_adata.obs.loc[:, [\"leiden\", k]].groupby(\"leiden\").mean().loc[:, k].to_dict()\n",
    "    groups = []\n",
    "    for i,j in tmp.items():\n",
    "        if i in neuronal_clusters:\n",
    "            if qc_dir[l] == \"gt\":\n",
    "                if j > qc_neuron_thresh[l]:\n",
    "                    groups.append(i)\n",
    "            else:\n",
    "                if j < qc_neuron_thresh[l]:\n",
    "                    groups.append(i)  \n",
    "        else:\n",
    "            if qc_dir[l] == \"gt\":\n",
    "                if j > qc_glia_thresh[l]:\n",
    "                    groups.append(i)\n",
    "            else:\n",
    "                if j < qc_glia_thresh[l]:\n",
    "                    groups.append(i)\n",
    "    print(groups)\n",
    "    external_adata.obs[\"cluster_\" + k + \"_flag\"] = \"False\"\n",
    "    external_adata.obs.loc[external_adata.obs[\"leiden\"].isin(groups), \"cluster_\" + k + \"_flag\"] = \"True\"\n",
    "    sc.pl.umap(\n",
    "        external_adata,\n",
    "        color=[\"leiden\"],\n",
    "        legend_loc=\"on data\",\n",
    "        frameon=False,\n",
    "        groups=groups,\n",
    "        na_in_legend=False,\n",
    "        sort_order=False\n",
    "    )\n",
    "    sc.pl.umap(\n",
    "        external_adata,\n",
    "        color=\"cluster_\" + k + \"_flag\",\n",
    "        frameon=False,\n",
    "        palette={\"True\": \"red\", \"False\": \"lightgrey\"},\n",
    "        sort_order=False\n",
    "    )\n",
    "    sc.pl.umap(\n",
    "        external_adata,\n",
    "        color=k,\n",
    "        frameon=False,\n",
    "        cmap=\"YlGnBu\",\n",
    "        sort_order=False\n",
    "    )\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dfe95d6-aa18-423a-8088-59a3d6a1bc1c",
   "metadata": {},
   "source": [
    "### Remove low quality cells and re-compute low dimensional representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1121c969-eccd-44bf-af94-3ef82391c2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = external_adata[(external_adata.obs.loc[:, external_adata.obs.columns.str.endswith(\"_flag\")] == \"True\").sum(axis=1) == 0, :].copy()\n",
    "\n",
    "rsc.pp.neighbors(external_adata, use_rep=\"X_scVI\")\n",
    "print(str(datetime.now()) + \" -- Computed neighbors\")\n",
    "rsc.tl.umap(external_adata, min_dist=0.3)\n",
    "print(str(datetime.now()) + \" -- Computed UMAP\")\n",
    "rsc.tl.leiden(external_adata, resolution=1.5)\n",
    "print(str(datetime.now()) + \" -- Computed clustering\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b239ed93-c086-4828-a0d2-ae4c73de2bbd",
   "metadata": {},
   "source": [
    "### Associate Subclass and Supertype colors with the AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36cba52c-7b44-45f7-aab8-96ba572b8391",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_colors = pd.read_csv(os.path.join(\"input\", \"cluster_order_and_colors.csv\"))\n",
    "subclass_colors = subclass_colors.loc[:, [\"subclass_label\", \"subclass_color\"]]\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"]\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "supertype_colors = pd.read_csv(os.path.join(\"input\", \"cluster_order_and_colors.csv\"))\n",
    "supertype_colors = supertype_colors.loc[:, [\"cluster_label\", \"cluster_color\"]]\n",
    "supertype_colors.index = supertype_colors[\"cluster_label\"]\n",
    "supertype_colors = supertype_colors[\"cluster_color\"].to_dict()\n",
    "\n",
    "sc.pl.umap(external_adata, color = [\"Subclass\"], legend_loc=\"on data\", palette=subclass_colors)\n",
    "sc.pl.umap(external_adata, color = [\"Supertype\"], legend_loc=\"on data\", palette=supertype_colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b31bb93-602f-475c-912b-794693f94c42",
   "metadata": {},
   "source": [
    "### Subset by neighborhood for additional QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2acbd58b",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2c77d917-96bf-4d64-823c-3e21e141f70a",
   "metadata": {},
   "source": [
    "### Save final mapping results and QC for neuronal subclasses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d610ce8b-8d2b-47d7-83c5-8b046842229d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in external_adata.obs[\"Source\"].cat.categories:\n",
    "    print(i)\n",
    "    print(external_adata.obs.loc[(external_adata.obs[\"Class\"] != \"Non-neuronal and Non-neural\") & (external_adata.obs[\"Source\"] == i), :].shape)\n",
    "    external_adata.obs.loc[(external_adata.obs[\"Class\"] != \"Non-neuronal and Non-neural\") & (external_adata.obs[\"Source\"] == i), :].to_csv(os.path.join(pwd, \"output\", \"PFC_\" + i + \"_external_metadata_keepers_neurons.2023-11-16.csv\"))\n",
    "\n",
    "external_adata.obs.loc[(external_adata.obs[\"Class\"] == \"Non-neuronal and Non-neural\"), :].to_csv(os.path.join(pwd, \"output\", \"PFC_external_non-neurons_intermediate_results.2023-11-16.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3a61b9e-876a-4eee-b11e-553bdce4377c",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Rebuild non-neuronal AnnData files with all genes to identify novel populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4532b161-5de9-45e9-85dd-cca7c46aeadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = []\n",
    "external_datasets = glob.glob(os.path.join(pwd, \"input\", \"*_external_singleomeCR6.2023-10-12.h5ad\"))\n",
    "external_datasets.extend([os.path.join(pwd, \"input\", \"SEAAD_A9_RNAseq_final-nuclei.2023-07-19.h5ad\")])\n",
    "\n",
    "non_neuronal_cells = pd.read_csv(os.path.join(pwd, \"output\", \"PFC_external_non-neurons_intermediate_results.2023-11-16.csv\"), index_col=0)\n",
    "\n",
    "for k in external_datasets:\n",
    "    print(os.path.basename(k))\n",
    "    tmp = sc.read_h5ad(k)\n",
    "    tmp = tmp[np.intersect1d(tmp.obs_names, non_neuronal_cells.index), :].copy()\n",
    "    tmp.X = tmp.layers[\"UMIs\"].copy()\n",
    "    del tmp.layers[\"UMIs\"]\n",
    "    if \"external\" in k:\n",
    "        tmp.obs[\"Source\"] = os.path.basename(k).replace(\"PFC_\", \"\").replace(\"_external_singleomeCR6.2023-10-12.h5ad\", \"\")\n",
    "    else:\n",
    "        tmp.obs[\"Source\"] = \"Allen\" \n",
    "    \n",
    "    external_adata.append(tmp)\n",
    "\n",
    "first = external_adata.pop()\n",
    "external_adata = first.concatenate(external_adata, index_unique=None)\n",
    "\n",
    "for i in external_adata.obs.columns[external_adata.obs.isna().sum(axis=0) > 0]:\n",
    "    if any(external_adata.obs[i].notna()) == False:\n",
    "        print(\"Dropping no-value column \" + i)\n",
    "        external_adata.obs.drop([i], axis=1, inplace=True)\n",
    "            \n",
    "    else:            \n",
    "        replace_with = \"\"\n",
    "        \n",
    "        if isinstance(external_adata.obs.loc[external_adata.obs[i].notna(), i][0], np.float64) == True or isinstance(external_adata.obs.loc[external_adata.obs[i].notna(), i][0], np.float32) == True:\n",
    "            replace_with = 0.0\n",
    "\n",
    "        if isinstance(external_adata.obs[i].dtype, pd.core.dtypes.dtypes.CategoricalDtype) == True:\n",
    "            external_adata.obs[i] = external_adata.obs[i].astype(\"object\")\n",
    "\n",
    "        if external_adata.obs[i].dtype == \"bool\" or external_adata.obs[i].dtype == \"boolean\":\n",
    "                external_adata.obs[i] = external_adata.obs[i].astype(\"object\")\n",
    "                external_adata.obs[i] = [str(l) for l in external_adata.obs[i]]\n",
    "        \n",
    "        print(\"Replacing NaNs with \" + str(replace_with) + \" for \" + i + \" with dtype \" + str(type(external_adata.obs.loc[external_adata.obs[i].notna(), i][0])))\n",
    "        external_adata.obs.loc[external_adata.obs[i].isna(), i] = replace_with\n",
    "        \n",
    "        if isinstance(external_adata.obs.loc[(external_adata.obs[i].notna()) & (external_adata.obs[i] != \"\"), i][0], bool) == True:\n",
    "            external_adata.obs[i] = external_adata.obs[i].astype(\"str\")\n",
    "            \n",
    "external_adata.obs[\"Donor ID\"] = external_adata.obs[\"Donor ID\"].astype(\"str\")\n",
    "external_adata.obs[\"Age at Death\"] = external_adata.obs[\"Age at Death\"].astype(\"str\")\n",
    "for i in external_adata.obs.columns[(external_adata.obs == \"Reference\").sum(axis=0) > 0]:\n",
    "    external_adata.obs[i] = external_adata.obs[i].astype(\"str\")\n",
    "\n",
    "external_adata.write(\"output\", \"PFC_external_SEAAD_singleomeCR6_UMIs_only_non-neuronal.2023-11-20.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972f2592-9b83-4458-9361-890a9f52bc7f",
   "metadata": {},
   "source": [
    "### Load the non-neuronal data and mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a22766-4f67-4a58-8b48-6a84890cd968",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = sc.read_h5ad(\"output\", \"PFC_external_SEAAD_singleomeCR6_UMIs_only_non-neuronal.2023-11-20.h5ad\")\n",
    "external_adata.obs.loc[external_adata.obs[\"Class\"] == \"\", [\"Class\", \"Class confidence\", \"Subclass\", \"Subclass confidence\", \"Supertype (non-expanded)\",  \"Supertype confidence\"]] = non_neuronal_cells.loc[external_adata.obs[\"Class\"] == \"\", [\"Class\", \"Class confidence\", \"Subclass\", \"Subclass confidence\", \"Supertype (non-expanded)\",  \"Supertype confidence\"]].copy()\n",
    "external_adata.obs[\"Subclass\"] = external_adata.obs[\"Subclass\"].cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1ec50d8-0934-486c-89bb-1cc468abe173",
   "metadata": {},
   "source": [
    "### Identify novel populations in non-neuronal neighborhoods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b341568-3dde-4a60-94f1-aebb1f375d5d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "external_adata.obs[\"Used in analysis\"] = True\n",
    "to_update = external_adata.obs.loc[:, [\"Used in analysis\", \"Subclass\", \"Supertype\", \"Supertype (non-expanded)\"]].copy()\n",
    "for j in [\"Oligodendrocyte\", \"OPC\", \"Astrocyte\", \"Microglia-PVM\", \"VLMC\", \"Endothelial\"]:\n",
    "    print(j)\n",
    "    # Subset the data, compute normalization\n",
    "    sub = external_adata[external_adata.obs[\"Subclass\"] == j, :].copy()\n",
    "    sub.layers[\"UMIs\"] = sub.X.copy()\n",
    "    sc.pp.normalize_total(sub, 1e5)\n",
    "    sc.pp.log1p(sub)\n",
    "\n",
    "    if os.path.exists(os.path.join(pwd, \"output\", \"External_Data_Models\", j.replace(\"/\", \" \"))) == False:\n",
    "        adata_ref = sub[sub.obs[\"Source\"] == \"Allen\", :].copy()\n",
    "        groupby = \"Supertype\"\n",
    "        n_top_genes = 2000\n",
    "        n_downsample_ref = 1000\n",
    "        n_ref_genes = 500\n",
    "\n",
    "        # Define model genes to use\n",
    "        markers = []\n",
    "        try:\n",
    "            sc.pp.highly_variable_genes(adata_ref, flavor=\"seurat_v3\", n_top_genes=n_top_genes, layer=\"UMIs\")\n",
    "        except:\n",
    "            sc.pp.highly_variable_genes(adata_ref, min_mean=1, min_disp=0.5)    \n",
    "        markers = adata_ref.var[adata_ref.var.highly_variable == True].index.to_list()\n",
    "        \n",
    "        ref_counts = adata_ref.obs[groupby].value_counts()\n",
    "        adata_ref = adata_ref[~(adata_ref.obs[groupby].isin(ref_counts[ref_counts < 15].index))]\n",
    "        \n",
    "        if np.setdiff1d(adata_ref.obs[groupby].cat.categories, \"Unknown\").shape[0] > 1:\n",
    "            \n",
    "            cells = []\n",
    "            for i in adata_ref.obs[groupby].cat.categories:\n",
    "                tmp_cells = adata_ref[adata_ref.obs[groupby] == i].obs_names.to_list()\n",
    "                if len(tmp_cells) > n_downsample_ref:\n",
    "                    cells = cells + random.sample(tmp_cells, k=n_downsample_ref)\n",
    "                else:\n",
    "                    cells.extend(tmp_cells)\n",
    "            adata_ref = adata_ref[cells]\n",
    "            \n",
    "            sc.tl.rank_genes_groups(adata_ref, method=\"wilcoxon\", tie_correct=True, groupby=groupby, pts=True)\n",
    "    \n",
    "            result = adata_ref.uns['rank_genes_groups']\n",
    "            groups = result['names'].dtype.names\n",
    "            marker_genes = {group: pd.DataFrame({key: result[key][group] for key in ['names', 'pvals_adj', 'logfoldchanges']}) for group in groups}\n",
    "    \n",
    "            for group in groups:\n",
    "                marker_genes[group]['pts'] = result['pts'][group][result['names'][group]].to_list()\n",
    "                marker_genes[group]['pts_rest'] = result['pts_rest'][group][result['names'][group]].to_list()\n",
    "                marker_genes[group].index = marker_genes[group].names\n",
    "                marker_genes[group].drop(columns=['names'], inplace=True)\n",
    "                tmp_genes = marker_genes[group].copy()\n",
    "                tmp_genes = tmp_genes[tmp_genes.pvals_adj < 0.05]\n",
    "                tmp_genes.sort_values(by=\"logfoldchanges\", axis=0, inplace=True, ascending=False)\n",
    "                markers.extend(tmp_genes.head(n_ref_genes).index.to_list())\n",
    "        else:\n",
    "            warnings.warn(groupby + \" contains only one label. Differentially expressed genes were NOT included in the model\")\n",
    "    \n",
    "        markers = np.unique(markers)\n",
    "        del adata_ref\n",
    "        \n",
    "        # Train the subclass models\n",
    "        model_args = {\n",
    "            \"n_layers\": 2,\n",
    "            \"n_latent\": 20,\n",
    "            \"dispersion\": \"gene-batch\",\n",
    "        }\n",
    "        sub_sub = sub[:, markers].copy()\n",
    "        scvi.model.SCVI.setup_anndata(\n",
    "            sub_sub,\n",
    "            layer=\"UMIs\",\n",
    "            batch_key=\"Source\",\n",
    "            categorical_covariate_keys=[\"library_prep\"]\n",
    "        )\n",
    "        subclass_model = scvi.model.SCVI(sub_sub, **model_args)\n",
    "        subclass_model.train(max_epochs=500, early_stopping=True, use_gpu=True)\n",
    "        subclass_model.save(os.path.join(pwd, \"output\", \"External_Data_Models\", j.replace(\"/\", \" \")))\n",
    "    else:\n",
    "        markers = torch.load(os.path.join(pwd, \"output\", \"External_Data_Models\", j.replace(\"/\", \" \"), \"model.pt\"))\n",
    "        markers = markers[\"var_names\"]\n",
    "        sub_sub = sub[:, markers].copy()\n",
    "        scvi.model.SCVI.setup_anndata(\n",
    "            sub_sub,\n",
    "            layer=\"UMIs\",\n",
    "            batch_key=\"Source\",\n",
    "            categorical_covariate_keys=[\"library_prep\"]\n",
    "        )\n",
    "        subclass_model = scvi.model.SCVI.load(os.path.join(pwd, \"output\", \"External_Data_Models\", j.replace(\"/\", \" \")), sub_sub)\n",
    "        \n",
    "    del sub_sub\n",
    "    sub.obsm[\"X_scVI\"] = subclass_model.get_latent_representation()\n",
    "    del sub.layers[\"UMIs\"]\n",
    "    print(str(datetime.now()) + \" -- Computed latent representation\")\n",
    "    rsc.pp.neighbors(sub, use_rep=\"X_scVI\", n_neighbors=200)\n",
    "    print(str(datetime.now()) + \" -- Computed neighbors\")\n",
    "    rsc.tl.umap(sub, min_dist=0.3)\n",
    "    print(str(datetime.now()) + \" -- Computed low dimensional embedding\")\n",
    "    rsc.tl.leiden(sub, resolution=1)\n",
    "    print(str(datetime.now()) + \" -- Computed clustering\")\n",
    "\n",
    "    # Add low abundance and SEA-AD specific cell types\n",
    "    if j == \"Oligodendrocyte\":\n",
    "        # Fix calling of low abundance Oligo_3\n",
    "        signature_score = sc.get.obs_df(\n",
    "            sub,\n",
    "            [\"FRY\", \"MDGA2\", \"CNTN1\", \"ERBB4\", \"SGCZ\"],\n",
    "        )\n",
    "        signature_score = ((signature_score - signature_score.mean(axis=0)) / signature_score.std(axis=0)).mean(axis=1)\n",
    "        to_update.loc[sub.obs_names[signature_score > 1.5], \"Supertype (non-expanded)\"] = \"Oligo_3\"\n",
    "        sub.obs.loc[signature_score > 1.5, \"Supertype (non-expanded)\"] = \"Oligo_3\"\n",
    "        \n",
    "    elif j == \"OPC\":\n",
    "        # Fix calling of low abundance OPC_1\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"1\", \"13\", \"180\", \"212\"])], \"Supertype (non-expanded)\"] = \"OPC_1\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"1\", \"13\", \"180\", \"212\"]), \"Supertype (non-expanded)\"] = \"OPC_1\"\n",
    "\n",
    "        # Add SEA-AD specific populations\n",
    "        signature_score = sc.get.obs_df(\n",
    "            sub,\n",
    "            [\"ARHGEF3\", \"TPST1\", \"SLC24A2\", \"RAB3IP\", \"CAMK2D\"],\n",
    "        )\n",
    "        signature_score = ((signature_score - signature_score.mean(axis=0)) / signature_score.std(axis=0)).mean(axis=1)\n",
    "        to_update[\"Supertype (non-expanded)\"] = to_update[\"Supertype (non-expanded)\"].cat.add_categories([\"OPC_2_2-SEAAD\"])\n",
    "        to_update.loc[sub.obs_names[signature_score > (2/3)], \"Supertype (non-expanded)\"] = \"OPC_2_2-SEAAD\"\n",
    "        sub.obs[\"Supertype (non-expanded)\"] = sub.obs[\"Supertype (non-expanded)\"].cat.add_categories([\"OPC_2_2-SEAAD\"])\n",
    "        sub.obs.loc[signature_score > (2/3), \"Supertype (non-expanded)\"] = \"OPC_2_2-SEAAD\"\n",
    "        \n",
    "    elif j == \"Astrocyte\":\n",
    "        # Add SEA-AD specific populations\n",
    "        signature_score = sc.get.obs_df(\n",
    "            sub,\n",
    "            [\"HILPDA\", \"CD44\", \"ARHGEF3\", \"TPST1\", \"SERPINA3\"],\n",
    "        )\n",
    "        signature_score = ((signature_score - signature_score.mean(axis=0)) / signature_score.std(axis=0)).mean(axis=1)\n",
    "        to_update[\"Supertype (non-expanded)\"] = to_update[\"Supertype (non-expanded)\"].cat.add_categories([\"Astro_6-SEAAD\"])\n",
    "        to_update.loc[sub.obs_names[signature_score > 1.5], \"Supertype (non-expanded)\"] = \"Astro_6-SEAAD\"\n",
    "        sub.obs[\"Supertype (non-expanded)\"] = sub.obs[\"Supertype (non-expanded)\"].cat.add_categories([\"Astro_6-SEAAD\"])\n",
    "        sub.obs.loc[signature_score > 1.5, \"Supertype (non-expanded)\"] = \"Astro_6-SEAAD\"\n",
    "\n",
    "        # Fix calling of Astro_3\n",
    "        signature_score = sc.get.obs_df(\n",
    "            sub,\n",
    "            [\"LINC00609\", \"AC012405.1\", \"L3MBTL4\", \"ADAMTSL3\", \"DCLK1\"],\n",
    "        )\n",
    "        signature_score = ((signature_score - signature_score.mean(axis=0)) / signature_score.std(axis=0)).mean(axis=1)\n",
    "        selected_cells = (signature_score < 1.5) & (sub.obs[\"leiden\"] != \"15\") & (sub.obs[\"Supertype (non-expanded)\"] == \"Astro_3\")\n",
    "        to_update.loc[sub.obs_names[selected_cells], \"Supertype (non-expanded)\"] = \"Astro_2\"\n",
    "        sub.obs.loc[selected_cells, \"Supertype (non-expanded)\"] = \"Astro_2\"\n",
    "        \n",
    "    elif j == \"Microglia-PVM\":\n",
    "        # Fix calling of low abundance Micro-PVM_1\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"11\"])], \"Supertype (non-expanded)\"] = \"Micro-PVM_1\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"11\"]), \"Supertype (non-expanded)\"] = \"Micro-PVM_1\"\n",
    "        \n",
    "        # Add SEA-AD specific populations\n",
    "        to_update[\"Supertype (non-expanded)\"] = to_update[\"Supertype (non-expanded)\"].cat.add_categories([\"Monocyte\", \"Micro-PVM_2_1-SEAAD\", \"Micro-PVM_2_3-SEAAD\", \"Micro-PVM_3-SEAAD\", \"Micro-PVM_4-SEAAD\"])\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"3\"])], \"Supertype (non-expanded)\"] = \"Monocyte\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"14\"])], \"Supertype (non-expanded)\"] = \"Micro-PVM_2_1-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"6\"])], \"Supertype (non-expanded)\"] = \"Micro-PVM_3-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"10\"])], \"Supertype (non-expanded)\"] = \"Micro-PVM_4-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"12\", \"17\"])], \"Supertype (non-expanded)\"] = \"Micro-PVM_2_3-SEAAD\"\n",
    "\n",
    "\n",
    "        sub.obs[\"Supertype (non-expanded)\"] = sub.obs[\"Supertype (non-expanded)\"].cat.add_categories([\"Monocyte\", \"Micro-PVM_2_1-SEAAD\", \"Micro-PVM_2_3-SEAAD\", \"Micro-PVM_3-SEAAD\", \"Micro-PVM_4-SEAAD\"])\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"3\"]), \"Supertype (non-expanded)\"] = \"Monocyte\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"14\"]), \"Supertype (non-expanded)\"] = \"Micro-PVM_2_1-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"6\"]), \"Supertype (non-expanded)\"] = \"Micro-PVM_3-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"10\"]), \"Supertype (non-expanded)\"] = \"Micro-PVM_4-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"12\", \"17\"]), \"Supertype (non-expanded)\"] = \"Micro-PVM_2_3-SEAAD\"\n",
    "\n",
    "    elif j == \"VLMC\":        \n",
    "        # Add SEA-AD specific populations\n",
    "        to_update[\"Supertype (non-expanded)\"] = to_update[\"Supertype (non-expanded)\"].cat.add_categories([\"SMC-SEAAD\", \"VLMC_2-SEAAD\", \"VLMC_3-SEAAD\", \"Pericyte_1\", \"Pericyte_2-SEAAD\"])\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"12\"])], \"Supertype (non-expanded)\"] = \"SMC-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"12\"])], \"Supertype (non-expanded)\"] = \"SMC-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"7\"])], \"Supertype (non-expanded)\"] = \"VLMC_2-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"4\"])], \"Supertype (non-expanded)\"] = \"VLMC_3-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"5\", \"9\"])], \"Supertype (non-expanded)\"] = \"Pericyte_2-SEAAD\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"Supertype (non-expanded)\"] == \"VLMC_2\"], \"Supertype (non-expanded)\"] = \"Pericyte_1\"\n",
    "\n",
    "        sub.obs[\"Supertype (non-expanded)\"] = sub.obs[\"Supertype (non-expanded)\"].cat.add_categories([\"SMC-SEAAD\", \"VLMC_2-SEAAD\", \"VLMC_3-SEAAD\", \"Pericyte_1\", \"Pericyte_2-SEAAD\"])\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"12\"]), \"Supertype (non-expanded)\"] = \"SMC-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"7\"]), \"Supertype (non-expanded)\"] = \"VLMC_2-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"4\"]), \"Supertype (non-expanded)\"] = \"VLMC_3-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"5\", \"9\"]), \"Supertype (non-expanded)\"] = \"Pericyte_2-SEAAD\"\n",
    "        sub.obs.loc[sub.obs[\"Supertype (non-expanded)\"] == \"VLMC_2\", \"Supertype (non-expanded)\"] = \"Pericyte_1\"\n",
    "\n",
    "    elif j == \"Endothelial\":        \n",
    "        # Add SEA-AD specific populations\n",
    "        to_update.loc[sub.obs_names, \"Supertype (non-expanded)\"] = \"Endo_2\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"2\"])], \"Supertype (non-expanded)\"] = \"Endo_1\"\n",
    "        to_update.loc[sub.obs_names[sub.obs[\"leiden\"].isin([\"1\", \"5\"])], \"Supertype (non-expanded)\"] = \"Endo_3\"\n",
    "\n",
    "        sub.obs.loc[:, \"Supertype (non-expanded)\"] = \"Endo_2\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"2\"]), \"Supertype (non-expanded)\"] = \"Endo_1\"\n",
    "        sub.obs.loc[sub.obs[\"leiden\"].isin([\"1\", \"5\"]), \"Supertype (non-expanded)\"] = \"Endo_3\"\n",
    "\n",
    "        \n",
    "    plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "    sc.pl.umap(sub, color=[\"Source\"], size=1)\n",
    "    sc.pl.umap(sub, color=[\"Source\"], groups=[\"Allen\"], size=1)\n",
    "    sc.pl.umap(sub, color=[\"Source\"], groups=[\"Mathys_2023\"], size=1)\n",
    "    sc.pl.umap(sub, color=[\"Source\"], groups=[\"Green_2023\"], size=1)\n",
    "    sc.pl.umap(sub, color=[\"Subclass\", \"Supertype (non-expanded)\", \"Supertype\"], legend_loc=\"on data\", ncols=1, size=1)\n",
    "    qc_vars = [\"Doublet score\", \"Genes detected\", \"Fraction mitochondrial UMIs\"]\n",
    "    for l,k in enumerate(qc_vars):\n",
    "        sc.pl.umap(\n",
    "            sub,\n",
    "            color=k,\n",
    "            frameon=False,\n",
    "            cmap=\"YlGnBu\",\n",
    "            sort_order=False,\n",
    "            size=1\n",
    "        )\n",
    "\n",
    "    sc.pl.umap(sub, color=[\"leiden\"], legend_loc=\"on data\", ncols=1)\n",
    "\n",
    "external_adata.obs[\"Supertype (non-expanded)\"] = to_update[\"Supertype (non-expanded)\"].copy()\n",
    "external_adata.obs[\"Supertype\"] = to_update[\"Supertype (non-expanded)\"].copy()\n",
    "external_adata = external_adata[(to_update[\"Used in analysis\"] == True)].copy()\n",
    "external_adata.obs[\"Used in analysis\"] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afa336c6-f922-429d-9dd3-f0952333bb1e",
   "metadata": {},
   "source": [
    "### Save final mapping results and QC for non-neuronal subclasses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fc3e845-74e3-4272-ab39-fccac98e498c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in external_adata.obs[\"Source\"].cat.categories:\n",
    "    print(i)\n",
    "    external_adata.obs.loc[(external_adata.obs[\"Source\"] == i), :].to_csv(os.path.join(pwd, \"output\", \"PFC_Model\", \"PFC_\" + i + \"_external_metadata_keepers_non-neurons.2023-12-11.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7883e986-5f5f-4d0f-bee1-225f0ad1866e",
   "metadata": {},
   "source": [
    "### Build combined dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fceea23-ecb5-4628-a62c-df9664aa968f",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = []\n",
    "external_datasets = glob.glob(os.path.join(pwd, \"input\", \"*_external_singleomeCR6.2023-10-12.h5ad\"))\n",
    "external_datasets.extend([os.path.join(pwd, \"input\", \"SEAAD_A9_RNAseq_final-nuclei.2023-07-19.h5ad\")])\n",
    "\n",
    "de_genes = pd.read_csv(os.path.join(pwd, \"output\", \"de_genes.csv\"), header=None)\n",
    "de_genes = de_genes[0].to_list()\n",
    "\n",
    "for k in external_datasets:\n",
    "    print(os.path.basename(k))\n",
    "    tmp = sc.read_h5ad(k)\n",
    "    tmp = tmp[:, de_genes].copy()\n",
    "    tmp.X = tmp.layers[\"UMIs\"].copy()\n",
    "    del tmp.layers[\"UMIs\"]\n",
    "    if \"external\" in k:\n",
    "        tmp.obs[\"Source\"] = os.path.basename(k).replace(\"PFC_\", \"\").replace(\"_external_singleomeCR6.2023-10-12.h5ad\", \"\")\n",
    "    else:\n",
    "        tmp.obs[\"Source\"] = \"Allen\"\n",
    "        tmp.obs = tmp.obs.drop([\"Class\", \"Class confidence\", \"Subclass\", \"Subclass confidence\", \"Supertype\", \"Supertype (non-expanded)\", \"Supertype confidence\"], axis=1)\n",
    "    \n",
    "    print(tmp.shape[0])\n",
    "    neuronal_annotations = pd.read_csv(os.path.join(pwd, \"output\", \"PFC_\" + tmp.obs[\"Source\"].iloc[0] + \"_external_metadata_keepers_neurons.2023-11-16.csv\"), index_col=0)\n",
    "    neuronal_annotations = neuronal_annotations.loc[:, [\"Class\", \"Class confidence\", \"Subclass\", \"Subclass confidence\", \"Supertype (non-expanded)\", \"Supertype confidence\"]].copy()\n",
    "    glial_annotations = pd.read_csv(os.path.join(pwd, \"output\", \"PFC_\" + tmp.obs[\"Source\"].iloc[0] + \"_external_metadata_keepers_non-neurons.2023-12-11.csv\"), index_col=0)\n",
    "    glial_annotations = glial_annotations.loc[:, [\"Class\", \"Class confidence\", \"Subclass\", \"Subclass confidence\", \"Supertype (non-expanded)\", \"Supertype confidence\"]].copy()\n",
    "    annotations = pd.concat([neuronal_annotations, glial_annotations], axis=0)\n",
    "    \n",
    "    tmp.obs = tmp.obs.merge(annotations, left_index=True, right_index=True, how=\"left\")\n",
    "    \n",
    "    external_adata.append(tmp)\n",
    "\n",
    "first = external_adata.pop()\n",
    "external_adata = first.concatenate(external_adata, index_unique=None)\n",
    "\n",
    "for i in external_adata.obs.columns[external_adata.obs.isna().sum(axis=0) > 0]:\n",
    "    if any(external_adata.obs[i].notna()) == False:\n",
    "        print(\"Dropping no-value column \" + i)\n",
    "        external_adata.obs.drop([i], axis=1, inplace=True)\n",
    "            \n",
    "    else:            \n",
    "        replace_with = \"\"\n",
    "        \n",
    "        if isinstance(external_adata.obs.loc[external_adata.obs[i].notna(), i][0], np.float64) == True or isinstance(external_adata.obs.loc[external_adata.obs[i].notna(), i][0], np.float32) == True:\n",
    "            replace_with = 0.0\n",
    "\n",
    "        if isinstance(external_adata.obs[i].dtype, pd.core.dtypes.dtypes.CategoricalDtype) == True:\n",
    "            external_adata.obs[i] = external_adata.obs[i].astype(\"object\")\n",
    "\n",
    "        if external_adata.obs[i].dtype == \"bool\" or external_adata.obs[i].dtype == \"boolean\":\n",
    "                external_adata.obs[i] = external_adata.obs[i].astype(\"object\")\n",
    "                external_adata.obs[i] = [str(l) for l in external_adata.obs[i]]\n",
    "        \n",
    "        print(\"Replacing NaNs with \" + str(replace_with) + \" for \" + i + \" with dtype \" + str(type(external_adata.obs.loc[external_adata.obs[i].notna(), i][0])))\n",
    "        external_adata.obs.loc[external_adata.obs[i].isna(), i] = replace_with\n",
    "        \n",
    "        if isinstance(external_adata.obs.loc[(external_adata.obs[i].notna()) & (external_adata.obs[i] != \"\"), i][0], bool) == True:\n",
    "            external_adata.obs[i] = external_adata.obs[i].astype(\"str\")\n",
    "            \n",
    "external_adata.obs[\"Donor ID\"] = external_adata.obs[\"Donor ID\"].astype(\"str\")\n",
    "external_adata.obs[\"Age at Death\"] = external_adata.obs[\"Age at Death\"].astype(\"str\")\n",
    "for i in external_adata.obs.columns[(external_adata.obs == \"Reference\").sum(axis=0) > 0]:\n",
    "    external_adata.obs[i] = external_adata.obs[i].astype(\"str\")\n",
    "    \n",
    "external_adata.obs[\"Used in analysis\"] = (external_adata.obs[\"Class\"] != \"\").astype(\"str\")\n",
    "\n",
    "external_adata.write(\"output\", \"PFC_external_SEAAD_singleomeCR6_UMIs_only_de_genes_final.2023-12-12.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aedfabbe-49e0-4b54-b836-3e754ecee34c",
   "metadata": {},
   "source": [
    "### Obtain representation of the whole dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "babfe2a9-6829-4de5-ae7e-8d866fe870bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_adata = sc.read_h5ad(\"output\", \"PFC_external_SEAAD_singleomeCR6_UMIs_expanded_de_genes_final.2023-12-12.h5ad\")\n",
    "selected_cells = (external_adata.obs[\"Cognitive Status\"] != \"Reference\") & (external_adata.obs[\"Used in analysis\"] == \"True\")\n",
    "\n",
    "external_adata = external_adata[selected_cells, :].copy()\n",
    "var_names = pd.read_csv(os.path.join(pwd, \"output\", \"External_Data_Models\", \"Full\", \"var_names.csv\"), header=None).iloc[:, 0].to_list()\n",
    "sub =  external_adata[:, var_names].copy()\n",
    "model = scvi.model.SCVI.load(os.path.join(pwd, \"output\", \"External_Data_Models\", \"Full\"), sub)\n",
    "del sub\n",
    "\n",
    "external_adata.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "print(str(datetime.now()) + \" -- Computed latent representation\")\n",
    "rsc.pp.neighbors(external_adata, use_rep=\"X_scVI\", n_neighbors=50)\n",
    "print(str(datetime.now()) + \" -- Computed neighbors\")\n",
    "rsc.tl.umap(external_adata, init_pos=\"random\")\n",
    "print(str(datetime.now()) + \" -- Computed UMAP\")\n",
    "\n",
    "sc.pp.subsample(external_adata, fraction=1)\n",
    "\n",
    "subclass_colors = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "subclass_colors = subclass_colors.loc[:, [\"subclass_label\", \"subclass_color\"]]\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"]\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "supertype_colors = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "supertype_colors = supertype_colors.loc[:, [\"cluster_label\", \"cluster_color\"]]\n",
    "supertype_colors.index = supertype_colors[\"cluster_label\"]\n",
    "supertype_colors = supertype_colors[\"cluster_color\"].to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1383754f-2c8a-44e9-88f3-9d809a658e5c",
   "metadata": {},
   "source": [
    "### Obtain final neighborhood representations signature scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13304c96-c50e-41e0-a9d7-146924959eb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "neighborhoods = {\n",
    "    \"MGE\": [\"Sst Chodl\", \"Sst\", \"Pvalb\", \"Chandelier\"], # 4\n",
    "    \"CGE\": [\"Lamp5\", \"Lamp5 Lhx6\", \"Pax6\", \"Sncg\", \"Vip\"], # 5\n",
    "    \"IT\": [\"L2/3 IT\", \"L4 IT\", \"L5 IT\", \"L6 IT\", \"L6 IT Car3\"], # 5\n",
    "    \"Deep Projecting\": [\"L5 ET\", \"L5/6 NP\", \"L6b\", \"L6 CT\"], # 4\n",
    "    \"Glia Refined\": [\"Oligodendrocyte\", \"OPC\", \"Astrocyte\"], # 3\n",
    "    \"Vascular and Immune Refined\": [\"Endothelial\", \"VLMC\", \"Microglia-PVM\"] # 3\n",
    "}\n",
    "\n",
    "fingerprint_correlation = {}\n",
    "for i,j in neighborhoods.items():\n",
    "    print(i)\n",
    "    neighborhood_model_args = {\n",
    "        \"n_layers\": 2,\n",
    "        \"n_latent\": 20,\n",
    "        \"dispersion\": \"gene-batch\",\n",
    "    }\n",
    "    neighborhood_external_adata = external_adata[external_adata.obs[\"Subclass\"].isin(j)].copy()\n",
    "    scvi.model.SCVI.setup_anndata(\n",
    "        neighborhood_external_adata,\n",
    "        layer=None,\n",
    "        batch_key=\"Source\",\n",
    "        categorical_covariate_keys=[\"library_prep\"]\n",
    "    )\n",
    "    if os.path.exists(os.path.join(pwd, \"output\", \"External_Data_Models\", i)) == False:\n",
    "        neighborhood_model = scvi.model.SCVI(neighborhood_external_adata, **neighborhood_model_args)\n",
    "        print(neighborhood_model)\n",
    "        print(neighborhood_model.view_anndata_setup())\n",
    "        neighborhood_model.train(max_epochs=500, early_stopping=True, use_gpu=True)\n",
    "        neighborhood_model.save(os.path.join(pwd, \"output\", \"External_Data_Models\", i))\n",
    "    else:\n",
    "        if os.path.exists(os.path.join(pwd, \"output\", \"External_Data_Models\", i, \"var_names.csv\")) == False:\n",
    "            tmp = torch.load(os.path.join(pwd, \"output\", \"External_Data_Models\", \"model.pt\"))\n",
    "            pd.DataFrame(tmp[\"var_names\"]).to_csv(os.path.join(pwd, \"output\", \"External_Data_Models\", i, \"var_names.csv\"), index=False, header=False)\n",
    "        var_names = pd.read_csv(os.path.join(pwd,\"output\", \"External_Data_Models\", i, \"var_names.csv\"), header=None).iloc[:, 0].to_list()\n",
    "        sub = neighborhood_external_adata[:, var_names].copy()\n",
    "        neighborhood_model = scvi.model.SCVI.load(os.path.join(pwd, \"output\", \"External_Data_Models\", i), sub)\n",
    "        del sub\n",
    "\n",
    "    neighborhood_external_adata.obsm[\"X_scVI\"] = neighborhood_model.get_latent_representation()\n",
    "    print(str(datetime.now()) + \" -- Computed latent representation\")\n",
    "    rsc.pp.neighbors(neighborhood_external_adata, use_rep=\"X_scVI\", n_neighbors=50)\n",
    "    print(str(datetime.now()) + \" -- Computed neighbors\")\n",
    "    rsc.tl.umap(neighborhood_external_adata)\n",
    "    print(str(datetime.now()) + \" -- Computed UMAP\")\n",
    "\n",
    "    sc.pp.normalize_total(neighborhood_external_adata, 1e5)\n",
    "    sc.pp.log1p(neighborhood_external_adata)\n",
    "    sc.tl.rank_genes_groups(neighborhood_external_adata, groupby=\"Supertype\", groups=np.setdiff1d(neighborhood_external_adata.obs[\"Supertype\"], \"\").tolist())\n",
    "    sc.pp.scale(neighborhood_external_adata)\n",
    "    signature_genes = pd.DataFrame(neighborhood_external_adata.uns['rank_genes_groups']['names']).head(10)\n",
    "    \n",
    "    fingerprint_correlation[i] = pd.DataFrame(\n",
    "        np.zeros((len(neighborhood_external_adata.obs[\"Source\"].cat.categories), len(np.setdiff1d(neighborhood_external_adata.obs[\"Supertype\"], \"\").tolist()))),\n",
    "        index=neighborhood_external_adata.obs[\"Source\"].cat.categories,\n",
    "        columns=np.setdiff1d(neighborhood_external_adata.obs[\"Supertype\"], \"\").tolist(),\n",
    "    )\n",
    "    \n",
    "    for z in np.setdiff1d(neighborhood_external_adata.obs[\"Supertype\"], \"\").tolist():\n",
    "        neighborhood_external_adata.obs[\"Signature Score\"] = neighborhood_external_adata[:, signature_genes[z]].X.mean(axis=1)\n",
    "        signature_scores = sc.get.obs_df(neighborhood_external_adata, [\"Source\", \"Supertype\", \"Signature Score\"])\n",
    "        signature_scores = signature_scores.groupby([\"Source\", \"Supertype\"]).mean().reset_index().pivot(index=\"Source\", columns=\"Supertype\", values=\"Signature Score\").fillna(0)\n",
    "        for y in neighborhood_external_adata.obs[\"Source\"].cat.categories:\n",
    "            fingerprint_correlation[i].loc[y, z] = sp_stats.spearmanr(signature_scores.loc[\"Allen\", :], signature_scores.loc[y, :])[0]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
