{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8568aff-eaab-4658-b3a6-3aaa27c551d4",
   "metadata": {},
   "source": [
    "## Load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0801d1-e029-4555-bf96-c250f809ecde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "import scvi\n",
    "import anndata\n",
    "import scipy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d23c69-4d87-46b8-b608-f0bca0335d99",
   "metadata": {},
   "source": [
    "## Load in anndata generated by pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede4156c-4a8b-421a-acd8-ce0ac9f09574",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad = anndata.read_h5ad(filename=\"multi_Endo.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a5ef12f-859e-4cbb-9df2-d10159d5fa63",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_rna = ad[ad.obs.modality == \"rna\", ad.var.modality == \"Gene Expression\"].copy()\n",
    "ad_paired = ad[ad.obs.modality == \"paired\"].copy()\n",
    "ad_atac = ad[ad.obs.modality == \"atac\", ad.var.modality == \"Peaks\"].copy()\n",
    "\n",
    "ad_p = anndata.AnnData(X=ad_paired.X, obs=ad_paired.obs, var=ad_paired.var)\n",
    "ad_a = anndata.AnnData(X=ad_atac.X, obs=ad_atac.obs, var=ad_atac.var)\n",
    "\n",
    "rna_seaad = anndata.read_h5ad(filename='/allen/programs/celltypes/workgroups/hct/SEA-AD/Integration/multivi_subclasses/singleomeRNA_MTG_Full/Endo.h5ad')\n",
    "\n",
    "var_names = ad_rna.var.index.values\n",
    "print(var_names)\n",
    "\n",
    "ad_rna_new = rna_seaad[:, var_names].copy()\n",
    "\n",
    "print(sum(ad_rna_new.var.index.values != ad_rna.var.index.values))\n",
    "\n",
    "ad_r = anndata.AnnData(X=ad_rna_new.X, obs=ad_rna_new.obs, var=ad_rna_new.var)\n",
    "\n",
    "# We can now use the organizing method from scvi to concatenate these anndata\n",
    "adata_mvi = scvi.data.organize_multiome_anndatas(ad_p, ad_r, ad_a)\n",
    "\n",
    "adata_mvi = adata_mvi[:, adata_mvi.var[\"modality\"].argsort()].copy()\n",
    "\n",
    "del ad, ad_rna, ad_paired, ad_atac, rna_seaad, ad_rna_new, ad_r, ad_p, ad_a\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2425378-a765-48a8-b9e3-60e8b6a77629",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a mapping of values to replace\n",
    "replace_dict = {'Female': 'F', 'Male': 'M'}\n",
    "\n",
    "# replace the values using replace method\n",
    "adata_mvi.obs['sex'] = adata_mvi.obs['sex'].replace(replace_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af58411e-43be-49ce-983f-1033630311f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi.model.MULTIVI.setup_anndata(adata_mvi, batch_key='modality', categorical_covariate_keys=[\"donor_name\", \"sex\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ee84d14-ee45-4bc8-bbeb-97eb60b3d72d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvi = scvi.model.MULTIVI(\n",
    "    adata_mvi,\n",
    "    n_genes=(adata_mvi.var['modality']=='Gene Expression').sum(),\n",
    "    n_regions=(adata_mvi.var['modality']=='Peaks').sum(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b1e8929-7504-48a9-a972-47ebe4a28689",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvi.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44e20592-546f-4e32-94fb-f1a73f68c9f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.obsm[\"X_MultiVI\"] = mvi.get_latent_representation()\n",
    "sc.pp.neighbors(adata_mvi, use_rep=\"X_MultiVI\")\n",
    "sc.tl.umap(adata_mvi, min_dist=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c222c2d6-9a03-4c7c-8022-f2fb59c6e306",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='modality')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b6e9e28-a2b8-4422-a9bc-9252e95378c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color='sex')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb45207-8623-4c07-a44f-7fac34d0ca29",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color='donor_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5151ac0-8772-4134-bb09-e7ccbd0469e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_label_purity_slot(adata_mvi, cell_idx, slot=\"subclass_scANVI\"):\n",
    "    idx = np.where(adata_mvi.uns['neighbors']['connectivities'][cell_idx].todense()>0)[1]\n",
    "    df = adata_mvi.obs[[slot]].iloc[idx]\n",
    "    df = df.loc[~df[slot].isnull()]\n",
    "    u, c = np.unique(df, return_counts=True)\n",
    "    if np.size(c) == 0:\n",
    "        ratio = 0\n",
    "        label = np.nan\n",
    "    else:\n",
    "        ratio = c[np.argmax(c)] / c.sum()\n",
    "        label = u[np.argmax(c)]\n",
    "    \n",
    "    return ratio, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43f42cc1-aa1c-469a-991c-e12fde994df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata_mvi.shape[0]\n",
    "purity_ratios = np.zeros(cells)\n",
    "labels = adata_mvi.obs[\"supertype_scANVI\"].to_numpy()\n",
    "for i in np.arange(cells):\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI\")\n",
    "    purity_ratios[i] = ratio\n",
    "    labels[i] = label_\n",
    "\n",
    "adata_mvi.obs[\"supertype_scANVI_purity_ratio\"] = purity_ratios\n",
    "adata_mvi.obs[\"supertype_scANVI_label_transfer\"] = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "356070ce-6974-4513-964d-0e356add7eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in np.where(adata_mvi.obs[\"supertype_scANVI_label_transfer\"].isnull())[0]:\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_label_transfer\")\n",
    "    adata_mvi.obs[\"supertype_scANVI_purity_ratio\"].loc[i] = ratio\n",
    "    adata_mvi.obs[\"supertype_scANVI_label_transfer\"].loc[i] = label_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9f3809-4747-407e-b90d-61c3945dd6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata_mvi.obs[\"supertype_scANVI_label_transfer\"].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624fd5f9-b17c-443d-8359-d6dfa49b3a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata_mvi.shape[0]\n",
    "purity_ratios = np.zeros(cells)\n",
    "labels = adata_mvi.obs[\"supertype_scANVI_leiden\"].to_numpy()\n",
    "for i in np.arange(cells):\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_leiden\")\n",
    "    purity_ratios[i] = ratio\n",
    "    labels[i] = label_\n",
    "\n",
    "adata_mvi.obs[\"supertype_scANVI_leiden_purity_ratio\"] = purity_ratios\n",
    "adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"] = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84fe814-cc6e-43b3-b240-357898661519",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in np.where(adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].isnull())[0]:\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_leiden_label_transfer\")\n",
    "    adata_mvi.obs[\"supertype_scANVI_leiden_purity_ratio\"].loc[i] = ratio\n",
    "    adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].loc[i] = label_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b4b9257-da8c-4017-b31f-636a92f1c7da",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e97811fb-2275-49bb-8974-b2703bec6d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden_label_transfer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dc98c29-97f1-4b82-926e-e3523bbc30ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color=\"supertype_scANVI_leiden\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ece63aef-80fc-45bd-8578-169b573aedc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from joblib import parallel_backend\n",
    "sc.settings.n_jobs=32\n",
    "with parallel_backend('threading', n_jobs=32):\n",
    "    sc.tl.leiden(adata_mvi, key_added = \"leiden_1.0\", resolution = 1.0) # default resolution in 1.0\n",
    "## Scanpy Clustering Leiden (https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/scanpy/scanpy_04_clustering.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93938f61-ca89-43a8-8b18-2402010235e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color=['leiden_1.0'], legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef7c1761-9061-4d74-86d7-dbdeae3eb17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## replace old names to new names.\n",
    "new_names = pd.read_csv('/allen/programs/celltypes/workgroups/hct/SEA-AD/RNAseq/diff_prop/input/MTG/old_names_to_new_names.csv', index_col=0)\n",
    "\n",
    "# create a dictionary\n",
    "name_dict = dict(zip(new_names['old_names'], new_names['new_names']))\n",
    "\n",
    "# replace specific values in a DataFrame column with other values, use .replace()\n",
    "adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'] = adata_mvi.obs['supertype_scANVI_leiden_label_transfer'].replace(name_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c3cc0fc-c9a4-4c0a-96e3-0440bac6ee9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden_label_transfer_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6954189-0040-4786-8ec2-cdece149cd3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_peaks = np.asarray(np.sum(adata_mvi[:, adata_mvi.var.modality == 'Peaks'].X, axis=1)).reshape(-1)\n",
    "adata_mvi.obs['total_peak_count'] = total_peaks\n",
    "meta = adata_mvi.obs.copy()\n",
    "df2 = meta.groupby(['leiden_1.0'])['total_peak_count'].mean()\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ff5c79-c2aa-4532-afc0-32c3e4de50bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data types of all columns.\n",
    "dtypes = adata_mvi.obs.dtypes\n",
    "# Find all non-string columns\n",
    "non_string_cols = dtypes[dtypes == 'object'].index.tolist()\n",
    "adata_mvi.obs[non_string_cols] = adata_mvi.obs[non_string_cols].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36498313-79c1-4a8e-9355-c2f2b212a689",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.write(\"multivi_AD_Endo.h5ad\")\n",
    "mvi.save(\"trained_multivi_AD_Endo\", save_anndata=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a26eb798-131e-40ca-bfca-2c5368555c03",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a03bba3-2355-4d6f-8009-95e229895854",
   "metadata": {},
   "outputs": [],
   "source": [
    "# column contains the pattern 'Endo' or is null\n",
    "idx = adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'].isnull() | adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'].str.contains('Endo') \n",
    "\n",
    "# Subset the anndata object using the boolean index\n",
    "ad = adata_mvi[idx, :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf3b5d5-94db-48a0-91c2-0d0da3eda061",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(ad.obs['supertype_scANVI_leiden_label_transfer_new'].isnull())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ba6d4da-5a4b-43c0-ad3d-efc6b147930e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(ad, color='supertype_scANVI_leiden_label_transfer_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8b5fa7-de0a-4f7f-b739-9433f640324c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ad.obs.groupby([\"leiden_1.0\", \"supertype_scANVI_leiden_label_transfer_new\"]).size().reset_index(name=\"Counts\")\n",
    "df_wide = df.pivot_table(index=\"leiden_1.0\", columns=\"supertype_scANVI_leiden_label_transfer_new\", values=\"Counts\")\n",
    "import seaborn as sns\n",
    "sns.heatmap(df_wide.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0a2bc3-b89c-4092-a5b2-f24f599c942b",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_folder = os.path.abspath(os.path.join(os.getcwd(), os.pardir))\n",
    "csv_file_path = os.path.join(parent_folder, 'donor_name_ch_cognitivestatus_binary.csv')\n",
    "selected_data = pd.read_csv(csv_file_path)\n",
    "\n",
    "selected_dict = dict(zip(selected_data['donor_name'], selected_data['ch_cognitivestatus_binary']))\n",
    "adata_mvi.obs['ch_cognitivestatus_binary'] = adata_mvi.obs['donor_name'].map(selected_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90c2b1f8-7702-4c09-b2da-554d02ce2ba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='nCount_RNA')\n",
    "sc.pl.umap(adata_mvi, color='nFeature_RNA')\n",
    "sc.pl.umap(adata_mvi, color='ch_cognitivestatus_binary')\n",
    "sc.pl.umap(adata_mvi, color='total_peak_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83d871f6-7c65-46a9-9569-320ebbe89840",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi = anndata.read_h5ad(\"multivi_AD_Endo.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53d9a575-757b-41c9-a15a-9655619bb17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = adata_mvi.obs.copy()\n",
    "meta = meta.reset_index()\n",
    "index_col_name = meta.columns[0]\n",
    "meta = meta.rename(columns={index_col_name: 'index_name'})\n",
    "meta.to_csv(\"multivi_AD_Endo.h5ad.obs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6830cc-035d-4c9c-af39-0dee846295b1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
