{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8568aff-eaab-4658-b3a6-3aaa27c551d4",
   "metadata": {},
   "source": [
    "## Load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0801d1-e029-4555-bf96-c250f809ecde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "import scvi\n",
    "import anndata\n",
    "import scipy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d23c69-4d87-46b8-b608-f0bca0335d99",
   "metadata": {},
   "source": [
    "## Load in anndata generated by pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede4156c-4a8b-421a-acd8-ce0ac9f09574",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad = anndata.read_h5ad(filename=\"multi_L6b.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e67b41a-d694-43cd-a1c4-7fcf102f55ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_rna = ad[ad.obs.modality == \"rna\", ad.var.modality == \"Gene Expression\"].copy()\n",
    "ad_paired = ad[ad.obs.modality == \"paired\"].copy()\n",
    "ad_atac = ad[ad.obs.modality == \"atac\", ad.var.modality == \"Peaks\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bc6508e-f07e-4963-8590-d03a0ddff002",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We also filter features to remove those that appear in fewer than 1% of the cells\n",
    "# This instruction removes genes or peaks that are not expressed/accessible in more than 1% of the cells.\n",
    "print(ad_atac.shape)\n",
    "min_cells = int(ad_atac.shape[0] * 0.01)\n",
    "sc.pp.filter_genes(ad_atac, min_cells=min_cells)\n",
    "print(ad_atac.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e4470ec-99e8-4610-87f4-0142a8ba188e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seaad = anndata.read_h5ad(filename='/allen/programs/celltypes/workgroups/hct/SEA-AD/Integration/multivi_subclasses/singleomeRNA_MTG_Full/L6b.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8956cd03-c090-4b2c-996d-9b1706ded521",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_names = ad_rna.var.index.values\n",
    "print(var_names)\n",
    "## merged multi_L6b.h5ad only has part of genes.\n",
    "ad_rna_new = rna_seaad[:, var_names].copy()\n",
    "print(sum(ad_rna_new.var.index.values != ad_rna.var.index.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81428c43-b328-444d-895e-36637b10f234",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ad_rna_new.shape)\n",
    "min_cells = int(ad_rna_new.shape[0] * 0.01)\n",
    "sc.pp.filter_genes(ad_rna_new, min_cells=min_cells)\n",
    "print(ad_rna_new.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1683e221-3eb4-4f9b-9bed-6bb5ff832129",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_p = anndata.AnnData(X=ad_paired.X, obs=ad_paired.obs, var=ad_paired.var)\n",
    "ad_a = anndata.AnnData(X=ad_atac.X, obs=ad_atac.obs, var=ad_atac.var)\n",
    "ad_r = anndata.AnnData(X=ad_rna_new.X, obs=ad_rna_new.obs, var=ad_rna_new.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49655d1d-0e9e-4ae8-b0f7-aa41b2325b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = ad_a.var_names\n",
    "genes = ad_r.var_names\n",
    "features = genes.union(peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8e81474-c15c-4106-bed7-9a881057487e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ad_p.shape)\n",
    "ad_p = ad_p[:, features].copy()\n",
    "print(ad_p.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c37b452-9396-4706-81a1-b99b51b5bd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can now use the organizing method from scvi to concatenate these anndata\n",
    "adata_mvi = scvi.data.organize_multiome_anndatas(ad_p, ad_r, ad_a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "377be546-1f5c-459a-9b4d-ba0a49b05d1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi = adata_mvi[:, adata_mvi.var[\"modality\"].argsort()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f07c82b-968d-4a55-9394-a512caac16c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "del ad, ad_rna, ad_paired, ad_atac, rna_seaad, ad_rna_new, ad_r, ad_p, ad_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c09cb88f-bf84-4f00-bf37-f90b99423fa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a mapping of values to replace\n",
    "replace_dict = {'Female': 'F', 'Male': 'M'}\n",
    "\n",
    "# replace the values using replace method\n",
    "adata_mvi.obs['sex'] = adata_mvi.obs['sex'].replace(replace_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c37055-1244-4dff-85a3-3a230119564a",
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi.model.MULTIVI.setup_anndata(adata_mvi, batch_key='modality', categorical_covariate_keys=[\"donor_name\", \"sex\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8bab66-dce3-4fe0-b68d-f2f3616eebd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvi = scvi.model.MULTIVI(\n",
    "    adata_mvi,\n",
    "    n_genes=(adata_mvi.var['modality']=='Gene Expression').sum(),\n",
    "    n_regions=(adata_mvi.var['modality']=='Peaks').sum(),\n",
    "    n_layers_encoder=1, n_layers_decoder=1, n_hidden=256,\n",
    "    n_latent=8,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "853e4458-6bb7-4934-bbdd-8a9c96579323",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvi.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb383059-65ff-4700-96f3-0520a76aa3dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.obsm[\"X_MultiVI\"] = mvi.get_latent_representation()\n",
    "sc.pp.neighbors(adata_mvi, use_rep=\"X_MultiVI\")\n",
    "sc.tl.umap(adata_mvi, min_dist=0.2)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='modality')\n",
    "\n",
    "sc.pl.umap(adata_mvi, color='sex')\n",
    "\n",
    "sc.pl.umap(adata_mvi, color='donor_name')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31171169-f9c6-4016-94ed-0f906efbc3da",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_label_purity_slot(adata_mvi, cell_idx, slot=\"subclass_scANVI\"):\n",
    "    idx = np.where(adata_mvi.uns['neighbors']['connectivities'][cell_idx].todense()>0)[1]\n",
    "    df = adata_mvi.obs[[slot]].iloc[idx]\n",
    "    df = df.loc[~df[slot].isnull()]\n",
    "    u, c = np.unique(df, return_counts=True)\n",
    "    if np.size(c) == 0:\n",
    "        ratio = 0\n",
    "        label = np.nan\n",
    "    else:\n",
    "        ratio = c[np.argmax(c)] / c.sum()\n",
    "        label = u[np.argmax(c)]\n",
    "    \n",
    "    return ratio, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af45e35c-a874-4764-b5f0-35186ac8fd74",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata_mvi.shape[0]\n",
    "purity_ratios = np.zeros(cells)\n",
    "labels = adata_mvi.obs[\"supertype_scANVI\"].to_numpy()\n",
    "for i in np.arange(cells):\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI\")\n",
    "    purity_ratios[i] = ratio\n",
    "    labels[i] = label_\n",
    "\n",
    "adata_mvi.obs[\"supertype_scANVI_purity_ratio\"] = purity_ratios\n",
    "adata_mvi.obs[\"supertype_scANVI_label_transfer\"] = labels\n",
    "\n",
    "for i in np.where(adata_mvi.obs[\"supertype_scANVI_label_transfer\"].isnull())[0]:\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_label_transfer\")\n",
    "    adata_mvi.obs[\"supertype_scANVI_purity_ratio\"].loc[i] = ratio\n",
    "    adata_mvi.obs[\"supertype_scANVI_label_transfer\"].loc[i] = label_\n",
    "    \n",
    "print(adata_mvi.obs[\"supertype_scANVI_label_transfer\"].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "591e1b97-e064-4c11-901f-5d5da3f012a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata_mvi.shape[0]\n",
    "purity_ratios = np.zeros(cells)\n",
    "labels = adata_mvi.obs[\"supertype_scANVI_leiden\"].to_numpy()\n",
    "for i in np.arange(cells):\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_leiden\")\n",
    "    purity_ratios[i] = ratio\n",
    "    labels[i] = label_\n",
    "\n",
    "adata_mvi.obs[\"supertype_scANVI_leiden_purity_ratio\"] = purity_ratios\n",
    "adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"] = labels\n",
    "\n",
    "for i in np.where(adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].isnull())[0]:\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_leiden_label_transfer\")\n",
    "    adata_mvi.obs[\"supertype_scANVI_leiden_purity_ratio\"].loc[i] = ratio\n",
    "    adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].loc[i] = label_\n",
    "    \n",
    "print(adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f094b3-687b-4268-8ded-a0d8928b2e5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden_label_transfer')\n",
    "sc.pl.umap(adata_mvi, color=\"supertype_scANVI_leiden\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a694988d-b209-4381-88f5-9522cea72ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.obs['supertype_scANVI_leiden_label_transfer'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b3f4e39-7861-4518-93e2-bf7f26e43bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from joblib import parallel_backend\n",
    "sc.settings.n_jobs=32\n",
    "with parallel_backend('threading', n_jobs=32):\n",
    "    sc.tl.leiden(adata_mvi, key_added = \"leiden_1.0\") # default resolution in 1.0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50105bc8-eedc-45e9-9c1e-000f3e26cb04",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color=['leiden_1.0'], legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c41488c8-21ca-4e7c-92c0-4fda0c02caf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_peaks = np.asarray(np.sum(adata_mvi[:, adata_mvi.var.modality == 'Peaks'].X, axis=1)).reshape(-1)\n",
    "adata_mvi.obs['total_peak_count'] = total_peaks\n",
    "meta = adata_mvi.obs.copy()\n",
    "df2 = meta.groupby(['leiden_1.0'])['total_peak_count'].mean()\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e4af77-98e4-45f8-9300-5ceea0a1afd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_names = pd.read_csv('/allen/programs/celltypes/workgroups/hct/SEA-AD/RNAseq/diff_prop/input/MTG/old_names_to_new_names.csv', index_col=0)\n",
    "\n",
    "# create a dictionary\n",
    "name_dict = dict(zip(new_names['old_names'], new_names['new_names']))\n",
    "\n",
    "# replace specific values in a DataFrame column with other values, use .replace()\n",
    "adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'] = adata_mvi.obs['supertype_scANVI_leiden_label_transfer'].replace(name_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffb016c6-9c5e-47fd-8f17-71a025358dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden_label_transfer_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6bae332-bbf2-4199-97bb-7409ac2ed060",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data types of all columns.\n",
    "dtypes = adata_mvi.obs.dtypes\n",
    "# Find all non-string columns\n",
    "non_string_cols = dtypes[dtypes == 'object'].index.tolist()\n",
    "adata_mvi.obs[non_string_cols] = adata_mvi.obs[non_string_cols].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24ac78c5-a1f2-43d9-adb2-04676f898874",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.write(\"multivi_AD_L6b.h5ad\")\n",
    "mvi.save(\"trained_multivi_AD_L6b\", save_anndata=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "290ef29f-9098-425e-bf00-10dd68e2a29e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cut_off = (adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'].isnull()) | (adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'].str.contains('L6b_'))\n",
    "ad = adata_mvi[cut_off].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30c6f9ca-79ce-47d7-a893-f64a0596db1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(ad, color='supertype_scANVI_leiden_label_transfer_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ccefff4-d995-43f2-b59e-20f9dee7dbe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ad.obs.groupby([\"leiden_1.0\", \"supertype_scANVI_leiden_label_transfer_new\"]).size().reset_index(name=\"Counts\")\n",
    "df_wide = df.pivot_table(index=\"leiden_1.0\", columns=\"supertype_scANVI_leiden_label_transfer_new\", values=\"Counts\")\n",
    "import seaborn as sns\n",
    "sns.heatmap(df_wide.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffa6af7e-d8cf-40d5-8f0d-344084a19fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_folder = os.path.abspath(os.path.join(os.getcwd(), os.pardir))\n",
    "csv_file_path = os.path.join(parent_folder, 'donor_name_ch_cognitivestatus_binary.csv')\n",
    "selected_data = pd.read_csv(csv_file_path)\n",
    "\n",
    "selected_dict = dict(zip(selected_data['donor_name'], selected_data['ch_cognitivestatus_binary']))\n",
    "adata_mvi.obs['ch_cognitivestatus_binary'] = adata_mvi.obs['donor_name'].map(selected_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "770a1a2c-ad3b-463e-b775-75415297f469",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='nCount_RNA')\n",
    "sc.pl.umap(adata_mvi, color='nFeature_RNA')\n",
    "sc.pl.umap(adata_mvi, color='ch_cognitivestatus_binary')\n",
    "sc.pl.umap(adata_mvi, color='total_peak_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bce4e354-9ec6-47f7-9b1c-1d4010fc1aaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = adata_mvi.obs.copy()\n",
    "meta = meta.reset_index()\n",
    "index_col_name = meta.columns[0]\n",
    "meta = meta.rename(columns={index_col_name: 'index_name'})\n",
    "meta.to_csv(\"multivi_AD_L6b.h5ad.obs.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
