{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8568aff-eaab-4658-b3a6-3aaa27c551d4",
   "metadata": {},
   "source": [
    "## Load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0801d1-e029-4555-bf96-c250f809ecde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "import scvi\n",
    "import anndata\n",
    "import scipy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d23c69-4d87-46b8-b608-f0bca0335d99",
   "metadata": {},
   "source": [
    "## Load in anndata generated by pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede4156c-4a8b-421a-acd8-ce0ac9f09574",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad = anndata.read_h5ad(filename=\"multi_Micro-PVM.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e67b41a-d694-43cd-a1c4-7fcf102f55ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split to three datasets by modality (RNA, ATAC, Multiome), and corrupt data\n",
    "# by remove some data to create single-modality data\n",
    "ad_rna = ad[ad.obs.modality == \"rna\", ad.var.modality == \"Gene Expression\"].copy()\n",
    "ad_paired = ad[ad.obs.modality == \"paired\"].copy()\n",
    "ad_atac = ad[ad.obs.modality == \"atac\", ad.var.modality == \"Peaks\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5952eb51-a59e-40b4-aab7-e09055f36499",
   "metadata": {},
   "outputs": [],
   "source": [
    "## add filter steps here to reduce memory usage.\n",
    "# We also filter features to remove those that appear in fewer than 1% of the cells\n",
    "# This instruction removes genes or peaks that are not expressed/accessible in more than 1% of the cells.\n",
    "print(ad_atac.shape)\n",
    "min_cells = int(ad_atac.shape[0] * 0.01)\n",
    "sc.pp.filter_genes(ad_atac, min_cells=min_cells)\n",
    "print(ad_atac.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7db0a14e-410e-4170-8d65-40acc1d482f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seaad = anndata.read_h5ad(filename='/allen/programs/celltypes/workgroups/hct/SEA-AD/Integration/multivi_subclasses/singleomeRNA_MTG_Full/Micro-PVM.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42b6fe89-53f3-400d-9a03-3dfa38cea0f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_names = ad_rna.var.index.values\n",
    "print(var_names)\n",
    "## merged multi_Micro-PVM.h5ad only has part of genes.\n",
    "ad_rna_new = rna_seaad[:, var_names].copy()\n",
    "print(sum(ad_rna_new.var.index.values != ad_rna.var.index.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71c6703e-0162-4816-b78d-2a6510d21f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ad_rna_new.shape)\n",
    "min_cells = int(ad_rna_new.shape[0] * 0.01)\n",
    "sc.pp.filter_genes(ad_rna_new, min_cells=min_cells)\n",
    "print(ad_rna_new.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c31dc7-def7-452f-b7de-bd54bf7dcf0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_p = anndata.AnnData(X=ad_paired.X, obs=ad_paired.obs, var=ad_paired.var)\n",
    "ad_a = anndata.AnnData(X=ad_atac.X, obs=ad_atac.obs, var=ad_atac.var)\n",
    "ad_r = anndata.AnnData(X=ad_rna_new.X, obs=ad_rna_new.obs, var=ad_rna_new.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77b310be-a90b-444c-9811-5974a855a482",
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = ad_a.var_names\n",
    "genes = ad_r.var_names\n",
    "features = genes.union(peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e00dd29a-e439-48f3-9d53-c7047acc1aaa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(ad_p.shape)\n",
    "ad_p = ad_p[:, features].copy()\n",
    "print(ad_p.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d41e88-4822-4879-a62a-4707f37cc7a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can now use the organizing method from scvi to concatenate these anndata\n",
    "adata_mvi = scvi.data.organize_multiome_anndatas(ad_p, ad_r, ad_a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bb1fc9c-b0e4-4dd8-ba29-0db3e66eff7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi = adata_mvi[:, adata_mvi.var[\"modality\"].argsort()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b9e9f17-8f3c-43c9-abb4-dd5b46c11e77",
   "metadata": {},
   "outputs": [],
   "source": [
    "del ad, ad_rna, ad_paired, ad_atac, rna_seaad, ad_rna_new, ad_r, ad_p, ad_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3be675-e76d-4160-8e3f-1761da23eb40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a mapping of values to replace\n",
    "replace_dict = {'Female': 'F', 'Male': 'M'}\n",
    "\n",
    "# replace the values using replace method\n",
    "adata_mvi.obs['sex'] = adata_mvi.obs['sex'].replace(replace_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "718ad19a-06a1-4229-ba2a-5845a3ac0da9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata_mvi.obs.modality.unique())\n",
    "print(sum(adata_mvi.obs.donor_name.isnull()))\n",
    "print(sum(adata_mvi.obs.sex.isnull()))\n",
    "print(adata_mvi.obs.sex.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a97d7dd-b38c-4bbb-8449-288ddeeae3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi.model.MULTIVI.setup_anndata(adata_mvi, batch_key='modality', categorical_covariate_keys=[\"donor_name\", \"sex\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1211a0-ea18-48d8-a954-2dbfa6a1dbfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvi = scvi.model.MULTIVI(\n",
    "    adata_mvi,\n",
    "    n_genes=(adata_mvi.var['modality']=='Gene Expression').sum(),\n",
    "    n_regions=(adata_mvi.var['modality']=='Peaks').sum(),\n",
    "    n_layers_encoder=1, n_layers_decoder=1, n_hidden=256,\n",
    "    n_latent=8, dropout_rate=0.1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9fb277e-e24f-481c-a8d5-669c61555432",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvi.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b93ef97-2fb5-44b9-8a5b-25792e864192",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_mvi.obsm[\"X_MultiVI\"] = mvi.get_latent_representation()\n",
    "sc.pp.neighbors(adata_mvi, use_rep=\"X_MultiVI\")\n",
    "sc.tl.umap(adata_mvi, min_dist=0.2)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='modality')\n",
    "\n",
    "sc.pl.umap(adata_mvi, color='sex')\n",
    "\n",
    "sc.pl.umap(adata_mvi, color='donor_name')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70135903-4656-46e3-8815-eb78a6ae3be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_label_purity_slot(adata_mvi, cell_idx, slot=\"subclass_scANVI\"):\n",
    "    idx = np.where(adata_mvi.uns['neighbors']['connectivities'][cell_idx].todense()>0)[1]\n",
    "    df = adata_mvi.obs[[slot]].iloc[idx]\n",
    "    df = df.loc[~df[slot].isnull()]\n",
    "    u, c = np.unique(df, return_counts=True)\n",
    "    if np.size(c) == 0:\n",
    "        ratio = 0\n",
    "        label = np.nan\n",
    "    else:\n",
    "        ratio = c[np.argmax(c)] / c.sum()\n",
    "        label = u[np.argmax(c)]\n",
    "    \n",
    "    return ratio, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a22a39cb-c3a8-4822-ab9f-e6fd9bd6f6d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata_mvi.shape[0]\n",
    "purity_ratios = np.zeros(cells)\n",
    "labels = adata_mvi.obs[\"supertype_scANVI\"].to_numpy()\n",
    "for i in np.arange(cells):\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI\")\n",
    "    purity_ratios[i] = ratio\n",
    "    labels[i] = label_\n",
    "\n",
    "adata_mvi.obs[\"supertype_scANVI_purity_ratio\"] = purity_ratios\n",
    "adata_mvi.obs[\"supertype_scANVI_label_transfer\"] = labels\n",
    "\n",
    "for i in np.where(adata_mvi.obs[\"supertype_scANVI_label_transfer\"].isnull())[0]:\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_label_transfer\")\n",
    "    adata_mvi.obs[\"supertype_scANVI_purity_ratio\"].loc[i] = ratio\n",
    "    adata_mvi.obs[\"supertype_scANVI_label_transfer\"].loc[i] = label_\n",
    "    \n",
    "print(adata_mvi.obs[\"supertype_scANVI_label_transfer\"].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de9ff889-8f93-4242-b4dd-be99ca39e413",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata_mvi.shape[0]\n",
    "purity_ratios = np.zeros(cells)\n",
    "labels = adata_mvi.obs[\"supertype_scANVI_leiden\"].to_numpy()\n",
    "for i in np.arange(cells):\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_leiden\")\n",
    "    purity_ratios[i] = ratio\n",
    "    labels[i] = label_\n",
    "\n",
    "adata_mvi.obs[\"supertype_scANVI_leiden_purity_ratio\"] = purity_ratios\n",
    "adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"] = labels\n",
    "\n",
    "for i in np.where(adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].isnull())[0]:\n",
    "    ratio, label_ = compute_label_purity_slot(adata_mvi, i, slot=\"supertype_scANVI_leiden_label_transfer\")\n",
    "    adata_mvi.obs[\"supertype_scANVI_leiden_purity_ratio\"].loc[i] = ratio\n",
    "    adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].loc[i] = label_\n",
    "    \n",
    "print(adata_mvi.obs[\"supertype_scANVI_leiden_label_transfer\"].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25ed1f39-de88-4b0e-96ad-345c96f4f66c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden_label_transfer')\n",
    "sc.pl.umap(adata_mvi, color=\"supertype_scANVI_leiden\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679a60f5-bc5a-43dc-877a-3aad43792803",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.obs['supertype_scANVI_leiden_label_transfer'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22776818-0998-4541-92c1-b99d9073e70b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from joblib import parallel_backend\n",
    "sc.settings.n_jobs=32\n",
    "with parallel_backend('threading', n_jobs=32):\n",
    "    sc.tl.leiden(adata_mvi, key_added = \"leiden_2.0\", resolution = 2.0) # default resolution in 1.0\n",
    "## Scanpy Clustering Leiden (https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/scanpy/scanpy_04_clustering.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d52e8a-b9ea-4d1a-8ccc-d3e63f2ce674",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color=['leiden_2.0'], legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "052edef5-b99e-4126-8a0f-014fb882cf6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_peaks = np.asarray(np.sum(adata_mvi[:, adata_mvi.var.modality == 'Peaks'].X, axis=1)).reshape(-1)\n",
    "adata_mvi.obs['total_peak_count'] = total_peaks\n",
    "meta = adata_mvi.obs.copy()\n",
    "df2 = meta.groupby(['leiden_1.0'])['total_peak_count'].mean()\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "926e9280-0aa0-4fcb-83d1-f35bebdaafa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "with parallel_backend('threading', n_jobs=32):\n",
    "    sc.tl.leiden(adata_mvi, key_added = \"leiden_1.0\", resolution = 1.0) # default resolution in 1.0\n",
    "## Scanpy Clustering Leiden (https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/scanpy/scanpy_04_clustering.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df24f3e7-1124-49e8-a27e-f130b51e14b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color=['leiden_1.0'], legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6c4821e-9083-458c-9c49-c286b99a3d0a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color=['leiden_1.0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf563050-8349-44d2-aedb-ba4d75984409",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data types of all columns.\n",
    "dtypes = adata_mvi.obs.dtypes\n",
    "# Find all non-string columns\n",
    "non_string_cols = dtypes[dtypes == 'object'].index.tolist()\n",
    "adata_mvi.obs[non_string_cols] = adata_mvi.obs[non_string_cols].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aa43940-07bd-453c-a36d-8778cc5f554f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.write(\"multivi_AD_MicroPVM.h5ad\")\n",
    "mvi.save(\"trained_multivi_AD_MicroPVM\", save_anndata=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7ab0bf7-fd67-4866-a3f3-01a2bbddb5ca",
   "metadata": {},
   "source": [
    "## Replace old names to new (multiome used old names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1b892eb-733f-46e9-94a6-ac7ae735cfcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi = anndata.read_h5ad(filename=\"multivi_AD_MicroPVM.h5ad\")\n",
    "new_names = pd.read_csv('/allen/programs/celltypes/workgroups/hct/SEA-AD/RNAseq/diff_prop/input/MTG/old_names_to_new_names.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb04409-b184-4edd-8642-51c3819dba99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary\n",
    "name_dict = dict(zip(new_names['old_names'], new_names['new_names']))\n",
    "\n",
    "# replace specific values in a DataFrame column with other values, use .replace()\n",
    "adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'] = adata_mvi.obs['supertype_scANVI_leiden_label_transfer'].replace(name_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f92947bd-03e1-4ab4-8f76-a23029b5e54c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden_label_transfer_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29ec954f-01ba-496c-b739-af68ecbc3bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_mvi, color='supertype_scANVI_leiden')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0006d58a-6b05-4b51-b0ed-590feb2cd414",
   "metadata": {},
   "outputs": [],
   "source": [
    "## output supertype annotation in csv\n",
    "meta = adata_mvi.obs.copy()\n",
    "meta = meta.reset_index()\n",
    "index_col_name = meta.columns[0]\n",
    "meta = meta.rename(columns={index_col_name: 'index_name'})\n",
    "meta.to_csv(\"multivi_AD_MicroPVM.h5ad.obs.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7348831f-c127-4d02-a3cd-9f37486973a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi.write(\"multivi_AD_MicroPVM_NewNames.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fc3e05a-2604-4391-81c7-8e25cc40a9dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_labels = ['cluster_doublet_score_flag', 'cluster_fraction_ribo_flag']\n",
    "cut_off = (~adata_mvi.obs['supertype_scANVI_leiden_label_transfer_new'].isin(bad_labels))\n",
    "ad = adata_mvi[cut_off].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f03e37-b6e3-4054-9f4c-118a7354b7cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(ad, color='supertype_scANVI_leiden_label_transfer_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5433ce43-f55a-4ef0-a527-94beb2da753d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ad.obs.groupby([\"leiden_1.0\", \"supertype_scANVI_leiden_label_transfer_new\"]).size().reset_index(name=\"Counts\")\n",
    "df_wide = df.pivot_table(index=\"leiden_1.0\", columns=\"supertype_scANVI_leiden_label_transfer_new\", values=\"Counts\")\n",
    "import seaborn as sns\n",
    "sns.heatmap(df_wide.T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9998bc-7889-494e-a586-9a7963fa7e1f",
   "metadata": {},
   "source": [
    "## reload anndata and plot QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b11f3247-7529-43a8-8824-23a4910819dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_mvi = anndata.read_h5ad(filename=\"multivi_AD_MicroPVM_NewNames.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2fb1619-1971-401d-a087-29dc1de4cabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_folder = os.path.abspath(os.path.join(os.getcwd(), os.pardir))\n",
    "csv_file_path = os.path.join(parent_folder, 'donor_name_ch_cognitivestatus_binary.csv')\n",
    "selected_data = pd.read_csv(csv_file_path)\n",
    "\n",
    "selected_dict = dict(zip(selected_data['donor_name'], selected_data['ch_cognitivestatus_binary']))\n",
    "adata_mvi.obs['ch_cognitivestatus_binary'] = adata_mvi.obs['donor_name'].map(selected_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f5b6f1-c9ca-4be2-a099-357e3bfb345c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (5, 5)\n",
    "sc.pl.umap(adata_mvi, color='nCount_RNA')\n",
    "sc.pl.umap(adata_mvi, color='nFeature_RNA')\n",
    "sc.pl.umap(adata_mvi, color='ch_cognitivestatus_binary')\n",
    "sc.pl.umap(adata_mvi, color='total_peak_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a63a0374-a478-4ec1-b24a-f8887df56655",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
