{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "09f33caa-3ddc-4c8b-ad09-fb0e3222edea",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37dce2f1-2427-4808-938f-e51cd8cc5d4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import colors\n",
    "%matplotlib inline\n",
    "import scanpy as sc\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "plt.ioff()\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea87b7a5-b8d2-4ddd-a00f-c4b96e418e75",
   "metadata": {},
   "source": [
    "### Define AnnData preparation function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f191cc77-08bc-472f-85e5-ddbdd62cfc03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_anndata(adata, region, modality, tag, metadata_to_color, annotation_color=True):\n",
    "    print(str(datetime.datetime.now()) + \" -- Starting export\")\n",
    "\n",
    "    # Add uns elements required by the schema\n",
    "    adata.uns[\"title\"] = region + \": Seattle Alzheimer's Disease Atlas (SEA-AD)\"\n",
    "    adata.uns[\"X_normalization\"] = \"ln(UP10K+1)\"\n",
    "    adata.uns[\"batch_condition\"] = [\"Specimen ID\"]\n",
    "    adata.uns[\"default_embedding\"] = \"X_umap\"\n",
    "    try:\n",
    "        del adata.uns[\"_scvi\"]\n",
    "        del adata.uns[\"subclass_scANVI_colors\"]\n",
    "        del adata.uns[\"supertype_scANVI_leiden_colors\"]\n",
    "\n",
    "        # Remove unwanted obsm values\n",
    "        del adata.obsm[\"_scvi_extra_categoricals\"]\n",
    "        del adata.obsm[\"_scvi_extra_continuous\"]\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    # add obs elements required by the schema\n",
    "    adata.obs_names = adata.obs[\"exp_component_name\"].copy()\n",
    "    \n",
    "    # Metadata tweaks for public release\n",
    "    adata.obs[\"ch_hispanicorlatino\"] = adata.obs[\"ch_hispanicorlatino\"].cat.rename_categories(\n",
    "        {\"\": \"Unknown\"},\n",
    "    )\n",
    "    adata.obs[\"late_stage\"] = adata.obs[\"late_stage\"].cat.rename_categories(\n",
    "        {\"Staging Precluded by FTLD with TDP43 or ALS/MND or TDP-43 pathology is unclassifiable\": \"Unclassifiable\"},\n",
    "    )\n",
    "    \n",
    "    adata.obs[\"alignment\"] = adata.obs[\"alignment\"].astype(\"object\")\n",
    "    adata.obs.loc[adata.obs[\"alignment\"] == \"Cell Ranger 3.0\", \"alignment\"] = \"Cell Ranger 6.0\"\n",
    "    adata.obs[\"alignment\"] = adata.obs[\"alignment\"].astype(\"category\")\n",
    "\n",
    "    \n",
    "    adata.obs.loc[adata.obs[\"age_at_death\"] > 89, \"age_at_death\"] = \"90+\"\n",
    "    adata.obs[\"age_at_death\"] = [str(i).replace(\".0\", \"\") for i in adata.obs[\"age_at_death\"]]\n",
    "    \n",
    "    adata.obs[\"ch_lastmocascore_2\"] = adata.obs[\"ch_lastmocascore_2\"].str.replace(\"/30\", \"\")\n",
    "\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoestatuses\"].str.contains(\"4\")\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoe_four\"].fillna(\"Reference\")\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoe_four\"].astype(\"category\")\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoe_four\"].cat.rename_categories({True: \"Y\", False: \"N\"})\n",
    "    \n",
    "    categories_to_fix = [\n",
    "        \"ch_race___1\",\n",
    "        \"ch_race___2\",\n",
    "        \"ch_race___3\",\n",
    "        \"ch_race___4\",\n",
    "        \"ch_race___5\",\n",
    "        \"ch_race___6\",\n",
    "        \"ch_race___97\",\n",
    "        \"ch_raceother\",\n",
    "        \"ch_hispanicorlatino\",\n",
    "        \"ch_education\",\n",
    "        \"adneurochange\",\n",
    "        \"thal\",\n",
    "        \"braak\",\n",
    "        \"cscore\",\n",
    "        \"caascore\",\n",
    "        \"lewybodydisease\",\n",
    "        \"ge_atherosclerosis_id\",\n",
    "        \"micro_arteriolosclerosis_id\",\n",
    "        \"late_stage\",\n",
    "        \"ch_cognitivestatus_binary\",\n",
    "        \"primary_studyname\",\n",
    "        \"secondary_studyname\"\n",
    "    ]\n",
    "    \n",
    "    for i in categories_to_fix:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"category\")\n",
    "        adata.obs[i] = adata.obs[i].cat.add_categories(\"Reference\")\n",
    "        adata.obs[i] = adata.obs[i].fillna(\"Reference\")\n",
    "\n",
    "\n",
    "    numerics_to_fix = [\n",
    "        \"ch_education_years\",\n",
    "        \"pmi_date\",\n",
    "        \"ap_freshbrainweight\",\n",
    "        \"ap_brainph\",\n",
    "        \"micro_totalmicroinfarcts\",\n",
    "        \"micro_microvascularbrain\",\n",
    "        \"ch_lastcasiscore\",\n",
    "        \"ch_casi_interval\",\n",
    "        \"ch_lastmmsescore\",\n",
    "        \"ch_mmse_interval\",\n",
    "        \"ch_lastmocascore_2\",\n",
    "        \"ch_moca_interval\",\n",
    "    ]\n",
    "\n",
    "    for i in numerics_to_fix:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"object\")\n",
    "        adata.obs.loc[adata.obs[\"reference_cell\"] == 1, i] = \"Reference\"\n",
    "        adata.obs[i] = [str(i) for i in adata.obs[i]]\n",
    "\n",
    "    \n",
    "    adata.obs[\"reference_cell\"] = adata.obs[\"reference_cell\"].astype(\"category\")\n",
    "    adata.obs[\"reference_cell\"] = adata.obs[\"reference_cell\"].cat.rename_categories(\n",
    "        {\n",
    "            0: \"False\",\n",
    "            1: \"True\",\n",
    "        },\n",
    "    )\n",
    "\n",
    "    dend_order = pd.read_csv(os.path.join(pwd, \"input\", region + \"_\" + modality, \"cluster_order_and_colors.csv\"))\n",
    "    \n",
    "    if tag == \"final\":\n",
    "        class_order = [j for j in dend_order[\"class_label\"].drop_duplicates() if j in adata.obs[\"class_scANVI\"].cat.categories]\n",
    "        adata.obs[\"class_scANVI\"] = adata.obs[\"class_scANVI\"].cat.reorder_categories(class_order)\n",
    "        \n",
    "        class_order = [j for j in dend_order[\"subclass_label\"].drop_duplicates() if j in adata.obs[\"subclass_scANVI\"].cat.categories]\n",
    "        adata.obs[\"subclass_scANVI\"] = adata.obs[\"subclass_scANVI\"].cat.reorder_categories(class_order)\n",
    "\n",
    "        class_order = [j for j in dend_order[\"cluster_label\"].drop_duplicates() if j in adata.obs[\"supertype_scANVI_leiden\"].cat.categories]\n",
    "        adata.obs[\"supertype_scANVI_leiden\"] = adata.obs[\"supertype_scANVI_leiden\"].cat.reorder_categories(class_order)\n",
    "        \n",
    "\n",
    "    sc.pp.subsample(adata, fraction=1)\n",
    "    \n",
    "    adata.obs[\"ch_cognitivestatus_binary\"] = adata.obs[\"ch_cognitivestatus_binary\"].cat.reorder_categories([\"Reference\", \"No dementia\", \"Dementia\"])\n",
    "    adata.obs[\"adneurochange\"] = adata.obs[\"adneurochange\"].cat.reorder_categories([\"Reference\", \"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "    adata.obs[\"braak\"] = adata.obs[\"braak\"].cat.reorder_categories([\"Reference\", \"Braak 0\", \"Braak II\", \"Braak III\", \"Braak IV\", \"Braak V\", \"Braak VI\"])\n",
    "    adata.obs[\"thal\"] = adata.obs[\"thal\"].cat.reorder_categories([\"Reference\", \"Thal 0\", \"Thal 1\", \"Thal 2\", \"Thal 3\", \"Thal 4\", \"Thal 5\"])\n",
    "    adata.obs[\"cscore\"] = adata.obs[\"cscore\"].cat.reorder_categories([\"Reference\", 'Absent', 'Sparse', 'Moderate', 'Frequent'])\n",
    "    adata.obs[\"lewybodydisease\"] = adata.obs[\"lewybodydisease\"].cat.reorder_categories([\"Reference\", 'Not Identified (olfactory bulb not assessed)', 'Not Identified (olfactory bulb assessed)', 'Olfactory bulb only', 'Amygdala-predominant', 'Brainstem-predominant', 'Limbic (Transitional)', 'Neocortical (Diffuse)'])\n",
    "    adata.obs[\"late_stage\"] = adata.obs[\"late_stage\"].cat.reorder_categories([\"Reference\", 'Unclassifiable', 'Not Identified', 'LATE Stage 1', 'LATE Stage 2', 'LATE Stage 3'])\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoe_four\"].cat.reorder_categories([\"Reference\", \"N\", \"Y\"])\n",
    "\n",
    "    # Keep only display metadata\n",
    "    adata.obs = adata.obs.loc[\n",
    "        :,\n",
    "        [\n",
    "            \"sample_id\",\n",
    "            \"reference_cell\",\n",
    "            \"donor_name\",\n",
    "            \"organism\",\n",
    "            \"roi\",\n",
    "            \"sex\",\n",
    "            \"gender\",\n",
    "            \"age_at_death\",\n",
    "            \"ch_race___1\",\n",
    "            \"ch_race___2\",\n",
    "            \"ch_race___3\",\n",
    "            \"ch_race___4\",\n",
    "            \"ch_race___5\",\n",
    "            \"ch_race___6\",\n",
    "            \"ch_race___97\",\n",
    "            \"ch_raceother\",\n",
    "            \"ch_hispanicorlatino\",\n",
    "            \"ch_education\",\n",
    "            \"ch_education_years\",\n",
    "            \"pmi_date\",\n",
    "            \"ap_freshbrainweight\",\n",
    "            \"ap_brainph\",\n",
    "            \"adneurochange\",\n",
    "            \"thal\",\n",
    "            \"braak\",\n",
    "            \"cscore\",\n",
    "            \"caascore\",\n",
    "            \"lewybodydisease\",\n",
    "            \"micro_totalmicroinfarcts\",\n",
    "            \"micro_microvascularbrain\",\n",
    "            \"ge_atherosclerosis_id\",\n",
    "            \"micro_arteriolosclerosis_id\",\n",
    "            \"late_stage\",\n",
    "            \"ch_cognitivestatus_binary\",\n",
    "            \"ch_lastcasiscore\",\n",
    "            \"ch_casi_interval\",\n",
    "            \"ch_lastmmsescore\",\n",
    "            \"ch_mmse_interval\",\n",
    "            \"ch_lastmocascore_2\",\n",
    "            \"ch_moca_interval\",\n",
    "            \"ch_apoe_four\", \n",
    "            \"primary_studyname\",\n",
    "            \"secondary_studyname\",\n",
    "            \"cell_prep_type\",\n",
    "            \"facs_population_plan\",\n",
    "            \"rna_amplification\",\n",
    "            \"sample_name\",\n",
    "            \"sample_quantity_count\",\n",
    "            \"expc_cell_capture\",\n",
    "            \"method\",\n",
    "            \"pcr_cycles\",\n",
    "            \"percent_cdna_longer_than_400bp\",\n",
    "            \"rna_amplification_pass_fail\",\n",
    "            \"amplified_quantity_ng\",           \n",
    "            \"load_name\",\n",
    "            \"library_prep\",\n",
    "            \"library_input_ng\",\n",
    "            \"r1_index\",\n",
    "            \"avg_size_bp\",\n",
    "            \"quantification_fmol\",\n",
    "            \"library_prep_pass_fail\",\n",
    "            \"exp_component_vendor_name\",\n",
    "            \"batch_vendor_name\",\n",
    "            \"experiment_component_failed\",\n",
    "            \"alignment\",\n",
    "            \"Genome\",\n",
    "            \"ar_id\",\n",
    "            \"bc\",\n",
    "            \"Estimated_number_of_cells\",\n",
    "            \"number_of_reads\",\n",
    "            \"sequencing_saturation\",\n",
    "            \"GEX_Mean_raw_reads_per_cell\",\n",
    "            \"GEX_Q30_bases_in_barcode\",\n",
    "            \"GEX_Q30_bases_in_read_2\",\n",
    "            \"GEX_Q30_bases_in_UMI\",\n",
    "            \"GEX_Percent_duplicates\",\n",
    "            \"GEX_Q30_bases_in_sample_index_i1\",\n",
    "            \"GEX_Q30_bases_in_sample_index_i2\",\n",
    "            \"GEX_Reads_with_TSO\",\n",
    "            \"GEX_Sequenced_read_pairs\",\n",
    "            \"GEX_Valid_UMIs\",\n",
    "            \"GEX_Valid_barcodes\",\n",
    "            \"GEX_Reads_mapped_to_genome\",\n",
    "            \"GEX_Reads_mapped_confidently_to_genome\",\n",
    "            \"GEX_Reads_mapped_confidently_to_intergenic_regions\",\n",
    "            \"GEX_Reads_mapped_confidently_to_intronic_regions\",\n",
    "            \"GEX_Reads_mapped_confidently_to_exonic_regions\",\n",
    "            \"GEX_Reads_mapped_confidently_to_transcriptome\",\n",
    "            \"GEX_Reads_mapped_antisense_to_gene\",\n",
    "            \"GEX_Fraction_of_transcriptomic_reads_in_cells\",\n",
    "            \"GEX_Total_genes_detected\",\n",
    "            \"GEX_Median_UMI_counts_per_cell\",\n",
    "            \"GEX_Median_genes_per_cell\",\n",
    "            \"Feature_linkages_detected\",\n",
    "            \"Linked_genes\",\n",
    "            \"Linked_peaks\",\n",
    "            \"ATAC_Confidently_mapped_read_pairs\",\n",
    "            \"ATAC_Fraction_of_genome_in_peaks\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_in_cells\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_TSS\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_peaks\",\n",
    "            \"ATAC_Fraction_of_transposition_events_in_peaks_in_cells\",\n",
    "            \"ATAC_Mean_raw_read_pairs_per_cell\",\n",
    "            \"ATAC_Median_high.quality_fragments_per_cell\",\n",
    "            \"ATAC_Non.nuclear_read_pairs\",\n",
    "            \"ATAC_Number_of_peaks\",\n",
    "            \"ATAC_Percent_duplicates\",\n",
    "            \"ATAC_Q30_bases_in_barcode\",\n",
    "            \"ATAC_Q30_bases_in_read_1\",\n",
    "            \"ATAC_Q30_bases_in_read_2\",\n",
    "            \"ATAC_Q30_bases_in_sample_index_i1\",\n",
    "            \"ATAC_Sequenced_read_pairs\",\n",
    "            \"ATAC_TSS_enrichment_score\",\n",
    "            \"ATAC_Unmapped_read_pairs\",\n",
    "            \"ATAC_Valid_barcodes\",\n",
    "            \"mapped_reads\",\n",
    "            \"unmapped_reads\",\n",
    "            \"nonconf_mapped_reads\",\n",
    "            \"total.reads\",\n",
    "            \"nCount_RNA\",\n",
    "            \"nFeature_RNA\",\n",
    "            \"doublet_score\",\n",
    "            \"fraction_mito\",\n",
    "            \"for_analysis\",\n",
    "            \"class_conf_scANVI\",\n",
    "            \"class_scANVI\",\n",
    "            \"subclass_conf_scANVI\",\n",
    "            \"subclass_scANVI\",\n",
    "            \"supertype_conf_scANVI\",\n",
    "            \"supertype_scANVI\",\n",
    "            \"supertype_scANVI_leiden\"\n",
    "        ]\n",
    "    ]\n",
    "    adata.obs = adata.obs.rename(\n",
    "        {\n",
    "            \"reference_cell\": \"Neurotypical reference\",\n",
    "            \"donor_name\": \"Donor ID\",\n",
    "            \"organism\": \"Organism\",\n",
    "            \"roi\": \"Brain Region\",\n",
    "            \"sex\": \"Sex\",\n",
    "            \"gender\": \"Gender\",\n",
    "            \"age_at_death\": \"Age at Death\",\n",
    "            \"ch_race___1\": \"Race (choice=White)\",\n",
    "            \"ch_race___2\": \"Race (choice=Black/ African American)\",\n",
    "            \"ch_race___3\": \"Race (choice=Asian)\",\n",
    "            \"ch_race___4\": \"Race (choice=American Indian/ Alaska Native)\",\n",
    "            \"ch_race___5\": \"Race (choice=Native Hawaiian or Pacific Islander)\",\n",
    "            \"ch_race___6\": \"Race (choice=Unknown or unreported)\",\n",
    "            \"ch_race___97\": \"Race (choice=Other)\",\n",
    "            \"ch_raceother\": \"specify other race\",\n",
    "            \"ch_hispanicorlatino\": \"Hispanic/Latino\",\n",
    "            \"ch_education\": \"Highest level of education\",\n",
    "            \"ch_education_years\": \"Years of education\",\n",
    "            \"pmi_date\": \"PMI\",\n",
    "            \"ap_freshbrainweight\": \"Fresh Brain Weight\",\n",
    "            \"ap_brainph\": \"Brain pH\",\n",
    "            \"adneurochange\": \"Overall AD neuropathological Change\",\n",
    "            \"thal\": \"Thal\",\n",
    "            \"braak\": \"Braak\",\n",
    "            \"cscore\": \"CERAD score\",\n",
    "            \"caascore\": \"Overall CAA Score\",\n",
    "            \"lewybodydisease\": \"Highest Lewy Body Disease\",\n",
    "            \"micro_totalmicroinfarcts\": \"Total Microinfarcts (not observed grossly)\",\n",
    "            \"micro_microvascularbrain\": \"Total microinfarcts in screening sections\",\n",
    "            \"ge_atherosclerosis_id\": \"Atherosclerosis\",\n",
    "            \"micro_arteriolosclerosis_id\": \"Arteriolosclerosis\",\n",
    "            \"late_stage\": \"LATE\",\n",
    "            \"ch_cognitivestatus_binary\": \"Cognitive Status\",\n",
    "            \"ch_lastcasiscore\": \"Last CASI Score\",\n",
    "            \"ch_casi_interval\": \"Interval from last CASI in months\",\n",
    "            \"ch_lastmmsescore\": \"Last MMSE Score\",\n",
    "            \"ch_mmse_interval\": \"Interval from last MMSE in months\",\n",
    "            \"ch_lastmocascore_2\": \"Last MOCA Score\",\n",
    "            \"ch_moca_interval\": \"Interval from last MOCA in months\",\n",
    "            \"ch_apoe_four\": \"APOE4 Status\",\n",
    "            \"primary_studyname\": \"Primary Study Name\",\n",
    "            \"secondary_studyname\": \"Secondary Study Name\",\n",
    "            \"Estimated_number_of_cells\": \"GEX_Estimated_number_of_cells\",\n",
    "            \"number_of_reads\": \"GEX_number_of_reads\",\n",
    "            \"sequencing_saturation\": \"GEX_sequencing_saturation\",\n",
    "            \"Feature_linkages_detected\": \"Multiome_Feature_linkages_detected\",\n",
    "            \"Linked_genes\": \"Multiome_Linked_genes\",\n",
    "            \"Linked_peaks\": \"Multiome_Linked_peaks\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_in_cells\": \"ATAC_Fraction_of_high_quality_fragments_in_cells\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_TSS\": \"ATAC_Fraction_of_high_quality_fragments_overlapping_TSS\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_peaks\": \"ATAC_Fraction_of_high_quality_fragments_overlapping_peaks\",\n",
    "            \"ATAC_Median_high.quality_fragments_per_cell\": \"ATAC_Median_high_quality_fragments_per_cell\",\n",
    "            \"ATAC_Non.nuclear_read_pairs\": \"ATAC_Non-nuclear_read_pairs\",\n",
    "            \"mapped_reads\": \"Number of mapped reads\",\n",
    "            \"unmapped_reads\": \"Number of unmapped reads\",\n",
    "            \"nonconf_mapped_reads\": \"Number of multimapped reads\",\n",
    "            \"total.reads\": \"Number of reads\",\n",
    "            \"nCount_RNA\": \"Number of UMIs\",\n",
    "            \"nFeature_RNA\": \"Genes detected\",\n",
    "            \"doublet_score\": \"Doublet score\",\n",
    "            \"fraction_mito\": \"Fraction mitochondrial UMIs\",\n",
    "            \"for_analysis\": \"Used in analysis\",\n",
    "            \"class_conf_scANVI\": \"Class confidence\",\n",
    "            \"class_scANVI\": \"Class\",\n",
    "            \"subclass_conf_scANVI\": \"Subclass confidence\",\n",
    "            \"subclass_scANVI\": \"Subclass\",\n",
    "            \"supertype_conf_scANVI\": \"Supertype confidence\",\n",
    "            \"supertype_scANVI\": \"Supertype (non-expanded)\",\n",
    "            \"supertype_scANVI_leiden\": \"Supertype\",\n",
    "        },\n",
    "        axis=1,\n",
    "    )\n",
    "    metadata_colors = {}\n",
    "    for i in metadata_to_color:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"category\")\n",
    "        a = adata.obs[i].cat.categories.to_list()\n",
    "        cmap = plt.get_cmap('viridis', len(a)+1)\n",
    "        b = [colors.rgb2hex(cmap(j)) for j in range(cmap.N)]\n",
    "        b.pop(0)\n",
    "        if i == \"Sex\":\n",
    "            b = [\"pink\", \"dodgerblue\"]\n",
    "        for A, B in zip(a, b):\n",
    "            try:\n",
    "                metadata_colors[i][A] = B\n",
    "            except:\n",
    "                metadata_colors[i] = {}\n",
    "                metadata_colors[i][A] = B\n",
    "        try:\n",
    "            del adata.uns[i + \"_colors\"]\n",
    "        except:\n",
    "            pass\n",
    "        sc.pl.umap(adata,\n",
    "            color=i,\n",
    "            title=\"\",\n",
    "            palette=metadata_colors[i],\n",
    "            size=10,\n",
    "            legend_fontsize=12\n",
    "        )\n",
    "    \n",
    "    if annotation_color == True:\n",
    "        tmp = dend_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "        tmp.index = tmp[\"subclass_label\"].copy()\n",
    "        tmp = tmp.drop(\"subclass_label\", axis=1)\n",
    "        tmp = tmp.to_dict()[\"subclass_color\"]\n",
    "        sc.pl.umap(adata,\n",
    "            color=\"Subclass\",\n",
    "            title=\"\",\n",
    "            palette=tmp,\n",
    "            size=10,\n",
    "            legend_fontsize=12\n",
    "        )\n",
    "\n",
    "        tmp = dend_order.loc[:, [\"cluster_label\", \"cluster_color\"]].drop_duplicates()\n",
    "        tmp.index = tmp[\"cluster_label\"].copy()\n",
    "        tmp = tmp.drop(\"cluster_label\", axis=1)\n",
    "        tmp = tmp.to_dict()[\"cluster_color\"]\n",
    "        sc.pl.umap(adata,\n",
    "            color=\"Supertype\",\n",
    "            title=\"\",\n",
    "            palette=tmp,\n",
    "            size=10,\n",
    "            legend_fontsize=12\n",
    "        )\n",
    "    \n",
    "    for i in adata.uns.keys():\n",
    "        if i.endswith(\"_colors\"):\n",
    "            adata.uns[i] = [re.sub(\"ff$\", \"\", j) for j in adata.uns[i]]\n",
    "            \n",
    "    adata.obs[\"Supertype\"] = adata.obs[\"Supertype\"].cat.rename_categories(\n",
    "        {\n",
    "            \"VLMC_2\": \"Pericyte_1\",\n",
    "            \"VLMC_2_1-SEAAD\": \"SMC-SEAAD\",\n",
    "            \"VLMC_2_2-SEAAD\": \"Pericyte_2-SEAAD\",\n",
    "            \"Micro-PVM_2_2-SEAAD\": \"Lymphocyte\",\n",
    "            \"Micro-PVM_1_1-SEAAD\": \"Monocyte\"\n",
    "        },\n",
    "    )\n",
    "    \n",
    "    adata.write(os.path.join(pwd, \"output\", region, \"SEAAD_\" + region + \"_\" + modality + \"_\" + tag + \"-nuclei.\" + str(datetime.datetime.date(datetime.datetime.now())) + \".h5ad\"), compression=\"gzip\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570e72c5-6bc9-420d-bd71-0037ee48f013",
   "metadata": {},
   "source": [
    "### Prepare and export objects for AWS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa77b5c-e72a-4e82-9090-25b576506d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = \"DLPFC\"\n",
    "modality = \"RNAseq\"\n",
    "date = \"2023-07-12\"\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", region, \"final.\" + date + \".h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54295d2f-70d1-4f26-a0b8-7e7463d321b4",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "metadata_to_color = [\n",
    "    \"Sex\",\n",
    "    \"Cognitive Status\",\n",
    "    \"Overall AD neuropathological Change\",\n",
    "    \"Braak\",\n",
    "    \"Thal\",\n",
    "    \"CERAD score\",\n",
    "    \"Highest Lewy Body Disease\",\n",
    "    \"LATE\",\n",
    "    \"APOE4 Status\",\n",
    "]\n",
    "\n",
    "prepare_anndata(\n",
    "    adata=adata,\n",
    "    region=region,\n",
    "    modality=modality,\n",
    "    tag=\"final\",\n",
    "    metadata_to_color=metadata_to_color\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a089a02c-22db-49c8-8ebb-77ac46204fad",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = \"DLPFC\"\n",
    "modality = \"RNAseq\"\n",
    "date = \"2023-07-12\"\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", region, \"raw.\" + date + \".h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71cc595b-c953-4e57-aa47-4cff72ea84b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_anndata(\n",
    "    adata=adata,\n",
    "    region=region,\n",
    "    modality=modality,\n",
    "    tag=\"all\",\n",
    "    metadata_to_color=[],\n",
    "    annotation_color=False\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
