{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed1422ad-877d-4a83-8528-21616bf43ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import re\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "%matplotlib inline\n",
    "import scanpy as sc\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "plt.ioff()\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "609269d1-552f-43b2-8252-60718ca82c4d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Define AnnData preparation function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeb9b251-7bcb-4f64-8228-b4b2812d83ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_anndata(adata, region, modality, tag, metadata_to_color, annotation_color=True):\n",
    "    # Add uns elements required by the schema\n",
    "    adata.uns[\"title\"] = region + \": Seattle Alzheimer's Disease Atlas (SEA-AD)\"\n",
    "    adata.uns[\"X_normalization\"] = \"None\"\n",
    "    adata.uns[\"batch_condition\"] = [\"Specimen ID\"]\n",
    "    adata.uns[\"default_embedding\"] = \"X_umap\"\n",
    "    \n",
    "    del adata.uns[\"_scvi\"]\n",
    "    del adata.uns[\"Doublet_or_LowQuality_colors\"]\n",
    "    del adata.uns[\"donor_name_colors\"]\n",
    "    del adata.uns[\"leiden_1.0_colors\"]\n",
    "    del adata.uns[\"modality_colors\"]\n",
    "    del adata.uns[\"sex_colors\"]\n",
    "    del adata.uns[\"subclass_scANVI_colors\"]\n",
    "    del adata.obsm[\"_scvi_extra_categoricals\"]\n",
    "    \n",
    "    # Metadata tweaks for public release\n",
    "    adata.obs[\"sex\"] = adata.obs[\"gender\"].copy()\n",
    "    adata.obs.loc[adata.obs[\"uwa\"] == \"UWA 7186\", \"sex\"] = \"Male\"\n",
    "    adata.obs.loc[adata.obs[\"donor_name\"] == \"H20.33.038\", \"ap_freshbrainweight\"] = \"Unavailable\"\n",
    "    adata.obs[\"ch_hispanicorlatino\"].cat.rename_categories(\n",
    "        {\"\": \"Unknown\"},\n",
    "        inplace=True\n",
    "    )\n",
    "    adata.obs[\"late_stage\"].cat.rename_categories(\n",
    "        {\"Staging Precluded by FTLD with TDP43 or ALS/MND or TDP-43 pathology is unclassifiable\": \"Unclassifiable\"},\n",
    "        inplace=True\n",
    "    )\n",
    "    adata.obs.loc[adata.obs[\"donor_name\"] == \"H20.33.026\", \"primary_studyname\"] = \"ADRC Clinical Core\"\n",
    "    \n",
    "    adata.obs.loc[adata.obs[\"age_at_death\"] > 89, \"age_at_death\"] = \"90+\"\n",
    "    adata.obs[\"age_at_death\"] = [str(i).replace(\".0\", \"\") for i in adata.obs[\"age_at_death\"]]\n",
    "    \n",
    "    adata.obs[\"ch_lastmocascore_2\"] = adata.obs[\"ch_lastmocascore_2\"].str.replace(\"/30\", \"\")\n",
    "    \n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoestatuses\"].str.contains(\"4\")\n",
    "    adata.obs[\"ch_apoe_four\"].fillna(\"Reference\", inplace=True)\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoe_four\"].astype(\"category\")\n",
    "    adata.obs[\"ch_apoe_four\"].cat.rename_categories({True: \"Y\", False: \"N\"}, inplace=True)\n",
    "    \n",
    "    categories_to_fix = [\n",
    "        \"ch_race___1\",\n",
    "        \"ch_race___2\",\n",
    "        \"ch_race___3\",\n",
    "        \"ch_race___4\",\n",
    "        \"ch_race___5\",\n",
    "        \"ch_race___6\",\n",
    "        \"ch_race___97\",\n",
    "        \"ch_raceother\",\n",
    "        \"ch_hispanicorlatino\",\n",
    "        \"ch_education\",\n",
    "        \"adneurochange\",\n",
    "        \"thal\",\n",
    "        \"braak\",\n",
    "        \"cscore\",\n",
    "        \"caascore\",\n",
    "        \"lewybodydisease\",\n",
    "        \"ge_atherosclerosis_id\",\n",
    "        \"micro_arteriolosclerosis_id\",\n",
    "        \"late_stage\",\n",
    "        \"ch_cognitivestatus_binary\",\n",
    "        \"primary_studyname\",\n",
    "        \"secondary_studyname\"\n",
    "    ]\n",
    "    \n",
    "    for i in categories_to_fix:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"category\")\n",
    "        adata.obs[i].cat.add_categories(\"Reference\", inplace=True)\n",
    "        adata.obs[i].fillna(\"Reference\", inplace=True)\n",
    "    \n",
    "    \n",
    "    numerics_to_fix = [\n",
    "        \"ch_education_years\",\n",
    "        \"pmi_date\",\n",
    "        \"ap_freshbrainweight\",\n",
    "        \"ap_brainph\",\n",
    "        \"micro_totalmicroinfarcts\",\n",
    "        \"micro_microvascularbrain\",\n",
    "        \"ch_lastcasiscore\",\n",
    "        \"ch_casi_interval\",\n",
    "        \"ch_lastmmsescore\",\n",
    "        \"ch_mmse_interval\",\n",
    "        \"ch_lastmocascore_2\",\n",
    "        \"ch_moca_interval\",\n",
    "    ]\n",
    "    \n",
    "    for i in numerics_to_fix:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"object\")\n",
    "        adata.obs.loc[adata.obs[\"reference_cell\"] == 1, i] = \"Reference\"\n",
    "        adata.obs[i] = [str(i) for i in adata.obs[i]]\n",
    "    \n",
    "    \n",
    "    adata.obs[\"reference_cell\"] = adata.obs[\"reference_cell\"].astype(\"category\")\n",
    "    adata.obs[\"reference_cell\"].cat.rename_categories(\n",
    "        {\n",
    "            0: \"False\",\n",
    "            1: \"True\",\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "    \n",
    "    dend_order = pd.read_csv(os.path.join(pwd, \"input\", region + \"_\" + modality, \"cluster_order_and_colors.csv\"))\n",
    "    \n",
    "    adata.obs[\"class_scANVI\"].cat.rename_categories(\n",
    "        {\n",
    "            \"exc\": \"Neuronal: Glutamatergic\",\n",
    "            \"inh\": \"Neuronal: GABAergic\",\n",
    "            \"glia\": \"Non-neuronal and Non-neural\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "    \n",
    "    \n",
    "    adata.obs[\"subclass_scANVI\"].cat.rename_categories(\n",
    "        {\n",
    "            \"Lamp5_Lhx6\": \"Lamp5 Lhx6\",\n",
    "            \"Astro\": \"Astrocyte\",\n",
    "            \"Oligo\": \"Oligodendrocyte\",\n",
    "            \"Endo\": \"Endothelial\",\n",
    "            \"Micro-PVM\": \"Microglia-PVM\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "    \n",
    "    adata.obs[\"supertype_scANVI_leiden\"].cat.rename_categories(\n",
    "        {\n",
    "            \"Astro_Unknown_25\": \"Astro_6-SEAAD\",\n",
    "            \"Oligo_Unknown_15\": \"Oligo_2_1-SEAAD\",\n",
    "            \"OPC_Unknown_25\": \"OPC_2_2-SEAAD\",\n",
    "            \"OPC_Unknown_18\": \"OPC_2_1-SEAAD\",\n",
    "            \"VLMC_Unknown_24\": \"VLMC_2_1-SEAAD\",\n",
    "            \"VLMC_Unknown_32\": \"VLMC_2_2-SEAAD\",\n",
    "            \"Micro-PVM_Unknown_0\": \"Micro-PVM_3-SEAAD\",\n",
    "            \"Micro-PVM_Unknown_9\": \"Micro-PVM_2_3-SEAAD\",\n",
    "            \"Micro-PVM_Unknown_10\": \"Micro-PVM_4-SEAAD\",\n",
    "            \"Micro-PVM_Unknown_116\": \"Micro-PVM_2_1-SEAAD\",\n",
    "            \"Micro-PVM_Unknown_135\": \"Micro-PVM_1_1-SEAAD\",\n",
    "            \"Micro-PVM_Unknown_200\": \"Micro-PVM_2_2-SEAAD\",\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "        \n",
    "    adata.obs[\"supertype_scANVI_leiden\"].cat.rename_categories(\n",
    "        {\n",
    "            \"VLMC_2\": \"Pericyte_1\",\n",
    "            \"VLMC_2_1-SEAAD\": \"SMC-SEAAD\",\n",
    "            \"VLMC_2_2-SEAAD\": \"Pericyte_2-SEAAD\",\n",
    "            \"Micro-PVM_2_2-SEAAD\": \"Lymphocyte\",\n",
    "            \"Micro-PVM_1_1-SEAAD\": \"Monocyte\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "\n",
    "    if tag == \"final\":\n",
    "        class_order = [j for j in dend_order[\"class_label\"].drop_duplicates() if j in adata.obs[\"class_scANVI\"].cat.categories]\n",
    "        adata.obs[\"class_scANVI\"] = adata.obs[\"class_scANVI\"].cat.reorder_categories(class_order)\n",
    "        \n",
    "        class_order = [j for j in dend_order[\"subclass_label\"].drop_duplicates() if j in adata.obs[\"subclass_scANVI\"].cat.categories]\n",
    "        adata.obs[\"subclass_scANVI\"] = adata.obs[\"subclass_scANVI\"].cat.reorder_categories(class_order)\n",
    "\n",
    "        class_order = [j for j in dend_order[\"cluster_label\"].drop_duplicates() if j in adata.obs[\"supertype_scANVI_leiden\"].cat.categories]\n",
    "        adata.obs[\"supertype_scANVI_leiden\"] = adata.obs[\"supertype_scANVI_leiden\"].cat.reorder_categories(class_order)\n",
    "    \n",
    "    sc.pp.subsample(adata, fraction=1)\n",
    "    \n",
    "    adata.obs[\"ch_cognitivestatus_binary\"].cat.reorder_categories([\"Reference\", \"No dementia\", \"Dementia\"], inplace=True)\n",
    "    adata.obs[\"adneurochange\"].cat.reorder_categories([\"Reference\", \"Not AD\", \"Low\", \"Intermediate\", \"High\"], inplace=True)\n",
    "    adata.obs[\"braak\"].cat.reorder_categories([\"Reference\", \"Braak 0\", \"Braak II\", \"Braak III\", \"Braak IV\", \"Braak V\", \"Braak VI\"], inplace=True)\n",
    "    adata.obs[\"thal\"].cat.reorder_categories([\"Reference\", \"Thal 0\", \"Thal 1\", \"Thal 2\", \"Thal 3\", \"Thal 4\", \"Thal 5\"], inplace=True)\n",
    "    adata.obs[\"cscore\"].cat.reorder_categories([\"Reference\", 'Absent', 'Sparse', 'Moderate', 'Frequent'], inplace=True)\n",
    "    adata.obs[\"lewybodydisease\"].cat.reorder_categories([\"Reference\", 'Not Identified (olfactory bulb not assessed)', 'Not Identified (olfactory bulb assessed)', 'Olfactory bulb only', 'Amygdala-predominant', 'Brainstem-predominant', 'Limbic (Transitional)', 'Neocortical (Diffuse)'], inplace=True)\n",
    "    adata.obs[\"late_stage\"].cat.reorder_categories([\"Reference\", 'Unclassifiable', 'Not Identified', 'LATE Stage 1', 'LATE Stage 2', 'LATE Stage 3'], inplace=True)\n",
    "    adata.obs[\"ch_apoe_four\"].cat.reorder_categories([\"Reference\", \"N\", \"Y\"], inplace=True)\n",
    "\n",
    "    # Keep only display metadata\n",
    "    adata.obs = adata.obs.loc[\n",
    "        :,\n",
    "        [\n",
    "            \"sample_id\",\n",
    "            \"reference_cell\",\n",
    "            \"donor_name\",\n",
    "            \"organism\",\n",
    "            \"roi\",\n",
    "            \"sex\",\n",
    "            \"gender\",\n",
    "            \"age_at_death\",\n",
    "            \"ch_race___1\",\n",
    "            \"ch_race___2\",\n",
    "            \"ch_race___3\",\n",
    "            \"ch_race___4\",\n",
    "            \"ch_race___5\",\n",
    "            \"ch_race___6\",\n",
    "            \"ch_race___97\",\n",
    "            \"ch_raceother\",\n",
    "            \"ch_hispanicorlatino\",\n",
    "            \"ch_education\",\n",
    "            \"ch_education_years\",\n",
    "            \"pmi_date\",\n",
    "            \"ap_freshbrainweight\",\n",
    "            \"ap_brainph\",\n",
    "            \"adneurochange\",\n",
    "            \"thal\",\n",
    "            \"braak\",\n",
    "            \"cscore\",\n",
    "            \"caascore\",\n",
    "            \"lewybodydisease\",\n",
    "            \"micro_totalmicroinfarcts\",\n",
    "            \"micro_microvascularbrain\",\n",
    "            \"ge_atherosclerosis_id\",\n",
    "            \"micro_arteriolosclerosis_id\",\n",
    "            \"late_stage\",\n",
    "            \"ch_cognitivestatus_binary\",\n",
    "            \"ch_lastcasiscore\",\n",
    "            \"ch_casi_interval\",\n",
    "            \"ch_lastmmsescore\",\n",
    "            \"ch_mmse_interval\",\n",
    "            \"ch_lastmocascore_2\",\n",
    "            \"ch_moca_interval\",\n",
    "            \"ch_apoe_four\", \n",
    "            \"primary_studyname\",\n",
    "            \"secondary_studyname\",\n",
    "            \"cell_prep_type\",\n",
    "            \"facs_population_plan\",\n",
    "            \"rna_amplification\",\n",
    "            \"sample_name\",\n",
    "            \"sample_quantity_count\",\n",
    "            \"expc_cell_capture\",\n",
    "            \"method\",\n",
    "            \"pcr_cycles\",\n",
    "            \"percent_cdna_longer_than_400bp\",\n",
    "            \"rna_amplification_pass_fail\",\n",
    "            \"load_name\",\n",
    "            \"library_prep\",\n",
    "            \"library_input_ng\",\n",
    "            \"r1_index\",\n",
    "            \"avg_size_bp\",\n",
    "            \"quantification_fmol\",\n",
    "            \"library_prep_pass_fail\",\n",
    "            \"exp_component_vendor_name\",\n",
    "            \"batch_vendor_name\",\n",
    "            \"experiment_component_failed\",\n",
    "            \"alignment\",\n",
    "            \"Genome\",\n",
    "            \"ar_id\",\n",
    "            \"bc\",\n",
    "            \"Estimated_number_of_cells\",\n",
    "            \"number_of_reads\",\n",
    "            \"GEX_Mean_raw_reads_per_cell\",\n",
    "            \"GEX_Q30_bases_in_barcode\",\n",
    "            \"GEX_Q30_bases_in_read_2\",\n",
    "            \"GEX_Q30_bases_in_UMI\",\n",
    "            \"GEX_Percent_duplicates\",\n",
    "            \"GEX_Q30_bases_in_sample_index_i1\",\n",
    "            \"GEX_Q30_bases_in_sample_index_i2\",\n",
    "            \"GEX_Reads_with_TSO\",\n",
    "            \"GEX_Sequenced_read_pairs\",\n",
    "            \"GEX_Valid_UMIs\",\n",
    "            \"GEX_Valid_barcodes\",\n",
    "            \"GEX_Reads_mapped_to_genome\",\n",
    "            \"GEX_Reads_mapped_confidently_to_genome\",\n",
    "            \"GEX_Reads_mapped_confidently_to_intergenic_regions\",\n",
    "            \"GEX_Reads_mapped_confidently_to_intronic_regions\",\n",
    "            \"GEX_Reads_mapped_confidently_to_exonic_regions\",\n",
    "            \"GEX_Reads_mapped_confidently_to_transcriptome\",\n",
    "            \"GEX_Reads_mapped_antisense_to_gene\",\n",
    "            \"GEX_Fraction_of_transcriptomic_reads_in_cells\",\n",
    "            \"GEX_Total_genes_detected\",\n",
    "            \"GEX_Median_UMI_counts_per_cell\",\n",
    "            \"GEX_Median_genes_per_cell\",\n",
    "            \"Feature_linkages_detected\",\n",
    "            \"Linked_genes\",\n",
    "            \"Linked_peaks\",\n",
    "            \"ATAC_Confidently_mapped_read_pairs\",\n",
    "            \"ATAC_Fraction_of_genome_in_peaks\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_in_cells\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_TSS\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_peaks\",\n",
    "            \"ATAC_Fraction_of_transposition_events_in_peaks_in_cells\",\n",
    "            \"ATAC_Mean_raw_read_pairs_per_cell\",\n",
    "            \"ATAC_Median_high.quality_fragments_per_cell\",\n",
    "            \"ATAC_Non.nuclear_read_pairs\",\n",
    "            \"ATAC_Number_of_peaks\",\n",
    "            \"ATAC_Percent_duplicates\",\n",
    "            \"ATAC_Q30_bases_in_barcode\",\n",
    "            \"ATAC_Q30_bases_in_read_1\",\n",
    "            \"ATAC_Q30_bases_in_read_2\",\n",
    "            \"ATAC_Q30_bases_in_sample_index_i1\",\n",
    "            \"ATAC_Sequenced_read_pairs\",\n",
    "            \"ATAC_TSS_enrichment_score\",\n",
    "            \"ATAC_Unmapped_read_pairs\",\n",
    "            \"ATAC_Valid_barcodes\",\n",
    "            \"nCount_RNA\",\n",
    "            \"nFeature_RNA\",\n",
    "            \"doublet_score\",\n",
    "            \"fraction_mito\",\n",
    "            \"for_analysis\",\n",
    "            \"class_conf_scANVI\",\n",
    "            \"class_scANVI\",\n",
    "            \"subclass_conf_scANVI\",\n",
    "            \"subclass_scANVI\",\n",
    "            \"supertype_conf_scANVI\",\n",
    "            \"supertype_scANVI\",\n",
    "            \"supertype_scANVI_leiden\",\n",
    "            \"rna_neighbors_qc_ratio_new\",\n",
    "            \"leiden_1.0\",\n",
    "        ]\n",
    "    ]\n",
    "    adata.obs.rename(\n",
    "        {\n",
    "            \"reference_cell\": \"Neurotypical reference\",\n",
    "            \"donor_name\": \"Donor ID\",\n",
    "            \"organism\": \"Organism\",\n",
    "            \"roi\": \"Brain Region\",\n",
    "            \"sex\": \"Sex\",\n",
    "            \"gender\": \"Gender\",\n",
    "            \"age_at_death\": \"Age at Death\",\n",
    "            \"ch_race___1\": \"Race (choice=White)\",\n",
    "            \"ch_race___2\": \"Race (choice=Black/ African American)\",\n",
    "            \"ch_race___3\": \"Race (choice=Asian)\",\n",
    "            \"ch_race___4\": \"Race (choice=American Indian/ Alaska Native)\",\n",
    "            \"ch_race___5\": \"Race (choice=Native Hawaiian or Pacific Islander)\",\n",
    "            \"ch_race___6\": \"Race (choice=Unknown or unreported)\",\n",
    "            \"ch_race___97\": \"Race (choice=Other)\",\n",
    "            \"ch_raceother\": \"specify other race\",\n",
    "            \"ch_hispanicorlatino\": \"Hispanic/Latino\",\n",
    "            \"ch_education\": \"Highest level of education\",\n",
    "            \"ch_education_years\": \"Years of education\",\n",
    "            \"pmi_date\": \"PMI\",\n",
    "            \"ap_freshbrainweight\": \"Fresh Brain Weight\",\n",
    "            \"ap_brainph\": \"Brain pH\",\n",
    "            \"adneurochange\": \"Overall AD neuropathological Change\",\n",
    "            \"thal\": \"Thal\",\n",
    "            \"braak\": \"Braak\",\n",
    "            \"cscore\": \"CERAD score\",\n",
    "            \"caascore\": \"Overall CAA Score\",\n",
    "            \"lewybodydisease\": \"Highest Lewy Body Disease\",\n",
    "            \"micro_totalmicroinfarcts\": \"Total Microinfarcts (not observed grossly)\",\n",
    "            \"micro_microvascularbrain\": \"Total microinfarcts in screening sections\",\n",
    "            \"ge_atherosclerosis_id\": \"Atherosclerosis\",\n",
    "            \"micro_arteriolosclerosis_id\": \"Arteriolosclerosis\",\n",
    "            \"late_stage\": \"LATE\",\n",
    "            \"ch_cognitivestatus_binary\": \"Cognitive Status\",\n",
    "            \"ch_lastcasiscore\": \"Last CASI Score\",\n",
    "            \"ch_casi_interval\": \"Interval from last CASI in months\",\n",
    "            \"ch_lastmmsescore\": \"Last MMSE Score\",\n",
    "            \"ch_mmse_interval\": \"Interval from last MMSE in months\",\n",
    "            \"ch_lastmocascore_2\": \"Last MOCA Score\",\n",
    "            \"ch_moca_interval\": \"Interval from last MOCA in months\",\n",
    "            \"ch_apoe_four\": \"APOE4 Status\",\n",
    "            \"primary_studyname\": \"Primary Study Name\",\n",
    "            \"secondary_studyname\": \"Secondary Study Name\",\n",
    "            \"Estimated_number_of_cells\": \"GEX_Estimated_number_of_cells\",\n",
    "            \"number_of_reads\": \"GEX_number_of_reads\",\n",
    "            \"Feature_linkages_detected\": \"Multiome_Feature_linkages_detected\",\n",
    "            \"Linked_genes\": \"Multiome_Linked_genes\",\n",
    "            \"Linked_peaks\": \"Multiome_Linked_peaks\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_in_cells\": \"ATAC_Fraction_of_high_quality_fragments_in_cells\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_TSS\": \"ATAC_Fraction_of_high_quality_fragments_overlapping_TSS\",\n",
    "            \"ATAC_Fraction_of_high.quality_fragments_overlapping_peaks\": \"ATAC_Fraction_of_high_quality_fragments_overlapping_peaks\",\n",
    "            \"ATAC_Median_high.quality_fragments_per_cell\": \"ATAC_Median_high_quality_fragments_per_cell\",\n",
    "            \"ATAC_Non.nuclear_read_pairs\": \"ATAC_Non-nuclear_read_pairs\",\n",
    "            \"nCount_RNA\": \"Number of UMIs\",\n",
    "            \"nFeature_RNA\": \"Genes detected\",\n",
    "            \"doublet_score\": \"Doublet score\",\n",
    "            \"fraction_mito\": \"Fraction mitochondrial UMIs\",\n",
    "            \"for_analysis\": \"Used in analysis\",\n",
    "            \"class_conf_scANVI\": \"Class confidence\",\n",
    "            \"class_scANVI\": \"Class\",\n",
    "            \"subclass_conf_scANVI\": \"Subclass confidence\",\n",
    "            \"subclass_scANVI\": \"Subclass\",\n",
    "            \"supertype_conf_scANVI\": \"Supertype confidence\",\n",
    "            \"supertype_scANVI\": \"Supertype (non-expanded)\",\n",
    "            \"supertype_scANVI_leiden\": \"Supertype\",\n",
    "            \"rna_neighbors_qc_ratio_new\": \"RNA Quality Control Score\",\n",
    "            \"leiden_1.0\": \"Quality Control Clusters\",\n",
    "        },\n",
    "        axis=1,\n",
    "        inplace=True\n",
    "    )\n",
    "\n",
    "    metadata_colors = {}\n",
    "    for i in metadata_to_color:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"category\")\n",
    "        a = adata.obs[i].cat.categories.to_list()\n",
    "        cmap = plt.get_cmap('viridis', len(a)+1)\n",
    "        b = [colors.rgb2hex(cmap(j)) for j in range(cmap.N)]\n",
    "        b.pop(0)\n",
    "        if i == \"Sex\":\n",
    "            b = [\"pink\", \"dodgerblue\"]\n",
    "        for A, B in zip(a, b):\n",
    "            try:\n",
    "                metadata_colors[i][A] = B\n",
    "            except:\n",
    "                metadata_colors[i] = {}\n",
    "                metadata_colors[i][A] = B\n",
    "        try:\n",
    "            del adata.uns[i + \"_colors\"]\n",
    "        except:\n",
    "            pass\n",
    "        sc.pl.umap(adata,\n",
    "            color=i,\n",
    "            title=\"\",\n",
    "            palette=metadata_colors[i],\n",
    "            size=10,\n",
    "            legend_fontsize=12\n",
    "        )\n",
    "    \n",
    "    if annotation_color == True:\n",
    "        tmp = dend_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "        tmp.index = tmp[\"subclass_label\"].copy()\n",
    "        tmp.drop(\"subclass_label\", axis=1, inplace=True)\n",
    "        tmp = tmp.to_dict()[\"subclass_color\"]\n",
    "        sc.pl.umap(adata,\n",
    "            color=\"Subclass\",\n",
    "            title=\"\",\n",
    "            palette=tmp,\n",
    "            size=10,\n",
    "            legend_fontsize=12\n",
    "        )\n",
    "    \n",
    "    for i in adata.uns.keys():\n",
    "        if i.endswith(\"_colors\"):\n",
    "            adata.uns[i] = [re.sub(\"ff$\", \"\", j) for j in adata.uns[i]]\n",
    "\n",
    "\n",
    "    CPS = pd.read_csv(os.path.join(pwd, \"input\", region, \"donor_name_CPS.csv\"), index_col=0)\n",
    "    adata.obs = adata.obs.merge(CPS, left_on=\"Donor ID\", right_index=True, how=\"left\")\n",
    "\n",
    "    adata.write(os.path.join(pwd, \"output\", region, \"SEAAD_\" + region + \"_\" + modality + \"_\" + tag + \" -nuclei.\" + str(datetime.datetime.date(datetime.datetime.now())) + \".h5ad\"), compression=\"gzip\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9040eab-d32b-40a0-88f8-1fdffb6fcdf4",
   "metadata": {},
   "source": [
    "### Read in snATACseq data and match format with snRNAseq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc55da2-93a7-434b-9910-472e3099a392",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = \"MTG\"\n",
    "modalit = \"ATACseq\"\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", region + \"_\" + modality, \"multivi_AD_3cohorts_RNA_ATAC_Multiome_v2.0_annotation_updated.h5ad\"))\n",
    "\n",
    "# Read in metadata variables that are also present in snRNAseq\n",
    "del adata.obsm[\"X_scVI\"]\n",
    "del adata.layers[\"UMIs\"]\n",
    "adata = adata[adata.obs[\"method\"] != \"10Xv3.1\"].copy()\n",
    "adata = adata[:, adata.var[\"modality\"] == \"Peaks\"].copy\n",
    "\n",
    "ATAC_metadata = pd.read_csv(os.path.join(pwd, \"input\", region + \"_\" + modality, \"ATAC_AD_Center_Grant_update.csv\"))\n",
    "ATAC_ref_metadata = pd.read_csv(os.path.join(pwd, \"input\", region + \"_\" + modality, \"ATAC_Hu_Ref_Brain_ATX-576/ATAC_LIMS_Hu_Ref_Brain_update.csv\"))\n",
    "ATAC_ref_metadata = ATAC_ref_metadata.loc[[i in ['L8AT_210427_01_A05', 'L8AT_210427_01_B05'] for i in ATAC_ref_metadata[\"library_prep\"]], ATAC_metadata.columns]\n",
    "ATAC_metadata = pd.concat([ATAC_metadata, ATAC_ref_metadata], axis=0)\n",
    "to_drop_singleome = [\n",
    "    \"Estimated_bulk_library_complexity\",\n",
    "    \"Fragments_flanking_a_single_nucleosome\",\n",
    "    \"Fragments_in_nucleosome-free_regions\",\n",
    "    \"Pipeline_version\",\n",
    "    \"Sample_ID\",\n",
    "    \"lib_method\",\n",
    "    \"percent_dinucleosome_analyzer\",\n",
    "    \"percent_mononucleosome_analyzer\",\n",
    "    \"percent_multinucleated_fragments_analyzer\",\n",
    "    \"percent_nucleosome_free_analyzer\",\n",
    "    \"Sequencing_saturation\"\n",
    "]\n",
    "to_rename_singleome = {\n",
    "    \"Confidently_mapped_read_pairs\": \"ATAC_Confidently_mapped_read_pairs\",\n",
    "    \"Fraction_of_genome_in_peaks\": \"ATAC_Fraction_of_genome_in_peaks\",\n",
    "    \"Fraction_of_high-quality_fragments_in_cells\": \"ATAC_Fraction_of_high.quality_fragments_in_cells\",\n",
    "    \"Fraction_of_high-quality_fragments_overlapping_TSS\": \"ATAC_Fraction_of_high.quality_fragments_overlapping_TSS\",\n",
    "    \"Fraction_of_high-quality_fragments_overlapping_peaks\": \"ATAC_Fraction_of_high.quality_fragments_overlapping_peaks\",\n",
    "    \"Fraction_of_transposition_events_in_peaks_in_cells\": \"ATAC_Fraction_of_transposition_events_in_peaks_in_cells\",\n",
    "    \"Mean_raw_read_pairs_per_cell\": \"ATAC_Mean_raw_read_pairs_per_cell\",\n",
    "    \"Median_high-quality_fragments_per_cell\": \"ATAC_Median_high.quality_fragments_per_cell\",\n",
    "    \"Non-nuclear_read_pairs\": \"ATAC_Non.nuclear_read_pairs\",\n",
    "    \"Number_of_peaks\": \"ATAC_Number_of_peaks\",\n",
    "    \"Percent_duplicates\": \"ATAC_Percent_duplicates\",\n",
    "    \"Q30_bases_in_barcode\": \"ATAC_Q30_bases_in_barcode\",\n",
    "    \"Q30_bases_in_read_1\": \"ATAC_Q30_bases_in_read_1\",\n",
    "    \"Q30_bases_in_read_2\": \"ATAC_Q30_bases_in_read_2\",\n",
    "    \"Q30_bases_in_sample_index_i1\": \"ATAC_Q30_bases_in_sample_index_i1\",\n",
    "    \"Sequenced_read_pairs\": \"ATAC_Sequenced_read_pairs\",\n",
    "    \"TSS_enrichment_score\": \"ATAC_TSS_enrichment_score\",\n",
    "    \"Unmapped_read_pairs\": \"ATAC_Unmapped_read_pairs\",\n",
    "    \"Valid_barcodes\": \"ATAC_Valid_barcodes\",\n",
    "}\n",
    "\n",
    "ATAC_metadata.drop(to_drop_singleome, axis=1, inplace=True)\n",
    "ATAC_metadata.rename(columns=to_rename_singleome, inplace=True)\n",
    "\n",
    "ATAC_metadata[\"alignment\"] = \"cellranger-arc-2.0.0\"\n",
    "\n",
    "tmp = adata.obs.merge(ATAC_metadata, left_on=\"library_prep\", right_on=\"library_prep\", how=\"left\")\n",
    "for i in tmp.columns[tmp.columns.str.endswith(\"_x\")]:\n",
    "    tmp[i] = tmp[i].astype(\"object\")\n",
    "    tmp[i].fillna(tmp[i.replace(\"_x\", \"_y\")], inplace=True)\n",
    "    tmp[i.replace(\"_x\", \"\")] = tmp[i].copy()\n",
    "    tmp.drop([i, i.replace(\"_x\", \"_y\")], axis=1, inplace=True)\n",
    "adata.obs = tmp.loc[:, adata.obs.columns]\n",
    "\n",
    "clinical_metadata = pd.read_spss(os.path.join(pwd, \"input\", region + \"_\" + modality, \"NeuropathMetadata_version103_20220415.KT.sav\"))\n",
    "tmp = adata.obs.merge(clinical_metadata, left_on=\"external_donor_name\", right_on=\"uwa\", how=\"left\")\n",
    "for i in tmp.columns[tmp.columns.str.endswith(\"_x\")]:\n",
    "    tmp[i] = tmp[i].astype(\"object\")\n",
    "    tmp[i].fillna(tmp[i.replace(\"_x\", \"_y\")], inplace=True)\n",
    "    tmp[i.replace(\"_x\", \"\")] = tmp[i].copy()\n",
    "    tmp.drop([i, i.replace(\"_x\", \"_y\")], axis=1, inplace=True)\n",
    "adata.obs = tmp.loc[:, adata.obs.columns]\n",
    "adata.obs.index = adata.obs[\"sample_id\"].copy()\n",
    "adata.obs.index.name = \"index\"\n",
    "\n",
    "# Pre-surgery to remove low quality clusters identified in snATACseq notebooks\n",
    "bad_clusters = ['6', '10', '21', '22', '25', '29', '32', '33', '36', '39', '47', '37', '42', '43', '45', '46']\n",
    "for_analysis = (adata.obs[\"rna_neighbors_qc_ratio_new\"] <= 0.2) & ([i not in bad_clusters for i in adata.obs[\"leiden_1.0\"]]) & ~(adata.obs[\"label_transfer\"].isna())\n",
    "for_analysis = np.array([str(i) for i in for_analysis])\n",
    "adata.obs[\"for_analysis\"] = for_analysis.copy()\n",
    "\n",
    "adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"subclass_scANVI\"] = adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"label_transfer\"]\n",
    "adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"subclass_conf_scANVI\"] = adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"subclass_purity_ratio\"]\n",
    "adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"class_scANVI\"] = \"exc\"\n",
    "adata.obs.loc[(adata.obs[\"method\"] == \"10xATAC_v1.1\") & ([i in [\"Lamp5_Lhx6\", \"Lamp5\", \"Sncg\", \"Pax6\", \"Vip\", \"Sst Chodl\", \"Sst\", \"Pvalb\", \"Chandelier\"] for i in adata.obs[\"subclass_scANVI\"]]), \"class_scANVI\"] = \"inh\"\n",
    "adata.obs.loc[(adata.obs[\"method\"] == \"10xATAC_v1.1\") & ([i in [\"Astro\", \"Oligo\", \"OPC\", \"Endo\", \"VLMC\", \"Micro-PVM\"] for i in adata.obs[\"subclass_scANVI\"]]), \"class_scANVI\"] = \"glia\"\n",
    "adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"class_conf_scANVI\"] = adata.obs.loc[adata.obs[\"method\"] == \"10xATAC_v1.1\", \"subclass_purity_ratio\"]\n",
    "\n",
    "adata.obs[\"reference_cell\"].fillna(0, inplace=True)\n",
    "adata.obs.loc[[i in ['L8AT_210427_01_A05', 'L8AT_210427_01_B05'] for i in adata.obs[\"library_prep\"]], \"reference_cell\"] = 1\n",
    "\n",
    "adata.obs[\"bc\"] = [re.sub(\"^([ATGC]+)-(.*)$\", \"\\\\1\", i) for i in adata.obs[\"sample_id\"]]\n",
    "\n",
    "adata.var.drop(\"modality\", axis=1, inplace=True)\n",
    "\n",
    "adata.obs[\"nCount_RNA\"] = adata.X.sum(axis=1)\n",
    "\n",
    "adata.obs[\"age\"] = adata.obs[\"age\"].astype(\"object\")\n",
    "tmp = [np.float32(i.replace(\" yrs\", \"\")) for i in adata.obs.loc[adata.obs[\"reference_cell\"] == 1, \"age\"]]\n",
    "adata.obs.loc[adata.obs[\"reference_cell\"] == 1, \"age_at_death\"] = tmp.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0f31110-307a-48b2-9ebf-03259103a93f",
   "metadata": {},
   "source": [
    "### Prepare and export objects for AWS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d4364b1-3cc0-4a83-853c-0f492e1b3831",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = \"MTG\"\n",
    "modality = \"ATACseq\"\n",
    "\n",
    "prepare_anndata(\n",
    "    adata=adata,\n",
    "    region=region,\n",
    "    modality=modality,\n",
    "    tag=\"all\",\n",
    "    metadata_to_color=[],\n",
    "    annotation_color=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5396e69e-ce26-4fc3-a626-3cea5e340f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = \"MTG\"\n",
    "modality = \"ATACseq\"\n",
    "\n",
    "metadata_to_color = [\n",
    "    \"Sex\",\n",
    "    \"Cognitive Status\",\n",
    "    \"Overall AD neuropathological Change\",\n",
    "    \"Braak\",\n",
    "    \"Thal\",\n",
    "    \"CERAD score\",\n",
    "    \"Highest Lewy Body Disease\",\n",
    "    \"LATE\",\n",
    "    \"APOE4 Status\",\n",
    "]\n",
    "\n",
    "prepare_anndata(\n",
    "    adata=adata,\n",
    "    region=region,\n",
    "    modality=modality,\n",
    "    tag=\"final\",\n",
    "    metadata_to_color=metadata_to_color\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
