{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6cf99f84-a0c9-4f4e-af2e-7f76c6d98eac",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b33daad-a1c5-4b2a-997c-3952481b51f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import random\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "from helper_functions import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=100, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8099bed3-1558-451b-a91d-0ab2b07a882d",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c162463-6072-4355-934e-e28c87cddd73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "great_ape_colors = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Great_Ape_subclass_cluster_colors.csv\"))\n",
    "\n",
    "# From https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"SEAAD_MTG_RNAseq_final-nuclei_no_data_signature_scores.2024-02-13.h5ad\"))\n",
    "adata_ref = adata[adata.obs[\"Neurotypical reference\"] == \"True\", :].copy()\n",
    "\n",
    "# Results from one cluster self projection run from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "self_proj = pd.read_csv(os.path.join(\"input\", \"Extended Data Figure 5\", \"cluster_iterative_scANVI_results.2022-03-21.csv\"), index_col=0)\n",
    "\n",
    "# Results from one supertype self projection run from  https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "supertype_self_proj = pd.read_csv(os.path.join(\"input\", \"Extended Data Figure 5\", \"supertype_iterative_scANVI_results.2022-03-24.csv\"), index_col=0)\n",
    "\n",
    "# Coordinates from one self projection run for all IT Excitatory neurons and L5 IT\n",
    "Excitatory_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Reference Excitatory obs_names.csv\"), index_col=0).index\n",
    "Excitatory_umap = np.load(os.path.join(pwd, \"input\", \"Extended Data Figure 5\",\"Reference Excitatory X_umap.npy\"))\n",
    "\n",
    "L5_IT_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Reference L5 IT obs_names.csv\"), index_col=0).index\n",
    "L5_IT_umap = np.load(os.path.join(pwd, \"input\", \"Extended Data Figure 5\",\"Reference L5 IT X_umap.npy\"))\n",
    "\n",
    "# Great Apes metadata, from https://www.science.org/doi/10.1126/science.adf6812\n",
    "great_ape_metadata = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Great_ApesMetadata_version101_20220321.csv\"), index_col=0)\n",
    "\n",
    "# Initial mapping and QC AnnData for SEA-AD Microglia from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "Micro_PVM = sc.read_h5ad(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Micro-PVM_scANVI_no_data.2022-04-08.h5ad\"))\n",
    "\n",
    "# Subclass UMAP coordinates from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "Microglia_PVM_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Microglia-PVM\", \"obs_names.csv\"), index_col=0).index\n",
    "Microglia_PVM_umap = np.load(os.path.join(pwd, \"input\", \"Extended Data Figure 5\", \"Microglia-PVM\", \"X_umap.npy\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be172ecb-6ef3-4972-8b00-9f95a9a173ed",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f152dfdd-924f-4fba-81de-ecddbc8a39aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref.obs[\"cluster\"] = great_ape_metadata.loc[adata_ref.obs_names, \"cluster\"].to_list()\n",
    "adata_ref.obs[\"cluster\"] = adata_ref.obs[\"cluster\"].astype(\"category\")\n",
    "\n",
    "low_confidence = [\"L2/3 IT_4\", \"L2/3 IT_9\", \"L2/3 IT_11\",\n",
    "                  \"L5 IT_4\",\n",
    "                  \"L5/6 NP_5\",\n",
    "                  \"Micro-PVM_3\",\n",
    "                  \"Pvalb_4\", \"Pvalb_11\",\n",
    "                  \"Sncg_7\",\n",
    "                  \"Sst_6\", \"Sst_8\", \"Sst_14\", \"Sst_15\", \"Sst_16\", \"Sst_17\", \"Sst_18\", \"Sst_21\", \"Sst_24\", \"Sst_26\", \n",
    "                  \"Vip_3\", \"Vip_7\",  \"Vip_8\", \"Vip_10\", \"Vip_17\", \"Vip_20\", \"Vip_22\"]\n",
    "adata_ref.obs[\"supertype\"] = adata_ref.obs[\"cluster\"].copy()\n",
    "adata_ref.obs[\"supertype\"] = adata_ref.obs[\"supertype\"].astype(\"object\")\n",
    "for i in low_confidence:\n",
    "    adata_ref.obs.loc[adata_ref.obs[\"cluster\"] == i, \"supertype\"] = \"Unknown\"\n",
    "adata_ref.obs[\"supertype\"] = adata_ref.obs[\"supertype\"].astype(\"category\")\n",
    "\n",
    "adata_ref.obs[\"class_held\"] = self_proj.loc[:, \"class_held\"]\n",
    "adata_ref.obs[\"class_held\"] = adata_ref.obs[\"class_held\"].astype(\"category\")\n",
    "\n",
    "adata_ref.obs[\"subclass_held\"] = self_proj.loc[:, \"subclass_held\"]\n",
    "adata_ref.obs[\"subclass_held\"] = adata_ref.obs[\"subclass_held\"].astype(\"category\")\n",
    "\n",
    "adata_ref.obs[\"cluster_held\"] = self_proj.loc[:, \"cluster_held\"]\n",
    "adata_ref.obs[\"cluster_held\"] = adata_ref.obs[\"cluster_held\"].astype(\"category\")\n",
    "\n",
    "adata_ref.obs[\"supertype_held\"] = supertype_self_proj.loc[:, \"supertype_held\"]\n",
    "adata_ref.obs[\"supertype_held\"] = adata_ref.obs[\"supertype_held\"].astype(\"category\")\n",
    "\n",
    "adata_ref.obs[\"class_held_scANVI\"] = self_proj.loc[:, \"class_held_scANVI\"]\n",
    "adata_ref.obs[\"class_held_scANVI\"] = adata_ref.obs[\"class_held_scANVI\"].astype(\"category\")\n",
    "adata_ref.obs[\"class_held_conf_scANVI\"] = self_proj.loc[:, \"class_held_conf_scANVI\"]\n",
    "\n",
    "adata_ref.obs[\"subclass_held_scANVI\"] = self_proj.loc[:, \"subclass_held_scANVI\"]\n",
    "adata_ref.obs[\"subclass_held_scANVI\"] = adata_ref.obs[\"subclass_held_scANVI\"].astype(\"category\")\n",
    "adata_ref.obs[\"subclass_held_conf_scANVI\"] = self_proj.loc[:, \"subclass_held_conf_scANVI\"]\n",
    "\n",
    "adata_ref.obs[\"cluster_held_scANVI\"] = self_proj.loc[:, \"cluster_held_scANVI\"]\n",
    "adata_ref.obs[\"cluster_held_scANVI\"] = adata_ref.obs[\"cluster_held_scANVI\"].astype(\"category\")\n",
    "adata_ref.obs[\"cluster_held_conf_scANVI\"] = self_proj.loc[:, \"cluster_held_conf_scANVI\"]\n",
    "\n",
    "adata_ref.obs[\"supertype_held_scANVI\"] = supertype_self_proj.loc[:, \"supertype_held_scANVI\"]\n",
    "adata_ref.obs[\"supertype_held_scANVI\"] = adata_ref.obs[\"supertype_held_scANVI\"].astype(\"category\")\n",
    "adata_ref.obs[\"supertype_held_conf_scANVI\"] = supertype_self_proj.loc[:, \"supertype_held_conf_scANVI\"]\n",
    "\n",
    "adata_ref.obs[\"class_held\"] = adata_ref.obs[\"class_held\"].cat.rename_categories(\n",
    "    {\n",
    "        \"exc\": \"Neuronal: Glutamatergic\",\n",
    "        \"inh\": \"Neuronal: GABAergic\",\n",
    "        \"glia\": \"Non-neuronal and Non-neural\"\n",
    "    },\n",
    ")\n",
    "adata_ref.obs[\"class_held_scANVI\"] = adata_ref.obs[\"class_held_scANVI\"].cat.rename_categories(\n",
    "    {\n",
    "        \"exc\": \"Neuronal: Glutamatergic\",\n",
    "        \"inh\": \"Neuronal: GABAergic\",\n",
    "        \"glia\": \"Non-neuronal and Non-neural\"\n",
    "    },\n",
    ")\n",
    "\n",
    "adata_ref.obs[\"subclass_held\"] = adata_ref.obs[\"subclass_held\"].cat.rename_categories(\n",
    "    {\n",
    "        \"Lamp5_Lhx6\": \"Lamp5 Lhx6\",\n",
    "        \"Astro\": \"Astrocyte\",\n",
    "        \"Oligo\": \"Oligodendrocyte\",\n",
    "        \"Endo\": \"Endothelial\",\n",
    "        \"Micro-PVM\": \"Microglia-PVM\"\n",
    "    },\n",
    ")\n",
    "adata_ref.obs[\"subclass_held_scANVI\"] = adata_ref.obs[\"subclass_held_scANVI\"].cat.rename_categories(\n",
    "    {\n",
    "        \"Lamp5_Lhx6\": \"Lamp5 Lhx6\",\n",
    "        \"Astro\": \"Astrocyte\",\n",
    "        \"Oligo\": \"Oligodendrocyte\",\n",
    "        \"Endo\": \"Endothelial\",\n",
    "        \"Micro-PVM\": \"Microglia-PVM\"\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91d89961-8a16-4b51-85d8-1c9ef37f5d6a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Class plots\n",
    "sub = adata_ref.copy()\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "class_colors = {}\n",
    "class_colors[\"Neuronal: Glutamatergic\"] = color_order.loc[color_order[\"subclass_label\"] == \"L5 IT\", \"subclass_color\"].iloc[0]\n",
    "class_colors[\"Neuronal: GABAergic\"] = color_order.loc[color_order[\"subclass_label\"] == \"Sst\", \"subclass_color\"].iloc[0]\n",
    "class_colors[\"Non-neuronal and Non-neural\"] = color_order.loc[color_order[\"subclass_label\"] == \"OPC\", \"subclass_color\"].iloc[0]\n",
    "class_colors[\"Unknown\"] = \"#EEEEEE99\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Class\"],\n",
    "    size=3,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=class_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference Whole Taxonomy_Class.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference Whole Taxonomy_Class.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference Whole Taxonomy_Class.pdf\"))\n",
    "\n",
    "df_self_proj = sub.obs.loc[:, [\"Class\", \"class_held_scANVI\", \"class_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "_ = plot_confusion(df=df_self_proj, figsize=(1.5,1.5))\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 5b_heatmap_Reference Whole Taxonomy_Class versus Class_scANVI.pdf\"), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "477cf61c-3a39-4773-99ef-4cc706884292",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Excitatory Subclass plots\n",
    "obs_names = adata_ref.obs_names.copy()\n",
    "adata_ref.obs.index = adata_ref.obs[\"sample_id\"].copy()\n",
    "sub = adata_ref[Excitatory_obs_names, :].copy()\n",
    "adata_ref.obs_names = obs_names.copy()\n",
    "sub.obsm[\"X_umap\"] = Excitatory_umap\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "    \n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "subclass_colors[\"Unknown\"] = \"#EEEEEE99\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Subclass\"],\n",
    "    size=3,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=subclass_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference Excitatory_Subclass.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference Excitatory_Subclass.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference Excitatory_Subclass.pdf\"))\n",
    "\n",
    "df_self_proj = sub.obs.loc[:, [\"Subclass\", \"subclass_held_scANVI\", \"subclass_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "_ = plot_confusion(df=df_self_proj, figsize=(4,4))\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 5b_heatmap_Reference Excitatory_Subclass versus Subclass_scANVI.pdf\"), bbox_inches='tight')\n",
    "\n",
    "\n",
    "# L5 IT Supertype plots\n",
    "\n",
    "obs_names = adata_ref.obs_names.copy()\n",
    "adata_ref.obs.index = adata_ref.obs[\"sample_id\"].copy()\n",
    "sub = adata_ref[L5_IT_obs_names, :].copy()\n",
    "adata_ref.obs_names = obs_names.copy()\n",
    "sub.obsm[\"X_umap\"] = L5_IT_umap\n",
    "sub = sub[sub.obs[\"cluster\"].str.startswith(\"L5 IT\"), :].copy()\n",
    "sub = sub[sub.obs[\"supertype_held_scANVI\"].str.startswith(\"L5 IT\"), :].copy()\n",
    "\n",
    "\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "    \n",
    "cluster_colors = great_ape_colors.loc[:, [\"cluster\", \"cluster_color\"]].drop_duplicates()\n",
    "cluster_colors.index = cluster_colors[\"cluster\"].copy()\n",
    "cluster_colors = cluster_colors[\"cluster_color\"].to_dict()\n",
    "cluster_colors[\"Unknown\"] = \"#EEEEEE99\"\n",
    "\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"cluster\"],\n",
    "    size=20,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=cluster_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference L5 IT_cluster.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference L5 IT_cluster.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference L5 IT_cluster.pdf\"))\n",
    "\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"cluster_held\"],\n",
    "    size=20,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=cluster_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference L5 IT_cluster_held.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference L5 IT_cluster_held.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference L5 IT_cluster_held.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"cluster_held_scANVI\"],\n",
    "    size=20,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=cluster_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference L5 IT_cluster_held_scANVI.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference L5 IT_cluster_held_scANVI.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference L5 IT_cluster_held_scANVI.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"cluster_held_conf_scANVI\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    cmap=\"YlGnBu_r\",\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference L5 IT_cluster_held_conf_scANVI.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference L5 IT_cluster_held_conf_scANVI.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference L5 IT_cluster_held_conf_scANVI.pdf\"))\n",
    "\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"supertype_held\"],\n",
    "    size=20,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=cluster_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference L5 IT_supertype_held.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference L5 IT_supertype_held.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference L5 IT_supertype_held.pdf\"))\n",
    "\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"supertype_held_scANVI\"],\n",
    "    size=20,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    palette=cluster_colors,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Reference L5 IT_supertype_held_scANVI.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Reference L5 IT_supertype_held_scANVI.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5b_umap_Reference L5 IT_supertype_held_scANVI.pdf\"))\n",
    "\n",
    "df_self_proj = sub.obs.loc[:, [\"cluster\", \"cluster_held_scANVI\", \"cluster_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "_ = plot_confusion(df=df_self_proj, figsize=(4,4))\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 5b_heatmap_Reference L5 IT_cluster versus cluster_scANVI.pdf\"), bbox_inches='tight')\n",
    "\n",
    "\n",
    "df_self_proj = sub.obs.loc[:, [\"supertype\", \"supertype_held_scANVI\", \"supertype_held_conf_scANVI\"]]\n",
    "df_self_proj = df_self_proj.loc[df_self_proj[\"supertype\"] != \"Unknown\", :]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "df_self_proj[\"true\"] = df_self_proj[\"true\"].cat.remove_unused_categories()\n",
    "df_self_proj[\"pred\"] = df_self_proj[\"pred\"].cat.remove_unused_categories()\n",
    "_ = plot_confusion(df=df_self_proj, figsize=(4,4))\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 5b_heatmap_Reference L5 IT_supertype versus supertype_scANVI.pdf\"), bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9938aaa4-2a7e-49aa-8050-250a4c23b0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# F1 score boxplots\n",
    "df_self_proj = adata_ref.obs.loc[:, [\"cluster\", \"cluster_held_scANVI\", \"cluster_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "result_cluster = get_scores(df=df_self_proj)\n",
    "result_cluster[\"model_version\"] = \"Clusters\"\n",
    "\n",
    "df_self_proj = adata_ref.obs.loc[:, [\"supertype\", \"Supertype (non-expanded)\", \"Supertype confidence\"]]\n",
    "df_self_proj = df_self_proj.loc[df_self_proj[\"supertype\"] != \"Unknown\", :]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "df_self_proj[\"true\"] = df_self_proj[\"true\"].cat.remove_unused_categories()\n",
    "df_self_proj[\"pred\"] = df_self_proj[\"pred\"].cat.remove_unused_categories()\n",
    "result_supertype = get_scores(df=df_self_proj)\n",
    "result_supertype[\"model_version\"] = \"Supertypes\"\n",
    "\n",
    "result = pd.concat([result_cluster, result_supertype], axis=0)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (1.5,4)\n",
    "ax = sns.boxplot(\n",
    "    data=result,\n",
    "    x=\"model_version\",\n",
    "    y=\"f1\",\n",
    "    showfliers=False,\n",
    "    palette=\"Greys\"\n",
    ");\n",
    "ax = sns.stripplot(\n",
    "    data=result,\n",
    "    x=\"model_version\",\n",
    "    y=\"f1\",\n",
    "    color=\"0.25\",\n",
    "    edgecolor=\"white\",\n",
    "    linewidth=0.5,\n",
    "    alpha=0.5\n",
    ");\n",
    "ax.axhline(0.7, linestyle=\"dashed\", color=\"black\")\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "ax.set(xlabel=\"\", ylabel=\"F1 Score\");\n",
    "\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 5b_boxplot_cluster_versus_supertype_F1_scores.pdf\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9a95678-fd0c-4028-884f-4386f64061c4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22795828-efbb-4823-87eb-8a7d21ceacba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Micro_PVM.obs[\"Uncertainty\"] = 1 - Micro_PVM.obs[\"supertype_conf_scANVI\"]\n",
    "supertype_colors = color_order.loc[:, [\"cluster_label\", \"cluster_color\"]].drop_duplicates()\n",
    "supertype_colors.index = supertype_colors[\"cluster_label\"].copy()\n",
    "supertype_colors.drop(\"cluster_label\", axis=1, inplace=True)\n",
    "supertype_colors = supertype_colors[\"cluster_color\"].to_dict()\n",
    "supertype_colors[\"SEA-AD\"] = \"#EEEEEE99\"\n",
    "\n",
    "Micro_PVM.obs[\"supertype\"] = Micro_PVM.obs[\"supertype\"].cat.rename_categories(\n",
    "    {\"Unknown\": \"SEA-AD\"},\n",
    ")\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"supertype\"],\n",
    "    size=20,\n",
    "    palette=supertype_colors,\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_supertype.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_supertype.pdf\"))\n",
    "\n",
    "\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"supertype_scANVI\"],\n",
    "    size=20,\n",
    "    palette=supertype_colors,\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_supertype_scANVI.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_supertype_scANVI.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_supertype_scANVI.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"Uncertainty\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    vmin=-0.05,\n",
    "    cmap=\"YlGnBu\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Uncertainty.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Uncertainty.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_Uncertainty.pdf\"))\n",
    "\n",
    "colors = list(supertype_colors.values())\n",
    "np.random.shuffle(colors)\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"leiden_merged\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    palette=colors,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Cluster.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Cluster.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_Cluster.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"cluster_doublet_score\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    vmin=-0.05,\n",
    "    cmap=\"YlGnBu\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Cluster Doublet score.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Cluster Doublet score.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_Cluster Doublet score.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"cluster_fraction_mito\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    cmap=\"YlGnBu\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Cluster Fraction of Mitochondrial UMIs.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Cluster Fraction of Mitochondrial UMIs.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_Cluster Fraction of Mitochondrial UMIs.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"cluster_donor_split\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    cmap=\"YlGnBu\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Cluster Donor entropy.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Cluster Donor entropy.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_Cluster Donor entropy.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"nFeature_RNA\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    cmap=\"YlGnBu\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Genes detected.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Genes detected.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5c_umap_Microglia_Genes detected.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce333259-6a1e-47d7-b642-977cde7d43f0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "197a840d-4bfc-44dc-8940-63c53acfb760",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (6,4)\n",
    "class_colors = {}\n",
    "class_colors[\"Neuronal: Glutamatergic\"] = color_order.loc[color_order[\"subclass_label\"] == \"L5 IT\", \"subclass_color\"].iloc[0]\n",
    "class_colors[\"Neuronal: GABAergic\"] = color_order.loc[color_order[\"subclass_label\"] == \"Sst\", \"subclass_color\"].iloc[0]\n",
    "class_colors[\"Non-neuronal and Non-neural\"] = color_order.loc[color_order[\"subclass_label\"] == \"OPC\", \"subclass_color\"].iloc[0]\n",
    "class_colors[\"Unknown\"] = \"#EEEEEE99\"\n",
    "\n",
    "df = sc.get.obs_df(adata, [\"library_prep\", \"Class\", \"Supertype confidence\", \"Supertype Signature Score\", \"Used in analysis\", \"Continuous Pseudo-progression Score\", \"Neurotypical reference\"])\n",
    "df = df.loc[(df[\"Used in analysis\"] == True) & (df[\"Neurotypical reference\"] == \"False\"), :].groupby([\"library_prep\", \"Class\", \"Continuous Pseudo-progression Score\", \"Used in analysis\", \"Neurotypical reference\"]).mean().reset_index().dropna()\n",
    "\n",
    "ax = sns.lmplot(\n",
    "    data=df,\n",
    "    x=\"Continuous Pseudo-progression Score\",\n",
    "    y=\"Supertype confidence\",\n",
    "    scatter_kws={\"alpha\": 0.2},\n",
    "    hue=\"Class\",\n",
    "    palette=class_colors\n",
    ");\n",
    "ax.set(ylim=(0.95,1));\n",
    "plt.ylabel(\"Supertype probabilities\");\n",
    "plt.xlabel(\"CPS\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 5d_regplot_Supertype confidence across CPS_splitby_class.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9ac3fc2-c7bc-444e-bfd3-0030876a4276",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34e66ed6-37b6-44a9-9b8e-3ed9b320db04",
   "metadata": {},
   "outputs": [],
   "source": [
    "supertype_colors = color_order.loc[:, [\"cluster_label\", \"cluster_color\"]].drop_duplicates()\n",
    "supertype_colors.index = supertype_colors[\"cluster_label\"].copy()\n",
    "supertype_colors = supertype_colors[\"cluster_color\"].to_dict()\n",
    "supertype_colors[\"Unknown\"] = \"#EEEEEE\"\n",
    "\n",
    "Micro_PVM = Micro_PVM[\n",
    "    (Micro_PVM.obs[\"cluster_doublet_score\"] < 0.1) &\n",
    "    (Micro_PVM.obs[\"cluster_fraction_mito\"] < 0.05) &\n",
    "    (Micro_PVM.obs[\"cluster_donor_split\"] < 0.4) &\n",
    "    (Micro_PVM.obs[\"nFeature_RNA\"] > 1000)\n",
    "].copy()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    Micro_PVM,\n",
    "    color=[\"supertype_scANVI\"],\n",
    "    size=20,\n",
    "    palette=supertype_colors,\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_subsetted_supertype_scANVI.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_subsetted_supertype_scANVI.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_subsetted_supertype_scANVI.pdf\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7de9e13d-a947-4058-b075-060447ca4797",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub.obs.loc[:, [\"leiden_merged2\", \"Neurotypical reference\"]].groupby([\"leiden_merged2\"]).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99b3875d-b4f0-4c10-8c69-e4761d1aebff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sub = adata[Microglia_PVM_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = Microglia_PVM_umap\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "sub.obs[\"Reference Supertype\"] = sub.obs[\"Supertype (non-expanded)\"].copy()\n",
    "sub.obs[\"Reference Supertype\"] = sub.obs[\"Reference Supertype\"].cat.add_categories([\"SEA-AD\"])\n",
    "sub.obs.loc[sub.obs[\"Neurotypical reference\"] == \"False\", \"Reference Supertype\"] = \"SEA-AD\"\n",
    "\n",
    "sub.obs[\"leiden_merged2\"] = sub.obs[\"Supertype\"].copy()\n",
    "sub.obs[\"leiden_merged2\"] = sub.obs[\"leiden_merged2\"].cat.rename_categories(\n",
    "    {\n",
    "        \"Micro-PVM_3-SEAAD\": \"0\",\n",
    "        \"Micro-PVM_1\": \"7\",\n",
    "        \"Micro-PVM_2_3-SEAAD\": \"9\",\n",
    "        \"Micro-PVM_4-SEAAD\": \"10\",\n",
    "        \"Micro-PVM_2\": \"85\",\n",
    "        \"Micro-PVM_2_1-SEAAD\": \"116\",\n",
    "        \"Monocyte\": \"135\",\n",
    "        \"Lymphocyte\": \"200\"\n",
    "    },\n",
    ")\n",
    "\n",
    "splitby = \"Supertype (non-expanded)\"\n",
    "fractions = pd.DataFrame(columns=[\"Max supertype fraction\", \"Fraction total reference\"])\n",
    "sub.obs[\"Neurotypical reference\"] = sub.obs[\"Neurotypical reference\"] == \"True\"\n",
    "supertype_counts = sub.obs.groupby([\"leiden_merged2\", splitby]).size()\n",
    "reference_counts = sub.obs.loc[:, [\"leiden_merged2\", \"Neurotypical reference\"]].groupby([\"leiden_merged2\"]).sum() / sub.obs[\"Neurotypical reference\"].value_counts().loc[True]\n",
    "for k in sub.obs[\"leiden_merged2\"].cat.categories:\n",
    "    l = np.nanmax(supertype_counts[k] / sub.obs.groupby([splitby]).size())\n",
    "    m = reference_counts.loc[k, :].max()\n",
    "    fractions = pd.concat([fractions, pd.DataFrame([[l, m]], index=[k], columns=[\"Max supertype fraction\", \"Fraction total reference\"])], axis=0)\n",
    "sub.obs = sub.obs.merge(fractions, left_on=\"leiden_merged2\", right_index=True, how=\"left\")\n",
    "\n",
    "supertype_colors = color_order.loc[:, [\"cluster_label\", \"cluster_color\"]].drop_duplicates()\n",
    "supertype_colors.index = supertype_colors[\"cluster_label\"].copy()\n",
    "supertype_colors = supertype_colors[\"cluster_color\"].to_dict()\n",
    "supertype_colors[\"SEA-AD\"] = \"#EEEEEE99\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Reference Supertype\"],\n",
    "    size=20,\n",
    "    palette=supertype_colors,\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_supertype.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_supertype.pdf\"))\n",
    "\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype (non-expanded)\"],\n",
    "    size=20,\n",
    "    palette=supertype_colors,\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_supertype_scANVI.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_supertype_scANVI.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_supertype_scANVI.pdf\"))\n",
    "\n",
    "\n",
    "np.random.seed(100)\n",
    "colors = list(supertype_colors.values())\n",
    "np.random.shuffle(colors)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"leiden_merged2\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    palette=colors,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Cluster.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Cluster.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_Cluster.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Max supertype fraction\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    cmap=\"YlGnBu\",\n",
    "    vmin=-0.05,\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Max supertype fraction.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Max supertype fraction.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_Max supertype fraction.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Fraction total reference\"],\n",
    "    size=20,\n",
    "    frameon=False,\n",
    "    cmap=\"YlGnBu\",\n",
    "    vmin=-0.05,\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_Fraction total reference.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_Fraction total reference.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_Fraction total reference.pdf\"))\n",
    "\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=20,\n",
    "    palette=supertype_colors,\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia_supertype_scANVI_expanded.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia_supertype_scANVI_expanded.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 5e_umap_Microglia_supertype_scANVI_expanded.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "247e765e-4173-4121-ae53-e10c587919d2",
   "metadata": {},
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bc3c1c4-d38a-4a34-8852-05939f1ee22f",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
