{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e3cf4ed-d88e-42e7-8bfb-c1148c7778de",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34ba3fe4-5e21-4db6-95dd-71f904797038",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import re\n",
    "import copy\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import rpy2\n",
    "%load_ext rpy2.ipython\n",
    "import warnings\n",
    "from scipy import stats as sp_stats\n",
    "from helper_functions import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=100, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78695622-034b-47e6-adf6-bdf575d5b025",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "961444e6-11ef-409f-89c3-38a19887cd53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# From Manuscript Supplementary Table 1\n",
    "Donors = pd.read_excel(os.path.join(pwd, \"input\", \"Extended Data Figure 7\", \"Supplementary Table 1.xlsx\"), sheet_name=\"SEA-AD_Cohort_Metadata\")\n",
    "\n",
    "# Results from one supertype self projection run from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "RNAseq_reference = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 7\", \"supertype_iterative_scANVI_results.2022-03-24.csv\"), index_col=0)\n",
    "\n",
    "# Results from one supertype self projection run for MERFISH genes from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "MERFISH_reference = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 7\", \"supertype_MERFISH_iterative_scANVI_results.2022-07-01.csv\"), index_col=0)\n",
    "\n",
    "# Great Apes metadata, from https://www.science.org/doi/10.1126/science.adf6812\n",
    "great_ape_metadata = pd.read_csv(os.path.join(pwd, \"input\", \"Extended Data Figure 7\", \"Great_ApesMetadata_version101_20220321.csv\"), index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0842bd5f-116a-4725-ba4f-4564a2415f5d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 7a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac75ccf-9bff-498e-a5a6-ab14305dc386",
   "metadata": {},
   "outputs": [],
   "source": [
    "Donors = Donors.loc[Donors[\"MERFISH2\"] == \"Y\", :]\n",
    "\n",
    "Donors[\"APOE4 Status\"] = Donors[\"APOE Genotype\"].str.contains(\"4\")\n",
    "Donors[\"Co-morbidity\"] = \"None\"\n",
    "Donors.loc[(Donors[\"Overall CAA Score\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Overall CAA Score\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"CAA\"\n",
    "\n",
    "Donors.loc[(Donors[\"Highest Lewy Body Disease\"].isin([\"Neocortical (Diffuse)\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Highest Lewy Body Disease\"].isin([\"Neocortical (Diffuse)\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"LBD\"\n",
    "\n",
    "Donors.loc[(Donors[\"LATE\"].isin([\"LATE Stage 3\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"LATE\"].isin([\"LATE Stage 3\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"LATE-NC\"\n",
    "\n",
    "Donors.loc[(Donors[\"Arteriolosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Arteriolosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"Arteriolosclerosis\"\n",
    "\n",
    "Donors.loc[(Donors[\"Atherosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Atherosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"Atherosclerosis\"\n",
    "\n",
    "Donors.loc[(Donors[\"Total Microinfarcts (not observed grossly)\"] > 6) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Total Microinfarcts (not observed grossly)\"] > 6) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"Microinfarcts\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3a1544a-a7a1-47ef-a1b1-1f78496a185d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i Donors\n",
    "# Code comes from Mike H\n",
    "\n",
    "library(pheatmap)\n",
    "\n",
    "vars=colnames(Donors)\n",
    "ustage=unique(Donors[,\"Overall AD neuropathological Change\"])\n",
    "\n",
    "# bracketed by Overall AD neuropathological Change\n",
    "datnad=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"Not AD\",]\n",
    "datlow=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"Low\",]\n",
    "datmid=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"Intermediate\",]\n",
    "dathigh=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"High\",]\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"Male\",\"Female\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[1,1]=table(datnad[,\"Sex\"])[2]\n",
    "mfmat[2,1]=table(datnad[,\"Sex\"])[1]\n",
    "mfmat[1,2]=table(datlow[,\"Sex\"])[2]\n",
    "mfmat[2,2]=table(datlow[,\"Sex\"])[1]\n",
    "mfmat[1,3]=table(datmid[,\"Sex\"])[2]\n",
    "mfmat[2,3]=table(datmid[,\"Sex\"])[1]\n",
    "mfmat[1,4]=table(dathigh[,\"Sex\"])[2]\n",
    "mfmat[2,4]=table(dathigh[,\"Sex\"])[1]\n",
    "mfper=mfmat\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_barplot_Sex_by_Overall AD neuropathological Change.pdf\",width=5,height=5)\n",
    "par(mfrow=c(1,1))\n",
    "barplot(mfper,horiz=T,col=c(\"lightblue\",\"antiquewhite2\"))\n",
    "dev.off()\n",
    "\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"APOE4\",\"No APOE4\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[1,1]=table(datnad[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,1]=table(datnad[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,2]=table(datlow[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,2]=table(datlow[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,3]=table(datmid[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,3]=table(datmid[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,4]=table(dathigh[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,4]=table(dathigh[,\"APOE4 Status\"])[1]\n",
    "mfper=mfmat\n",
    "mfper[1,1]=0; mfper[1,2]=0\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"APOE4\",\"No APOE4\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[2,1]=table(datnad[,\"APOE4 Status\"])[1]\n",
    "mfmat[2,2]=table(datlow[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,3]=table(datmid[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,3]=table(datmid[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,4]=table(dathigh[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,4]=table(dathigh[,\"APOE4 Status\"])[1]\n",
    "mfper=mfmat\n",
    "mfper[1,1]=0; mfper[1,2]=0\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_barplot_APOE4 Status_by_Overall AD neuropathological Change.pdf\",width=5,height=5)\n",
    "par(mfrow=c(1,1))\n",
    "barplot(mfper,horiz=T,col=c(\"darkseagreen\",\"gray90\"))\n",
    "dev.off()\n",
    "\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"Dementia\",\"No Dementia\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[2,1]=table(datnad[,\"Cognitive Status\"])[1]\n",
    "\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_barplot_Cognitive Status_by_Overall AD neuropathological Change.pdf\",width=5,height=5)\n",
    "par(mfrow=c(1,1))\n",
    "barplot(mfper,horiz=T,col=c(\"red\",\"gray90\"))\n",
    "dev.off()\n",
    "\n",
    "par(las=2)\n",
    "\n",
    "# Braak\n",
    "ubraak=unique(Donors[,\"Braak\"])\n",
    "ustage=unique(Donors[,\"Overall AD neuropathological Change\"])\n",
    "\n",
    "brakorder=c(\"Braak 0\",\"Braak II\",\"Braak III\",\"Braak IV\",\"Braak V\",\"Braak VI\")\n",
    "stageorder=c(\"High\",\"Intermediate\",\"Low\",\"Not AD\")\n",
    "\n",
    "brakmat=matrix(0,4,6,dimnames=list(stageorder,brakorder))\n",
    "\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(brakorder)) {\n",
    "    brakmat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"Braak\"]==brakorder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "brakper=brakmat\n",
    "for (j in 1:6) {\n",
    "  brakper[,j]=round(brakper[,j]/sum(brakper[,j]),2)\n",
    "}\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_heatmap_Braak_by_Overall AD neuropathological Change.pdf\",width=6,height=4)\n",
    "mpal=colorRampPalette(c(\"white\",\"indianred\",\"brown3\",\"darkred\"))(50)\n",
    "pheatmap(brakmat,cluster_rows=F,cluster_cols=F,display_numbers=T,legend=F,number_color = \"black\",fontsize_number=12,col=mpal)\n",
    "dev.off()\n",
    " \n",
    "# Thal\n",
    "uthal=unique(Donors[,\"Thal\"])\n",
    "thalorder=c(\"Thal 0\",\"Thal 1\",\"Thal 2\",\"Thal 3\",\"Thal 4\",\"Thal 5\")\n",
    "\n",
    "thalmat=matrix(0,4,6,dimnames=list(stageorder,thalorder))\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(thalorder)) {\n",
    "    thalmat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"Thal\"]==thalorder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "thalper=thalmat\n",
    "for (j in 1:6) {\n",
    "  thalper[,j]=round(thalper[,j]/sum(thalper[,j]),2)\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_heatmap_Thal_by_Overall AD neuropathological Change.pdf\",width=6,height=4)\n",
    "mpal=colorRampPalette(c(\"white\",\"darkseagreen1\",\"darkseagreen\",\"darkseagreen4\"))(50)\n",
    "pheatmap(thalmat,cluster_rows=F,cluster_cols=F,display_numbers=T,legend=F,number_color = \"black\",fontsize_number=12,col=mpal)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "\n",
    "# CERAD Score\n",
    "ucscore=unique(Donors[,\"CERAD score\"])\n",
    "cscoreorder=c(\"Absent\",\"Sparse\",\"Moderate\",\"Frequent\")\n",
    "\n",
    "cscoremat=matrix(0,4,4,dimnames=list(stageorder,cscoreorder))\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(cscoreorder)) {\n",
    "    cscoremat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"CERAD score\"]==cscoreorder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "cscoreper=cscoremat\n",
    "for (j in 1:4) {\n",
    "  cscoreper[,j]=round(cscoreper[,j]/sum(cscoreper[,j]),2)\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_heatmap_CERAD score_by_Overall AD neuropathological Change.pdf\",width=6,height=4)\n",
    "mpal=colorRampPalette(c(\"white\",\"skyblue1\",\"deepskyblue2\",\"deepskyblue4\"))(50)\n",
    "pheatmap(cscoremat,cluster_rows=F,cluster_cols=F,display_numbers=T,legend=F,number_color = \"black\",fontsize_number=12,col=mpal)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_boxplot_Age at Death_by_Overall AD neuropathological Change.pdf\",width=6,height=8)\n",
    "par(mfrow=c(4,2))\n",
    "agerange=c(min(Donors[,\"Age at Death\"]),max(Donors[,\"Age at Death\"]))\n",
    "boxplot(datnad[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"No AD\",horizontal=T)\n",
    "stripchart(datnad[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)        \n",
    "\n",
    "boxplot(datlow[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"Low\",horizontal=T)\n",
    "stripchart(datlow[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)    \n",
    "\n",
    "boxplot(datmid[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"Intermediate\",horizontal=T)\n",
    "stripchart(datmid[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)     \n",
    "\n",
    "boxplot(dathigh[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"High\",horizontal=T)\n",
    "stripchart(dathigh[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)             \n",
    "dev.off()\n",
    "\n",
    "# Co-morbidities\n",
    "umorb=unique(Donors[,\"Co-morbidity\"])\n",
    "\n",
    "umorborder=c(\"None\",\"LBD\",\"Atherosclerosis\",\"Arteriolosclerosis\",\"Microinfarcts\",\"LATE-NC\",\"Multiple\")\n",
    "\n",
    "morbmat=matrix(0,4,7,dimnames=list(stageorder,umorborder))\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(umorborder)) {\n",
    "    morbmat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"Co-morbidity\"]==umorborder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "morbper=morbmat\n",
    "for (i in 1:4) {\n",
    "  morbper[i,]=round(morbper[i,]/sum(morbper[i,]),2)\n",
    "}\n",
    "\n",
    "pdf(\"output/Extended Data Figure 7a_barplot_Co-morbidity_by_Overall AD neuropathological Change.pdf\",width=6,height=8)\n",
    "barplot(t(morbper[4:1,]),horiz=T)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b83ab09-b078-4b36-9fd1-9fcc60a90338",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 7c-e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec16808-bca4-4d1a-a5dd-25d11c790e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1352dda2-db9b-4402-b260-db8e327efc6d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 7f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ea30afc-d867-4ef8-a968-d7d0911c9f3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "MERFISH_reference = MERFISH_reference.loc[:, [\"supertype_held_scANVI\", \"supertype_held_conf_scANVI\", \"subclass_held_scANVI\", \"subclass_held_conf_scANVI\"]]\n",
    "MERFISH_reference.columns = \"MERFISH_\" + MERFISH_reference.columns\n",
    "RNAseq_reference = RNAseq_reference.merge(MERFISH_reference, left_index=True, right_index=True, how=\"left\")\n",
    "RNAseq_reference = RNAseq_reference.merge(great_ape_metadata.loc[:, [\"subclass\", \"cluster\"]], how=\"left\", left_index=True, right_index=True)\n",
    "\n",
    "low_confidence = [\"L2/3 IT_4\", \"L2/3 IT_9\", \"L2/3 IT_11\",\n",
    "                  \"L5 IT_4\",\n",
    "                  \"L5/6 NP_5\",\n",
    "                  \"Micro-PVM_3\",\n",
    "                  \"Pvalb_4\", \"Pvalb_11\",\n",
    "                  \"Sncg_7\",\n",
    "                  \"Sst_6\", \"Sst_8\", \"Sst_14\", \"Sst_15\", \"Sst_16\", \"Sst_17\", \"Sst_18\", \"Sst_21\", \"Sst_24\", \"Sst_26\", \n",
    "                  \"Vip_3\", \"Vip_7\",  \"Vip_8\", \"Vip_10\", \"Vip_17\", \"Vip_20\", \"Vip_22\"]\n",
    "RNAseq_reference[\"supertype\"] = RNAseq_reference[\"cluster\"].copy()\n",
    "RNAseq_reference[\"supertype\"] = RNAseq_reference[\"supertype\"].astype(\"object\")\n",
    "for i in low_confidence:\n",
    "    RNAseq_reference.loc[RNAseq_reference[\"cluster\"] == i, \"supertype\"] = \"Unknown\"\n",
    "RNAseq_reference[\"supertype\"] = RNAseq_reference[\"supertype\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aee3dc6-2bb4-43a4-8cf5-d1bf95ba70e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_self_proj = RNAseq_reference.loc[:, [\"subclass\", \"subclass_held_scANVI\", \"subclass_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "df_self_proj[\"true\"] = df_self_proj[\"true\"].astype(\"category\")\n",
    "df_self_proj[\"pred\"] = df_self_proj[\"pred\"].astype(\"category\")\n",
    "result_subclass = get_scores(df=df_self_proj)\n",
    "result_subclass[\"model_version\"] = \"Subclass (Full)\"\n",
    "df_self_proj = RNAseq_reference.loc[:, [\"subclass\", \"MERFISH_subclass_held_scANVI\", \"MERFISH_subclass_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "df_self_proj[\"true\"] = df_self_proj[\"true\"].astype(\"category\")\n",
    "df_self_proj[\"pred\"] = df_self_proj[\"pred\"].astype(\"category\")\n",
    "result_merfish = get_scores(df=df_self_proj)\n",
    "result_merfish[\"model_version\"] = \"Subclass (MERFISH)\"\n",
    "result = pd.concat([result_subclass, result_merfish], axis=0)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (1.5,4)\n",
    "ax = sns.boxplot(\n",
    "    data=result,\n",
    "    x=\"model_version\",\n",
    "    y=\"f1\",\n",
    "    showfliers=False,\n",
    "    palette=\"Greys\"\n",
    ");\n",
    "ax = sns.stripplot(\n",
    "    data=result,\n",
    "    x=\"model_version\",\n",
    "    y=\"f1\",\n",
    "    color=\"0.25\",\n",
    "    edgecolor=\"white\",\n",
    "    linewidth=0.5,\n",
    "    alpha=0.5\n",
    ");\n",
    "ax.axhline(0.7, linestyle=\"dashed\", color=\"black\")\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "ax.set(xlabel=\"\", ylabel=\"F1 Score\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 7f_boxplot_Subclass F1 scores with all genes versus MERFISH panel.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "df_self_proj = RNAseq_reference.loc[:, [\"supertype\", \"supertype_held_scANVI\", \"supertype_held_conf_scANVI\"]]\n",
    "df_self_proj = df_self_proj.loc[df_self_proj[\"supertype\"] != \"Unknown\", :]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "df_self_proj[\"true\"] = df_self_proj[\"true\"].astype(\"category\")\n",
    "df_self_proj[\"pred\"] = df_self_proj[\"pred\"].astype(\"category\")\n",
    "result_supertype = get_scores(df=df_self_proj)\n",
    "result_supertype[\"model_version\"] = \"Supertype (Full)\"\n",
    "df_self_proj = RNAseq_reference.loc[:, [\"supertype\", \"MERFISH_supertype_held_scANVI\", \"MERFISH_supertype_held_conf_scANVI\"]]\n",
    "df_self_proj.columns = [\"true\", \"pred\", \"conf\"]\n",
    "df_self_proj[\"true\"] = df_self_proj[\"true\"].astype(\"category\")\n",
    "df_self_proj[\"pred\"] = df_self_proj[\"pred\"].astype(\"category\")\n",
    "result_merfish = get_scores(df=df_self_proj)\n",
    "result_merfish[\"model_version\"] = \"Supertype (MERFISH)\"\n",
    "result = pd.concat([result_supertype, result_merfish], axis=0)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (1.5,4)\n",
    "ax = sns.boxplot(\n",
    "    data=result,\n",
    "    x=\"model_version\",\n",
    "    y=\"f1\",\n",
    "    showfliers=False,\n",
    "    palette=\"Greys\"\n",
    ");\n",
    "ax = sns.stripplot(\n",
    "    data=result,\n",
    "    x=\"model_version\",\n",
    "    y=\"f1\",\n",
    "    color=\"0.25\",\n",
    "    edgecolor=\"white\",\n",
    "    linewidth=0.5,\n",
    "    alpha=0.5\n",
    ");\n",
    "ax.axhline(0.7, linestyle=\"dashed\", color=\"black\")\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "ax.set(xlabel=\"\", ylabel=\"F1 Score\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 7f_boxplot_Supertype F1 scores with all genes versus MERFISH panel.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86dfe88f-cd9f-45d9-b8df-e819dad973ba",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf9fdb61-4f7a-4423-9876-78a2fee7a06e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aec33af4-691b-40ba-a75c-f8563221f38b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdae322d-20b5-4c22-bc5c-b21e68975d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a612a79-0041-46c2-a077-61fedb64d31d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 5i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e86250d-cc58-4668-93d8-a99523ea82e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
