{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9b4001c3-96ba-4317-840a-5c91db8907a8",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d8a6c9-19ef-4b78-b0e5-d719d215c84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "from scipy import stats as sp_stats\n",
    "from helper_functions import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=100, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "281f64e8-d4ee-428e-8587-d6bbe4381051",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9f867ab-02a0-4e19-bf5a-209c0f7ea3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "\n",
    "# From 00_build_input_files.py\n",
    "Sst_Pvalb = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"SEAAD_MTG_RNAseq_final-nuclei_limited.2024-02-13.h5ad\"))\n",
    "\n",
    "# Neighborhood UMAP coordinates from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "Sst_Pvalb_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"Sst_Pvalb\", \"obs_names.csv\"), index_col=0).index\n",
    "Sst_Pvalb_umap = np.load(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"Sst_Pvalb\", \"X_umap.npy\"))\n",
    "Sst_Pvalb = Sst_Pvalb[Sst_Pvalb_obs_names, :].copy()\n",
    "Sst_Pvalb.obsm[\"X_umap\"] = Sst_Pvalb_umap\n",
    "\n",
    "# scCODA results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "results_table = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"MTG_RNAseq\", \"Continuous_Pseudo-progression_Score\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_results.csv\"), index_col=0)\n",
    "results_table = results_table.loc[results_table[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", :].loc[:, [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean()\n",
    "Sst_Pvalb.obs = Sst_Pvalb.obs.merge(results_table, left_on=\"Supertype\", right_index=True, how=\"left\")\n",
    "\n",
    "# Mean expression table from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "mean_expression = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"mean_expression.h5ad\"))\n",
    "\n",
    "# MERFISH object from https://sea-ad-spatial-transcriptomics.s3.amazonaws.com/index.html#middle-temporal-gyrus/\n",
    "MERFISH = sc.read_h5ad(os.path.join(pwd, \"input\", \"SEAAD_MTG_MERFISH.2024-02-13.h5ad\"))\n",
    "MERFISH = MERFISH[MERFISH.obs[\"Used in analysis\"] == True, :].copy()\n",
    "MERFISH.obs = MERFISH.obs.merge(results_table, left_on=\"Supertype\", right_index=True, how=\"left\")\n",
    "\n",
    "# Cell type specific marker genes from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "effect_sizes_vs_all = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"effect_sizes_vs_all.h5ad\"))\n",
    "                    \n",
    "# Effect size table from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "effect_size_table = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 6 and Extended Data Figure 11\", \"effect_size_table.csv\"), index_col=0)\n",
    "level = \"Supertype\"\n",
    "tmp = effect_size_table.loc[effect_size_table[\"Taxonomy Level\"] == level, [\"Gene\", \"Mean expression (natural log UMIs per 10k plus 1)\"]]\n",
    "tmp = tmp.groupby([\"Gene\"])[\"Mean expression (natural log UMIs per 10k plus 1)\"].transform(lambda x : sp_stats.zscore(x,ddof=1)).fillna(0)\n",
    "effect_size_table[\"Mean expression z-score\"] = tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6b28f5-af26-46e7-b878-06c22a1efe12",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 6a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d26f94e-fdf3-4be0-b8c8-35b4b1332f01",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.subsample(Sst_Pvalb, fraction=1)\n",
    "\n",
    "df = color_order.loc[color_order[\"subclass_label\"].isin([\"Sst\", \"Pvalb\"]), [\"cluster_label\", \"cluster_color\"]]\n",
    "df.index = df[\"cluster_label\"]\n",
    "df = df[\"cluster_color\"].to_dict()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    Sst_Pvalb,\n",
    "    color=\"Supertype\",\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    size=40,\n",
    "    palette=df,\n",
    "    save=\"_Sst_Pvalb_Supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Sst_Pvalb_Supertype.pdf\"), os.path.join(pwd, \"output\", \"Figure 6a_umap_Sst_Pvalb_Supertype.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    Sst_Pvalb,\n",
    "    color=\"Final Parameter\",\n",
    "    color_map=\"YlGnBu_r\",\n",
    "    sort_order=False,\n",
    "    frameon=False,\n",
    "    size=40,\n",
    "    save=\"_Sst_Pvalb_Effect_Size along CPS.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Sst_Pvalb_Effect_Size along CPS.pdf\"), os.path.join(pwd, \"output\", \"Figure 6a_umap_Sst_Pvalb_Effect_Size along CPS.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcfb252e-1030-460f-bd9f-e0ba658b87ee",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 11a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c8a9af-b986-4d37-86df-8aadacd4dcc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sst_Pvalb.obs[\"Supertype\"] = Sst_Pvalb.obs[\"Supertype\"].astype(\"category\").cat.reorder_categories(\n",
    "    [\"Sst_1\", \"Sst_4\", \"Sst_5\", \"Sst_7\", \"Sst_10\", \"Sst_13\", \"Sst_12\", \"Sst_9\",  \"Sst_19\", \"Sst_3\", \"Sst_11\", \"Sst_20\", \"Sst_22\", \"Sst_23\", \"Sst_25\", \"Sst_2\",\n",
    "     \"Pvalb_6\", \"Pvalb_5\", \"Pvalb_8\", \"Pvalb_14\", \"Pvalb_15\", \"Pvalb_2\", \"Pvalb_3\", \"Pvalb_1\", \"Pvalb_7\", \"Pvalb_12\", \"Pvalb_10\", \"Pvalb_13\", \"Pvalb_9\"],\n",
    ")\n",
    "\n",
    "sns.heatmap(\n",
    "    pd.DataFrame(mean_expression[:, Sst_Pvalb.obs[\"Supertype\"].cat.categories].X).corr(),\n",
    "    cmap=\"RdBu_r\",\n",
    "    xticklabels=Sst_Pvalb.obs[\"Supertype\"].cat.categories,\n",
    "    yticklabels=Sst_Pvalb.obs[\"Supertype\"].cat.categories,\n",
    ");\n",
    "plt.title(\"Correlation of cluster means\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 11a_Sst_Pvalb_supertype_mean_expression_correlation.pdf\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be851be4-3be9-4100-a127-fea5e76ed1d8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 11b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a54ab51-3c9c-4e43-b98d-8a1df4f35bc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "affected_supertypes = [\n",
    "    \"Sst_3\",\n",
    "    \"Sst_19\",\n",
    "    \"Sst_9\",\n",
    "    \"Sst_11\",\n",
    "    \"Sst_20\",\n",
    "    \"Sst_23\",\n",
    "    \"Sst_25\",\n",
    "    \"Sst_2\",\n",
    "    \"Pvalb_6\",\n",
    "    \"Pvalb_5\",\n",
    "    \"Pvalb_8\",\n",
    "    \"Pvalb_3\",\n",
    "    \"Pvalb_2\",\n",
    "    \"Pvalb_15\",\n",
    "    \"Pvalb_14\",\n",
    "]\n",
    "effect_sizes_vs_all.var[\"Abundance Change\"] = effect_sizes_vs_all.var[\"Supertype\"].isin(affected_supertypes)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (6,6)\n",
    "Sst = effect_sizes_vs_all[:, (effect_sizes_vs_all.var[\"Subclass\"] == \"Sst\") & (effect_sizes_vs_all.var[\"Abundance Change\"] == True)].X.mean(axis=1).tolist()\n",
    "Pvalb = effect_sizes_vs_all[:, (effect_sizes_vs_all.var[\"Subclass\"] == \"Pvalb\") & (effect_sizes_vs_all.var[\"Abundance Change\"] == True)].X.mean(axis=1).tolist()\n",
    "df = pd.DataFrame([Sst, Pvalb], index=[\"Sst\", \"Pvalb\"], columns=effect_sizes_vs_all.obs_names).T\n",
    "df[\"multiplied\"] = df[\"Sst\"] * df[\"Pvalb\"]\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Sst\",\n",
    "    y=\"Pvalb\",\n",
    "    color=\"lightgrey\",\n",
    "    alpha=0.5,\n",
    "    rasterized=True\n",
    ");\n",
    "ax.axhline(0, linestyle=\"--\", color=\"grey\");\n",
    "ax.axvline(0, linestyle=\"--\", color=\"grey\");\n",
    "ax.set(xlim=(-55,40), ylim=(-55,40));\n",
    "plt.legend(\"\", frameon=False);\n",
    "plt.xlabel(\"Enrichment (Effect Size/SE) in vuln. Sst\")\n",
    "plt.ylabel(\"Enrichment (Effect Size/SE) in vuln. Pvalb\")\n",
    "plt.title(\"Correlation=\" + str(np.round(sp_stats.pearsonr(Sst, Pvalb)[0],2)));\n",
    "texts = [plt.text(df.loc[i, \"Sst\"], df.loc[i, \"Pvalb\"], i) for i in df.loc[(np.sign(df[\"Sst\"]) == 1), :].sort_values(by=[\"multiplied\"], ascending=False).iloc[:7].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Sst\"], df.loc[i, \"Pvalb\"], i) for i in df.loc[(np.sign(df[\"Sst\"]) == -1), :].sort_values(by=[\"multiplied\"], ascending=False).iloc[:7].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Sst\"], df.loc[i, \"Pvalb\"], i) for i in df.loc[(np.sign(df[\"Sst\"]) == 1), :].sort_values(by=[\"multiplied\"], ascending=True).iloc[:5].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Sst\"], df.loc[i, \"Pvalb\"], i) for i in df.loc[(np.sign(df[\"Sst\"]) == -1), :].sort_values(by=[\"multiplied\"], ascending=True).iloc[:5].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color=\"gray\"));\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 11b_scatterplot_vulnerable_Sst_and_Pvalb_markers.pdf\"), bbox_inches=\"tight\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "143dd4e2-87f3-4079-908f-fa484e15a747",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 6b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b5bf772-5123-4e6e-aa8f-404eb39b0c5f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subclass_colors_by_supertype = color_order.loc[:, [\"cluster_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors_by_supertype.index = subclass_colors_by_supertype[\"cluster_label\"].copy()\n",
    "subclass_colors_by_supertype = subclass_colors_by_supertype[\"subclass_color\"].to_dict()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,4)\n",
    "df = MERFISH.obs.loc[\n",
    "    ([i in [\"L2/3 IT\", \"L4 IT\", \"L5 IT\", \"L6 IT\", \"L6 IT Car3\", \"L5 ET\", \"L5/6 NP\", \"L6 CT\", \"L6b\"] for i in MERFISH.obs[\"Subclass\"]]) &\n",
    "    (MERFISH.obs[\"Supertype confidence\"] > 0.4), \n",
    "     :\n",
    "]\n",
    "df[\"Supertype\"] = df[\"Supertype\"].astype(\"category\").cat.remove_unused_categories()\n",
    "ax = sns.lineplot(\n",
    "    data=df,\n",
    "    x=\"Final Parameter\",\n",
    "    y=\"Depth from pia\",\n",
    "    hue=\"Supertype\",\n",
    "    palette=subclass_colors_by_supertype,\n",
    "    estimator=np.median,\n",
    "    ci=\"sd\",\n",
    ");\n",
    "df = df.loc[:, [\"Supertype\", \"Final Parameter\", \"Depth from pia\"]].groupby([\"Supertype\"]).median()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Final Parameter\",\n",
    "    y=\"Depth from pia\",\n",
    "    hue=\"Supertype\",\n",
    "    size=10,\n",
    "    palette=subclass_colors_by_supertype,\n",
    ");\n",
    "ax.set_xlabel(\"Effect size across\\npseudoprogression\");\n",
    "ax.set_ylabel(\"Cortical depth\");\n",
    "plt.title(\"Excitatory neurons\");\n",
    "plt.axvline(x=0, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=500, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=1000, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=1700, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=2150, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=2850, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.xticks(rotation=90);\n",
    "plt.ylim(4000, 0);\n",
    "plt.xlim(0.5, -1.75);\n",
    "plt.legend('',frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 6b_scatterplot_Excitatory_Effect_Size along CPS_versus_MERFISH_layer_depth.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show();\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,4)\n",
    "df = MERFISH.obs.loc[\n",
    "    ([i in [\"Sst\", \"Pvalb\"] for i in MERFISH.obs[\"Subclass\"]]) &\n",
    "    (MERFISH.obs[\"Supertype confidence\"] > 0.4), \n",
    "     :\n",
    "]\n",
    "df[\"Supertype\"] = df[\"Supertype\"].astype(\"category\").cat.remove_unused_categories()\n",
    "ax = sns.lineplot(\n",
    "    data=df,\n",
    "    x=\"Final Parameter\",\n",
    "    y=\"Depth from pia\",\n",
    "    hue=\"Supertype\",\n",
    "    palette=subclass_colors_by_supertype,\n",
    "    estimator=np.median,\n",
    "    ci=\"sd\",\n",
    ");\n",
    "df = df.loc[:, [\"Supertype\", \"Final Parameter\", \"Depth from pia\"]].groupby([\"Supertype\"]).median()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Final Parameter\",\n",
    "    y=\"Depth from pia\",\n",
    "    hue=\"Supertype\",\n",
    "    size=10,\n",
    "    palette=subclass_colors_by_supertype,\n",
    ");\n",
    "ax.set_xlabel(\"Effect size across\\npseudoprogression\");\n",
    "ax.set_ylabel(\"Cortical depth\");\n",
    "plt.title(\"MGE inhibitory neurons\");\n",
    "plt.axvline(x=0, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=500, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=1000, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=1700, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=2150, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=2850, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.xticks(rotation=90);\n",
    "plt.ylim(4000, 0);\n",
    "plt.xlim(0.5, -1.75);\n",
    "plt.legend('',frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 6b_scatterplot_MGE_Effect_Size along CPS_versus_MERFISH_layer_depth.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show();\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,4)\n",
    "df = MERFISH.obs.loc[\n",
    "    ([i in [\"Lamp5\", \"Sncg\", \"Pax6\", \"Vip\"] for i in MERFISH.obs[\"Subclass\"]]) &\n",
    "    (MERFISH.obs[\"Supertype confidence\"] > 0.4), \n",
    "     :\n",
    "]\n",
    "df[\"Supertype\"] = df[\"Supertype\"].astype(\"category\").cat.remove_unused_categories()\n",
    "ax = sns.lineplot(\n",
    "    data=df,\n",
    "    x=\"Final Parameter\",\n",
    "    y=\"Depth from pia\",\n",
    "    hue=\"Supertype\",\n",
    "    palette=subclass_colors_by_supertype,\n",
    "    estimator=np.median,\n",
    "    ci=\"sd\",\n",
    ");\n",
    "df = df.loc[:, [\"Supertype\", \"Final Parameter\", \"Depth from pia\"]].groupby([\"Supertype\"]).median()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Final Parameter\",\n",
    "    y=\"Depth from pia\",\n",
    "    hue=\"Supertype\",\n",
    "    size=10,\n",
    "    palette=subclass_colors_by_supertype,\n",
    ");\n",
    "ax.set_xlabel(\"Effect size across\\npseudoprogression\");\n",
    "ax.set_ylabel(\"Cortical depth\");\n",
    "plt.title(\"CGE inhibitory neurons\");\n",
    "plt.axvline(x=0, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=500, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=1000, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=1700, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=2150, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.axhline(y=2850, color=\"grey\", linewidth=0.5, linestyle=\"dashed\");\n",
    "plt.xticks(rotation=90);\n",
    "plt.ylim(4000, 0);\n",
    "plt.xlim(0.5, -1.75);\n",
    "plt.legend('',frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 6b_scatterplot_CGE_Effect_Size along CPS_versus_MERFISH_layer_depth.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a4bc42-50cb-4bf2-9a73-521ed55d9889",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 11d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99a99cfe-776b-4a8f-8f9f-a57ba1c2a21d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca35765f-17fc-4d1f-b51a-36fb6b597f80",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 6e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2afa77e-f26c-4fd2-bf87-5d49de5aca76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d29497-f305-402a-888c-bddd47a0d7f4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 6f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552b2916-a103-41d7-94a2-ad2279583239",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a8fa17-6986-4fb5-815f-b1161f15b9f7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 6h,i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb89fe9-4743-4223-ba25-a071a11d1008",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.Series(effect_size_table[\"Gene\"].unique())\n",
    "# Define gene lists\n",
    "gene_lists = {}\n",
    "# Electron transport chain components, based on GO:0022900\n",
    "gene_lists[\"Electron transport chain components\"] = []\n",
    "#complexes = [\"NDUF\", \"SDH\", \"UQC\", \"COX\", \"ATP5\"]\n",
    "complexes = [\"NDUF\", \"COX\", \"ATP5\"]\n",
    "\n",
    "for j,i in enumerate(complexes):\n",
    "    gene_lists[\"Electron transport chain components\"].extend(genes[(genes.str.startswith(i)) & ~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\"))])\n",
    "\n",
    "# Glycolytic enzymes\n",
    "# From https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-400&detail-level=2\n",
    "gene_lists[\"Glycolytic enzymes\"] = []\n",
    "gene_lists[\"Glycolytic enzymes\"].extend([\n",
    "    \"HK1\", \"HK2\", \"HK3\", \"GCK\",\n",
    "    \"GPI\",\n",
    "    \"PFKL\", \"PFKM\", \"PFKP\",\n",
    "    \"ALDOA\", \"ALDOB\", \"ALDOC\",\n",
    "    \"TPI1\",\n",
    "    \"GAPDH\", \"GAPDHS\",\n",
    "    \"PGK1\", \"PGK2\",\n",
    "    \"PGAM1\", \"PGAM2\", \"PGAM4\",\n",
    "    \"BPGM\",\n",
    "    \"ENO1\", \"ENO2\", \"ENO3\", \"ENO4\",\n",
    "    \"PKLR\",\n",
    "    \"PKM\",\n",
    "])\n",
    "# Cholesterol biosynthesis\n",
    "# From https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-341\n",
    "# https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-3\n",
    "# and https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-4\n",
    "gene_lists[\"Cholesterol biosynthesis\"] = []\n",
    "gene_lists[\"Cholesterol biosynthesis\"].extend([\n",
    "    \"FDFT1\",\n",
    "    \"SQLE\",\n",
    "    \"LSS\",\n",
    "    \"CYP51A1\",\n",
    "    \"LBR\",\n",
    "    \"TM7SF2\",\n",
    "    \"MSMO1\",\n",
    "    \"NSDHL\",\n",
    "    \"HSD17B1\",\n",
    "    \"EBP\",\n",
    "    \"DHCR24\",\n",
    "    \"SC5D\",\n",
    "    \"DHCR7\",\n",
    "])\n",
    "\n",
    "# Based on UniProt Keywords\n",
    "gene_lists[\"Fatty Acid metabolism\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"uniprotkb_keyword_KW_0276_AND_reviewed_2024_02_06.tsv\"), sep=\"\\t\")\n",
    "tmp = [i.split(\" \")[0] for i in tmp[\"Gene Names\"]]\n",
    "gene_lists[\"Fatty Acid metabolism\"].extend(np.intersect1d(tmp, genes))\n",
    "\n",
    "gene_lists[\"Sphingolipid metabolism\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"uniprotkb_keyword_KW_0746_AND_reviewed_2024_02_06.tsv\"), sep=\"\\t\")\n",
    "tmp = [i.split(\" \")[0] for i in tmp[\"Gene Names\"]]\n",
    "gene_lists[\"Sphingolipid metabolism\"].extend(np.intersect1d(tmp, genes))\n",
    "\n",
    "gene_lists[\"Steroid metabolism\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"uniprotkb_keyword_KW_0753_AND_reviewed_2024_02_06.tsv\"), sep=\"\\t\")\n",
    "tmp = [i.split(\" \")[0] for i in tmp[\"Gene Names\"]]\n",
    "gene_lists[\"Steroid metabolism\"].extend(np.intersect1d(tmp, genes))\n",
    "\n",
    "gene_lists[\"Phospholipid metabolism\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"uniprotkb_keyword_KW_1208_AND_reviewed_2024_02_06.tsv\"), sep=\"\\t\")\n",
    "tmp = [i.split(\" \")[0] for i in tmp[\"Gene Names\"]]\n",
    "gene_lists[\"Phospholipid metabolism\"].extend(np.intersect1d(tmp, genes))\n",
    "\n",
    "gene_lists[\"Lipid metabolism\"] = np.union1d(np.union1d(np.union1d(np.union1d(gene_lists[\"Cholesterol biosynthesis\"], gene_lists[\"Fatty Acid metabolism\"]), gene_lists[\"Sphingolipid metabolism\"]), gene_lists[\"Steroid metabolism\"]), gene_lists[\"Phospholipid metabolism\"])\n",
    "\n",
    "# Ribosomal proteins, based on GO:0006412\n",
    "gene_lists[\"Ribosomal proteins\"] = []\n",
    "gene_lists[\"Ribosomal proteins\"].extend([i for i in genes if re.match(\"RP[SL]([0-9]{1,2}[ABLXY]?[0-9]?|A|P[0-9]{1})$\", i) != None])\n",
    "\n",
    "# Eukaryotic initiation, elongation, and termination factors, based on GO:0006412\n",
    "gene_lists[\"Translation factors\"] = []\n",
    "gene_lists[\"Translation factors\"].extend([i for i in genes if re.match(\"^E[IET]F[0-9]{1,2}[A-Z]?[0-9]?[BL]?$\", i) != None])\n",
    "\n",
    "# DNA-templated transcription, based on GO:0006351\n",
    "gene_lists[\"Transcription machinery\"] = []\n",
    "gene_lists[\"Transcription machinery\"].extend(genes[(genes.str.startswith(\"POLR2\"))])\n",
    "gene_lists[\"Transcription machinery\"].extend(genes[(genes.str.startswith(\"GTF2\")) & ~(genes.str.contains(\"RD\"))])\n",
    "gene_lists[\"Transcription machinery\"].extend(genes[(genes.str.startswith(\"MED\")) & ~(genes.str.contains(\"OS\"))])\n",
    "gene_lists[\"Transcription machinery\"].extend([\n",
    "    \"TFB\", # General transcription factor\n",
    "    \"CCNC\", # Mediator\n",
    "])\n",
    "gene_lists[\"Transcription machinery\"] = np.intersect1d(gene_lists[\"Transcription machinery\"], genes[~(genes.str.contains(\"-AS\"))])\n",
    "\n",
    "# Protein ubiquitination, based on UbiNet 2.0 (https://academic.oup.com/database/article/doi/10.1093/database/baab010/6162640)\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"Categorized_human_E3_ligases.csv\"))\n",
    "for i in tmp[\"Category\"].unique():\n",
    "    gene_lists[\"Ubiquitination-E3-\" + i] = []\n",
    "    gene_lists[\"Ubiquitination-E3-\" + i].extend(np.intersect1d(tmp.loc[tmp[\"Category\"] == i, \"E3\"], genes))\n",
    "\n",
    "gene_lists[\"Ubiquitin\"] = []\n",
    "gene_lists[\"Ubiquitin\"].extend(genes[(genes.str.startswith(\"UBA\"))]) # Ubiquitin E1 ligases\n",
    "gene_lists[\"Ubiquitin\"].extend([\n",
    "    \"UBB\", \"UBC\", \"RPS27A\", # Ubiquitin\n",
    "])\n",
    "gene_lists[\"Ubiquitination-E1E2\"] = []\n",
    "gene_lists[\"Ubiquitination-E1E2\"].extend([\n",
    "    \"UBB\", \"UBC\", \"RPS27A\", # Ubiquitin\n",
    "    \"NAE1\", \"SAE1\", \"ATG7\", # E1 ligases\n",
    "    \"UFC1\", \"BIRC6\", \"ATG3\", # E2 Ligases\n",
    "])\n",
    "                                            \n",
    "gene_lists[\"Ubiquitination\"] = []\n",
    "for i in gene_lists.keys():\n",
    "    if \"Ubiquitin\" in i:\n",
    "        gene_lists[\"Ubiquitination\"] = np.union1d(gene_lists[\"Ubiquitination\"], gene_lists[i]).tolist()\n",
    "                  \n",
    "# Protein phosphorylation, based on KinHub (https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1433-7)\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"KinHub_database.csv\"))\n",
    "for i in tmp[\"Group\"].unique():\n",
    "    gene_lists[\"Phosphorylation-\" + i] = []\n",
    "    gene_lists[\"Phosphorylation-\" + i].extend(np.intersect1d(tmp.loc[tmp[\"Group\"] == i, \"HGNC Name\"].to_list(), genes))\n",
    "\n",
    "gene_lists[\"Phosphorylation\"] = []\n",
    "for i in gene_lists.keys():\n",
    "    if \"Phosphorylation-\" in i:\n",
    "        gene_lists[\"Phosphorylation\"] = np.union1d(gene_lists[\"Phosphorylation\"], gene_lists[i]).tolist()\n",
    "    \n",
    "# Voltage gated ion channels, from Guide to Pharmacology\n",
    "gene_lists[\"Voltage gated ion channels\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"GtP_receptors.tsv\"), sep=\"\\t\")\n",
    "gene_lists[\"Voltage gated ion channels\"].extend(np.intersect1d(tmp.loc[tmp[\"Type\"] == \"vgic\", \"HGNC symbol\"].dropna().unique(), genes))\n",
    "\n",
    "# Ligand gated ion channels, from Guide to Pharmacology\n",
    "gene_lists[\"Ligand gated ion chanels\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"GtP_receptors.tsv\"), sep=\"\\t\")\n",
    "gene_lists[\"Ligand gated ion chanels\"].extend(np.intersect1d(tmp.loc[tmp[\"Type\"] == \"lgic\", \"HGNC symbol\"].dropna().unique(), genes))\n",
    "\n",
    "# Nuclear hormone receptors, from Guide to Pharmacology\n",
    "gene_lists[\"Nuclear hormone receptors\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"GtP_receptors.tsv\"), sep=\"\\t\")\n",
    "gene_lists[\"Nuclear hormone receptors\"].extend(np.intersect1d(tmp.loc[tmp[\"Type\"] == \"nhr\", \"HGNC symbol\"].dropna().unique(), genes))\n",
    "\n",
    "# Transcription factors, from Animal TFDB\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"AnimalTFDB.tsv\"), sep=\"\\t\")\n",
    "for i in tmp[\"Family\"].unique():\n",
    "    gene_lists[\"Transcription factors-\" + i] = []\n",
    "    gene_lists[\"Transcription factors-\" + i].extend(np.intersect1d(tmp.loc[tmp[\"Family\"] == i, \"Symbol\"].to_list(), genes))\n",
    "\n",
    "gene_lists[\"Transcription factors\"] = []\n",
    "for i in gene_lists.keys():\n",
    "    if \"Transcription factors-\" in i:\n",
    "        gene_lists[\"Transcription factors\"] = np.union1d(gene_lists[\"Transcription factors\"], gene_lists[i]).tolist()\n",
    "    \n",
    "# GWAS genes, from Bellenguez et al (2022).\n",
    "gene_lists[\"GWAS\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"AD_GWAS.csv\"))\n",
    "gene_lists[\"GWAS\"].extend(np.intersect1d(tmp[\"Gene\"], genes))\n",
    "\n",
    "# Cell adhesion, based on GO:0007155\n",
    "gene_lists[\"Cell adhesion\"] = []\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CNTN\")) & ~(genes.str.contains(\"AP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"NINJ\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"NLGN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"DLG\")) & ~(genes.str.contains(\"AP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"ADGR\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"PKP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"TENM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"DSCAM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"NECTIN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CADM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"SIGLEC\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CTNN\")) & ~(genes.str.contains(\"IP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"ICAM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CEACAM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CLDN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CLSTN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CELSR\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"DSG\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CDH\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"PCDH\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"ITG\")) & ~(genes.str.contains(\"BP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend([\n",
    "    \"DSC1\", \"DSC2\", \"DSC3\",\n",
    "    \"DCHS1\", \"DCHS2\",\n",
    "    \"FAT1\", \"FAT2\", \"FAT3\", \"FAT4\",\n",
    "    \"SELE\", \"SELL\", \"SELP\",\n",
    "    \"VCAM1\", \"PECAM1\", \"MADCAM1\"\n",
    "    \"L1CAM\", \"NRCAM\", \"CHL1\", \"NFASC\",\n",
    "    \"MAG\", \"F11R\", \"JAM2\", \"JAM3\", \"JAML\",\n",
    "    \"VSIG1\", \"VSIG2\", \"IGSF11\", \"VSIG4\", \"ESAM\",\n",
    "    \"CXADR\", \"PVR\", \"CD2\", \"HEPACAM\", \"HEPACAM2\",\n",
    "    \"BCAM\", \"EPCAM\", \"ALCAM\", \"CERCAM\", \"MCAM\",\n",
    "    \"UNC5A\", \"UNC5B\", \"UNC5C\", \"UNC5D\",\n",
    "    \"VCL\", \"JUP\", \"MSN\", \"CD209\", \n",
    "])\n",
    "gene_lists[\"Cell adhesion\"] = np.intersect1d(gene_lists[\"Cell adhesion\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "\n",
    "# Actin-Spectrin-Septin cytoskeletal components and modulators, based on GO:0005200, GO:0003779, GO:0030507, and GO:0031105\n",
    "gene_lists[\"Actin cytoskeleton\"] = []\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"ACT\"))]) # Microfilament (Actin and Actinin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"POTE\"))]) # Microfilament (POTEs)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"AFAP\"))]) # Microfilament (Actin associated proteins)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"CORO\"))]) # Microfilament (Coronin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"FLN\"))]) # Microfilament (Filamin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"LMOD\"))]) # Microfilament (Leiomodin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"TMOD\"))]) # Microfilament (Tropomodulin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SNT\"))]) # Microfilament (Syntrophin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"TNN\")) & ~(genes.str.contains(\"K\"))]) # Microfilament (Troponin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SSH\"))]) # Microfilament (Slingshot)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"PARV\"))]) # Microfilament (Parvin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"MICAL\"))]) # Microfilament (F-Actin-Monooxygenase)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"TNS\"))]) # Microfilament (Tensin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"PLEK\"))]) # Microfilament (Pleckstrin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SYNE\"))]) # Microfilament (Nesprin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SYNPO\"))]) # Microfilament (Synaptopodin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"PFN\"))]) # Microfilament (Profilin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"ARP\")) & ~(genes.str.contains(\"PP\"))]) # Microfilament (ARP2/3)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"FMN\"))]) # Microfilament (Formin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"FSCN\"))]) # Microfilament (Fascin)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SEPTIN\"))]) # Septin\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SPTA\"))]) # Spectrin\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SPTB\"))]) # Spectrin\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SORBS\"))]) # Sorbin\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"KLHL\"))]) # Microfilament (Kelch like)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"PDLIM\"))]) # Microfilament (PDZ and LIM containing)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"WAS\")) & ~(genes.str.contains(\"IR\"))]) # Microfilament (Actin binders/remodelers)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"PHACTR\"))]) # Microfilament (Phosphatase and Actin Regulator 4)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SHANK\"))]) # Microfilament (SH3 And Multiple Ankyrin Repeat Containing)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"EPB41\"))]) # Microfilament (Erythrocyte Membrane Protein 4.1)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"BAIAP\"))]) # Microfilament (Brain-specific angiogenesis Inhibitor)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"GMF\"))]) # Microfilament (Glial maturation factors)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"SHROOM\"))]) # Microfilament (SHROOM/Actin remodeler)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"GAS2\"))]) # Microfilament (Growth arrest specific 2)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend(genes[(genes.str.startswith(\"FGD\"))]) # Microfilament (FYVE, RhoGEF And PH Domain Containing)\n",
    "gene_lists[\"Actin cytoskeleton\"].extend([\n",
    "    \"ABL1\", \"ABL2\",\n",
    "    \"VASP\", \"INF2\", \"FHDC1\", \"DIAPH1\", \"DIAPH2\", \"DIAPH3\", # Microfilament (Filament end tracking)\n",
    "    \"PLS1\", \"LCP1\", \"PLS3\", \"TMSB4X\", \"TMSB4Y\", # Microfilament (Cross-linker)\n",
    "    \"CAP1\", \"CAP2\", \"CAPG\", \"CAPZA1\", \"CAPZA2\", \"CAPZA3\", \"CAPZB\", \"TWF1\", \"TWF2\", # Microfilament (Capping proteins)\n",
    "    \"GSN\", \"SCIN\",  \"CFL1\", \"CFL2\", \"DTSN\", # Microfilament (Filament severing, depolymerizing)\n",
    "    \"ANK1\", \"ANK2\", \"ANK3\", # Microfilament (Ankyrin)\n",
    "    \"CNN1\", \"CNN2\", \"CNN3\", # Microfilament (Calponin)\n",
    "    \"TTN\", \"MOYZ1\", \"MYOZ2\", \"MYOZ3\", \"MYORG\", \"MYOM1\", \"MYOM2\", \"MYOM3\", # Titin complex\n",
    "    \"VIL\", \"VIL1\", \"VILL\", \"SVIL\", \"AVIL\", \"EVL\", \"ENAH\", \"DMTN\", \"ABLIM1\", \"ABLIM2\", \"ABLIM3\", \"DBN1\", \"DBNL\", \"ALKBH4\", \"UTRN\", # Actin binders/remodelers\n",
    "    \"WDR1\", \"COTL1\", \"ANLN\", \"ADD1\", \"ADD2\", \"ADD3\", \"MYOT\", \"MYPN\", \"PALLD\", \"DAAM1\", \"DAAM2\", \"TLN1\", \"TLN2\", # Actin binders/remodelers\n",
    "    \"WIPF1\", \"WIPF2\", \"WIPF3\", \"IPP\", \"FAM107A\", \"FAM107B\", \"DMD\", \"EMD\", \"CSRP1\", \"CSRP2\", \"CSRP3\", \"JMY\", # Actin binders/remodelers\n",
    "    \"EPB41L1\", \"EZR\", \"CTTN\", \"DAG1\", \"EPS8\", \"EPS8L1\", \"EPS8L2\", \"EPS8L3\", \"ENC1\", \"RDX\", \"MSN\", \"WHAMM\", \"CGN\", # Actin binders/remodelers\n",
    "    \"CCDC88A\", \"MICAL3\", \"MAEA\", \"SETD3\", \"NEB\", \"NRAP\", \"PPP1R18\", \"SPIRE1\", \"SPIRE2\", \"MACF1\", \"DST\", \"PLEC\", \"TAGLN\", \"FKBP15\",  # Actin binders/remodelers\n",
    "])\n",
    "gene_lists[\"Actin cytoskeleton\"] = np.intersect1d(gene_lists[\"Actin cytoskeleton\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\"))])\n",
    "\n",
    "# Intermediate filament/Neurofilament cytoskeleton, based on GO:0005200, GO:0045111, and GO:0019215\n",
    "gene_lists[\"Intermediate filament\"] = []\n",
    "gene_lists[\"Intermediate filament\"].extend(genes[(genes.str.startswith(\"LMN\"))]) # Lamins\n",
    "gene_lists[\"Intermediate filament\"].extend([\n",
    "    \"DES\", \"GFAP\",  \"VIM\", \"SYNC\", \"SYNM\", # Desmin, GFAP, Peripherin, Vimentin, Syncoilin, Synemin\n",
    "    \"INA\", \"PRPH\", \"NEFL\", \"NEFM\", \"NEFH\",  # Neurofilament\n",
    "    \"BFSP1\", \"BFSP2\", \"NES\", # Beaded filaments\n",
    "])\n",
    "gene_lists[\"Intermediate filament\"] = np.intersect1d(gene_lists[\"Intermediate filament\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\"))])\n",
    "\n",
    "# Microtubule cytoskeleton, based on GO:0005200, GO:0015630, and GO:0008017\n",
    "gene_lists[\"Microtubule cytoskeleton\"] = []\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TUBA\"))]) # Tubulin\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TUBB\"))]) # Tubulin\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TUBD\"))]) # Tubulin\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TUBE\"))]) # Tubulin\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TUBG\"))]) # Tubulin\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"CROC\"))]) # Ciliary Rootlet Coiled-Coil, Rootletin Famil\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"DCDC\"))]) # Doublecortin domain containing\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"SPEF\"))]) # Sperm flagellar protein\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TPG\"))]) # Tubulin Polyglutamylase Complex\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"RMD\"))]) # Regulator Of Microtubule Dynamics\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"CKAP\"))]) # Cytoskeleton Associated Protein\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"EML\"))]) # Echinoderm Microtubule Associated Protein Like\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"HAUS\"))]) # HAUS Augmin Like Complex\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"MAST\"))]) # Microtubule Associated Serine/Threonine Kinase\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"MAPRE\"))]) # Microtubule Associated Protein RP/EB Family\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"DLGAP\"))]) # Discs Large Homolog Associated Protein\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TBC\")) & ~(genes.str.startswith(\"TBC1\"))]) # Tubulin Folding Cofactor\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"CAMSAP\"))]) # Calmodulin Regulated Spectrin Associated Protein Family\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"MAP1L\"))]) # Microtubule associated protein\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"NDE\"))]) # NudE Neurodevelopment Protein\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"NUDC\"))]) # NudC\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"CETN\"))]) # Centrin\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TTLL\"))]) # Tubulin Polyglutamylase Complex\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"NEK\"))]) # NIMA Related Kinase\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"TPPP\"))]) # Tubulin Polymerization Promoting Protein Family\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"ODF\"))]) # Outer Dense Fiber Of Sperm Tails\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend(genes[(genes.str.startswith(\"KATN\"))]) # Katanin Regulatory Subunit\n",
    "gene_lists[\"Microtubule cytoskeleton\"].extend([\n",
    "    \"MAP1A\", \"MAP1B\", \"MAP1S\", \"MAP10\", \"MAP2\", \"MAPT\", \"MAP4\", \"MAP6\", \"MAP6D1\", \"MAP7\", \"MAP9\"  # Microtubule associated protein\n",
    "    \"SPC24\", \"SPC25\", \"NDC80\", \"NUF2\", # Component Of NDC80 Kinetochore Complex\n",
    "    \"TOGARAM1\", \"TOGARAM2\" # TOG Array Regulator Of Axonemal Microtubules\n",
    "    \"MTUS1\", \"MTUS2\", # Microtubule Associated Scaffold Protein\n",
    "    \"NUMA1\", # Nuclear Mitotic Apparatus Protein\n",
    "    \"SKA1\", \"SKA2\", \"SKA3\", # Spindle And Kinetochore Associated Complex\n",
    "    \"DCX\", # Doublecortin\n",
    "    \"SPATA4\", \"SPAG6\", \"SPAG17\", # Microtubule related sperm proteins\n",
    "    \"APC\", \"APC2\", # Adenomatous polyposis coli\n",
    "    \"PRC1\", # Protein Regulator Of Cytokinesis\n",
    "    \"SAXO1\", \"SAXO2\", # Stabilizer Of Axonemal Microtubules\n",
    "    \"CCDC69\", \"CEP57\", \"CEP57L1\", \"CRIPT\", \"MACO1\", \"CCDC181\", \"MAST2\", \"ALMS1\", \"CCDC66\", # Microtubule binding proteins\n",
    "    \"CCDC170\", \"ABRAXAS1\", \"ABRAXAS2\", \"CCSAP\", \"CFAP157\", \"CEP295NL\", \"CEP295\", \"CCSER2\", \"CCSER1\" # Microtubule binding proteins\n",
    "    \"FSD2\", \"MID1\", \"MID2\", \"CDK5RAP2\", \"NUSAP1\", \"MTCL1\", \"SPAST\", \"RGS14\", \"FIGN\", \"FIGNL1\", \"FIGNL2\", # Microtubule binding proteins\n",
    "    \"NAV1\", \"NAV2\", \"NAV3\", \"RACGAP1\", \"ZNF207\", \"TPGS1\", \"TPGS2\", \"HDAC6\", \"FKBP4\", \"RCC2\", \"PCNT\", # Microtubule binding proteins\n",
    "    \"RP1\", \"RP1L1\", \"FGF13\", \"CCDC66\", \"NDRG1\", \"MDM1\", \"SIRT2\", \"FAM110C\", \"CEP170\", \"MATCAP1\", \"MATCAP2\" # Microtubule binding proteins\n",
    "    \"CSTPP1\", \"LRRC49\", \"MID1IP1\", \"INO80\", \"FSD1\", \"LZTS2\", \"BEX4\", \"FEZ1\", \"FEZ2\", \"NICN1\", \"NIN\", \"NINL\" # Microtubule binding proteins\n",
    "    \"ARL2\", \"CLASP1\", \"CLASP2\"\n",
    "\n",
    "])\n",
    "gene_lists[\"Microtubule cytoskeleton\"] = np.intersect1d(gene_lists[\"Microtubule cytoskeleton\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "# Vesicle trafficking, based on https://www.sciencedirect.com/science/article/pii/S0896627310007816\n",
    "gene_lists[\"Vesicle trafficking\"] = []\n",
    "gene_lists[\"Vesicle trafficking\"].extend(genes[(genes.str.startswith(\"KIF\"))]) # Kinesin\n",
    "gene_lists[\"Vesicle trafficking\"].extend(genes[(genes.str.startswith(\"DYN\"))]) # Dyenin\n",
    "gene_lists[\"Vesicle trafficking\"].extend(genes[(genes.str.startswith(\"DNAH\")) & ~(genes.str.contains(\"OS\"))]) # Dyenin\n",
    "gene_lists[\"Vesicle trafficking\"].extend(genes[(genes.str.startswith(\"DNAI\"))]) # Dyenin\n",
    "gene_lists[\"Vesicle trafficking\"].extend(genes[(genes.str.startswith(\"MYH\"))]) # Kinesin\n",
    "gene_lists[\"Vesicle trafficking\"].extend(genes[(genes.str.startswith(\"MYL\"))]) # Kinesin\n",
    "gene_lists[\"Vesicle trafficking\"].extend([\n",
    "    \"MYO10\", \"MYO15A\", \"MYO15B\", \"MYO16\", \"MYO18A\", # Myosin\n",
    "    \"MYO18B\", \"MYO19\", \"MYO1A\", \"MYO1B\", \"MYO1C\", \"MYO1D\", \"MYO1E\", # Myosin\n",
    "    \"MYO1F\", \"MYO1G\", \"MYO1H\", \"MYO3A\", \"MYO3B\", \"MYO5A\", \"MYO5B\", # Myosin\n",
    "    \"MYO5C\", \"MYO6\", \"MYO7A\", \"MYO7B\", \"MYO9A\", \"MYO9B\", # Myosin\n",
    "    \"STARD9\" # Others\n",
    "])\n",
    "gene_lists[\"Vesicle trafficking\"] = np.intersect1d(gene_lists[\"Vesicle trafficking\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "# Adapator proteins/trafficking proteins, based on https://www.sciencedirect.com/science/article/pii/S0896627310007816, GO:0030705, and GO:0016192\n",
    "gene_lists[\"Trafficking adaptors\"] = []\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"TRAK\"))]) # Milton\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"RHOT\"))]) # Miro\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"GRIP\"))]) # AMPA receptor\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SYT\"))]) # Synaptogamin, for TrkB+ vesicles\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"RAB\"))]) # RABs\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"STX\"))]) # Syntaxin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SNAP\")) & ~(genes.str.startswith(\"SNAPC\"))]) # SNARE\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"VAMP\"))]) # Synaptobrevin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"VAP\"))]) # Synaptobrevin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"CLIP\"))]) # CAP-Gly Domain Containing Linker\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"GIPC\"))]) # GIPC PDZ Domain Containing Family\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"AP1\"))]) # Adapter Protein Complex 1\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"AP2\"))]) # Adapter Protein Complex 2\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"AP3\"))]) # Adapter Protein Complex 3\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"AP4\"))]) # Adapter Protein Complex 4\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"DTN\"))]) # Dystobrevin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"BLOC\"))]) # Biogenesis Of Lysosomal Organelles Complex\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"BORC\"))]) # Biogenesis Of Lysosomal Organelles Complex\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SNX\"))]) # Adapter Protein Complex 4\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"ARF\"))]) # ADP Ribosylation Factor\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"RIMS\"))]) # Regulating Synaptic Membrane Exocytosis\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"COP\")) & ~(genes.str.startswith(\"COPS\"))]) # COPI\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SEC13\"))]) # COPII\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SEC16\"))]) # COPII\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SEC2\"))]) # COPII\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SEC3\"))]) # COPII\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"PACSIN\"))]) # Protein Kinase C And Casein Kinase Substrate In Neurons\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SYNGR\"))]) # Synaptogyrin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SNC\"))]) # Synuclein\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"COG\"))]) # COG\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"VTI\"))]) # VTI\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"HOOK\"))]) # HOOK\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"CYTH\"))]) # Cytohesin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"TRAPP\"))]) # TRAPP Complex\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"GOLG\")) & ~(genes.str.startswith(\"GOLGA6\")) & ~(genes.str.startswith(\"GOLGA7\")) & ~(genes.str.startswith(\"GOLGA8\"))]) # Golgin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"LIN7\"))]) # NMDA receptor cargo adaptor\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"DCTN\"))]) # Dynactin\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"TMED\"))]) # TMED\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"UNC13\"))]) # UND13\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"GOSR\"))]) # Golgi SNAP\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"WASHC\"))]) # WASH complex\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"SCAMP\"))]) # SCAMP\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"EXOC\"))]) # Exocyst\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"TBC1D\"))]) # TBC1D Family\n",
    "gene_lists[\"Trafficking adaptors\"].extend(genes[(genes.str.startswith(\"GOLPH\"))]) # Golgi Phosphoproteins\n",
    "\n",
    "\n",
    "gene_lists[\"Trafficking adaptors\"].extend([\n",
    "    \"MADD\", \"PPFIA1\", \"PPFIA2\", \"SYP\", # Pre-synaptic vesicle adaptor\n",
    "    \"KIFBP\", \"SYBU\", \"MGARP\", # Mitochondria\n",
    "    \"MAPK8IP1\", \"MAPK8IP2\", \"MAPK8IP3\", \"APP\", # Vesicles\n",
    "    \"HAP1\", # GABA receptor vesicles\n",
    "    \"HTT\", # BDNF vesicles\n",
    "    \"HSPA8\", # Slow cargo\n",
    "    \"SPEN\", # NR2B vesicles,\n",
    "    \"GPHN\", # Glycine vesicles\n",
    "    \"RILP\", # NGF-TrkA Endosomes\n",
    "    \"BSN\", \"PCLO\", \"CPLX1\", \"CPLX2\", \"CPLX3\", \"CPLX4\",  # Active-zone-protein veiscles\n",
    "    \"DBNDD1\", \"DBNDD2\", # Dystobrevin\n",
    "    \"PEX5\", \"PEX14\", # Perioxisomes\n",
    "    \"F8A1\", \"TMEM108\", \"TANC1\", \"TANC2\", \"STARD3\", \"STARD3NL\",\n",
    "    \"FYCOV1\", \"RUFY1\", \"RUFY2\", \"RUFY3\", \"RUFY4\",\n",
    "    \"CLN3\", \"CLN5\", \"BECN1\", \"BECN2\", \"ARL1\", \"SYCN\", \"TMEM87A\",\n",
    "    \"VPS26A\", \"VPS26B\", \"VPS29\", \"VPS35\", \"VPS35L\", # Retromer\n",
    "    \"VPS11\", \"VPS16\", \"VPS18\", \"VPS33A\", \"VPS8\", \"VPS39\", \"VPS41\", #  CORVET And HOPS Complexes\n",
    "    \"VPS33B\", \"VPS50\", \"VPS51\", \"VPS52\", \"VPS53\", \"VPS54\",\n",
    "    \"CCDC93\", \"CCDC91\", \"EHBP1\", \"EHBP1L1\", \"RBSN\", \"EXPH5\", \"MLPH\", \"MYRIP\", \"OPTN\",\n",
    "    \"BLTP1\", \"DNM1\", \"DNM2\", \"DNM3\", \"TMEM115\", \"APBA1\", \"APBA2\", \"APLP1\"\n",
    "    \"ANKFY1\", \"RGP1\", \"SAR1A\", \"SAR1B\", \"SCRN1\", \"SCRN2\", \"SCRN3\", \"BAIAP3\",\n",
    "    \"PHETA1\", \"PHETA2\", \"RHOBTB1\", \"RHOBTB2\", \"RHOBTB3\", \"RIN1\", \"RIN2\", \"RIN3\", \n",
    "    \"ENTR1\", \"ITSN1\", \"ITSN2\", \"SCFD1\", \"SCFD2\", \"SORT1\", \"SORL1\", \"RIC1\", \n",
    "    \"LAMP1\", \"GCC1\", \"GCC2\", \"AHI1\", \"LYST\", \"NBEA\", \"TMEM108\", \"LLGL1\", \"LLGL2\",\n",
    "    \"STON1\", \"STON2\", \"ZFYVE16\", \"ZFYVE9\", \"ZFYVE27\", \"TFG\", \"RUBCN\", \"RUBCNL\", \"BICD1\", \"BICD2\", \"EEA1\", \n",
    "    \"OTOF\", \"GBF1\", \"FAM91A1\", \"WDR11\", \"TMCC1\", \"TMCC2\", \"FCHSD1\", \"FCHSD2\",\n",
    "    \"USP6NL\", \"NBAS\", \"PREB\", \"GOPC\", \"BRSK1\", \"BRSK2\", \"EIPR1\", \"TRIP11\", \"TRIP10\",\n",
    "    \"DOC2A\", \"DOC2B\", \"C17orf75\", \"GGA1\", \"GGA2\", \"GGA3\", \"C9orf72\", \"PICK1\",\n",
    "    \"RILP\", \"RILPL1\", \"RILPL2\", \"SYNRG\", \"DAB1\", \"DAB2\", \"GAPVD1\", \"LAMTOR1\", \"SURF4\",\n",
    "    \"TOM1\", \"TOM1L2\", \"USO1\", \"SNPH\", \"NSF\", \"VTA1\", \"HIP1\", \"HIP1R\"\n",
    "    \n",
    "    \n",
    "]) \n",
    "gene_lists[\"Trafficking adaptors\"] = np.intersect1d(gene_lists[\"Trafficking adaptors\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "# Axonal guidance, based on GO:0097485\n",
    "gene_lists[\"Axonal guidance\"] = []\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"SEMA\"))]) # Semaphorin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"UNC5\")) & ~(genes.str.contains(\"50\"))]) # Netrin receptor\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"EPH\"))]) # Ephrin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"EFN\"))]) # Ephrin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"LGI\"))]) # Leucine Repeat (Slit-like)\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NOTCH\"))]) # Notch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"DLL\"))]) # Notch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"JAG\"))]) # Notch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"SLIT\"))]) # Slit\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"LRTM\"))]) # Slit-like\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"ROBO\"))]) # Robo\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"PLXN\"))]) # Plexin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FLRT\"))]) # Fibronectin Leucine Rich Transmembrane\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"VEGF\"))]) # Vascular endothelial growth factor\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NCAM\"))]) # NCAM\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"CNTN\")) & ~(genes.str.contains(\"AP\"))]) # Contactin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"PTCH\"))]) # Patch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FEZ\"))]) # FEZ\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"ERBB\"))]) # Erb-B2 Receptor Tyrosine Kinase\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FZD\"))]) # Fzd\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NRP\"))]) # Neuropilin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"BMP\"))]) # BMP\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FGF\")) & ~(genes.str.contains(\"P\"))]) # FGF\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NTN\"))]) # Netrin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"LGR\"))]) # LGRs\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"WNT\"))]) # Wnt\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NTRK\"))]) # Ntrk\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NECTIN\"))]) # Nectin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"DSCAM\"))]) # DSCAM\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NRXN\"))]) # Neurexin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"GFR\"))]) # GDNF Receptor\n",
    "\n",
    "# Based on https://elifesciences.org/articles/47889\n",
    "gene_lists[\"Neuropeptides\"] = []\n",
    "gene_lists[\"Neuropeptides\"].extend([\n",
    "    \"SST\", \"CORT\", \"NPY\", \"GRP\", \"CCK\", \"NTS\", \"NMB\", \"RLN1\", \"TRH\", \"ADCYAP1\",\n",
    "    \"VIP\", \"CRH\", \"PNOC\", \"TAC1\", \"TAC3\", \"PDYN\", \"PTHLH\", \"PENK\"\n",
    "])\n",
    "\n",
    "gene_lists[\"Neuropeptide receptors\"] = []\n",
    "gene_lists[\"Neuropeptide receptors\"].extend([\n",
    "    \"NPY1R\", \"NPY2R\", \"NPY5R\", \"SSTR1\", \"SSTR2\", \"SSTR3\", \"SSTR4\", \"SSTR5\",\n",
    "    \"VIPR1\", \"VIPR2\", \"TACR1\", \"TACR2\", \"TACR3\", \"OPRD1\", \"OPRM1\", \"OPRK1\",\n",
    "    \"CRHR1\", \"CRHR2\", \"PTH1R\", \"OPRL1\", \"TRHR\", \"GRPR\", \"ADCYAP1R1\",\n",
    "    \"NTSR1\", \"NTSR2\", \"NMBR\"\n",
    "])\n",
    "\n",
    "# Myelin components and factors, from https://www.ncbi.nlm.nih.gov/books/NBK28221/#:~:text=The%20protein%20composition%20of%20CNS,are%20present%20to%20lesser%20extents.\n",
    "gene_lists[\"Myelin components\"] = []\n",
    "gene_lists[\"Myelin components\"].extend([\n",
    "    \"MOG\", \"MAG\", \"MBP\", \"MOBP\", \"MPZ\", \"OMG\", \"CNP\", \"PLP1\", \"PLLP\", \"OPALIN\"\n",
    "])\n",
    "\n",
    "# Re-myelination program, based on https://www.mdpi.com/2218-273X/11/2/283, https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3892435/, and https://www.sciencedirect.com/science/article/pii/S0896627300805152, and https://linkinghub.elsevier.com/retrieve/pii/S0962892415001646\n",
    "gene_lists[\"Remyelination\"] = []\n",
    "gene_lists[\"Remyelination\"].extend([\n",
    "    \"OLIG1\", \"OLIG2\", \"SOX10\", \"SOX8\", \"CSPG4\", \"DLL1\", \"DLL2\", \"DLL3\", \"ASCL1\", \"HES5\", \"ID2\", \"ID4\", \"HDAC1\", \"HDAC2\", \"MYRF\", \"TCF7L2\", \"GPR17\", \"NKX2-2\", \"ZFP488\", \"ZFP191\"\n",
    "])\n",
    "\n",
    "# Fc Receptors & MHCII\n",
    "gene_lists[\"Fc receptors and MHCII\"] = []\n",
    "gene_lists[\"Fc receptors and MHCII\"].extend(genes[(genes.str.startswith(\"FCGR\"))])\n",
    "gene_lists[\"Fc receptors and MHCII\"].extend(genes[(genes.str.startswith(\"HLA-D\"))])\n",
    "gene_lists[\"Fc receptors and MHCII\"].extend([\"CD74\"])\n",
    "gene_lists[\"Fc receptors and MHCII\"] = np.intersect1d(gene_lists[\"Fc receptors and MHCII\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "# Human plaque induced genes, from https://www.sciencedirect.com/science/article/pii/S0092867420308151?via%3Dihub\n",
    "gene_lists[\"Microglia plaque induced genes\"] = []\n",
    "gene_lists[\"Microglia plaque induced genes\"].extend([\n",
    "    \"AXL\", \"C1QA\", \"C1QB\", \"C1QC\", \"FCGR2A\", \"FCGR2B\", \"FCGR2C\", \"FCGR3A\",\n",
    "    \"CSF1R\", \"CTSH\", \"CTSS\", \"CTSD\", \"CYBA\", \"GRN\", \"LY86\", \"LYZ\", \"HEXB\",\n",
    "    \"LAPTM5\", \"OLFML3\", \"TYROBP\"\n",
    "])\n",
    "\n",
    "# Interferon stimulated genes, from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4313732/\n",
    "gene_lists[\"Interferon stimulated genes\"] = []\n",
    "gene_lists[\"Interferon stimulated genes\"].extend(genes[(genes.str.startswith(\"ISG\")) & ~(genes.str.contains(\"L\"))])\n",
    "gene_lists[\"Interferon stimulated genes\"].extend(genes[(genes.str.startswith(\"IFI\")) & ~(genes.str.contains(\"L\"))])\n",
    "gene_lists[\"Interferon stimulated genes\"].extend(genes[(genes.str.startswith(\"IRF\")) & ~(genes.str.contains(\"BP\"))])\n",
    "gene_lists[\"Interferon stimulated genes\"].extend(genes[(genes.str.startswith(\"JAK\")) & ~(genes.str.contains(\"IP\"))])\n",
    "gene_lists[\"Interferon stimulated genes\"].extend(genes[(genes.str.startswith(\"STAT\")) & ~(genes.str.contains(\"H\"))])\n",
    "gene_lists[\"Interferon stimulated genes\"].extend(genes[(genes.str.startswith(\"OAS\"))])\n",
    "gene_lists[\"Interferon stimulated genes\"].extend([\"MX1\", \"MX2\", \"NINJ1\", \"IL1B\", \"IL6\", \"RNASEL\", \"EIF2AK2\", \"RSAD2\", \"BST2\", \"CH25H\"])\n",
    "gene_lists[\"Interferon stimulated genes\"] = np.intersect1d(gene_lists[\"Interferon stimulated genes\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"uniprotkb_keyword_KW_0276_AND_reviewed_2024_02_06.tsv\"), sep=\"\\t\")\n",
    "tmp = [i.split(\" \")[0] for i in tmp[\"Gene Names\"]]\n",
    "gene_lists[\"Fatty Acid metabolism\"].extend(np.intersect1d(tmp, genes))\n",
    "\n",
    "affected_supertypes = [\n",
    "    \"Sst_3\",\n",
    "    \"Sst_19\",\n",
    "    \"Sst_9\",\n",
    "    \"Sst_11\",\n",
    "    \"Sst_20\",\n",
    "    \"Sst_23\",\n",
    "    \"Sst_25\",\n",
    "    \"Sst_2\",\n",
    "    \"Pvalb_6\",\n",
    "    \"Pvalb_5\",\n",
    "    \"Pvalb_8\",\n",
    "    \"Pvalb_3\",\n",
    "    \"Pvalb_2\",\n",
    "    \"Pvalb_15\",\n",
    "    \"Pvalb_14\",\n",
    "]\n",
    "effect_size_table[\"Abundance Change\"] = effect_size_table[\"Population\"].isin(affected_supertypes)\n",
    "\n",
    "Sst_Pvalb.obs[\"Abundance Change\"] = Sst_Pvalb.obs[\"Supertype\"].isin(affected_supertypes).astype(\"str\")\n",
    "Sst_Pvalb.obs[\"Subclass\"] = Sst_Pvalb.obs[\"Subclass\"].astype(\"str\")\n",
    "\n",
    "Sst_Pvalb.obs[\"Subclass_Abundance Change\"] = Sst_Pvalb.obs[\"Subclass\"] + \"_\" + Sst_Pvalb.obs[\"Abundance Change\"]\n",
    "Sst_Pvalb = Sst_Pvalb[Sst_Pvalb.obs[\"Neurotypical reference\"] == \"False\", :].copy()\n",
    "\n",
    "Sst_Pvalb.obs[\"Subclass\"] = Sst_Pvalb.obs[\"Subclass\"].astype(\"category\")\n",
    "Sst_Pvalb.obs[\"Subclass_Abundance Change\"] = Sst_Pvalb.obs[\"Subclass_Abundance Change\"].astype(\"category\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b3f708b-5429-46cc-a91c-742a090a252d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Scatterplot\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Supertype\")\n",
    "    & (effect_size_table[\"Population\"].str.contains(\"Sst_\"))\n",
    "].copy()\n",
    "mean_table = df.drop([\"Population\", \"Taxonomy Level\"], axis=1).groupby([\"Gene\", \"Abundance Change\"]).mean()\n",
    "mean_table = mean_table.reset_index()\n",
    "mean_table.index = mean_table[\"Gene\"].copy()\n",
    "mean_table.index.name = \"index\"\n",
    "diff_table = mean_table.groupby([\"Gene\"]).diff().dropna().drop(\"Abundance Change\", axis=1)\n",
    "diff_table.columns = [\"Diff_\" + i for i in diff_table.columns]\n",
    "\n",
    "mean_table = mean_table.drop(\"Gene\", axis=1).pivot(columns=[\"Abundance Change\"])\n",
    "mean_table.columns = [i[0] + \"_\" + str(i[1]) for i in mean_table.columns]\n",
    "mean_table = mean_table.merge(diff_table, left_index=True, right_index=True, how=\"left\")\n",
    "\n",
    "df = mean_table.copy()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "tmp = df.loc[\n",
    "    ~(\n",
    "        (df[\"Mean expression (natural log UMIs per 10k plus 1)_True\"] < 0.05) &\n",
    "        (df[\"Mean expression (natural log UMIs per 10k plus 1)_False\"] < 0.05)\n",
    "    ),\n",
    "    :\n",
    "].copy()\n",
    "\n",
    "indep = \"Effect size across early pseudoprogression_True\"\n",
    "dep = \"Effect size across early pseudoprogression_False\"\n",
    "hue = \"Diff_Effect size across early pseudoprogression\"\n",
    "\n",
    "ax = sns.scatterplot(\n",
    "    data=tmp,\n",
    "    x=indep,\n",
    "    y=dep,\n",
    "    alpha=0.05,\n",
    "    size=1,\n",
    "    color=\"lightgrey\",\n",
    ");\n",
    "\n",
    "colors = {\n",
    "    \"Electron transport chain components\": \"red\",\n",
    "    \"Ribosomal proteins\": \"orange\",\n",
    "    \"Ubiquitination\": \"blue\",\n",
    "    \"Phosphorylation\": \"green\",\n",
    "}\n",
    "\n",
    "try:\n",
    "    del texts\n",
    "except:\n",
    "    pass\n",
    "\n",
    "for m,n in gene_lists.items():\n",
    "    if m not in colors.keys():\n",
    "        continue\n",
    "    ax = sns.scatterplot(\n",
    "        data=tmp.loc[(tmp[hue].abs() > 1) & (tmp.index.isin(n)), :],\n",
    "        x=indep,\n",
    "        y=dep,\n",
    "        alpha=0.3,\n",
    "        size=1,\n",
    "        color=colors[m],\n",
    "    );\n",
    "    ax = sns.scatterplot(\n",
    "        data=tmp.loc[(tmp[hue].abs() <= 1) & (tmp.index.isin(n)), :],\n",
    "        x=indep,\n",
    "        y=dep,\n",
    "        alpha=0.05,\n",
    "        size=1,\n",
    "        color=colors[m],\n",
    "    );\n",
    "    if m in [\"Electron transport chain components\", \"Ribosomal proteins\"]:\n",
    "        try:\n",
    "            texts.extend([plt.text(tmp.loc[i, indep], tmp.loc[i, dep], i, size=8, color=\"grey\") for i in tmp[(tmp[dep] < -1) & (tmp[hue] > 1) & (tmp.index.isin(n))].sort_values(by=dep).index[:10]])\n",
    "        except:\n",
    "            texts = [plt.text(tmp.loc[i, indep], tmp.loc[i, dep], i, size=8, color=\"grey\") for i in tmp[(tmp[dep] < -1) & (tmp[hue] > 1) & (tmp.index.isin(n))].sort_values(by=dep).index[:10]]\n",
    "    if m in [\"Ubiquitination\", \"Phosphorylation\", \"Lipid metabolism\"]:\n",
    "        try:\n",
    "            texts.extend([plt.text(tmp.loc[i, indep], tmp.loc[i, dep], i, size=8, color=\"grey\", horizontalalignment=\"right\") for i in tmp[(tmp[indep] < -1) & (tmp[hue] < -1) & (tmp.index.isin(n))].sort_values(by=dep).index[:10]])\n",
    "        except:\n",
    "            texts = [plt.text(tmp.loc[i, indep], tmp.loc[i, dep], i, size=8, color=\"grey\", horizontalalignment=\"right\") for i in tmp[(tmp[indep] < -1) & (tmp[hue] < -1) & (tmp.index.isin(n))].sort_values(by=dep).index[:10]]\n",
    "\n",
    "x = np.linspace(-4,4,100)\n",
    "y = x - 1\n",
    "plt.plot(x, y, linestyle=\"--\", color=\"lightgrey\");\n",
    "y = x + 1\n",
    "plt.plot(x, y, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False)\n",
    "ax.set(xlim=(-4,4), ylim=(-4,4), xlabel=\"Mean early effect size in\\nvulnerable Sst supertypes\", ylabel=\"Mean early effect size in\\n unaffected Sst supertypes\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 6i_scatterplot_vulnerable versus unaffected Sst supertypes early effect sizes.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbee692c-9b30-4bcb-9dbb-1d087443bd7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deltaplots\n",
    "for i in [\"MAPK8\", \"NGF\", \"MME\", \"LNX2\", \"ATP5MPL\", \"RPL4\"]:\n",
    "    ax = delta_plot(\n",
    "        adata=Sst_Pvalb,\n",
    "        genes=[i],\n",
    "        groupby=\"Subclass\",\n",
    "        groupby_subset=[\"Sst\", \"Pvalb\"],\n",
    "        plotby=\"Subclass_Abundance Change\",\n",
    "        donor=\"Donor ID\",\n",
    "        across=\"Continuous Pseudo-progression Score\",\n",
    "        highlight=[\"Sst_True\", \"Pvalb_True\"],\n",
    "        colormap=None,\n",
    "        title=\"Expression of \" + i,\n",
    "        legend=False,\n",
    "        save=os.path.join(pwd, \"output\", \"Figure 6hi_deltaplot_{title}_in_Sst_Pvalb_splitby_subclass_groupby_Subclass_Abundance Change.pdf\")\n",
    "    )\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6267aff-1968-467e-b8e0-d7221f458c45",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 11h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1590d210-91a9-4118-a9be-d080bbcd6864",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Supertype\")\n",
    "    & (effect_size_table[\"Population\"].str.contains(\"Pvalb_\"))\n",
    "].copy()\n",
    "mean_table = df.drop([\"Population\", \"Taxonomy Level\"], axis=1).groupby([\"Gene\", \"Abundance Change\"]).mean()\n",
    "mean_table = mean_table.reset_index()\n",
    "mean_table.index = mean_table[\"Gene\"].copy()\n",
    "mean_table.index.name = \"index\"\n",
    "diff_table = mean_table.groupby([\"Gene\"]).diff().dropna().drop(\"Abundance Change\", axis=1)\n",
    "diff_table.columns = [\"Diff_\" + i for i in diff_table.columns]\n",
    "\n",
    "mean_table = mean_table.drop(\"Gene\", axis=1).pivot(columns=[\"Abundance Change\"])\n",
    "mean_table.columns = [i[0] + \"_\" + str(i[1]) for i in mean_table.columns]\n",
    "mean_table = mean_table.merge(diff_table, left_index=True, right_index=True, how=\"left\")\n",
    "\n",
    "df = mean_table.copy()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "# Unsupervised\n",
    "tmp = df.loc[\n",
    "    ~(\n",
    "        (df[\"Mean expression (natural log UMIs per 10k plus 1)_True\"] < 0.05) &\n",
    "        (df[\"Mean expression (natural log UMIs per 10k plus 1)_False\"] < 0.05)\n",
    "    ),\n",
    "    :\n",
    "].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=tmp,\n",
    "    x=indep,\n",
    "    y=dep,\n",
    "    alpha=0.05,\n",
    "    size=1,\n",
    "    color=\"lightgrey\",\n",
    ");\n",
    "\n",
    "colors = {\n",
    "    \"Electron transport chain components\": \"red\",\n",
    "    \"Ribosomal proteins\": \"orange\",\n",
    "    \"Ubiquitination\": \"blue\",\n",
    "    \"Phosphorylation\": \"green\",\n",
    "}\n",
    "\n",
    "try:\n",
    "    del texts\n",
    "except:\n",
    "    pass\n",
    "\n",
    "for m,n in gene_lists.items():\n",
    "    if m not in colors.keys():\n",
    "        continue\n",
    "\n",
    "    indep = \"Effect size across early pseudoprogression_True\"\n",
    "    dep = \"Effect size across early pseudoprogression_False\"\n",
    "    hue = \"Diff_Effect size across early pseudoprogression\"\n",
    "    ax = sns.scatterplot(\n",
    "        data=tmp.loc[(tmp[hue].abs() > 1) & (tmp.index.isin(n)), :],\n",
    "        x=indep,\n",
    "        y=dep,\n",
    "        alpha=0.3,\n",
    "        size=1,\n",
    "        color=colors[m],\n",
    "    );\n",
    "    ax = sns.scatterplot(\n",
    "        data=tmp.loc[(tmp[hue].abs() <= 1) & (tmp.index.isin(n)), :],\n",
    "        x=indep,\n",
    "        y=dep,\n",
    "        alpha=0.05,\n",
    "        size=1,\n",
    "        color=colors[m],\n",
    "    );\n",
    "\n",
    "x = np.linspace(-4,4,100)\n",
    "y = x - 1\n",
    "plt.plot(x, y, linestyle=\"--\", color=\"lightgrey\");\n",
    "y = x + 1\n",
    "plt.plot(x, y, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False)\n",
    "ax.set(xlim=(-4,4), ylim=(-4,4), xlabel=\"Mean early effect size in\\nvulnerable Pvalb supertypes\", ylabel=\"Mean early effect size in\\n unaffected Pvalb supertypes\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 11h_scatterplot_vulnerable versus unaffected Pvalb supertypes early effect sizes.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8c5eeb9-7c69-49df-8399-4afaba2a98ff",
   "metadata": {},
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a697b6-dba5-4670-863e-7fa228e3004d",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
