{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c4b3a822-57b7-4cee-9b0d-46b9d9b681e9",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d8a6c9-19ef-4b78-b0e5-d719d215c84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import copy as cp\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "import scipy.stats as sp_stats\n",
    "from helper_functions import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=100, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee2e13f8-8040-4f23-8c61-383618e98d8b",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df48c455-9a0f-4456-a63f-dadcc17378c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "\n",
    "# From 00_build_input_files.py\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 7\", \"SEAAD_MTG_RNAseq_final-nuclei_limited.2024-02-13.h5ad\"))\n",
    "\n",
    "# Subclass UMAP coordinates from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "Microglia_PVM_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"Microglia-PVM\", \"obs_names.csv\"), index_col=0).index\n",
    "Microglia_PVM_umap = np.load(os.path.join(pwd, \"input\", \"Figure 7\", \"Microglia-PVM\", \"X_umap.npy\"))\n",
    "\n",
    "Astrocyte_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"Astrocyte\", \"obs_names.csv\"), index_col=0).index\n",
    "Astrocyte_umap = np.load(os.path.join(pwd, \"input\", \"Figure 7\", \"Astrocyte\", \"X_umap.npy\"))\n",
    "\n",
    "# Green et al (2023) labels from personal communication and Mathys et al (2023) labels from https://compbio.mit.edu/\n",
    "# Green et al (2023) and Mathys et al (2023) mapping results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#PFC/RNAseq/public_datasets/\n",
    "green_labels = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\",\"Green_2023_author-annotation.csv\"), index_col=0)\n",
    "green_mapping = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"Green_2023-cell-annotation.2024-03-27.csv\"), index_col=0)\n",
    "\n",
    "mathys_label_files = glob.glob(os.path.join(pwd, \"input\", \"Figure 7\", \"Mathys_2023\", \"*.csv\"))\n",
    "mathys_labels = False\n",
    "for i in mathys_label_files:\n",
    "    tmp = pd.read_csv(i, index_col=0)\n",
    "    try:\n",
    "        mathys_labels = pd.concat([mathys_labels, tmp], axis=0)\n",
    "    except TypeError as i:\n",
    "        mathys_labels = tmp.copy()\n",
    "mathys_mapping = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"Mathys_2023-cell-annotation.2024-03-27.csv\"), index_col=0)\n",
    "\n",
    "# Mean expression table from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "mean_expression = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 7\", \"mean_expression.h5ad\"))\n",
    "\n",
    "# Cell type specific marker genes from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "effect_sizes_vs_all = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 7\", \"effect_sizes_vs_all.h5ad\"))\n",
    "      \n",
    "# Effect size table from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "effect_size_table = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"effect_size_table.csv\"), index_col=0)\n",
    "level = \"Subclass\"\n",
    "tmp = effect_size_table.loc[effect_size_table[\"Taxonomy Level\"] == level, [\"Gene\", \"Mean expression (natural log UMIs per 10k plus 1)\"]]\n",
    "tmp = tmp.groupby([\"Gene\"])[\"Mean expression (natural log UMIs per 10k plus 1)\"].transform(lambda x : sp_stats.zscore(x,ddof=1)).fillna(0)\n",
    "effect_size_table[\"Mean expression z-score\"] = tmp\n",
    "\n",
    "# GRN results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "eRegulons = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"eRegulon_metadata_filtered.csv\"), index_col=0)\n",
    "TFs = [i.replace(\"_extended\", \"\") for i in eRegulons[\"Gene_signature_name\"].unique()]\n",
    "TFs = pd.DataFrame([[i] + i.split(\"_\") for i in TFs], columns=[\"Gene_signature_name\", \"TF\", \"Direction\", \"Target Genes\"])\n",
    "TFs[\"Target Genes\"] = [np.int32(i.replace(\"(\", \"\").replace(\")\", \"\").replace(\"g\", \"\")) for i in TFs[\"Target Genes\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd81429-980c-4f2a-b135-dd2414d7ff09",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a229a78e-edf8-4b29-8aa2-20975085e180",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub = adata[Microglia_PVM_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = Microglia_PVM_umap\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=20,\n",
    "    palette=\"tab10\",\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia-PVM_Supertype_tab10.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia-PVM_Supertype_tab10.pdf\"), os.path.join(pwd, \"output\", \"Figure 7a_umap_Microglia-PVM_Supertype_tab10.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39760697-5fb2-45d9-9e53-d3e3e64df682",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58289cd8-cfe5-4e6a-b445-814a4fe5e34f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "effect_sizes_microglia = effect_sizes_vs_all[:, effect_sizes_vs_all.var[\"Subclass\"] == \"Microglia-PVM\"].copy()\n",
    "effect_sizes_microglia = pd.DataFrame(\n",
    "    effect_sizes_microglia.X,\n",
    "    index=effect_sizes_microglia.obs_names,\n",
    "    columns=effect_sizes_microglia.var_names\n",
    ")\n",
    "\n",
    "mean_expression_microglia = mean_expression[:, mean_expression.var[\"Subclass\"] == \"Microglia-PVM\"].copy()\n",
    "mean_expression_microglia = pd.DataFrame(\n",
    "    mean_expression_microglia.X,\n",
    "    index=mean_expression_microglia.obs_names,\n",
    "    columns=mean_expression_microglia.var_names\n",
    ")\n",
    "\n",
    "z_score_microglia = mean_expression_microglia.T\n",
    "z_score_microglia = (z_score_microglia - z_score_microglia.mean(axis=0)) / z_score_microglia.std(axis=0)\n",
    "z_score_microglia = z_score_microglia.T\n",
    "\n",
    "\n",
    "genes = []\n",
    "subclass = \"Microglia-PVM\"\n",
    "for i in effect_sizes_microglia.columns:\n",
    "    gene_n = 3\n",
    "    if i == \"Micro-PVM_3-SEAAD\":\n",
    "        gene_n = 30\n",
    "    new_genes = effect_sizes_microglia.loc[(effect_sizes_microglia.max(axis=1) == effect_sizes_microglia.loc[:, i]) & (mean_expression_microglia.max(axis=1) == mean_expression_microglia.loc[:, i]), :].sort_values(by=i, ascending=False).iloc[:gene_n].index.to_list()\n",
    "    genes.extend(z_score_microglia.loc[new_genes, i].sort_values(ascending=False).index.to_list())\n",
    "    \n",
    "plt.rcParams[\"figure.figsize\"] = (15,3)\n",
    "sns.heatmap(\n",
    "    z_score_microglia.loc[genes, :].T,\n",
    "    cmap=\"RdBu_r\",\n",
    "    center=0,\n",
    "    xticklabels=True,\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7b_heatmap_Microglia marker z scores.pdf\"), bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "517dfccf-8777-4ec2-be33-a4a52ab0ea59",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ff1e9c6-a8fa-4dfb-87e1-c50ac3482785",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Green_2023\n",
    "df = green_mapping.merge(green_labels.loc[:, [\"cell\", \"state\"]], how=\"inner\", left_on=\"Original Cell ID\", right_on=\"cell\")\n",
    "df = df.loc[:, [\"Supertype\", \"state\"]].groupby(\"state\").value_counts(normalize=True).reset_index().pivot(index=\"Supertype\", columns=\"state\", values=\"proportion\")\n",
    "df = df.fillna(0)\n",
    "plt.rcParams[\"figure.figsize\"] = (5,3)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[\n",
    "        [\"Micro-PVM_1\", \"Micro-PVM_2\", \"Micro-PVM_2_1-SEAAD\", \"Micro-PVM_2_3-SEAAD\", \"Micro-PVM_3-SEAAD\", \"Micro-PVM_4-SEAAD\", \"Monocyte\"],\n",
    "        [\"Macrophages\", \"Mic.2\", \"Mic.3\", \"Mic.9\", \"Mic.5\", \"Mic.4\", \"Mic.14\", \"Mic.16\", \"Mic.11\", \"Mic.1\", \"Mic.10\", \"Mic.8\", \"Mic.7\", \"Mic.12\", \"Mic.13\", \"Mic.6\", \"Mic.15\", \"Monocytes\"]\n",
    "    ],\n",
    "    cmap=\"YlGnBu\",\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    ");\n",
    "ax.set(xlabel=\"Green et al 2023\", ylabel=\"SEA-AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7c_heatmap_Green_Microglia_Comparison.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# Mathys_2023\n",
    "df = mathys_mapping.merge(mathys_labels.loc[:, [\"barcode\", \"cell_type_high_resolution\"]], how=\"inner\", left_on=\"Original Cell ID Alternative\", right_on=\"barcode\")\n",
    "df = df.loc[:, [\"Supertype\", \"cell_type_high_resolution\"]].groupby(\"cell_type_high_resolution\").value_counts(normalize=True).reset_index().pivot(index=\"Supertype\", columns=\"cell_type_high_resolution\", values=\"proportion\")\n",
    "df = df.fillna(0)\n",
    "plt.rcParams[\"figure.figsize\"] = (3,3)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[\n",
    "        [\"Micro-PVM_1\", \"Micro-PVM_2\", \"Micro-PVM_2_1-SEAAD\", \"Micro-PVM_2_3-SEAAD\", \"Micro-PVM_3-SEAAD\", \"Micro-PVM_4-SEAAD\", \"Monocyte\"],\n",
    "        [\"CAMs\", \"Mic P2RY12\", \"Mic TPT1\", \"Mic MKI67\"]\n",
    "    ],\n",
    "    cmap=\"YlGnBu\",\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    ");\n",
    "ax.set(xlabel=\"Mathys et al 2023\", ylabel=\"SEA-AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7c_heatmap_Mathys_Microglia_Comparison.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a44b3b61-3bc7-4f5d-bdc8-f3533b98082a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6936ff2b-ea1b-49b3-afa0-ed69f6f68540",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Microglia early\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Microglia-PVM\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axvline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axvline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = [\"IRF1\", \"IRF7\", \"IFI16\", \"FCGR1A\", \"FCGR1B\", \"FCGR2A\", \"CD74\", \"HLA-DRB5\", \"IL1B\", \"JAK3\", \"NINJ1\"]\n",
    "already_colored.extend([\"C1QC\", \"C1QB\", \"CTSC\", \"LY86\", \"CSF1R\", \"FCGR3A\"])\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"red\") for i in df.loc[df[\"Gene\"].isin([\"IRF1\", \"IRF7\", \"IFI16\", \"FCGR1A\", \"FCGR1B\", \"FCGR2A\", \"CD74\", \"HLA-DRB5\", \"IL1B\", \"JAK3\", \"NINJ1\"]), :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"blue\") for i in df.loc[df[\"Gene\"].isin([\"C1QC\", \"C1QB\", \"CTSC\", \"LY86\", \"CSF1R\", \"FCGR3A\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\", ascending=False).iloc[:10, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\").iloc[:10, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7d_scatterplot_Microglia_Early.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Microglia late\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Microglia-PVM\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axhline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axhline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = [\"APOE\", \"CTSD\", \"CTSS\", \"LYZ\", \"CD9\"]\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"blue\") for i in df.loc[df[\"Gene\"].isin([\"APOE\", \"CTSD\", \"CTSS\", \"LYZ\", \"CD9\"]), :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across late pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across late pseudoprogression\", ascending=False).iloc[:10, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across late pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across late pseudoprogression\").iloc[:10, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7d_scatterplot_Microglia_Late.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1d5efde-ecf4-41f8-980a-9147996478ff",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7728716-0ab7-49bb-847e-4e0e18f7345b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "\n",
    "# Scatterplot\n",
    "df = effect_size_table.loc[(effect_size_table[\"Gene\"].isin(TFs.loc[TFs[\"Direction\"] == \"+\", \"TF\"])) & (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") & (effect_size_table[\"Population\"].isin([\"Microglia-PVM\"])), :]\n",
    "sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Mean expression z-score\",\n",
    "    y=\"Effect size across early pseudoprogression\",\n",
    "    alpha=0.5,\n",
    ")\n",
    "texts = [plt.text(df.loc[i, \"Mean expression z-score\"], df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Gene\"], size=8) for i in df.loc[(df[\"Mean expression z-score\"] > 2) & (df[\"Effect size across early pseudoprogression\"] > 1.1), :].index]\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color=\"gray\"));\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7e_scatterplot_Micro-PVM transcription factor regulators.pdf\"), bbox_inches=\"tight\")\n",
    "MR = TFs.loc[TFs[\"TF\"].isin(df.loc[(df[\"Mean expression z-score\"] > 2) & (df[\"Effect size across early pseudoprogression\"] > 1.1), \"Gene\"]) & (TFs[\"Direction\"] == \"+\"), \"Gene_signature_name\"].to_list()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1223d3c-05aa-4fbe-8925-5b2230346e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# KDEplot\n",
    "Downstream_Genes = eRegulons.loc[(eRegulons[\"Gene_signature_name\"].isin(MR)), [\"Gene\", \"TF\"]].drop_duplicates().loc[:, \"Gene\"].value_counts()\n",
    "Downstream_Genes = Downstream_Genes[Downstream_Genes > Downstream_Genes.median() + 1].index\n",
    "\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Gene\"].isin(Downstream_Genes)) & (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") & (effect_size_table[\"Population\"].isin([\"Microglia-PVM\"])),\n",
    "    :\n",
    "].merge(\n",
    "    eRegulons.loc[(eRegulons[\"Gene_signature_name\"].isin(MR)), [\"Gene\", \"TF2G_importance\"]].drop_duplicates().groupby(\"Gene\").mean(),\n",
    "    left_on=\"Gene\",\n",
    "    right_on=\"Gene\",\n",
    "    how=\"left\"\n",
    ")\n",
    "df_else = effect_size_table.loc[\n",
    "    ~(effect_size_table[\"Gene\"].isin(Downstream_Genes)) & (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") & (effect_size_table[\"Population\"].isin([\"Microglia-PVM\"])),\n",
    "    :\n",
    "]\n",
    "if df.shape[0] < df_else.shape[0]:\n",
    "    df_else = df_else.loc[np.random.choice(df_else.index, size=df.shape[0], replace=False), :].copy()\n",
    "    \n",
    "sns.kdeplot(\n",
    "    data=df_else,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    color=\"orange\"\n",
    ");\n",
    "sns.kdeplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\"\n",
    ");\n",
    "print(\"p-value early: \" + str(sp_stats.ttest_ind(df[\"Effect size across early pseudoprogression\"], df_else[\"Effect size across early pseudoprogression\"], equal_var=False)[1]))\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7e_kdeplot_Micro-PVM downstream targets.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe288d20-584f-4feb-bd7b-f91975545a52",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "814d4413-7ef8-4611-8a7b-2705ec347844",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Deltaplots\n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "sub = adata[adata.obs[\"Neurotypical reference\"] == \"False\", :].copy()\n",
    "\n",
    "for i in [\"RUNX1\", \"CTSC\", \"JAK3\"]:\n",
    "    ax = delta_plot(\n",
    "        adata=sub,\n",
    "        genes=[i],\n",
    "        groupby=\"Class\",\n",
    "        groupby_subset=[\"Non-neuronal and Non-neural\"],\n",
    "        plotby=\"Subclass\",\n",
    "        donor=\"Donor ID\",\n",
    "        across=\"Continuous Pseudo-progression Score\",\n",
    "        highlight=[],\n",
    "        colormap=subclass_colors,\n",
    "        title=\"Expression of \" + i,\n",
    "        legend=False,\n",
    "        save=os.path.join(pwd, \"output\", \"Figure 7f_deltaplot_{title}_by non-neuronal Subclass.pdf\")\n",
    "    )\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "423de82d-7aa2-4846-b741-0f526d5597de",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbfbbd1-a61f-43ec-a780-13202f08dfd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub = adata[Astrocyte_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = Astrocyte_umap\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=20,\n",
    "    palette=\"tab10\",\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Astrocyte_Supertype_tab10.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Astrocyte_Supertype_tab10.pdf\"), os.path.join(pwd, \"output\", \"Figure 7g_umap_Astrocyte_Supertype_tab10.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c39b329-2bf5-4a6b-88a7-f8df3f1a620f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53d08eda-3c99-411f-9463-bef14e27fde1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Green_2023\n",
    "df = green_mapping.merge(green_labels.loc[:, [\"cell\", \"state\"]], how=\"inner\", left_on=\"Original Cell ID\", right_on=\"cell\")\n",
    "df = df.loc[:, [\"Supertype\", \"state\"]].groupby(\"state\").value_counts(normalize=True).reset_index().pivot(index=\"Supertype\", columns=\"state\", values=\"proportion\")\n",
    "df = df.fillna(0)\n",
    "plt.rcParams[\"figure.figsize\"] = (5,3)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[\n",
    "        [\"Astro_2\", \"Astro_1\", \"Astro_3\", \"Astro_4\", \"Astro_5\", \"Astro_6-SEAAD\"],\n",
    "        [\"Ast.1\", \"Ast.2\", \"Ast.3\", \"Ast.6\", \"Ast.8\", \"Ast.9\", \"Ast.10\", \"Ast.7\", \"Ast.4\", \"Ast.5\"]\n",
    "    ],\n",
    "    cmap=\"YlGnBu\",\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    ");\n",
    "ax.set(xlabel=\"Green et al 2023\", ylabel=\"SEA-AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7h_heatmap_Green_Astrocyte_Comparison.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Mathys_2023\n",
    "df = mathys_mapping.merge(mathys_labels.loc[:, [\"barcode\", \"cell_type_high_resolution\"]], how=\"inner\", left_on=\"Original Cell ID Alternative\", right_on=\"barcode\")\n",
    "df = df.loc[:, [\"Supertype\", \"cell_type_high_resolution\"]].groupby(\"cell_type_high_resolution\").value_counts(normalize=True).reset_index().pivot(index=\"Supertype\", columns=\"cell_type_high_resolution\", values=\"proportion\")\n",
    "df = df.fillna(0)\n",
    "plt.rcParams[\"figure.figsize\"] = (2,3)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[\n",
    "        [\"Astro_2\", \"Astro_1\", \"Astro_3\", \"Astro_4\", \"Astro_5\", \"Astro_6-SEAAD\"],\n",
    "        [\"Ast GRM3\", \"Ast DPP10\", \"Ast CHI3L1\"]\n",
    "    ],\n",
    "    cmap=\"YlGnBu\",\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    ");\n",
    "ax.set(xlabel=\"Mathys et al 2023\", ylabel=\"SEA-AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7c_heatmap_Mathys_Astrocyte_Comparison.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4940a6b0-01dd-4435-a0a8-f7c1eeec7587",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19146b3e-ed7e-433b-81cb-665bd7e9bf51",
   "metadata": {},
   "outputs": [],
   "source": [
    "effect_sizes_astrocyte = effect_sizes_vs_all[:, effect_sizes_vs_all.var[\"Subclass\"] == \"Astrocyte\"].copy()\n",
    "effect_sizes_astrocyte = pd.DataFrame(\n",
    "    effect_sizes_astrocyte.X,\n",
    "    index=effect_sizes_astrocyte.obs_names,\n",
    "    columns=effect_sizes_astrocyte.var_names\n",
    ")\n",
    "\n",
    "mean_expression_astrocyte = mean_expression[:, mean_expression.var[\"Subclass\"] == \"Astrocyte\"].copy()\n",
    "mean_expression_astrocyte = pd.DataFrame(\n",
    "    mean_expression_astrocyte.X,\n",
    "    index=mean_expression_astrocyte.obs_names,\n",
    "    columns=mean_expression_astrocyte.var_names\n",
    ")\n",
    "\n",
    "z_score_astrocyte = mean_expression_astrocyte.T\n",
    "z_score_astrocyte = (z_score_astrocyte - z_score_astrocyte.mean(axis=0)) / z_score_astrocyte.std(axis=0)\n",
    "z_score_astrocyte = z_score_astrocyte.T\n",
    "\n",
    "\n",
    "genes = []\n",
    "subclass = \"Astrocyte\"\n",
    "for i in effect_sizes_astrocyte.columns:\n",
    "    gene_n = 3\n",
    "    if i == \"Astro_2\":\n",
    "        gene_n = 30\n",
    "    new_genes = effect_sizes_astrocyte.loc[(effect_sizes_astrocyte.max(axis=1) == effect_sizes_astrocyte.loc[:, i]) & (mean_expression_astrocyte.max(axis=1) == mean_expression_astrocyte.loc[:, i]), :].sort_values(by=i, ascending=False).iloc[:gene_n].index.to_list()\n",
    "    genes.extend(z_score_astrocyte.loc[new_genes, i].sort_values(ascending=False).index.to_list())\n",
    "    \n",
    "plt.rcParams[\"figure.figsize\"] = (15,3)\n",
    "sns.heatmap(\n",
    "    z_score_astrocyte.loc[genes, :].T,\n",
    "    cmap=\"RdBu_r\",\n",
    "    center=0,\n",
    "    xticklabels=True,\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7i_heatmap_Astrocyte marker z scores.pdf\"), bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1328e88-2a82-4b49-9d62-a25c3ef10820",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10a5ca87-22a7-49cd-865b-91edcd3d2fc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.Series(effect_size_table[\"Gene\"].unique())\n",
    "# Define gene lists\n",
    "gene_lists = {}\n",
    "\n",
    "# Cell adhesion, based on GO:0007155\n",
    "gene_lists[\"Cell adhesion\"] = []\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CNTN\")) & ~(genes.str.contains(\"AP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"NINJ\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"NLGN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"DLG\")) & ~(genes.str.contains(\"AP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"ADGR\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"PKP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"TENM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"DSCAM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"NECTIN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CADM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"SIGLEC\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CTNN\")) & ~(genes.str.contains(\"IP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"ICAM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CEACAM\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CLDN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CLSTN\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CELSR\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"DSG\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"CDH\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"PCDH\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend(genes[(genes.str.startswith(\"ITG\")) & ~(genes.str.contains(\"BP\"))])\n",
    "gene_lists[\"Cell adhesion\"].extend([\n",
    "    \"DSC1\", \"DSC2\", \"DSC3\",\n",
    "    \"DCHS1\", \"DCHS2\",\n",
    "    \"FAT1\", \"FAT2\", \"FAT3\", \"FAT4\",\n",
    "    \"SELE\", \"SELL\", \"SELP\",\n",
    "    \"VCAM1\", \"PECAM1\", \"MADCAM1\"\n",
    "    \"L1CAM\", \"NRCAM\", \"CHL1\", \"NFASC\",\n",
    "    \"MAG\", \"F11R\", \"JAM2\", \"JAM3\", \"JAML\",\n",
    "    \"VSIG1\", \"VSIG2\", \"IGSF11\", \"VSIG4\", \"ESAM\",\n",
    "    \"CXADR\", \"PVR\", \"CD2\", \"HEPACAM\", \"HEPACAM2\",\n",
    "    \"BCAM\", \"EPCAM\", \"ALCAM\", \"CERCAM\", \"MCAM\",\n",
    "    \"UNC5A\", \"UNC5B\", \"UNC5C\", \"UNC5D\",\n",
    "    \"VCL\", \"JUP\", \"MSN\", \"CD209\", \n",
    "])\n",
    "gene_lists[\"Cell adhesion\"] = np.intersect1d(gene_lists[\"Cell adhesion\"], genes[~(genes.str.contains(\"-AS\")) & ~(genes.str.contains(\"-DT\")) & ~(genes.str.contains(\"-IT\"))])\n",
    "\n",
    "# Axonal guidance, based on GO:0097485\n",
    "gene_lists[\"Axonal guidance\"] = []\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"SEMA\"))]) # Semaphorin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"UNC5\")) & ~(genes.str.contains(\"50\"))]) # Netrin receptor\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"EPH\"))]) # Ephrin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"EFN\"))]) # Ephrin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"LGI\"))]) # Leucine Repeat (Slit-like)\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NOTCH\"))]) # Notch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"DLL\"))]) # Notch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"JAG\"))]) # Notch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"SLIT\"))]) # Slit\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"LRTM\"))]) # Slit-like\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"ROBO\"))]) # Robo\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"PLXN\"))]) # Plexin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FLRT\"))]) # Fibronectin Leucine Rich Transmembrane\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"VEGF\"))]) # Vascular endothelial growth factor\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NCAM\"))]) # NCAM\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"CNTN\")) & ~(genes.str.contains(\"AP\"))]) # Contactin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"PTCH\"))]) # Patch\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FEZ\"))]) # FEZ\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"ERBB\"))]) # Erb-B2 Receptor Tyrosine Kinase\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FZD\"))]) # Fzd\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NRP\"))]) # Neuropilin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"BMP\"))]) # BMP\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"FGF\")) & ~(genes.str.contains(\"P\"))]) # FGF\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NTN\"))]) # Netrin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"LGR\"))]) # LGRs\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"WNT\"))]) # Wnt\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NTRK\"))]) # Ntrk\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NECTIN\"))]) # Nectin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"DSCAM\"))]) # DSCAM\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"NRXN\"))]) # Neurexin\n",
    "gene_lists[\"Axonal guidance\"].extend(genes[(genes.str.startswith(\"GFR\"))]) # GDNF Receptor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8abdea7-eb89-42d9-b69c-b3f969d7d688",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Astrocyte early\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Astrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df2 = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Astrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    effect_size_table[\"Gene\"].isin(np.union1d(gene_lists[\"Cell adhesion\"], gene_lists[\"Axonal guidance\"])),\n",
    "    :\n",
    "]\n",
    "df = pd.concat([df, df2], axis=0)\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axvline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axvline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = cp.copy(np.union1d(gene_lists[\"Cell adhesion\"], gene_lists[\"Axonal guidance\"]).tolist())\n",
    "already_colored.extend([\"GFAP\"])\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"red\") for i in df.loc[(df[\"Gene\"].isin(np.union1d(gene_lists[\"Cell adhesion\"], gene_lists[\"Axonal guidance\"]))) & (df[\"Effect size across early pseudoprogression\"] > 1.5) , :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"blue\") for i in df.loc[df[\"Gene\"].isin([\"GFAP\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\", ascending=False).iloc[:5, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\").iloc[:10, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7j_scatterplot_Astrocyte_Early.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Astrocyte late\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Astrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df2 = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Astrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    effect_size_table[\"Gene\"].isin(np.union1d(gene_lists[\"Cell adhesion\"], gene_lists[\"Axonal guidance\"])),\n",
    "    :\n",
    "]\n",
    "df = pd.concat([df, df2], axis=0)\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axhline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axhline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = cp.copy(np.union1d(gene_lists[\"Cell adhesion\"], gene_lists[\"Axonal guidance\"]).tolist())\n",
    "already_colored.extend([\"APOE\"])\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"red\") for i in df.loc[(df[\"Gene\"].isin(np.union1d(gene_lists[\"Cell adhesion\"], gene_lists[\"Axonal guidance\"]))) & (df[\"Effect size across late pseudoprogression\"].abs() > 1.5) , :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"green\") for i in df.loc[df[\"Gene\"].isin([\"APOE\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across late pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across late pseudoprogression\", ascending=False).iloc[:10, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across late pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across late pseudoprogression\").iloc[:10, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 7j_scatterplot_Astrocyte_Late.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96316aa7-9523-409f-905f-25a85aa7e9ba",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 7k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad03cccf-068c-40b2-a134-c1f1effa3990",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deltaplots\n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "sub = adata[adata.obs[\"Neurotypical reference\"] == \"False\", :].copy()\n",
    "\n",
    "for i in [\"APOE\"]:\n",
    "    ax = delta_plot(\n",
    "        adata=sub,\n",
    "        genes=[i],\n",
    "        groupby=\"Class\",\n",
    "        groupby_subset=[\"Non-neuronal and Non-neural\"],\n",
    "        plotby=\"Subclass\",\n",
    "        donor=\"Donor ID\",\n",
    "        across=\"Continuous Pseudo-progression Score\",\n",
    "        highlight=[],\n",
    "        colormap=subclass_colors,\n",
    "        title=\"Expression of \" + i,\n",
    "        legend=False,\n",
    "        save=os.path.join(pwd, \"output\", \"Figure 7k_deltaplot_{title}_by non-neuronal Subclass.pdf\")\n",
    "    )\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1caf1ca0-4a32-4ddd-9c07-86cf81d8b1b5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a697b6-dba5-4670-863e-7fa228e3004d",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
