{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "efb50d42-b227-4930-8b2a-55450b2ed639",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14555b13-06ff-4dbf-be22-fcf91dbc0f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import copy\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "from scipy import stats as sp_stats\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=500, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "663338dc-e54a-4b76-a757-c4a218bae922",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cde7bc1-6724-483d-b994-ed3305a7ef52",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "\n",
    "# [NOT PUBLIC] We are working to share this object, it requires amending DUAs with RADC and Sage Bionetworks\n",
    "public_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_SEAAD_and_public_datasets_singleomeCR6_UMIs_only_expanded_de_genes_2.2023-12-12.h5ad\"))\n",
    "\n",
    "# [NOT PUBLIC] We are working to share these scores, it requires amending DUAs with RADC and Sage Bionetworks\n",
    "Signature_Scores = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"Signature Scores.csv\"), index_col=0)\n",
    "\n",
    "# scCODA results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "A9_RNAseq_neuronal = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"A9_RNAseq\", \"Continuous_Pseudo-progression_Score\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_results.csv\"), index_col=0)\n",
    "A9_RNAseq_neuronal_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"A9_RNAseq\", \"objects\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_abundances.h5ad\"))\n",
    "Mathys_RNAseq_neuronal = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Mathys_2023\", \"Overall_AD_neuropathological_Change_codes\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_results.csv\"), index_col=0)\n",
    "Mathys_RNAseq_neuronal_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Mathys_2023\", \"objects\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_abundances.h5ad\"))\n",
    "Green_RNAseq_neuronal = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Green_2023\", \"Overall_AD_neuropathological_Change_codes\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_results.csv\"), index_col=0)\n",
    "Green_RNAseq_neuronal_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Green_2023\", \"objects\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_abundances.h5ad\"))\n",
    "\n",
    "A9_RNAseq_non_neuronal = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"A9_RNAseq\", \"Continuous_Pseudo-progression_Score\", \"Non-neuronal and Non-neural_Supertype_results.csv\"), index_col=0)\n",
    "A9_RNAseq_non_neuronal_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"A9_RNAseq\", \"objects\", \"Non-neuronal and Non-neural_Supertype_abundances.h5ad\"))\n",
    "Mathys_RNAseq_non_neuronal = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Mathys_2023\", \"Overall_AD_neuropathological_Change_codes\", \"Non-neuronal and Non-neural_Supertype_results.csv\"), index_col=0)\n",
    "Mathys_RNAseq_non_neuronal_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Mathys_2023\", \"objects\", \"Non-neuronal and Non-neural_Supertype_abundances.h5ad\"))\n",
    "Green_RNAseq_non_neuronal = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Green_2023\", \"Overall_AD_neuropathological_Change_codes\", \"Non-neuronal and Non-neural_Supertype_results.csv\"), index_col=0)\n",
    "Green_RNAseq_non_neuronal_adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 4 and Extended Data Figure 9\", \"PFC_Green_2023\", \"objects\", \"Non-neuronal and Non-neural_Supertype_abundances.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7cb795a-7e6e-48f8-b98f-75fb1d6ea693",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a978ad8f-b2e6-4751-aa82-5a9521f5f6ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cells = (public_adata.obs[\"Cognitive Status\"] != \"Reference\") & (public_adata.obs[\"Used in analysis\"] == \"True\")\n",
    "ranks = pd.DataFrame(np.zeros((11, 0)), index=[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d22ddfd1-ca22-4858-b4e8-555a72d254e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# CERAD score\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"CERAD score\"]].drop_duplicates().loc[:, [\"Source\", \"CERAD score\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[~df[\"CERAD score\"].isin([\"Reference\", \"\"]), :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"CERAD score\"] != \"\", :].copy()\n",
    "df[\"CERAD score\"] = df[\"CERAD score\"].cat.remove_unused_categories()\n",
    "df[\"CERAD score\"] = df[\"CERAD score\"].cat.reorder_categories([\"Absent\", \"Sparse\", \"Moderate\", \"Frequent\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"CERAD score\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"CERAD score\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4a_barplot_cerad.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Braak stage\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Braak\"]].drop_duplicates().loc[:, [\"Source\", \"Braak\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Braak\"] != \"Reference\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"Braak\"] != \"\", :].copy()\n",
    "df[\"Braak\"] = df[\"Braak\"].cat.remove_unused_categories()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Braak\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"Braak\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4a_barplot_braak.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# ADNC\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Overall AD neuropathological Change\"]].drop_duplicates().loc[:, [\"Source\", \"Overall AD neuropathological Change\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Overall AD neuropathological Change\"] != \"Reference\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"Overall AD neuropathological Change\"] != \"\", :].copy()\n",
    "df[\"Overall AD neuropathological Change\"] = df[\"Overall AD neuropathological Change\"].cat.remove_unused_categories()\n",
    "df[\"Overall AD neuropathological Change\"] = df[\"Overall AD neuropathological Change\"].cat.reorder_categories([\"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Overall AD neuropathological Change\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"Overall AD neuropathological Change\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4a_barplot_adnc.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Fraction with APOE4\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"APOE4 Status\"]].drop_duplicates().loc[:, [\"Source\", \"APOE4 Status\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"APOE4 Status\"] == \"Y\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction APOE4+\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Fraction APOE4+\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4a_barplot_fraction apoe4.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Fraction with dementia\n",
    "\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Cognitive Status\"]].drop_duplicates().loc[:, [\"Source\", \"Cognitive Status\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Cognitive Status\"] == \"Dementia\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction dementia\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Fraction dementia\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4a_barplot_fraction dementia.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Severe co-morbidity\n",
    "public_adata.obs[\"Severe co-morbidity\"] = \"False\"\n",
    "public_adata.obs.loc[public_adata.obs[\"Highest Lewy Body Disease\"] == \"Neocortical (Diffuse)\", \"Severe co-morbidity\"] = \"True\"\n",
    "public_adata.obs.loc[public_adata.obs[\"LATE\"] == \"Late Stage 3\", \"Severe co-morbidity\"] = \"True\"\n",
    "public_adata.obs.loc[public_adata.obs[\"Overall CAA Score\"] == \"Severe\", \"Severe co-morbidity\"] = \"True\"\n",
    "public_adata.obs.loc[public_adata.obs[\"Atherosclerosis\"] == \"Severe\", \"Severe co-morbidity\"] = \"True\"\n",
    "public_adata.obs.loc[public_adata.obs[\"Arteriolosclerosis\"] == \"Severe\", \"Severe co-morbidity\"] = \"True\"\n",
    "public_adata.obs.loc[public_adata.obs[\"Source\"].isin([\"Lau_2020\", \"Leng_2021\", \"Morabito_2021\", \"Yang_2022\"]), \"Severe co-morbidity\"] = \"\"\n",
    "\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Severe co-morbidity\"]].drop_duplicates().loc[:, [\"Source\", \"Severe co-morbidity\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Severe co-morbidity\"] == \"True\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction of donors with a severe co-morbidity\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Fraction of donors with a severe co-morbidity\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4a_barplot_fraction of donors with a severe co-morbidity.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "648519b6-3125-4347-b408-6e110fa12821",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 9a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcf7aa02-a87d-4bb7-8754-40eb1323a8a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# LBD\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Highest Lewy Body Disease\"]].drop_duplicates()\n",
    "df.loc[df[\"Highest Lewy Body Disease\"] == \"Not Identified (olfactory bulb assessed)\", \"Highest Lewy Body Disease\"] = \"Not Identified (olfactory bulb not assessed)\"\n",
    "df.loc[df[\"Highest Lewy Body Disease\"] == \"Olfactory bulb only\", \"Highest Lewy Body Disease\"] = \"Not Identified (olfactory bulb not assessed)\"\n",
    "df.loc[df[\"Highest Lewy Body Disease\"] == \"Amygdala-predominant\", \"Highest Lewy Body Disease\"] = \"Not Identified (olfactory bulb not assessed)\"\n",
    "df[\"Highest Lewy Body Disease\"] = df[\"Highest Lewy Body Disease\"].cat.remove_unused_categories()\n",
    "df = df.loc[:, [\"Source\", \"Highest Lewy Body Disease\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Highest Lewy Body Disease\"] != \"Reference\", :].copy()\n",
    "df.loc[df[\"Highest Lewy Body Disease\"] == \"Not Identified (olfactory bulb assessed)\", \"Highest Lewy Body Disease\"] = \"Not Identified (olfactory bulb not assessed)\"\n",
    "\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"Highest Lewy Body Disease\"] != \"\", :].copy()\n",
    "df[\"Highest Lewy Body Disease\"] = df[\"Highest Lewy Body Disease\"].cat.remove_unused_categories()\n",
    "df[\"Highest Lewy Body Disease\"] = df[\"Highest Lewy Body Disease\"].cat.reorder_categories([\"Not Identified (olfactory bulb not assessed)\", \"Brainstem-predominant\", \"Limbic (Transitional)\", \"Neocortical (Diffuse)\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Highest Lewy Body Disease\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"Highest Lewy Body Disease\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_lbd.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# LATE-NC\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"LATE\"]].drop_duplicates().loc[:, [\"Source\", \"LATE\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"LATE\"] != \"Reference\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"LATE\"] != \"\", :].copy()\n",
    "df[\"LATE\"] = df[\"LATE\"].cat.remove_unused_categories()\n",
    "df[\"LATE\"] = df[\"LATE\"].cat.reorder_categories([\"Not Identified\", \"LATE Stage 1\", \"LATE Stage 2\", \"LATE Stage 3\", \"Unclassifiable\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"LATE\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"LATE\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_late.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# CAA\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Overall CAA Score\"]].drop_duplicates().loc[:, [\"Source\", \"Overall CAA Score\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Overall CAA Score\"] != \"Reference\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"Overall CAA Score\"] != \"\", :].copy()\n",
    "df[\"Overall CAA Score\"] = df[\"Overall CAA Score\"].cat.remove_unused_categories()\n",
    "df[\"Overall CAA Score\"] = df[\"Overall CAA Score\"].cat.reorder_categories([\"Not identified\", \"Mild\", \"Moderate\", \"Severe\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Overall CAA Score\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"Overall CAA Score\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_caa.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Atherosclerosis\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Atherosclerosis\"]].drop_duplicates().loc[:, [\"Source\", \"Atherosclerosis\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Atherosclerosis\"] != \"Reference\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"Atherosclerosis\"] != \"\", :].copy()\n",
    "df[\"Atherosclerosis\"] = df[\"Atherosclerosis\"].cat.remove_unused_categories()\n",
    "df[\"Atherosclerosis\"] = df[\"Atherosclerosis\"].cat.reorder_categories([\"None\", \"Mild\", \"Moderate\", \"Severe\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Atherosclerosis\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"Atherosclerosis\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_athero.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Arteriolosclerosis\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Arteriolosclerosis\"]].drop_duplicates().loc[:, [\"Source\", \"Arteriolosclerosis\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Arteriolosclerosis\"] != \"Reference\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction stage\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df = df.loc[df[\"Arteriolosclerosis\"] != \"\", :].copy()\n",
    "df[\"Arteriolosclerosis\"] = df[\"Arteriolosclerosis\"].cat.remove_unused_categories()\n",
    "df[\"Arteriolosclerosis\"] = df[\"Arteriolosclerosis\"].cat.reorder_categories([\"None\", \"Mild\", \"Moderate\", \"Severe\"])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Arteriolosclerosis\", \"Fraction stage\"]].pivot(index=\"Source\", columns=\"Arteriolosclerosis\", values=\"Fraction stage\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], :].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_arterio.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Fraction of female donors\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Sex\"]].drop_duplicates().loc[:, [\"Source\", \"Sex\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Sex\"] == \"Female\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction female\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Fraction female\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_fraction female.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Age at death\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"Age at Death\"]].drop_duplicates()\n",
    "df[\"Age at Death\"] = df[\"Age at Death\"].astype(\"object\")\n",
    "df.loc[df[\"Age at Death\"] == \"90+\", \"Age at Death\"] = 90\n",
    "df[\"Age at Death\"] = [np.int64(np.round(np.float64(i), 0)) for i in df[\"Age at Death\"]]\n",
    "df[\"Binned age at death\"] = df[\"Age at Death\"].copy()\n",
    "df.loc[df[\"Age at Death\"] >= 90, \"Binned age at death\"] = \">90\"\n",
    "df.loc[df[\"Age at Death\"] < 80, \"Binned age at death\"] = \"<80\"\n",
    "df.loc[(df[\"Age at Death\"] >= 80) & (df[\"Age at Death\"] < 90), \"Binned age at death\"] = \"80-90\"\n",
    "df[\"Binned age at death\"] = df[\"Binned age at death\"].astype(\"category\")\n",
    "df[\"Binned age at death\"] = df[\"Binned age at death\"].cat.reorder_categories([\"<80\", \"80-90\", \">90\"])\n",
    "df = df.loc[:, [\"Source\", \"Binned age at death\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df[\"Binned age at death\"] = df[\"Binned age at death\"].astype(\"str\")\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction age\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Binned age at death\", \"Fraction age\"]].pivot(index=\"Source\", columns=\"Binned age at death\", values=\"Fraction age\")\n",
    "df = df.loc[[\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"], [\"<80\", \"80-90\", \">90\"]].copy()\n",
    "ax = df.plot(kind=\"barh\", stacked=True, cmap=\"YlGnBu\", **{\"width\": 0.8});\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1));\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9a_barplot_age at death.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7237a7a7-8b1e-4be7-a495-65ef5df19653",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4b and Extended Data Figure 9b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5060f149-dc01-4c44-93da-0e1283ad9eae",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Number of donors\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\"]].drop_duplicates().loc[:, \"Source\"].value_counts(sort=False).reset_index()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"count\": \"No. donors\",\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "df = df.sort_values(by=\"Source\", ascending=False)\n",
    "print(\"Min: \" + str(df[\"No. donors\"].min()) + \" Max: \" + str(df[\"No. donors\"].max()))\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "ax = sns.scatterplot(data=df, x=\"No. donors\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_number of donors.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"No. donors\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_number of donors.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"No. donors\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"No. donors\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# Post mortem interval\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\", \"PMI\"]].drop_duplicates()\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"PMI\"] = [np.float64(i) for i in df[\"PMI\"].replace(\"\", np.nan)]\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "ax = sns.boxplot(data=df, x=\"PMI\", y=\"Source\", hue=\"SEA-AD\", showfliers=False);\n",
    "ax = sns.stripplot(data=df, x=\"PMI\", y=\"Source\", color=\"grey\", size=2, alpha=0.5, jitter=0.2);\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd,  \"output\", \"Extended Data Figure 9b_barplot_PMI.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "df = df.loc[:, [\"Source\", \"PMI\"]].groupby(\"Source\").mean().reset_index()\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "df = df.sort_values(by=\"Source\", ascending=False)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "print(\"Min: \" + str(df[\"PMI\"].min()) + \" Max: \" + str(df[\"PMI\"].max()))\n",
    "df[\"PMI\"] = df[\"PMI\"] * -1\n",
    "ax = sns.scatterplot(data=df, x=\"PMI\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_PMI.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"PMI\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"PMI\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# Cells in the PFC\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\"]].value_counts(sort=False).reset_index()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"count\": \"Cells in DLPFC\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "print(\"Min: \" + str(df[\"Cells in DLPFC\"].min()) + \" Max: \" + str(df[\"Cells in DLPFC\"].max()))\n",
    "ax = sns.scatterplot(data=df, x=\"Cells in DLPFC\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_cells in DLPFC.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Cells in DLPFC\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_cells in DLPFC.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"Cells in DLPFC\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"Cells in DLPFC\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# Cells per donor\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "df = public_adata.obs.loc[selected_cells, [\"Source\", \"Donor ID\"]].value_counts(sort=False).reset_index().drop([\"Donor ID\"], axis=1).groupby(\"Source\").mean().reset_index()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"count\": \"Cells per donor\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "print(\"Min: \" + str(df[\"Cells per donor\"].min()) + \" Max: \" + str(df[\"Cells per donor\"].max()))\n",
    "ax = sns.scatterplot(data=df, x=\"Cells per donor\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_cells per donor.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Cells per donor\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_cells per donor.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"Cells per donor\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"Cells per donor\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# UMIs per cell\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "df = public_adata.obs.loc[selected_cells, [\"Number of UMIs\", \"Source\"]].groupby(\"Source\").mean().reset_index()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"Number of UMIs\": \"UMIs per cell\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "print(\"Min: \" + str(df[\"UMIs per cell\"].min()) + \" Max: \" + str(df[\"UMIs per cell\"].max()))\n",
    "ax = sns.scatterplot(data=df, x=\"UMIs per cell\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_umis per cell.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"UMIs per cell\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_umis per cell.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"UMIs per cell\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"UMIs per cell\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# Genes per cell\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "df = public_adata.obs.loc[selected_cells, [\"Genes detected\", \"Source\"]].groupby(\"Source\").mean().reset_index()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"Genes detected\": \"Genes per cell\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "print(\"Min: \" + str(df[\"Genes per cell\"].min()) + \" Max: \" + str(df[\"Genes per cell\"].max()))\n",
    "ax = sns.scatterplot(data=df, x=\"Genes per cell\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_genes per cell.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Genes per cell\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_genes per cell.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"Genes per cell\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"Genes per cell\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# Fraction of mitochondrial UMIs per cell\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "df = public_adata.obs.loc[selected_cells, [\"Fraction mitochondrial UMIs\", \"Source\"]].groupby(\"Source\").mean().reset_index()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"Fraction mitochondrial UMIs\": \"Fraction MT UMIs per cell\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "print(\"Min: \" + str(df[\"Fraction MT UMIs per cell\"].min()) + \" Max: \" + str(df[\"Fraction MT UMIs per cell\"].max()))\n",
    "df[\"Fraction MT UMIs per cell\"] = df[\"Fraction MT UMIs per cell\"] * -1\n",
    "ax = sns.scatterplot(data=df, x=\"Fraction MT UMIs per cell\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "df[\"Fraction MT UMIs per cell\"] = df[\"Fraction MT UMIs per cell\"] * -1\n",
    "ax.set(xticklabels=[], yticklabels=[]);\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_fraction mt umis per cell.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Fraction MT UMIs per cell\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_fraction mt umis per cell.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"Fraction MT UMIs per cell\", ascending=True).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"Fraction MT UMIs per cell\"] = tmp[\"index\"]\n",
    "plt.show()\n",
    "\n",
    "# Fraction of cells that passed QC\n",
    "plt.rcParams[\"figure.figsize\"] = (2,0.4)\n",
    "df = public_adata.obs.loc[(public_adata.obs[\"Cognitive Status\"] != \"Reference\"), [\"Source\", \"Used in analysis\"]].groupby(\"Source\").value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Used in analysis\"] == \"True\", :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction Passed QC\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Size\"] = 1\n",
    "df.loc[df[\"Source\"] == \"SEA-AD\", \"Size\"] = 5\n",
    "print(\"Min: \" + str(df[\"Fraction Passed QC\"].min()) + \" Max: \" + str(df[\"Fraction Passed QC\"].max()))\n",
    "ax = sns.scatterplot(data=df, x=\"Fraction Passed QC\", y=1, hue=\"SEA-AD\", palette=\"tab10\", size=\"Size\");\n",
    "ax.set(xticklabels=[], yticklabels=[], xlim=(0,1));\n",
    "ax.tick_params(bottom=False, left=False);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4b_metricplot_fraction passed qc.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.barplot(data=df, x=\"Fraction Passed QC\", y=\"Source\", hue=\"SEA-AD\");\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 9b_barplot_fraction passed qc.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "tmp = df.sort_values(by=\"Fraction Passed QC\", ascending=False).reset_index(drop=True).reset_index()\n",
    "tmp.index = tmp[\"Source\"].copy()\n",
    "ranks.loc[tmp.index, \"Fraction Passed QC\"] = tmp[\"index\"]\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6bcaf07-5861-4aa5-ab99-b1422413a5d8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab8c4288-3c64-4918-83e6-e09cb13ee4de",
   "metadata": {},
   "outputs": [],
   "source": [
    "public_adata = public_adata[selected_cells, :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "075d1320-be37-429f-890c-d500e8b76957",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Signature_Scores[\"index\"] = Signature_Scores[\"index\"].astype(\"category\")\n",
    "Signature_Scores[\"index\"] = Signature_Scores[\"index\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "Signature_Scores[\"SEA-AD\"] = Signature_Scores[\"index\"] == \"SEA-AD\"\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "ax = sns.boxplot(data=Signature_Scores, x=\"value\", y=\"index\", hue=\"SEA-AD\", palette=\"tab10\", showfliers=False);\n",
    "ax = sns.stripplot(data=Signature_Scores, x=\"value\", y=\"index\", color=\"grey\", size=2, alpha=0.5, jitter=0.2);\n",
    "ax.set(ylabel=None, xlabel=\"Supertype signature scores\");\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4c_boxplot_Supertype signature scores.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91a71a4-ffbc-41cf-b0c2-118e343ee78b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = public_adata.obs.loc[:, [\"Source\", \"Supertype\", \"Supertype confidence\"]].groupby([\"Source\", \"Supertype\"]).mean().reset_index()\n",
    "df[\"Source\"] = df[\"Source\"].astype(\"category\")\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "ax = sns.boxplot(data=df, x=\"Supertype confidence\", y=\"Source\", hue=\"SEA-AD\", palette=\"tab10\", showfliers=False);\n",
    "ax = sns.stripplot(data=df, x=\"Supertype confidence\", y=\"Source\", color=\"grey\", size=2, alpha=0.5, jitter=0.2);\n",
    "ax.set(ylabel=None, xlabel=\"Supertype confidence\");\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4c_boxplot_Supertype confidence.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f1dd8b2-18da-46cb-ba36-e4447fd81f48",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "695c097c-74e6-43b7-8594-172f3f8d6942",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "\n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]]\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"]\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "sc.pl.umap(\n",
    "    public_adata,\n",
    "    color=[\"Source\"],\n",
    "    frameon=False,\n",
    "    size=0.05,\n",
    "    title=None,\n",
    "    save=\"_Public Source.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Public Source.pdf\"), os.path.join(pwd, \"output\", \"Figure 4d_umap_Public Source.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    public_adata,\n",
    "    color=[\"Subclass\"],\n",
    "    legend_loc=\"on data\",\n",
    "    palette=subclass_colors,\n",
    "    frameon=False,\n",
    "    size=0.05,\n",
    "    save=\"_Public Subclass.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Public Subclass.pdf\"), os.path.join(pwd, \"output\", \"Figure 4d_umap_Public Subclass.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ace5705-fde2-4632-9176-010c22b4813c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Figure 9c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc8353df-b809-4b4a-9e6c-f2602c5256cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "neighborhoods = {\n",
    "    \"MGE\": [\"Sst Chodl\", \"Sst\", \"Pvalb\", \"Chandelier\"], # 4\n",
    "    \"CGE\": [\"Lamp5\", \"Lamp5 Lhx6\", \"Pax6\", \"Sncg\", \"Vip\"], # 5\n",
    "    \"IT\": [\"L2/3 IT\", \"L4 IT\", \"L5 IT\", \"L6 IT\", \"L6 IT Car3\"], # 5\n",
    "    \"Deep Projecting\": [\"L5 ET\", \"L5/6 NP\", \"L6b\", \"L6 CT\"], # 4\n",
    "    \"Glia\": [\"Oligodendrocyte\", \"OPC\", \"Astrocyte\"], # 3\n",
    "    \"Vascular and Immune\": [\"Endothelial\", \"VLMC\", \"Microglia-PVM\"] # 3\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf9a927-5702-4b9d-beb5-b8f34899f0ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(\n",
    "    public_adata[public_adata.obs[\"Subclass\"].isin(neighborhoods[\"MGE\"]), :],\n",
    "    basis=\"X_MGE_umap\",\n",
    "    color=[\"Signature Score\"],\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    cmap=\"RdBu_r\",\n",
    "    size=1,\n",
    "    save=\"_Public MGE Signature Score.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"X_MGE_umap_Public MGE Signature Score.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 9c_umap_Public MGE Signature Score.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "268ab922-ddb8-46a8-be93-c2619a16d587",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Figure 9d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9de5c5c9-5f9c-462a-8a32-763e287204b6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "\n",
    "supertype_colors = color_order.loc[:, [\"cluster_label\", \"cluster_color\"]]\n",
    "supertype_colors.index = supertype_colors[\"cluster_label\"]\n",
    "supertype_colors = supertype_colors[\"cluster_color\"].to_dict()\n",
    "supertype_colors[\"VLMC_2-SEAAD\"] = supertype_colors['Pericyte_2-SEAAD']\n",
    "supertype_colors[\"VLMC_3-SEAAD\"] = supertype_colors['Micro-PVM_4-SEAAD']\n",
    "\n",
    "for i,j in neighborhoods.items():\n",
    "    sc.pl.embedding(\n",
    "        public_adata[public_adata.obs[\"Subclass\"].isin(j), :],\n",
    "        basis=\"X_\" + i + \"_umap\",\n",
    "        color=[\"Supertype\"],\n",
    "        legend_loc=\"on data\",\n",
    "        palette=supertype_colors,\n",
    "        frameon=False,\n",
    "        size=0.5,\n",
    "        save=\"_Public \" + i + \" Supertype.pdf\"\n",
    "    )\n",
    "    os.rename(os.path.join(pwd, \"figures\", \"X_\" + i + \"_umap_Public \" + i + \" Supertype.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 9d_umap_Public \" + i + \" Supertype.pdf\"))\n",
    "\n",
    "    sc.pl.embedding(\n",
    "        public_adata[public_adata.obs[\"Subclass\"].isin(j), :],\n",
    "        basis=\"X_\" + i + \"_umap\",\n",
    "        color=[\"Source\"],\n",
    "        legend_loc=\"on data\",\n",
    "        frameon=False,\n",
    "        size=0.5,\n",
    "        save=\"_Public \" + i + \" Source.pdf\"\n",
    "    )\n",
    "    os.rename(os.path.join(pwd, \"figures\", \"X_\" + i + \"_umap_Public \" + i + \" Source.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 9d_umap_Public \" + i + \" Source.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3be3bdb-1997-4a5f-9d23-66146a1b96d2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13183883-c1d6-498e-96ca-afd3b452d8d3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Neuronal\n",
    "cell_types = np.intersect1d(np.intersect1d(A9_RNAseq_neuronal[\"Cell Type\"].unique(), Mathys_RNAseq_neuronal[\"Cell Type\"].unique()), Green_RNAseq_neuronal[\"Cell Type\"].unique())\n",
    "A9_RNAseq_neuronal = A9_RNAseq_neuronal.loc[A9_RNAseq_neuronal[\"Cell Type\"].isin(cell_types), :].copy()\n",
    "Mathys_RNAseq_neuronal = Mathys_RNAseq_neuronal.loc[Mathys_RNAseq_neuronal[\"Cell Type\"].isin(cell_types), :].copy()\n",
    "Green_RNAseq_neuronal = Green_RNAseq_neuronal.loc[Green_RNAseq_neuronal[\"Cell Type\"].isin(cell_types), :].copy()\n",
    "\n",
    "df = pd.concat(\n",
    "    [\n",
    "        A9_RNAseq_neuronal.loc[A9_RNAseq_neuronal[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(), \n",
    "        Mathys_RNAseq_neuronal.loc[Mathys_RNAseq_neuronal[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(),\n",
    "        Green_RNAseq_neuronal.loc[Green_RNAseq_neuronal[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean()\n",
    "\n",
    "    ],\n",
    "    axis=1\n",
    ").reset_index()\n",
    "df.columns = [\"Cell Type\", \"SEA-AD\", \"Base Inclusion Probability\", \"Mathys_2023\", \"Mathys Inclusion Probability\", \"Green_2023\", \"Green Inclusion Probability\"]\n",
    "df[\"Cell Type\"] = df[\"Cell Type\"].astype(\"category\")\n",
    "df[\"Cell Type\"] = df[\"Cell Type\"].cat.reorder_categories(\n",
    "    color_order.loc[color_order[\"cluster_label\"].isin(df[\"Cell Type\"]), \"cluster_label\"]\n",
    ")\n",
    "\n",
    "df.index = df[\"Cell Type\"].copy()\n",
    "plt.rcParams[\"figure.figsize\"] = (20,2)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[df[\"Cell Type\"].cat.categories, [\"SEA-AD\", \"Mathys_2023\", \"Green_2023\"]].T,\n",
    "    cmap=\"RdBu_r\",\n",
    "    xticklabels=True,\n",
    "    center=0,\n",
    "    vmin=-1\n",
    ");\n",
    "ax.set(xlabel=None, title=\"Effect size across AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4e_heatmap_Neuronal supertype replication.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Non-neuronal\n",
    "cell_types = np.intersect1d(np.intersect1d(A9_RNAseq_non_neuronal[\"Cell Type\"].unique(), Mathys_RNAseq_non_neuronal[\"Cell Type\"].unique()), Green_RNAseq_non_neuronal[\"Cell Type\"].unique())\n",
    "A9_RNAseq_non_neuronal = A9_RNAseq_non_neuronal.loc[A9_RNAseq_non_neuronal[\"Cell Type\"].isin(cell_types), :].copy()\n",
    "Mathys_RNAseq_non_neuronal = Mathys_RNAseq_non_neuronal.loc[Mathys_RNAseq_non_neuronal[\"Cell Type\"].isin(cell_types), :].copy()\n",
    "Green_RNAseq_non_neuronal = Green_RNAseq_non_neuronal.loc[Green_RNAseq_non_neuronal[\"Cell Type\"].isin(cell_types), :].copy()\n",
    "\n",
    "df = pd.concat(\n",
    "    [\n",
    "        A9_RNAseq_non_neuronal.loc[A9_RNAseq_non_neuronal[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(), \n",
    "        Mathys_RNAseq_non_neuronal.loc[Mathys_RNAseq_non_neuronal[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(),\n",
    "        Green_RNAseq_non_neuronal.loc[Green_RNAseq_non_neuronal[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(), \n",
    "\n",
    "    ],\n",
    "    axis=1\n",
    ").reset_index()\n",
    "df.columns = [\"Cell Type\", \"SEA-AD\", \"Base Inclusion Probability\", \"Mathys_2023\", \"Mathys Inclusion Probability\", \"Green_2023\", \"Green Inclusion Probability\"]\n",
    "df[\"Cell Type\"] = df[\"Cell Type\"].astype(\"category\")\n",
    "df[\"Cell Type\"] = df[\"Cell Type\"].cat.reorder_categories(\n",
    "    color_order.loc[color_order[\"cluster_label\"].isin(df[\"Cell Type\"]), \"cluster_label\"]\n",
    ")\n",
    "\n",
    "df.index = df[\"Cell Type\"].copy()\n",
    "plt.rcParams[\"figure.figsize\"] = (4,2)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[df[\"Cell Type\"].cat.categories, [\"SEA-AD\", \"Mathys_2023\", \"Green_2023\"]].T,\n",
    "    cmap=\"RdBu_r\",\n",
    "    xticklabels=True,\n",
    "    center=0,\n",
    "    vmin=-1\n",
    ");\n",
    "ax.set(xlabel=None, title=\"Effect size across AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4e_heatmap_Non-neuronal supertype replication.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67c7f009-46ef-4a6e-a1e9-2e1e03b207cb",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 9e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63769c4-9495-492b-bf99-1937e7d738da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Green_2023\n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "subclass_colors[\"Not Significant\"] = \"lightgrey\"\n",
    "\n",
    "A9_RNAseq_neuronal = A9_RNAseq_neuronal.loc[A9_RNAseq_neuronal[\"Cell Type\"].isin(Green_RNAseq_neuronal[\"Cell Type\"]), :].copy()\n",
    "Green_RNAseq_neuronal = Green_RNAseq_neuronal.loc[Green_RNAseq_neuronal[\"Cell Type\"].isin(A9_RNAseq_neuronal[\"Cell Type\"]), :].copy()\n",
    "\n",
    "df = pd.concat(\n",
    "    [\n",
    "        A9_RNAseq_neuronal.loc[A9_RNAseq_neuronal[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(), \n",
    "        Green_RNAseq_neuronal.loc[Green_RNAseq_neuronal[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean()\n",
    "    ],\n",
    "    axis=1\n",
    ").reset_index()\n",
    "df.columns = [\"Cell Type\", \"Base Effect Size\", \"Base Inclusion Probability\", \"New Effect Size\", \"New Inclusion Probability\"]\n",
    "df[\"Subclass\"] = [re.sub(\"^(.*)_[0-9]+(_[0-9]+)?(-SEAAD)?$\", \"\\\\1\", i).replace(\"Lamp5_Lhx6\", \"Lamp5 Lhx6\") for i in df[\"Cell Type\"]]\n",
    "df[\"Subclass\"] = df[\"Subclass\"].astype(\"category\")\n",
    "subclass_order = color_order.loc[color_order[\"subclass_label\"].isin(df[\"Subclass\"]), \"subclass_label\"].drop_duplicates().to_list()\n",
    "subclass_order.append(\"Not Significant\")\n",
    "df[\"Subclass\"] = df[\"Subclass\"].cat.reorder_categories(subclass_order)\n",
    "slope, intercept, r_value, p_value, std_err = sp_stats.linregress(\n",
    "    df[\"Base Effect Size\"],\n",
    "    df[\"New Effect Size\"]\n",
    ")\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Base Effect Size\",\n",
    "    y=\"New Effect Size\",\n",
    "    hue=\"Subclass\",\n",
    "    palette=subclass_colors\n",
    ");\n",
    "ax.set(xlabel=\"Effect size in SEA-AD\", ylabel=\"Effect size in Green_2023\", title=\"Correlation=\" + str(np.round(r_value,2)));\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4e_scatterplot_Neuronal supertype replication Green.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# Mathys_2023\n",
    "\n",
    "A9_RNAseq_neuronal = A9_RNAseq_neuronal.loc[A9_RNAseq_neuronal[\"Cell Type\"].isin(Mathys_RNAseq_neuronal[\"Cell Type\"]), :].copy()\n",
    "Mathys_RNAseq_neuronal = Mathys_RNAseq_neuronal.loc[Mathys_RNAseq_neuronal[\"Cell Type\"].isin(A9_RNAseq_neuronal[\"Cell Type\"]), :].copy()\n",
    "\n",
    "df = pd.concat(\n",
    "    [\n",
    "        A9_RNAseq_neuronal.loc[A9_RNAseq_neuronal[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean(), \n",
    "        Mathys_RNAseq_neuronal.loc[Mathys_RNAseq_neuronal[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\", \"Inclusion probability\"]].groupby(\"Cell Type\").mean()\n",
    "    ],\n",
    "    axis=1\n",
    ").reset_index()\n",
    "df.columns = [\"Cell Type\", \"Base Effect Size\", \"Base Inclusion Probability\", \"New Effect Size\", \"New Inclusion Probability\"]\n",
    "df[\"Subclass\"] = [re.sub(\"^(.*)_[0-9]+(_[0-9]+)?(-SEAAD)?$\", \"\\\\1\", i).replace(\"Lamp5_Lhx6\", \"Lamp5 Lhx6\") for i in df[\"Cell Type\"]]\n",
    "df[\"Subclass\"] = df[\"Subclass\"].astype(\"category\")\n",
    "subclass_order = color_order.loc[color_order[\"subclass_label\"].isin(df[\"Subclass\"]), \"subclass_label\"].drop_duplicates().to_list()\n",
    "subclass_order.append(\"Not Significant\")\n",
    "df[\"Subclass\"] = df[\"Subclass\"].cat.reorder_categories(subclass_order)\n",
    "slope, intercept, r_value, p_value, std_err = sp_stats.linregress(\n",
    "    df[\"Base Effect Size\"],\n",
    "    df[\"New Effect Size\"]\n",
    ")\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Base Effect Size\",\n",
    "    y=\"New Effect Size\",\n",
    "    hue=\"Subclass\",\n",
    "    palette=subclass_colors\n",
    ");\n",
    "ax.set(xlabel=\"Effect size in SEA-AD\", ylabel=\"Effect size in Mathys_2023\", title=\"Correlation=\" + str(np.round(r_value,2)));\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4e_scatterplot_Neuronal supertype replication Mathys.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6289f6a0-f52a-4745-9398-7d9ac49688a4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "008c670f-bd38-4504-9e47-b5a11c06a6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### public_adata.obs[\"Source_Donor ID\"] = public_adata.obs[\"Source\"].astype(\"str\") + \"__\" + public_adata.obs[\"Donor ID\"].astype(\"str\")\n",
    "public_adata.obs[\"Supertype\"] = public_adata.obs[\"Supertype\"].cat.remove_unused_categories()\n",
    "df = public_adata.obs.loc[selected_cells & (public_adata.obs[\"Class\"] != 'Non-neuronal and Non-neural'), [\"Supertype\", \"Source_Donor ID\"]].groupby([\"Source_Donor ID\"]).value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Supertype\"].isin(public_adata.obs.loc[public_adata.obs[\"Class\"] != 'Non-neuronal and Non-neural', \"Supertype\"].unique()), :]\n",
    "df[\"Source\"] = [i[0] for i in df[\"Source_Donor ID\"].str.split(\"__\")]\n",
    "df[\"Donor ID\"] = [i[1] for i in df[\"Source_Donor ID\"].str.split(\"__\")]\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Relative abundance\",\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"Relative abundance\"] = df[\"Relative abundance\"] > 0\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Supertype\", \"Relative abundance\"]].groupby([\"Source\", \"Supertype\"]).value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Supertype\"].isin(public_adata.obs.loc[public_adata.obs[\"Class\"] != 'Non-neuronal and Non-neural', \"Supertype\"].unique()), :]\n",
    "df = df.loc[df[\"Relative abundance\"] == True, :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction of donors a neuronal supertype was captured in\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Source\"] = df[\"Source\"].astype(\"category\")\n",
    "df[\"Source\"] = df[\"Source\"].cat.add_categories([\"Olah_2020\", \"Yang_2022\"])\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "ax = sns.boxplot(data=df, x=\"Fraction of donors a neuronal supertype was captured in\", y=\"Source\", showfliers=False);\n",
    "ax = sns.stripplot(data=df, x=\"Fraction of donors a neuronal supertype was captured in\", y=\"Source\", color=\"grey\", size=2, alpha=0.5);\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4f_Donor neuronal supertype fraction.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "public_adata.obs[\"Source_Donor ID\"] = public_adata.obs[\"Source\"].astype(\"str\") + \"__\" + public_adata.obs[\"Donor ID\"].astype(\"str\")\n",
    "public_adata.obs[\"Supertype\"] = public_adata.obs[\"Supertype\"].cat.remove_unused_categories()\n",
    "df = public_adata.obs.loc[selected_cells & (public_adata.obs[\"Class\"] == 'Non-neuronal and Non-neural'), [\"Supertype\", \"Source_Donor ID\"]].groupby([\"Source_Donor ID\"]).value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Supertype\"].isin(public_adata.obs.loc[public_adata.obs[\"Class\"] == 'Non-neuronal and Non-neural', \"Supertype\"].unique()), :]\n",
    "df[\"Source\"] = [i[0] for i in df[\"Source_Donor ID\"].str.split(\"__\")]\n",
    "df[\"Donor ID\"] = [i[1] for i in df[\"Source_Donor ID\"].str.split(\"__\")]\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Relative abundance\",\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"Relative abundance\"] = df[\"Relative abundance\"] > 0\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (2,2)\n",
    "df = df.loc[:, [\"Source\", \"Supertype\", \"Relative abundance\"]].groupby([\"Source\", \"Supertype\"]).value_counts(normalize=True, sort=False).reset_index()\n",
    "df = df.loc[df[\"Supertype\"].isin(public_adata.obs.loc[public_adata.obs[\"Class\"] == 'Non-neuronal and Non-neural', \"Supertype\"].unique()), :]\n",
    "df = df.loc[df[\"Relative abundance\"] == True, :].copy()\n",
    "df = df.rename(\n",
    "    {\n",
    "        \"proportion\": \"Fraction of donors a non-neuronal supertype was captured in\"\n",
    "    },\n",
    "    axis=1\n",
    ")\n",
    "df[\"SEA-AD\"] = df[\"Source\"] == \"SEA-AD\"\n",
    "df[\"Source\"] = df[\"Source\"].astype(\"category\")\n",
    "df[\"Source\"] = df[\"Source\"].cat.reorder_categories(np.flip([\"SEA-AD\", \"Green_2023\", \"Mathys_2023\", \"Cain_2022\", \"Yang_2022\", \"Morabito_2021\", \"Leng_2021\", \"Lau_2020\", \"Zhou_2020\", \"Olah_2020\", \"Mathys_2019\"]))\n",
    "df = df.loc[~df[\"Supertype\"].isin([\"VLMC_2-SEAAD\", \"VLMC_3-SEAAD\", \"Pericyte_2-SEAAD\"]), :]\n",
    "ax = sns.boxplot(data=df, x=\"Fraction of donors a non-neuronal supertype was captured in\", y=\"Source\", showfliers=False);\n",
    "ax = sns.stripplot(data=df, x=\"Fraction of donors a non-neuronal supertype was captured in\", y=\"Source\", color=\"grey\", size=2, alpha=0.5);\n",
    "ax.set(ylabel=None);\n",
    "plt.legend([],[], frameon=False);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4f_Donor non-neuronal supertype fraction.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d027c1-b175-48a5-ae42-edfa933f995e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 4g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4247581-79a1-4bbe-8dfb-93e8d99e81ed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# SEA-AD\n",
    "neurons = copy.copy(A9_RNAseq_neuronal_adata)\n",
    "results_table_neurons = copy.copy(A9_RNAseq_neuronal)\n",
    "neurons = neurons[:, np.intersect1d(neurons.var_names, results_table_neurons[\"Cell Type\"])].copy()\n",
    "glia = copy.copy(A9_RNAseq_non_neuronal_adata)\n",
    "results_table_glia = copy.copy(A9_RNAseq_non_neuronal)\n",
    "glia = glia[:, np.intersect1d(glia.var_names, results_table_glia[\"Cell Type\"])].copy()\n",
    "x = ((neurons.X > 0).sum(axis=0) / neurons.shape[0]).tolist()\n",
    "x.extend(((glia.X > 0).sum(axis=0) / glia.shape[0]).tolist())\n",
    "y = results_table_neurons.loc[results_table_neurons[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list()\n",
    "y.extend(results_table_glia.loc[results_table_glia[\"Covariate\"] == \"Continuous_Pseudo-progression_Score\", [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list())\n",
    "y = np.abs(y)\n",
    "plt.rcParams[\"figure.figsize\"] = (1.5,3)\n",
    "ax = sns.scatterplot(\n",
    "    y=x,\n",
    "    x=y,\n",
    ");\n",
    "plt.axhline(0.75, linestyle=\"--\", color=\"lightgrey\")\n",
    "ax.set(ylabel=\"Fraction of donors with a supertype\", xlabel=\"Effect Size\", title=\"SEA-AD\", ylim=(0,1.1));\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4g_scatterplot_SEA-AD effect size versus zero inflation.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Green_2023\n",
    "neurons = copy.copy(Green_RNAseq_neuronal_adata)\n",
    "results_table_neurons = copy.copy(Green_RNAseq_neuronal)\n",
    "neurons = neurons[:, np.intersect1d(neurons.var_names, results_table_neurons[\"Cell Type\"])].copy()\n",
    "glia = copy.copy(Green_RNAseq_non_neuronal_adata)\n",
    "results_table_glia = copy.copy(Green_RNAseq_non_neuronal)\n",
    "glia = glia[:, np.intersect1d(glia.var_names, results_table_glia[\"Cell Type\"])].copy()\n",
    "x = ((neurons.X > 0).sum(axis=0) / neurons.shape[0]).tolist()\n",
    "x.extend(((glia.X > 0).sum(axis=0) / glia.shape[0]).tolist())\n",
    "y = results_table_neurons.loc[results_table_neurons[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list()\n",
    "y.extend(results_table_glia.loc[results_table_glia[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list())\n",
    "y = np.abs(y)\n",
    "plt.rcParams[\"figure.figsize\"] = (1.5,3)\n",
    "ax = sns.scatterplot(\n",
    "    y=x,\n",
    "    x=y,\n",
    ");\n",
    "plt.axhline(0.75, linestyle=\"--\", color=\"lightgrey\")\n",
    "ax.set(ylabel=\"Fraction of donors with a supertype\", xlabel=\"Effect Size\", title=\"Green_2023\", ylim=(0,1.1));\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4g_scatterplot_Green effect size versus zero inflation.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# Mathys_2023\n",
    "neurons = copy.copy(Mathys_RNAseq_neuronal_adata)\n",
    "results_table_neurons = copy.copy(Mathys_RNAseq_neuronal)\n",
    "neurons = neurons[:, np.intersect1d(neurons.var_names, results_table_neurons[\"Cell Type\"])].copy()\n",
    "glia = copy.copy(Mathys_RNAseq_non_neuronal_adata)\n",
    "results_table_glia = copy.copy(Mathys_RNAseq_non_neuronal)\n",
    "glia = glia[:, np.intersect1d(glia.var_names, results_table_glia[\"Cell Type\"])].copy()\n",
    "x = ((neurons.X > 0).sum(axis=0) / neurons.shape[0]).tolist()\n",
    "x.extend(((glia.X > 0).sum(axis=0) / glia.shape[0]).tolist())\n",
    "y = results_table_neurons.loc[results_table_neurons[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list()\n",
    "y.extend(results_table_glia.loc[results_table_glia[\"Covariate\"] == \"Overall_AD_neuropathological_Change_codes\", [\"Cell Type\", \"Final Parameter\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list())\n",
    "y = np.abs(y)\n",
    "plt.rcParams[\"figure.figsize\"] = (1.5,3)\n",
    "ax = sns.scatterplot(\n",
    "    y=x,\n",
    "    x=y,\n",
    ");\n",
    "plt.axhline(0.75, linestyle=\"--\", color=\"lightgrey\")\n",
    "ax.set(ylabel=\"Fraction of donors with a supertype\", xlabel=\"Effect Size\", title=\"Mathys_2023\", ylim=(0,1.1));\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 4g_scatterplot_Mathys effect size versus zero inflation.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c0a7bbc-0de8-43f0-98b2-288ba946239d",
   "metadata": {},
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bc3c1c4-d38a-4a34-8852-05939f1ee22f",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
