{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f04549b5-991f-4139-acc1-2cc23b47071c",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f36e257e-5f0d-42c3-80a5-136d6f8d3057",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import re\n",
    "import glob\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "from helper_functions import *\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=500, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "pwd = os.getcwd()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc7b14fa-d94e-4fe8-92ff-321d00aa1df3",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6e4426a-c520-46a3-a207-c841993ac349",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "\n",
    "# From 00_build_input_data.py\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", \"SEAAD_MTG_RNAseq_final-nuclei_no_data.2024-02-13.h5ad\"))\n",
    "\n",
    "# Subclass UMAP coordinates from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "Microglia_PVM_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", \"Microglia-PVM\", \"obs_names.csv\"), index_col=0).index\n",
    "Microglia_PVM_umap = np.load(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", \"Microglia-PVM\", \"X_umap.npy\"))\n",
    "\n",
    "L5_IT_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", \"L5 IT\", \"obs_names.csv\"), index_col=0).index\n",
    "L5_IT_umap = np.load(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", \"L5 IT\", \"X_umap.npy\"))\n",
    "\n",
    "# scCODA results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "datasets = glob.glob(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", \"*\", \"*\", \"*.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f97e8c-0abc-4fe5-8397-43272f4938d9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 3a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3973798-7988-45bc-a411-e7a74fcfd9e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"Reference Subclass\"] = adata.obs[\"Subclass\"].copy()\n",
    "adata.obs.loc[adata.obs[\"Neurotypical reference\"] == \"False\", \"Reference Subclass\"] = np.nan\n",
    "adata.uns[\"Reference Subclass_colors\"] = adata.uns[\"Subclass_colors\"].copy()\n",
    "\n",
    "adata.obs[\"Reference Supertype\"] = adata.obs[\"Supertype\"].copy()\n",
    "adata.obs.loc[adata.obs[\"Neurotypical reference\"] == \"False\", \"Reference Supertype\"] = np.nan\n",
    "adata.uns[\"Reference Supertype_colors\"] = adata.uns[\"Supertype_colors\"].copy()\n",
    "\n",
    "sc.pp.subsample(adata, fraction=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba3553f3-15c6-43cb-a041-b3f755b4628e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Subclass plots\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"Reference Subclass\"],\n",
    "    size=1,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    na_in_legend=False,\n",
    "    save=\"_Whole Taxonomy_Reference Subclass.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Whole Taxonomy_Reference Subclass.pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_Whole Taxonomy_Reference Subclass.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"Subclass\"],\n",
    "    size=1,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Whole Taxonomy_Subclass.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Whole Taxonomy_Subclass.pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_Whole Taxonomy_Subclass.pdf\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b57ecd-9b22-45cc-adc5-9b73e4a8e5d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# L5 IT supertype plot\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "\n",
    "sub = adata[L5_IT_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = L5_IT_umap\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Reference Supertype\"],\n",
    "    size=10,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    na_in_legend=False,\n",
    "    save=\"_L5 IT_Reference Supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_L5 IT_Reference Supertype.pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_L5 IT_Reference Supertype.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=10,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_L5 IT_Supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_L5 IT_Supertype.pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_L5 IT_Supertype.pdf\"))\n",
    "\n",
    "\n",
    "sub = adata[Microglia_PVM_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = Microglia_PVM_umap\n",
    "sub.obs.loc[~sub.obs[\"Reference Supertype\"].isin([\"Micro-PVM_1\", \"Micro-PVM_2\"]), \"Reference Supertype\"] = np.nan\n",
    "\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Reference Supertype\"],\n",
    "    size=10,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    na_in_legend=False,\n",
    "    save=\"_Microglia-PVM_Reference Supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia-PVM_Reference Supertype.pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_Microglia-PVM_Reference Supertype.pdf\"))\n",
    "\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype (non-expanded)\"],\n",
    "    size=10,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    palette=color_order.loc[color_order[\"cluster_label\"].isin([\"Micro-PVM_1\", \"Micro-PVM_2\"]), \"cluster_color\"].to_list(),\n",
    "    save=\"_Microglia-PVM_Supertype (non-expanded).pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia-PVM_Supertype (non-expanded).pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_Microglia-PVM_Supertype (non-expanded).pdf\"))\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=10,\n",
    "    legend_loc=\"on data\",\n",
    "    frameon=False,\n",
    "    legend_fontsize=16,\n",
    "    legend_fontweight=\"bold\",\n",
    "    title=\"\",\n",
    "    save=\"_Microglia-PVM_Supertype.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Microglia-PVM_Supertype.pdf\"), os.path.join(pwd, \"output\", \"Figure 3a_umap_Microglia-PVM_Supertype.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec6976a-8850-4806-9a15-2df9bd517815",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 3b and Extended Data Figure 8a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b4c1b9-b035-48d6-9a1f-ee73ba45f2c5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_size = {\n",
    "    \"A9_RNAseq\": [-1.75, 1.75],\n",
    "    \"MTG_RNAseq\": [-1.75, 1.75],\n",
    "    \"PFC_Mathys_2023\": [-0.5, 0.5],\n",
    "    \"PFC_Green_2023\": [-0.5, 0.5]\n",
    "\n",
    "}\n",
    "plot_order = {\n",
    "    \"Non-neuronal and Non-neural\": color_order.loc[color_order[\"class_label\"] == \"Non-neuronal and Non-neural\", \"cluster_label\"].to_list(),\n",
    "    \"Neuronal: Glutamatergic Neuronal: GABAergic\": color_order.loc[color_order[\"class_label\"] != \"Non-neuronal and Non-neural\", \"cluster_label\"].to_list(),\n",
    "}\n",
    "plot_colors = {\n",
    "    \"Non-neuronal and Non-neural\": color_order.loc[color_order[\"class_label\"] == \"Non-neuronal and Non-neural\", \"cluster_color\"].to_list(),\n",
    "    \"Neuronal: Glutamatergic Neuronal: GABAergic\": color_order.loc[color_order[\"class_label\"] != \"Non-neuronal and Non-neural\", \"cluster_color\"].to_list(),\n",
    "}\n",
    "plot_dims = {\n",
    "    \"Non-neuronal and Non-neural\": (10,4),\n",
    "    \"Neuronal: Glutamatergic Neuronal: GABAergic\": (30,4)\n",
    "}\n",
    "\n",
    "for i in datasets:\n",
    "    region = i.split(\"/\")[-3]\n",
    "    test = i.split(\"/\")[-2]\n",
    "    population = i.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    print(\"Region: \" + region + \" Test: \" + test + \" Population: \" + population)\n",
    "    \n",
    "    results_table = pd.read_csv(i, index_col=0)\n",
    "\n",
    "    missing_types = np.setdiff1d(plot_order[population], results_table[\"Cell Type\"])\n",
    "    for missing_type in missing_types:\n",
    "        to_add = results_table.loc[results_table[\"Cell Type\"] == results_table.loc[:, \"Cell Type\"].iloc[0], :]\n",
    "        to_add[\"Cell Type\"] = missing_type\n",
    "        to_add[\"Final Parameter\"] = 0\n",
    "        to_add[\"SD\"] = 0\n",
    "        to_add[\"Inclusion probability\"] = 0\n",
    "        results_table = pd.concat([results_table, to_add], axis=0)\n",
    "\n",
    "    extra_types = np.setdiff1d(results_table[\"Cell Type\"], plot_order[population])\n",
    "\n",
    "    if len(extra_types) > 0:\n",
    "        print(\"WARNING: Removing extra types \" + str(list(extra_types)))\n",
    "        results_table = results_table.loc[~results_table[\"Cell Type\"].isin(extra_types), :].copy()\n",
    "    \n",
    "    results_table = results_table.sort_values(by=\"Covariate\")\n",
    "\n",
    "    results_table[\"Credible\"] = results_table[\"Final Parameter\"] != 0\n",
    "    try:\n",
    "        results_table[\"Cell Type\"] = results_table[\"Cell Type\"].astype(\"category\")\n",
    "        results_table[\"Cell Type\"] = results_table[\"Cell Type\"].cat.reorder_categories(plot_order[population])\n",
    "\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    plt.rcParams['figure.figsize'] = plot_dims[population]\n",
    "    \n",
    "    df = results_table.loc[:, [\"Covariate\", \"Cell Type\", \"Final Parameter\"]].groupby([\"Covariate\", \"Cell Type\"]).mean().reset_index()\n",
    "    if (test == \"Continuous_Pseudo-progression_Score\" and region in [\"MTG_RNAseq\", \"A9_RNAseq\"]) or (test == \"Overall_AD_neuropathological_Change_codes\" and region in [\"PFC_Mathys_2023\", \"PFC_Green_2023\"]):\n",
    "        ax = sns.heatmap(df.pivot(index=\"Covariate\", columns=\"Cell Type\", values=\"Final Parameter\"), center=0, cmap=\"RdBu_r\", vmin=plot_size[region][0], vmax=plot_size[region][1]);\n",
    "        ax.set_ylabel(\"\");\n",
    "        ax.set_xlabel(\"\");\n",
    "        plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 8a_heatmap_\" + region + \"_\" + population + \"_along_\" + test + \".pdf\"), bbox_inches=\"tight\")\n",
    "        plt.show()\n",
    "\n",
    "    if region in [\"MTG_RNAseq\", \"A9_RNAseq\"]:\n",
    "        for covariate in np.unique(results_table[\"Covariate\"]):\n",
    "            if covariate not in [\"Cognitive_Status[T.No dementia]\", \"Overall_AD_neuropathological_Change_codes\", \"Continuous_Pseudo-progression_Score\"]:\n",
    "                continue\n",
    "    \n",
    "            print(\"Covariate: \" + covariate + \" Test: \" + test + \" Population: \" + population)\n",
    "    \n",
    "            df = results_table.loc[results_table[\"Covariate\"] == covariate, :]\n",
    "            \n",
    "            # Flip the sign of the effect size\n",
    "            if test == \"Cognitive_Status\" and covariate == \"Cognitive_Status[T.No dementia]\":\n",
    "                df[\"Final Parameter\"] = -1 * df[\"Final Parameter\"]\n",
    "                \n",
    "            ax = sns.barplot(data=df, x=\"Cell Type\", y=\"Final Parameter\", palette=plot_colors[population] if population in plot_colors.keys() else None);\n",
    "            ax.set_xlabel(\"\");\n",
    "            ax.set_ylabel(\"Effect size along\\n\" + covariate.replace(\"_\", \" \"));\n",
    "            ax.set_xticks(ax.get_xticks(), ax.get_xticklabels(), rotation=90)\n",
    "            ax.set_ylim(plot_size[region]);\n",
    "            plt.savefig(os.path.join(pwd, \"output\", \"Figure 3b_barplot_\" + region + \"_\" + population + \"_along_\" + test + \".pdf\"), bbox_inches=\"tight\")\n",
    "            plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9f0d70a-1457-4146-ac99-31be8eff0f04",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 8b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d334e58a-1c4e-49d1-a9ff-0506fd0f2759",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "test_pairs = [\n",
    "    [\"MTG_RNAseq\", \"MTG_ATACseq\"],\n",
    "    [\"MTG_RNAseq\", \"MTG_RNAseq_Donor\"],\n",
    "    [\"MTG_RNAseq\", \"MTG_RNAseq_No_SA_Donors\"],\n",
    "    [\"MTG_RNAseq\", \"MTG_RNAseq_PMI\"],\n",
    "]\n",
    "for test_pair in test_pairs:\n",
    "    test_x = test_pair[0]\n",
    "    test_y = test_pair[1]\n",
    "    test_var = \"Continuous_Pseudo-progression_Score\"\n",
    "    results_x = pd.concat(\n",
    "        [\n",
    "            pd.read_csv(os.path.join(\"input\", \"Figure 3 and Extended Data 8\", test_x, test_var, \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_results.csv\"), index_col=0),\n",
    "            pd.read_csv(os.path.join(\"input\", \"Figure 3 and Extended Data 8\", test_x, test_var, \"Non-neuronal and Non-neural_Supertype_results.csv\"), index_col=0)\n",
    "        ],\n",
    "        axis=0\n",
    "    )\n",
    "    results_y = pd.concat(\n",
    "        [\n",
    "            pd.read_csv(os.path.join(\"input\", \"Figure 3 and Extended Data 8\", test_y, test_var, \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_results.csv\"), index_col=0),\n",
    "            pd.read_csv(os.path.join(\"input\", \"Figure 3 and Extended Data 8\", test_y, test_var, \"Non-neuronal and Non-neural_Supertype_results.csv\"), index_col=0)\n",
    "        ],\n",
    "        axis=0\n",
    "    )\n",
    "    \n",
    "    common_types = np.intersect1d(results_x[\"Cell Type\"].unique(), results_y[\"Cell Type\"].unique())\n",
    "    \n",
    "    results_x = results_x.loc[results_x[\"Cell Type\"].isin(common_types), :].copy()\n",
    "    results_y = results_y.loc[results_y[\"Cell Type\"].isin(common_types), :].copy()\n",
    "    \n",
    "    \n",
    "    slope, intercept, r_value, p_value, std_err = sp_stats.linregress(\n",
    "        results_x.loc[results_x[\"Covariate\"] == test_var, [\"Final Parameter\", \"Cell Type\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list(),\n",
    "        results_y.loc[results_y[\"Covariate\"] == test_var, [\"Final Parameter\", \"Cell Type\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list()\n",
    "    )\n",
    "    \n",
    "    plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "    ax = sns.regplot(\n",
    "        x=results_x.loc[results_x[\"Covariate\"] == test_var, [\"Final Parameter\", \"Cell Type\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list(),\n",
    "        y=results_y.loc[results_y[\"Covariate\"] == test_var, [\"Final Parameter\", \"Cell Type\"]].groupby(\"Cell Type\").mean().loc[:, \"Final Parameter\"].to_list(),    \n",
    "    );\n",
    "    ax.set(xlabel=\"Effect size on CPS\\nin \" + test_x.replace(\"MTG_\", \"\"), ylabel=\"Effect size on CPS\\nin \" + test_y.replace(\"MTG_\", \"\"), title=\"Comparison across supertypes, r=\" + str(np.round(r_value,2)));\n",
    "    plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 8b_regplot_\" + test_x.replace(\"MTG_\", \"\") + \"_versus_\" + test_y.replace(\"MTG_\", \"\") + \"_effect_size_CPS.pdf\"), bbox_inches=\"tight\");\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5906e47-b004-4bd0-b6ef-6a3171f57630",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 3c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e0f6a2b-a9f3-4df4-81cc-c66e4dd8c46d",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = [\"MTG_RNAseq\", \"A9_RNAseq\"]\n",
    "for region in regions:\n",
    "    neurons = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", region, \"objects\", \"Neuronal: Glutamatergic Neuronal: GABAergic_Supertype_abundances.h5ad\"))\n",
    "    glia = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 3 and Extended Data 8\", region, \"objects\", \"Non-neuronal and Non-neural_Supertype_abundances.h5ad\"))\n",
    "\n",
    "    neurons.var[\"Subclass\"] = [re.sub(\"_[0-9]{1,3}\", \"\", c) for c in neurons.var.index]\n",
    "    neurons.var[\"Subclass\"] = neurons.var[\"Subclass\"].astype(\"category\")\n",
    "    neurons.var[\"Subclass\"] = neurons.var[\"Subclass\"].cat.rename_categories(\n",
    "        {\n",
    "            \"Lamp5_Lhx6\": \"Lamp5 Lhx6\",\n",
    "        },\n",
    "    )\n",
    "\n",
    "    glia.var[\"Subclass\"] = [re.sub(\"_[0-9]{1,3}(_[0-9])?(-SEAAD)?\", \"\", c) for c in glia.var.index]\n",
    "    glia.var[\"Subclass\"] = glia.var[\"Subclass\"].astype(\"category\")\n",
    "    glia.var[\"Subclass\"] = glia.var[\"Subclass\"].cat.rename_categories(\n",
    "        {\n",
    "            \"Astro\": \"Astrocyte\",\n",
    "            \"Oligo\": \"Oligodendrocyte\",\n",
    "            \"Endo\": \"Endothelial\",\n",
    "            \"Micro-PVM\": \"Microglia-PVM\"\n",
    "        },\n",
    "    )\n",
    "\n",
    "    # In all three tests (Cognitive Status, ADNC, and CPS)\n",
    "    affected_supertypes = [\n",
    "        \"Lamp5_3\",\n",
    "        \"Lamp5_5\",\n",
    "        \"Sncg_1\",\n",
    "        \"Sncg_8\",\n",
    "        \"Vip_2\",\n",
    "        \"Vip_11\",\n",
    "        \"Vip_13\",\n",
    "        \"Vip_1\",\n",
    "        \"Sst_3\",\n",
    "        \"Sst_19\",\n",
    "        \"Sst_9\",\n",
    "        \"Sst_11\",\n",
    "        \"Sst_20\",\n",
    "        \"Sst_23\",\n",
    "        \"Sst_25\",\n",
    "        \"Sst_2\",\n",
    "        \"Pvalb_6\",\n",
    "        \"Pvalb_5\",\n",
    "        \"Pvalb_8\",\n",
    "        \"Pvalb_3\",\n",
    "        \"Pvalb_2\",\n",
    "        \"Pvalb_15\",\n",
    "        \"Pvalb_14\",\n",
    "        \"Pvalb_10\",\n",
    "        \"L2/3 IT_1\",\n",
    "        \"L2/3 IT_6\",\n",
    "        \"L2/3 IT_7\",\n",
    "        \"L2/3 IT_5\",\n",
    "        \"L2/3 IT_13\",\n",
    "        \"L2/3 IT_10\",\n",
    "        \"L2/3 IT_8\",\n",
    "        \"L2/3 IT_12\",\n",
    "        \"L2/3 IT_3\",\n",
    "        \"Astro_2\",\n",
    "        \"OPC_2\",\n",
    "        \"Oligo_2\",\n",
    "        \"Micro-PVM_3-SEAAD\"\n",
    "    ]\n",
    "\n",
    "    subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "    subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "    subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "    ax = lmplots(\n",
    "        neurons,\n",
    "        feature_name=\"Continuous_Pseudo-progression_Score\",\n",
    "        figsize=(2.5,4),\n",
    "        y_scale=\"log_relative\",\n",
    "        to_plot=\"Subclass\",\n",
    "        cmap=subclass_colors,\n",
    "        to_plot_filter=[\"Lamp5\", \"Sncg\", \"Vip\", \"Sst\", \"Pvalb\", \"L2/3 IT\"],\n",
    "        celltype_filter=affected_supertypes,\n",
    "    )\n",
    "    ax.set_ylabel(\"log(relative abundance)\");\n",
    "    ax.set_xlabel(\"Pseudoprogression\");\n",
    "    plt.legend([],frameon=False);\n",
    "    plt.savefig(os.path.join(pwd, \"output\", \"Figure 3c_lmplplot_\" + region + \"_neuronal_affected_supertypes_relative_abundance_by_subclass_versus_CPS.pdf\"), bbox_inches=\"tight\")\n",
    "    plt.show();\n",
    "\n",
    "    ax = lmplots(\n",
    "        glia,\n",
    "        feature_name=\"Continuous_Pseudo-progression_Score\",\n",
    "        figsize=(2.5,4),\n",
    "        y_scale=\"log_relative\",\n",
    "        to_plot=\"Subclass\",\n",
    "        cmap=subclass_colors,\n",
    "        to_plot_filter=[\"Astrocyte\", \"Microglia-PVM\", \"Oligodendrocyte\", \"OPC\"],\n",
    "        celltype_filter=affected_supertypes,\n",
    "    )\n",
    "    ax.set_ylabel(\"log(relative abundance)\");\n",
    "    ax.set_xlabel(\"Pseudoprogression\");\n",
    "    plt.legend([],frameon=False);\n",
    "    plt.savefig(os.path.join(pwd, \"output\", \"Figure 3c_lmplplot_\" + region + \"_non-neuronal_affected_supertypes_relative_abundance_by_subclass_versus_CPS.pdf\"), bbox_inches=\"tight\")\n",
    "    plt.show();\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21b5a37c-6e8a-4343-9daa-2fa2c3dbcce0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 3d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb642714-ef83-47a2-9b45-235c9b7b4071",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to generate Spatial transcriptomics figures are in the Spatial Transcriptomics folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "077d042e-ff31-4a1e-9339-4c40d7022d0a",
   "metadata": {},
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a697b6-dba5-4670-863e-7fa228e3004d",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
