{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fff04a46-6ea5-4a79-a54f-84487afa952d",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f72c2c7-eb5a-4512-924d-3ea23d6e4a54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import copy\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import rpy2\n",
    "%load_ext rpy2.ipython\n",
    "import warnings\n",
    "from scipy import stats as sp_stats\n",
    "from helper_functions import *\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=500, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45d953e0-b344-4a0a-8c68-dc8f9fbfe59c",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f801357d-ce22-4663-9616-c0bae0cdb9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "\n",
    "# From https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/\n",
    "RNAseq = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 1 and Extended Data Figure 1 and 2\", \"SEAAD_MTG_RNAseq_all-nuclei_metadata.2024-02-13.csv\"), index_col=0)\n",
    "ATACseq = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 1 and Extended Data Figure 1 and 2\", \"SEAAD_MTG_ATACseq_all-nuclei_metadata.2024-02-13.csv\"), index_col=0)\n",
    "RNAseq[\"Subclass\"] = RNAseq[\"Subclass\"].astype(\"category\").cat.reorder_categories(color_order[\"subclass_label\"].unique())\n",
    "\n",
    "# From Manuscript Supplementary Table 1\n",
    "Donors = pd.read_excel(os.path.join(pwd, \"input\", \"Figure 1 and Extended Data Figure 1 and 2\", \"Supplementary Table 1.xlsx\"), sheet_name=\"SEA-AD_Cohort_Metadata\")\n",
    "\n",
    "# [NOT PUBLIC] From ACT database (must be requested from ACT)\n",
    "Memory = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 1 and Extended Data Figure 1 and 2\", \"long_mem_scores_2022-07-20.txt\"), sep=\"\\t\")\n",
    "\n",
    "# From https://portal.brain-map.org/explore/seattle-alzheimers-disease/seattle-alzheimers-disease-brain-cell-atlas-download\n",
    "quant_neuropath = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 1 and Extended Data Figure 1 and 2\", \"sea-ad_all_mtg_quant_neuropath_bydonorid_081122.csv\"), index_col=1)\n",
    "quant_neuropath = quant_neuropath.drop(\"Unnamed: 0\", axis=1)\n",
    "\n",
    "# From 00_build_input_data.py\n",
    "nuclear_cytosolic = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 1 and Extended Data Figure 1 and 2\", \"nuclear_cytosolic.csv\"), index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ebdd833-d450-457e-8689-5007ac25f1f1",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 1b-c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9486600-7999-4fa1-9ca2-ad9c84eef36b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Donors[\"APOE4 Status\"] = Donors[\"APOE Genotype\"].str.contains(\"4\")\n",
    "Donors[\"Co-morbidity\"] = \"None\"\n",
    "Donors.loc[(Donors[\"Overall CAA Score\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Overall CAA Score\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"CAA\"\n",
    "\n",
    "Donors.loc[(Donors[\"Highest Lewy Body Disease\"].isin([\"Neocortical (Diffuse)\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Highest Lewy Body Disease\"].isin([\"Neocortical (Diffuse)\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"LBD\"\n",
    "\n",
    "Donors.loc[(Donors[\"LATE\"].isin([\"LATE Stage 3\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"LATE\"].isin([\"LATE Stage 3\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"LATE-NC\"\n",
    "\n",
    "Donors.loc[(Donors[\"Arteriolosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Arteriolosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"Arteriolosclerosis\"\n",
    "\n",
    "Donors.loc[(Donors[\"Atherosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Atherosclerosis\"].isin([\"Severe\"])) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"Atherosclerosis\"\n",
    "\n",
    "Donors.loc[(Donors[\"Total Microinfarcts (not observed grossly)\"] > 6) & (Donors[\"Co-morbidity\"] != \"None\"), \"Co-morbidity\"] = \"Multiple\"\n",
    "Donors.loc[(Donors[\"Total Microinfarcts (not observed grossly)\"] > 6) & (Donors[\"Co-morbidity\"] == \"None\"), \"Co-morbidity\"] = \"Microinfarcts\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "595cf1af-e6a2-4a58-9429-34287e0d37d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i Donors\n",
    "# Code comes from Mike H\n",
    "\n",
    "library(pheatmap)\n",
    "\n",
    "vars=colnames(Donors)\n",
    "ustage=unique(Donors[,\"Overall AD neuropathological Change\"])\n",
    "\n",
    "# bracketed by Overall AD neuropathological Change\n",
    "datnad=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"Not AD\",]\n",
    "datlow=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"Low\",]\n",
    "datmid=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"Intermediate\",]\n",
    "dathigh=Donors[Donors[,\"Overall AD neuropathological Change\"]==\"High\",]\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"Male\",\"Female\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[1,1]=table(datnad[,\"Sex\"])[2]\n",
    "mfmat[2,1]=table(datnad[,\"Sex\"])[1]\n",
    "mfmat[1,2]=table(datlow[,\"Sex\"])[2]\n",
    "mfmat[2,2]=table(datlow[,\"Sex\"])[1]\n",
    "mfmat[1,3]=table(datmid[,\"Sex\"])[2]\n",
    "mfmat[2,3]=table(datmid[,\"Sex\"])[1]\n",
    "mfmat[1,4]=table(dathigh[,\"Sex\"])[2]\n",
    "mfmat[2,4]=table(dathigh[,\"Sex\"])[1]\n",
    "mfper=mfmat\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Figure 1b_barplot_Sex_by_Overall AD neuropathological Change.pdf\",width=5,height=5)\n",
    "par(mfrow=c(1,1))\n",
    "barplot(mfper,horiz=T,col=c(\"lightblue\",\"antiquewhite2\"))\n",
    "dev.off()\n",
    "\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"APOE4\",\"No APOE4\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[1,1]=table(datnad[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,1]=table(datnad[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,2]=table(datlow[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,2]=table(datlow[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,3]=table(datmid[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,3]=table(datmid[,\"APOE4 Status\"])[1]\n",
    "mfmat[1,4]=table(dathigh[,\"APOE4 Status\"])[2]\n",
    "mfmat[2,4]=table(dathigh[,\"APOE4 Status\"])[1]\n",
    "mfper=mfmat\n",
    "mfper[1,1]=0; mfper[1,2]=0\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "pdf(\"output/Figure 1b_barplot_APOE4 Status_by_Overall AD neuropathological Change.pdf\",width=5,height=5)\n",
    "par(mfrow=c(1,1))\n",
    "barplot(mfper,horiz=T,col=c(\"darkseagreen\",\"gray90\"))\n",
    "dev.off()\n",
    "\n",
    "\n",
    "mfmat=matrix(0,2,4,dimnames=list(c(\"Dementia\",\"No Dementia\"),c(\"No AD\",\"Low\",\"Intermediate\",\"High\")))\n",
    "mfmat[1,1]=table(datnad[,\"Cognitive Status\"])[1]\n",
    "mfmat[2,1]=table(datnad[,\"Cognitive Status\"])[2]\n",
    "mfmat[1,2]=table(datlow[,\"Cognitive Status\"])[1]\n",
    "mfmat[2,2]=table(datlow[,\"Cognitive Status\"])[2]\n",
    "mfmat[1,3]=table(datmid[,\"Cognitive Status\"])[1]\n",
    "mfmat[2,3]=table(datmid[,\"Cognitive Status\"])[2]\n",
    "mfmat[1,4]=table(dathigh[,\"Cognitive Status\"])[1]\n",
    "mfmat[2,4]=table(dathigh[,\"Cognitive Status\"])[2]\n",
    "mfmat[1,1]=0; mfmat[2,1]=9\n",
    "mfper=mfmat\n",
    "\n",
    "for (j in 1:4) {\n",
    "  mfper[,j]=mfper[,j]/sum(mfper[,j])\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Figure 1c_barplot_Cognitive Status_by_Overall AD neuropathological Change.pdf\",width=5,height=5)\n",
    "par(mfrow=c(1,1))\n",
    "barplot(mfper,horiz=T,col=c(\"red\",\"gray90\"))\n",
    "dev.off()\n",
    "\n",
    "par(las=2)\n",
    "\n",
    "# Braak\n",
    "ubraak=unique(Donors[,\"Braak\"])\n",
    "ustage=unique(Donors[,\"Overall AD neuropathological Change\"])\n",
    "\n",
    "brakorder=c(\"Braak 0\",\"Braak II\",\"Braak III\",\"Braak IV\",\"Braak V\",\"Braak VI\")\n",
    "stageorder=c(\"High\",\"Intermediate\",\"Low\",\"Not AD\")\n",
    "\n",
    "brakmat=matrix(0,4,6,dimnames=list(stageorder,brakorder))\n",
    "\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(brakorder)) {\n",
    "    brakmat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"Braak\"]==brakorder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "brakper=brakmat\n",
    "for (j in 1:6) {\n",
    "  brakper[,j]=round(brakper[,j]/sum(brakper[,j]),2)\n",
    "}\n",
    "\n",
    "pdf(\"output/Figure 1c_heatmap_Braak_by_Overall AD neuropathological Change.pdf\",width=6,height=4)\n",
    "mpal=colorRampPalette(c(\"white\",\"indianred\",\"brown3\",\"darkred\"))(50)\n",
    "pheatmap(brakmat,cluster_rows=F,cluster_cols=F,display_numbers=T,legend=F,number_color = \"black\",fontsize_number=12,col=mpal)\n",
    "dev.off()\n",
    " \n",
    "# Thal\n",
    "uthal=unique(Donors[,\"Thal\"])\n",
    "thalorder=c(\"Thal 0\",\"Thal 1\",\"Thal 2\",\"Thal 3\",\"Thal 4\",\"Thal 5\")\n",
    "\n",
    "thalmat=matrix(0,4,6,dimnames=list(stageorder,thalorder))\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(thalorder)) {\n",
    "    thalmat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"Thal\"]==thalorder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "thalper=thalmat\n",
    "for (j in 1:6) {\n",
    "  thalper[,j]=round(thalper[,j]/sum(thalper[,j]),2)\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Figure 1c_heatmap_Thal_by_Overall AD neuropathological Change.pdf\",width=6,height=4)\n",
    "mpal=colorRampPalette(c(\"white\",\"darkseagreen1\",\"darkseagreen\",\"darkseagreen4\"))(50)\n",
    "pheatmap(thalmat,cluster_rows=F,cluster_cols=F,display_numbers=T,legend=F,number_color = \"black\",fontsize_number=12,col=mpal)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "\n",
    "# CERAD Score\n",
    "ucscore=unique(Donors[,\"CERAD score\"])\n",
    "cscoreorder=c(\"Absent\",\"Sparse\",\"Moderate\",\"Frequent\")\n",
    "\n",
    "cscoremat=matrix(0,4,4,dimnames=list(stageorder,cscoreorder))\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(cscoreorder)) {\n",
    "    cscoremat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"CERAD score\"]==cscoreorder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "cscoreper=cscoremat\n",
    "for (j in 1:4) {\n",
    "  cscoreper[,j]=round(cscoreper[,j]/sum(cscoreper[,j]),2)\n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"output/Figure 1c_heatmap_CERAD score_by_Overall AD neuropathological Change.pdf\",width=6,height=4)\n",
    "mpal=colorRampPalette(c(\"white\",\"skyblue1\",\"deepskyblue2\",\"deepskyblue4\"))(50)\n",
    "pheatmap(cscoremat,cluster_rows=F,cluster_cols=F,display_numbers=T,legend=F,number_color = \"black\",fontsize_number=12,col=mpal)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "pdf(\"output/Figure 1b_boxplot_Age at Death_by_Overall AD neuropathological Change.pdf\",width=6,height=8)\n",
    "par(mfrow=c(4,2))\n",
    "agerange=c(min(Donors[,\"Age at Death\"]),max(Donors[,\"Age at Death\"]))\n",
    "boxplot(datnad[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"No AD\",horizontal=T)\n",
    "stripchart(datnad[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)        \n",
    "\n",
    "boxplot(datlow[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"Low\",horizontal=T)\n",
    "stripchart(datlow[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)    \n",
    "\n",
    "boxplot(datmid[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"Intermediate\",horizontal=T)\n",
    "stripchart(datmid[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)     \n",
    "\n",
    "boxplot(dathigh[,\"Age at Death\"],col = \"gray90\",outine=F,ylim=c(agerange[1],agerange[2]),main=\"High\",horizontal=T)\n",
    "stripchart(dathigh[,\"Age at Death\"],method = \"jitter\",pch = 19,col = 4,vertical = FALSE, add = TRUE)             \n",
    "dev.off()\n",
    "\n",
    "# Co-morbidities\n",
    "umorb=unique(Donors[,\"Co-morbidity\"])\n",
    "\n",
    "umorborder=c(\"None\",\"LBD\",\"Atherosclerosis\",\"Arteriolosclerosis\",\"Microinfarcts\",\"LATE-NC\",\"Multiple\")\n",
    "\n",
    "morbmat=matrix(0,4,7,dimnames=list(stageorder,umorborder))\n",
    "\n",
    "for (i in 1:length(stageorder)) {\n",
    "  for (j in 1:length(umorborder)) {\n",
    "    morbmat[i,j]=length(intersect(which(Donors[,\"Overall AD neuropathological Change\"]==stageorder[i]),which(Donors[,\"Co-morbidity\"]==umorborder[j])))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "morbper=morbmat\n",
    "for (i in 1:4) {\n",
    "  morbper[i,]=round(morbper[i,]/sum(morbper[i,]),2)\n",
    "}\n",
    "\n",
    "pdf(\"output/Figure 1c_barplot_Co-morbidity_by_Overall AD neuropathological Change.pdf\",width=6,height=8)\n",
    "barplot(t(morbper[4:1,]),horiz=T)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ddf04c-64b3-4022-b7e4-795f39f8a335",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31caf424-f831-4172-91c1-79efb0ed36ed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Boxplots\n",
    "plt.rcParams[\"figure.figsize\"] = (3,4)\n",
    "batch_var = \"Donor ID\"\n",
    "covariate = \"Overall AD neuropathological Change\"\n",
    "y_values = [\n",
    "    \"PMI\",\n",
    "    \"Fresh Brain Weight\",\n",
    "    \"Brain pH\",\n",
    "    \"RIN\",\n",
    "]\n",
    "\n",
    "to_get = copy.copy(y_values)\n",
    "to_get.append(batch_var)\n",
    "to_get.append(covariate)\n",
    "\n",
    "metadata = RNAseq.loc[RNAseq[covariate] != \"Reference\", to_get].copy()\n",
    "metadata = metadata.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "for i in y_values:\n",
    "    metadata[i] = metadata[i].replace(\"Unavailable\", np.nan)\n",
    "    metadata[i] = metadata[i].astype(\"float32\")\n",
    "    ax = sns.boxplot(data=metadata, x=covariate, y=i, showfliers=False, palette=\"tab20\");\n",
    "    ax = sns.swarmplot(data=metadata, x=covariate, y=i, color=\"0.25\", edgecolor=\"white\", linewidth=0.5);\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "    ax.set(xlabel=\"\")\n",
    "\n",
    "    plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 1b_boxplot_\" + i + \"_by_\" + covariate + \".pdf\"),  bbox_inches='tight')\n",
    "    plt.show()\n",
    "\n",
    "batch_var = \"library_prep\"\n",
    "covariate = \"Overall AD neuropathological Change\"\n",
    "y_values = [\n",
    "    \"Used in analysis\",\n",
    "    \"quantification_fmol\",\n",
    "    \"NeuN positive fraction on FANS\",\n",
    "]\n",
    "\n",
    "to_get = copy.copy(y_values)\n",
    "to_get.append(batch_var)\n",
    "to_get.append(covariate)\n",
    "\n",
    "metadata = RNAseq.loc[RNAseq[covariate] != \"Reference\", to_get].copy()\n",
    "metadata = metadata.drop(\n",
    "    \"Used in analysis\",\n",
    "    axis=1\n",
    ").merge(\n",
    "    metadata.loc[\n",
    "        :,\n",
    "        [\"Used in analysis\", \"library_prep\"]\n",
    "    ].groupby(\"library_prep\").mean(),\n",
    "    left_on=\"library_prep\",\n",
    "    right_index=True,\n",
    "    how=\"left\"\n",
    ").drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "for i in y_values:\n",
    "    metadata[i] = metadata[i].replace(\"Unavailable\", np.nan)\n",
    "    metadata[i] = metadata[i].astype(\"float32\")\n",
    "    ax = sns.boxplot(data=metadata, x=covariate, y=i, showfliers=False, palette=\"tab20\");\n",
    "    ax = sns.swarmplot(data=metadata, x=covariate, y=i, color=\"0.25\", edgecolor=\"white\", linewidth=0.5);\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "    ax.set(xlabel=\"\")\n",
    "\n",
    "    plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 1b_boxplot_\" + i + \"_by_\" + covariate + \".pdf\"),  bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c621c68d-f3fd-44e1-acc0-f423cfd0416c",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Principle component analysis on pre-sequencing metadata\n",
    "metadata = RNAseq.loc[\n",
    "    (RNAseq[\"Neurotypical reference\"] == False) & (RNAseq[\"method\"] == \"10Xv3.1\") & (RNAseq[\"Fresh Brain Weight\"] != \"Unavailable\"),\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"library_prep\",\n",
    "        \"Overall AD neuropathological Change\",\n",
    "        \"PMI\",\n",
    "        \"Fresh Brain Weight\",\n",
    "        \"Brain pH\",\n",
    "        \"NeuN positive fraction on FANS\",\n",
    "        \"library_input_ng\",\n",
    "        \"quantification_fmol\",\n",
    "        \"avg_size_bp\",\n",
    "        \"RIN\",\n",
    "    ]\n",
    "]\n",
    "for i in metadata.columns[3:]:\n",
    "    metadata[i] = metadata[i].astype(\"float32\")\n",
    "    \n",
    "metadata[\"Donor ID\"] = metadata[\"Donor ID\"].astype(\"category\")\n",
    "metadata.columns = [\n",
    "        \"Donor ID\",\n",
    "        \"library_prep\",\n",
    "        \"ADNC\",\n",
    "        \"Post mortem interval\",\n",
    "        \"Brain weight\",\n",
    "        \"Brain pH\",\n",
    "        \"NeuN positive by FACS\",\n",
    "        \"Amplified cDNA concentration\",\n",
    "        \"Library concentration\",\n",
    "        \"Library Insert size\",\n",
    "        \"RNA Quality (RIN)\",\n",
    "]\n",
    "metadata_X = metadata.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"Post mortem interval\",\n",
    "        \"Brain weight\",\n",
    "        \"Brain pH\",\n",
    "        \"NeuN positive by FACS\",\n",
    "        \"Amplified cDNA concentration\",\n",
    "        \"Library concentration\",\n",
    "        \"Library Insert size\",\n",
    "        \"RNA Quality (RIN)\",\n",
    "    ]\n",
    "].groupby([\"Donor ID\"]).mean()\n",
    "metadata_obs = metadata.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"ADNC\",\n",
    "    ]\n",
    "].drop_duplicates().reset_index()\n",
    "metadata_obs.index = metadata_obs[\"Donor ID\"].copy()\n",
    "metadata_obs = metadata_obs.drop(\"Donor ID\", axis=1)\n",
    "metadata_obs = metadata_obs.loc[metadata_X.index, :]\n",
    "metadata_obs[\"ADNC\"] = metadata_obs[\"ADNC\"].astype(\"category\")\n",
    "metadata_obs[\"ADNC\"] = metadata_obs[\"ADNC\"].cat.reorder_categories([\"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "metadata_ad = ad.AnnData(X=metadata_X, obs=metadata_obs)\n",
    "sc.pp.scale(metadata_ad)\n",
    "sc.pp.pca(metadata_ad)\n",
    "metadata_ad.obs[\"PC1\"] = metadata_ad.obsm[\"X_pca\"][:, 0]\n",
    "\n",
    "# Violin plots\n",
    "sc.pl.violin(metadata_ad, keys=[\"PC1\"], groupby=\"ADNC\", size=5, rotation=90, palette=\"tab20\", save=\"_pre-sequencing_metrics_PC1_by_ADNC.pdf\")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_pre-sequencing_metrics_PC1_by_ADNC.pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1b_violinplot_pre-sequencing_metrics_PC1_by_ADNC.pdf\"))\n",
    "                       \n",
    "# Heatmap\n",
    "features = metadata_ad.var_names.to_series()\n",
    "pc = 0\n",
    "g = metadata_ad.varm['PCs'][:, pc]\n",
    "o = np.argsort(g)\n",
    "sel = np.concatenate((o[:7], o[-1:])).tolist()\n",
    "emb = metadata_ad.obsm['X_pca'][:, pc]\n",
    "donors = np.argsort(emb)\n",
    "tempdata = metadata_ad[donors, features[sel].index.tolist()]\n",
    "tempdata.obs[\"Donor ID\"] = tempdata.obs.index.astype(\"category\")\n",
    "tempdata.obs[\"Donor ID\"] = tempdata.obs[\"Donor ID\"].cat.reorder_categories(pd.Index(metadata_ad.obs.iloc[donors,].index))\n",
    "donor_colors = tempdata.obs.loc[:, [\"ADNC\"]]\n",
    "donor_colors[\"ADNC\"] = donor_colors[\"ADNC\"].cat.rename_categories(\n",
    "    {\n",
    "        \"Not AD\": \"#226894\",\n",
    "        \"Low\": \"#aac0db\",\n",
    "        \"Intermediate\": \"#e47733\",\n",
    "        \"High\": \"#f0b382\",\n",
    "    },\n",
    ")    \n",
    "tmp = pd.DataFrame(tempdata.X.T)\n",
    "tmp.index = tempdata.var_names\n",
    "ax = sns.clustermap(\n",
    "    tmp,\n",
    "    row_cluster=False,\n",
    "    col_cluster=False,\n",
    "    cmap=\"RdBu_r\",\n",
    "    figsize=(16,4),\n",
    "    yticklabels=True,\n",
    "    xticklabels=False,\n",
    "    col_colors=donor_colors[\"ADNC\"].to_list(),\n",
    "    cbar_pos=(0.10, 0.18, 0.01, 0.4),\n",
    "    **{\"center\": 0}\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 1b_heatmap_pre-sequencing_metics_PC1.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d98f0048-0a93-443a-9950-0951b6afa15a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b3507c0-fe7c-451d-af86-05cc8dfd3bbe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (2.5,4)\n",
    "covariate = \"Overall AD neuropathological Change\"\n",
    "adata = ad.AnnData(obs=RNAseq.loc[(RNAseq[covariate] != \"Reference\") & (RNAseq[\"Used in analysis\"] == True), :])\n",
    "adata.obs[covariate] = adata.obs[covariate].cat.reorder_categories([\"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Genes detected\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    rotation=90,\n",
    "    save=\"_Genes detected_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_RNAseq_Genes detected_by_\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_RNAseq_Genes detected_by_\" + covariate + \".pdf\"))\n",
    "\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Number of UMIs\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    log=True,\n",
    "    rotation=90,\n",
    "    save=\"_Number of UMIs_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_RNAseq_Number of UMIs_by_\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_RNAseq_Number of UMIs_by_\" + covariate + \".pdf\"))\n",
    "\n",
    "\n",
    "adata = ad.AnnData(obs=ATACseq.loc[(ATACseq[covariate] != \"Reference\") & (ATACseq[\"Used in analysis\"] == True), :])\n",
    "adata.obs[covariate] = adata.obs[covariate].cat.reorder_categories([\"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Peaks detected\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    rotation=90,\n",
    "    save=\"_Peaks detected_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_ATACseq_Peaks detected_by_\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_ATACseq_Peaks detected_by_\" + covariate + \".pdf\"))\n",
    "\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Number of UMIs\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    log=True,\n",
    "    rotation=90,\n",
    "    save=\"_Number of UMIs_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_ATACseq_Number of UMIs_by\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_ATACseq_Number of UMIs_by_\" + covariate + \".pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc7c328-e3b4-41ae-b8b2-827cd9753256",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 1d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "453afbe6-24a9-49ba-a683-344deb292a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (2.5,4)\n",
    "covariate = \"method\"\n",
    "adata = ad.AnnData(obs=RNAseq.loc[(RNAseq[covariate] != \"Reference\") & (RNAseq[\"Used in analysis\"] == True), :])\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Genes detected\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    rotation=90,\n",
    "    save=\"_Genes detected_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_RNAseq_Genes detected_by_\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_RNAseq_Genes detected_by_\" + covariate + \".pdf\"))\n",
    "\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Number of UMIs\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    log=True,\n",
    "    rotation=90,\n",
    "    save=\"_Number of UMIs_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_RNAseq_Number of UMIs_by_\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_RNAseq_Number of UMIs_by_\" + covariate + \".pdf\"))\n",
    "\n",
    "\n",
    "adata = ad.AnnData(obs=ATACseq.loc[(ATACseq[covariate] != \"Reference\") & (ATACseq[\"Used in analysis\"] == True), :])\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Peaks detected\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    rotation=90,\n",
    "    save=\"_Peaks detected_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_ATACseq_Peaks detected_by_\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_ATACseq_Peaks detected_by_\" + covariate + \".pdf\"))\n",
    "\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    keys=[\"Number of UMIs\"],\n",
    "    groupby=covariate,\n",
    "    stripplot=False,\n",
    "    log=True,\n",
    "    rotation=90,\n",
    "    save=\"_Number of UMIs_by_\" + covariate + \".pdf\",\n",
    "    **{\"palette\": \"tab20\"}\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"violin_ATACseq_Number of UMIs_by\" + covariate + \".pdf\"), os.path.join(pwd, \"output\", \"Extended Data Figure 1c_violinplot_ATACseq_Number of UMIs_by_\" + covariate + \".pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7f7eda-3c8e-433c-84ad-0f4b054c1667",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 2a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd9ced97-d1d7-48be-9357-3c1253ec78ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run PCA on RNAseq library metrics\n",
    "RNAseq_metrics = RNAseq.loc[\n",
    "    (RNAseq[\"Neurotypical reference\"] == False) & (RNAseq[\"method\"] == \"10Xv3.1\"),\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"library_prep\",\n",
    "        \"Overall AD neuropathological Change\",\n",
    "        \"GEX_Fraction_of_transcriptomic_reads_in_cells\",\n",
    "        \"GEX_Mean_raw_reads_per_cell\",\n",
    "        \"GEX_Median_genes_per_cell\",\n",
    "        \"GEX_Median_UMI_counts_per_cell\",\n",
    "        \"GEX_Reads_mapped_confidently_to_genome\",\n",
    "        \"GEX_Reads_mapped_confidently_to_intronic_regions\",\n",
    "        \"GEX_Reads_mapped_confidently_to_exonic_regions\",\n",
    "        \"GEX_Reads_mapped_confidently_to_intergenic_regions\",\n",
    "        \"GEX_Reads_mapped_antisense_to_gene\",\n",
    "        \"GEX_Total_genes_detected\",\n",
    "    ]\n",
    "]\n",
    "RNAseq_metrics[\"Donor ID\"] = RNAseq_metrics[\"Donor ID\"].astype(\"category\")\n",
    "RNAseq_metrics.columns = [\n",
    "        \"Donor ID\",\n",
    "        \"library_prep\",\n",
    "        \"ADNC\",\n",
    "        \"Fraction of transcriptomic reads in cells\",\n",
    "        \"Mean raw reads per cell\",\n",
    "        \"Median genes per cell\",\n",
    "        \"Median UMI counts per cell\",\n",
    "        \"Fraction mapped uniquely to genome\",\n",
    "        \"Fraction mapped uniquely to intronic regions\",\n",
    "        \"Fraction mapped uniquely to exonic regions\",\n",
    "        \"Fraction mapped uniquely to intergenic regions\",\n",
    "        \"Fraction mapped antisense\",\n",
    "        \"Total genes detected in the library\",\n",
    "]\n",
    "RNAseq_metrics_X = RNAseq_metrics.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"Fraction of transcriptomic reads in cells\",\n",
    "        \"Mean raw reads per cell\",\n",
    "        \"Median genes per cell\",\n",
    "        \"Median UMI counts per cell\",\n",
    "        \"Fraction mapped uniquely to genome\",\n",
    "        \"Fraction mapped uniquely to intronic regions\",\n",
    "        \"Fraction mapped uniquely to intergenic regions\",\n",
    "        \"Fraction mapped antisense\",\n",
    "        \"Total genes detected in the library\",\n",
    "    ]\n",
    "].groupby([\"Donor ID\"]).mean()\n",
    "RNAseq_metrics_obs = RNAseq_metrics.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"ADNC\",\n",
    "    ]\n",
    "].drop_duplicates().reset_index()\n",
    "RNAseq_metrics_obs.index = RNAseq_metrics_obs[\"Donor ID\"].copy()\n",
    "RNAseq_metrics_obs = RNAseq_metrics_obs.drop(\"Donor ID\", axis=1)\n",
    "RNAseq_metrics_obs = RNAseq_metrics_obs.loc[RNAseq_metrics_X.index, :]\n",
    "RNAseq_metrics_obs[\"ADNC\"] = RNAseq_metrics_obs[\"ADNC\"].astype(\"category\")\n",
    "RNAseq_metrics_obs[\"ADNC\"] = RNAseq_metrics_obs[\"ADNC\"].cat.reorder_categories([\"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "RNAseq_metrics_ad = ad.AnnData(X=RNAseq_metrics_X, obs=RNAseq_metrics_obs)\n",
    "sc.pp.scale(RNAseq_metrics_ad)\n",
    "sc.pp.pca(RNAseq_metrics_ad)\n",
    "RNAseq_metrics_ad.obs[\"PC1\"] = RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0]\n",
    "\n",
    "# Plot the scaled values for the features\n",
    "genes = RNAseq_metrics_ad.var_names.to_series()\n",
    "pc = 0\n",
    "g = RNAseq_metrics_ad.varm['PCs'][:, pc]\n",
    "o = np.argsort(g)\n",
    "sel = np.concatenate((o[:7], o[-1:])).tolist()\n",
    "emb = RNAseq_metrics_ad.obsm['X_pca'][:, pc]\n",
    "donors = np.argsort(emb)\n",
    "tempdata = RNAseq_metrics_ad[donors, genes[sel].index.tolist()]\n",
    "tempdata.obs[\"Donor ID\"] = tempdata.obs.index.astype(\"category\")\n",
    "tempdata.obs[\"Donor ID\"] = tempdata.obs[\"Donor ID\"].cat.reorder_categories(pd.Index(RNAseq_metrics_ad.obs.iloc[donors,].index))\n",
    "donor_colors = tempdata.obs.loc[:, [\"ADNC\"]]\n",
    "donor_colors[\"ADNC\"] = donor_colors[\"ADNC\"].cat.rename_categories(\n",
    "    {\n",
    "        \"Not AD\": \"#226894\",\n",
    "        \"Low\": \"#aac0db\",\n",
    "        \"Intermediate\": \"#e47733\",\n",
    "        \"High\": \"#f0b382\",\n",
    "    },\n",
    ")    \n",
    "tmp = pd.DataFrame(tempdata.X.T)\n",
    "tmp.index = tempdata.var_names\n",
    "ax = sns.clustermap(\n",
    "    tmp,\n",
    "    row_cluster=False,\n",
    "    col_cluster=False,\n",
    "    cmap=\"RdBu_r\",\n",
    "    figsize=(16,4),\n",
    "    yticklabels=True,\n",
    "    xticklabels=False,\n",
    "    col_colors=donor_colors[\"ADNC\"].to_list(),\n",
    "    cbar_pos=(0.10, 0.18, 0.01, 0.4),\n",
    "    **{\"linewidths\": 1, \"linecolor\": \"lightgrey\", \"center\": 0}\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 2a_heatmap_RNAseq_metrics_PC1.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85aed63d-9549-485a-ae83-9450ff190693",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run PCA on ATACseq library metrics\n",
    "ATACseq_metrics = ATACseq.loc[\n",
    "    (ATACseq[\"Neurotypical reference\"] == False) & (ATACseq[\"method\"] == \"10xATAC_v1.1\"),\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"library_prep\",\n",
    "        \"Overall AD neuropathological Change\",\n",
    "        \"ATAC_Mean_raw_read_pairs_per_cell\",\n",
    "        \"ATAC_TSS_enrichment_score\",\n",
    "        \"ATAC_Confidently_mapped_read_pairs\",\n",
    "        \"ATAC_Fraction_of_genome_in_peaks\",\n",
    "        \"ATAC_Fraction_of_high_quality_fragments_overlapping_TSS\",\n",
    "        \"ATAC_Fraction_of_high_quality_fragments_overlapping_peaks\",\n",
    "        \"ATAC_Fraction_of_transposition_events_in_peaks_in_cells\",\n",
    "        \"ATAC_Median_high_quality_fragments_per_cell\",\n",
    "    ]\n",
    "]\n",
    "ATACseq_metrics.columns = [\n",
    "        \"Donor ID\",\n",
    "        \"library_prep\",\n",
    "        \"ADNC\",\n",
    "        \"Mean raw read pairs per cell\",\n",
    "        \"TSS enrichment score\",\n",
    "        \"Fraction of uniquely mapped read pairs\",\n",
    "        \"Fraction of genome in peaks\",\n",
    "        \"Fraction of fragments overlapping TSS\",\n",
    "        \"Fraction of fragments overlapping peaks\",\n",
    "        \"Fraction of transposition events in peaks in cells\",\n",
    "        \"Median fragments per cell\",\n",
    "]\n",
    "ATACseq_metrics_X = ATACseq_metrics.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"Mean raw read pairs per cell\",\n",
    "        \"TSS enrichment score\",\n",
    "        \"Fraction of uniquely mapped read pairs\",\n",
    "        \"Fraction of genome in peaks\",\n",
    "        \"Fraction of fragments overlapping TSS\",\n",
    "        \"Fraction of fragments overlapping peaks\",\n",
    "        \"Fraction of transposition events in peaks in cells\",\n",
    "        \"Median fragments per cell\",\n",
    "    ]\n",
    "].groupby([\"Donor ID\"]).mean()\n",
    "ATACseq_metrics_obs = ATACseq_metrics.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"Donor ID\",\n",
    "        \"ADNC\",\n",
    "    ]\n",
    "].drop_duplicates().reset_index().drop(\"index\", axis=1)\n",
    "ATACseq_metrics_obs.index = ATACseq_metrics_obs[\"Donor ID\"].copy()\n",
    "ATACseq_metrics_obs = ATACseq_metrics_obs.drop(\"Donor ID\", axis=1)\n",
    "ATACseq_metrics_obs = ATACseq_metrics_obs.loc[ATACseq_metrics_X.index, :]\n",
    "ATACseq_metrics_obs[\"ADNC\"] = ATACseq_metrics_obs[\"ADNC\"].astype(\"category\")\n",
    "ATACseq_metrics_obs[\"ADNC\"] = ATACseq_metrics_obs[\"ADNC\"].cat.reorder_categories([\"Not AD\", \"Low\", \"Intermediate\", \"High\"])\n",
    "ATACseq_metrics_ad = ad.AnnData(X=ATACseq_metrics_X, obs=ATACseq_metrics_obs)\n",
    "ATACseq_metrics_ad\n",
    "\n",
    "sc.pp.scale(ATACseq_metrics_ad)\n",
    "sc.pp.pca(ATACseq_metrics_ad)\n",
    "ATACseq_metrics_ad.obs[\"PC1\"] = ATACseq_metrics_ad.obsm[\"X_pca\"][:, 0]\n",
    "\n",
    "# Plot the scaled values for the features\n",
    "genes = ATACseq_metrics_ad.var_names.to_series()\n",
    "pc = 0\n",
    "g = ATACseq_metrics_ad.varm['PCs'][:, pc]\n",
    "o = np.argsort(g)\n",
    "sel = np.concatenate((o[:7],o[-1:])).tolist()\n",
    "emb = ATACseq_metrics_ad.obsm['X_pca'][:, pc]\n",
    "donors = np.argsort(emb)\n",
    "tempdata = ATACseq_metrics_ad[donors, genes[sel].index.tolist()]\n",
    "tempdata.obs[\"Donor ID\"] = tempdata.obs.index.astype(\"category\")\n",
    "tempdata.obs[\"Donor ID\"] = tempdata.obs[\"Donor ID\"].cat.reorder_categories(pd.Index(ATACseq_metrics_ad.obs.iloc[donors,].index))\n",
    "donor_colors = tempdata.obs.loc[:, [\"ADNC\"]]\n",
    "donor_colors[\"ADNC\"] = donor_colors[\"ADNC\"].cat.rename_categories(\n",
    "    {\n",
    "        \"Not AD\": \"#226894\",\n",
    "        \"Low\": \"#aac0db\",\n",
    "        \"Intermediate\": \"#e47733\",\n",
    "        \"High\": \"#f0b382\",\n",
    "    },\n",
    ")    \n",
    "tmp = pd.DataFrame(tempdata.X.T)\n",
    "tmp.index = tempdata.var_names\n",
    "ax = sns.clustermap(\n",
    "    tmp,\n",
    "    row_cluster=False,\n",
    "    col_cluster=False,\n",
    "    cmap=\"RdBu_r\",\n",
    "    figsize=(16,4),\n",
    "    yticklabels=True,\n",
    "    xticklabels=False,\n",
    "    col_colors=donor_colors[\"ADNC\"].to_list(),\n",
    "    cbar_pos=(0.10, 0.18, 0.01, 0.4),\n",
    "    **{\"linewidths\": 1, \"linecolor\": \"lightgrey\", \"center\": 0}\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 2a_heatmap_ATACseq_metrics_PC1.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "314174d3-706b-44ec-919f-becb4b6d4c4b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 1d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b68609c1-4427-43b3-8178-2f783026c4f6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Plot the RNAseq and ATACseq PCs against one another per donor\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "ax = sns.scatterplot(\n",
    "    x=RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0] - (RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min(),\n",
    "    y=ATACseq_metrics_ad.obsm[\"X_pca\"][:, 0] - (ATACseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min(),\n",
    "    hue=RNAseq_metrics_ad.obs[\"ADNC\"], palette=\"tab20\"\n",
    ");\n",
    "ax = sns.regplot(\n",
    "    x=RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0] - (RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min(),\n",
    "    y=ATACseq_metrics_ad.obsm[\"X_pca\"][:, 0] - (ATACseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min(),\n",
    "    scatter=False,\n",
    "    line_kws={\"color\": \"lightgrey\", \"lw\": 2}\n",
    ");\n",
    "ax.set_xlabel(\"snRNAseq library metrics\");\n",
    "ax.set_ylabel(\"snATACseq library metrics\");\n",
    "ax.axvline(x=1.5 - (RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min(), **{\"linestyle\": \"--\", \"color\": \"red\"});\n",
    "ax.axhline(y=2.5 - (ATACseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min(), **{\"linestyle\": \"--\", \"color\": \"red\"});\n",
    "ax.legend(bbox_to_anchor=(1.05, 1), ncol=1);\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 1d_scatterplot_RNAseq_metrics_PC1_versus_ATACseq_metrics_PC1.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72afe2dd-1186-4d87-8dfc-d8a477b50166",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 1e and Extended Data Figure 2b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ce6aa75-1ba2-4353-ba8a-c4dab20e94bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNAseq[\"Memory groups\"] = \"ADNC0-2\"\n",
    "RNAseq.loc[(RNAseq[\"Overall AD neuropathological Change\"] == \"High\") & (RNAseq[\"Severely Affected Donor\"] == \"N\"), \"Memory groups\"] = \"ADNC3\"\n",
    "RNAseq.loc[(RNAseq[\"Overall AD neuropathological Change\"] == \"High\") & (RNAseq[\"Severely Affected Donor\"] == \"Y\"), \"Memory groups\"] = \"SA\"\n",
    "RNAseq[\"Memory groups\"] = RNAseq[\"Memory groups\"].astype(\"category\")\n",
    "\n",
    "Memory = Memory.loc[~Memory[\"donor_name\"].isna(),:]\n",
    "Memory = Memory.loc[(Memory[\"visit\"] % 2) == 0, :]\n",
    "\n",
    "Memory[\"visit\"] = Memory[\"visit\"] / 2\n",
    "Memory = Memory.merge(RNAseq.loc[:, [\"Donor ID\", \"Memory groups\"]].drop_duplicates().reset_index(drop=True), left_on=\"donor_name\", right_on=\"Donor ID\", how=\"left\")\n",
    "to_keep = []\n",
    "Memory[\"normalized_visit\"] = 0\n",
    "for i in Memory[\"Donor ID\"].unique():\n",
    "    Memory.loc[Memory[\"Donor ID\"] == i, \"normalized_visit\"] = Memory.loc[Memory[\"Donor ID\"] == i, \"visit\"] - Memory.loc[Memory[\"Donor ID\"] == i, \"visit\"].max()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e746616-9872-412e-958c-0212023b7130",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Plot memory curves split by SA donors, other high pathology donors, and all other donors\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "colors = {\n",
    "    \"ADNC0-2\": \"grey\",\n",
    "    \"ADNC3\": \"goldenrod\",\n",
    "    \"SA\": \"dodgerblue\"\n",
    "}\n",
    "\n",
    "for j in [\"VSP_E\", \"LAN_E\", \"EXF_E\", \"MEM_E\"]:\n",
    "    for i in Memory[\"Memory groups\"].unique():\n",
    "        data = Memory[Memory[\"Memory groups\"] == i]\n",
    "        x = data[\"normalized_visit\"].to_numpy()\n",
    "        y = data[j].to_numpy()\n",
    "        xgrid = np.linspace(x.min(), x.max())\n",
    "        K = 1000\n",
    "        smooths = np.stack([smooth(x, y, xgrid) for k in range(K)]).T   \n",
    "        mean = np.nanmean(smooths, axis=1)\n",
    "        stderr = sp_stats.sem(smooths, axis=1)\n",
    "        stderr = np.nanstd(smooths, axis=1, ddof=0)\n",
    "        plt.fill_between(xgrid, mean - 2 * stderr, mean + 2 * stderr, color=colors[i], alpha=0.06)\n",
    "        plt.plot(xgrid, mean, color=colors[i])\n",
    "    plt.ylim((-1.5, 1.5));\n",
    "    plt.xlabel(\"Visits until death\");\n",
    "    plt.ylabel(\"Cognitive score in \" + j + \" domain\");\n",
    "    if j == \"MEM_E\":\n",
    "        plt.savefig(os.path.join(pwd, \"output\", \"Figure 1e_lmplot_Visits until death_versus_Cognitive score in \" + j + \" domain.pdf\"), bbox_inches=\"tight\");\n",
    "    else:\n",
    "        plt.savefig(os.path.join(pwd, \"output\", \"Extended Data Figure 2b_lmplot_Visits until death_versus_Cognitive score in \" + j + \" domain.pdf\"), bbox_inches=\"tight\");\n",
    "\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d49b2982-70c8-4e5c-a921-766a39a5ab9f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 1g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "844b74f1-c7c6-417a-b46b-f18b52efc86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNAseq_metrics_ad.obs = RNAseq_metrics_ad.obs.merge(Donors.loc[:, [\"Donor ID\", \"Severely Affected Donor\"]].fillna(\"N\"), how=\"left\", left_index=True, right_on=\"Donor ID\")\n",
    "RNAseq_metrics_ad.obs.index = RNAseq_metrics_ad.obs[\"Donor ID\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bfe61b8-c43d-4588-8823-99eb71d5eefe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot RNAseq PC1 versus quantitative neuropathology\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "ax = sns.scatterplot(\n",
    "    x=RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0] - (RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min() + 0.01,\n",
    "    y=quant_neuropath.loc[RNAseq_metrics_ad.obs_names, \"number of NeuN positive cells per area_Layer3\"].to_list(),\n",
    "    hue=RNAseq_metrics_ad.obs[\"Severely Affected Donor\"],\n",
    ");\n",
    "ax = sns.regplot(\n",
    "    x=RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0] - (RNAseq_metrics_ad.obsm[\"X_pca\"][:, 0]).min() + 0.01,\n",
    "    y=quant_neuropath.loc[RNAseq_metrics_ad.obs_names, \"number of NeuN positive cells per area_Layer3\"].to_list(),\n",
    "    scatter=False,\n",
    "    logistic=True,\n",
    "    line_kws={\"color\": \"lightgrey\", \"lw\": 2}\n",
    ");\n",
    "ax.set_ylabel(\"number of NeuN positive cells per area_Layer3\");\n",
    "ax.set_xlabel(\"snRNAseq library metrics\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 1g_lmplot_RNAseq_metrics_PC1_versus_number of NeuN positive cells per area_Layer3.pdf\"), bbox_inches=\"tight\");\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a6a5c3-1a93-4032-9caf-f03cb981e5fa",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 1h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f015e128-613a-4793-b6b2-5604a0069d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot unnormalized expression of cytosolic (mitochondrial) and nuclear RNA species\n",
    "plt.rcParams[\"figure.figsize\"] = (2,4)\n",
    "to_plot = [\n",
    "    \"MT-CO1\",\n",
    "    \"MT-ND3\",\n",
    "    \"MALAT1\",\n",
    "    \"MEG3\"\n",
    "]\n",
    "for i in to_plot:\n",
    "    sns.boxplot(\n",
    "        data=nuclear_cytosolic,\n",
    "        x=\"Severely Affected Donor\",\n",
    "        y=i,\n",
    "        showfliers=False,\n",
    "    );\n",
    "    plt.xlabel(\"Severely Affected Donor\");\n",
    "    plt.savefig(os.path.join(pwd, \"output\", \"Figure 1h_boxplot_\" + i + \"_by_Severely Affected Donor.pdf\"), bbox_inches=\"tight\");\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c905804e-2cc3-4802-8e4b-953aa16a72db",
   "metadata": {},
   "source": [
    "### Figure 1i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c4e443-1cba-448e-ae4f-5a441a79d551",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need code from Giuseppe/Mariano"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bd9f3dd-d661-42fc-a2b0-596fa6db7b3f",
   "metadata": {},
   "source": [
    "### Extended Data Figure 1c-e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f8c8bd6-e6c6-49c3-a7fd-23284c52b88d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need code from Giuseppe/Mariano"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606f5bdf-94c2-4cd9-84c9-3e01b50ee963",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extended Data Figure 1f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c681a47-ff70-4215-abae-c21bb32b9abc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (20,5)\n",
    "to_get = [\n",
    "    \"library_prep\",\n",
    "    \"Subclass\",\n",
    "    \"Severely Affected Donor\",\n",
    "    \"Fraction mitochondrial UMIs\",\n",
    "]\n",
    "metadata = RNAseq.loc[RNAseq[\"Overall AD neuropathological Change\"] != \"Reference\", to_get].copy()\n",
    "metadata[\"Mitochondrial Flag\"] = metadata[\"Fraction mitochondrial UMIs\"] > 0.05\n",
    "metadata = metadata.groupby([\"library_prep\", \"Subclass\", \"Severely Affected Donor\"]).mean().reset_index()\n",
    "\n",
    "sns.stripplot(\n",
    "    data=metadata,\n",
    "    x=\"Subclass\",\n",
    "    y=\"Mitochondrial Flag\",\n",
    "    alpha=0.5,\n",
    "    hue=\"Severely Affected Donor\",\n",
    "    dodge=True,\n",
    ");\n",
    "plt.xticks(rotation=90, ha=\"right\");\n",
    "plt.ylabel(\"Fraction removed for the\\nfraction of mitochondrial reads\");\n",
    "plt.legend(bbox_to_anchor=(1.01, 1.0), loc=\"upper left\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 1f_stripplot_Fraction removed for the fraction of mitochondrial reads_by_Subclass_and_Severely Affected Donor.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0d4262-1bf8-4f82-a9b5-a4fd8f4da94f",
   "metadata": {},
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bc3c1c4-d38a-4a34-8852-05939f1ee22f",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
