{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e10517e9-1c64-4305-bcb5-2cb3ac9c2bd6",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d8a6c9-19ef-4b78-b0e5-d719d215c84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import copy as cp\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "import scipy.stats as sp_stats\n",
    "from helper_functions import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "sc.set_figure_params(scanpy=True, dpi=100, dpi_save=500, frameon=False, vector_friendly=True, figsize=(10,10), format='png')\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af2c7f99-1087-48ef-87ca-7bd41baa8868",
   "metadata": {},
   "source": [
    "### Load needed datasets/data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df48c455-9a0f-4456-a63f-dadcc17378c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster order and colors from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "color_order = pd.read_csv(os.path.join(pwd, \"input\", \"cluster_order_and_colors.csv\"))\n",
    "\n",
    "# From 00_build_input_files.py\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 8\", \"SEAAD_MTG_RNAseq_final-nuclei_limited.2024-02-13.h5ad\"))\n",
    "\n",
    "# Subclass UMAP coordinates from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "OPC_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 8\", \"OPC\", \"obs_names.csv\"), index_col=0).index\n",
    "OPC_umap = np.load(os.path.join(pwd, \"input\", \"Figure 8\", \"OPC\", \"X_umap.npy\"))\n",
    "\n",
    "Oligodendrocyte_obs_names = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 8\", \"Oligodendrocyte\", \"obs_names.csv\"), index_col=0).index\n",
    "Oligodendrocyte_umap = np.load(os.path.join(pwd, \"input\", \"Figure 8\", \"Oligodendrocyte\", \"X_umap.npy\"))\n",
    "\n",
    "# Green et al (2023) labels from personal communication and Mathys et al (2023) labels from https://compbio.mit.edu/\n",
    "# Green et al (2023) and Mathys et al (2023) mapping results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#PFC/RNAseq/public_datasets/green_labels = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 8\",\"Green_2023_author-annotation.csv\"), index_col=0)\n",
    "green_mapping = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 8\", \"Green_2023-cell-annotation.2024-03-27.csv\"), index_col=0)\n",
    "\n",
    "mathys_label_files = glob.glob(os.path.join(pwd, \"input\", \"Figure 8\", \"Mathys_2023\", \"*.csv\"))\n",
    "mathys_labels = False\n",
    "for i in mathys_label_files:\n",
    "    tmp = pd.read_csv(i, index_col=0)\n",
    "    try:\n",
    "        mathys_labels = pd.concat([mathys_labels, tmp], axis=0)\n",
    "    except TypeError as i:\n",
    "        mathys_labels = tmp.copy()\n",
    "mathys_mapping = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 8\", \"Mathys_2023-cell-annotation.2024-03-27.csv\"), index_col=0)\n",
    "\n",
    "# Mean expression table from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "mean_expression = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 8\", \"mean_expression.h5ad\"))\n",
    "\n",
    "# Cell type specific marker genes from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "effect_sizes_vs_all = sc.read_h5ad(os.path.join(pwd, \"input\", \"Figure 8\", \"effect_sizes_vs_all.h5ad\"))\n",
    "      \n",
    "# Effect size table from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "effect_size_table = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 8\", \"effect_size_table.csv\"), index_col=0)\n",
    "level = \"Subclass\"\n",
    "tmp = effect_size_table.loc[effect_size_table[\"Taxonomy Level\"] == level, [\"Gene\", \"Mean expression (natural log UMIs per 10k plus 1)\"]]\n",
    "tmp = tmp.groupby([\"Gene\"])[\"Mean expression (natural log UMIs per 10k plus 1)\"].transform(lambda x : sp_stats.zscore(x,ddof=1)).fillna(0)\n",
    "effect_size_table[\"Mean expression z-score\"] = tmp\n",
    "\n",
    "# GRN results from https://sea-ad-single-cell-profiling.s3.amazonaws.com/index.html#MTG/RNAseq/Supplementary%20Information/\n",
    "eRegulons = pd.read_csv(os.path.join(pwd, \"input\", \"Figure 7\", \"eRegulon_metadata_filtered.csv\"), index_col=0)\n",
    "TFs = [i.replace(\"_extended\", \"\") for i in eRegulons[\"Gene_signature_name\"].unique()]\n",
    "TFs = pd.DataFrame([[i] + i.split(\"_\") for i in TFs], columns=[\"Gene_signature_name\", \"TF\", \"Direction\", \"Target Genes\"])\n",
    "TFs[\"Target Genes\"] = [np.int32(i.replace(\"(\", \"\").replace(\")\", \"\").replace(\"g\", \"\")) for i in TFs[\"Target Genes\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "648c53b4-04f2-4aca-89db-3f579570d3d2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81bcc447-898e-4253-ae7f-6763e1e47413",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub = adata[OPC_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = OPC_umap\n",
    "sub = sub[sub.obs[\"Supertype\"] != \"OPC_2_1-SEAAD\", :].copy()\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=20,\n",
    "    palette=\"tab10\",\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_OPC_Supertype_tab10.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_OPC_Supertype_tab10.pdf\"), os.path.join(pwd, \"output\", \"Figure 8a_umap_OPC_Supertype_tab10.pdf\"))\n",
    "\n",
    "\n",
    "sub = adata[Oligodendrocyte_obs_names, :].copy()\n",
    "sub.obsm[\"X_umap\"] = Oligodendrocyte_umap\n",
    "sub = sub[sub.obs[\"Supertype\"] != \"Oligo_2_1-SEAAD\", :].copy()\n",
    "sc.pp.subsample(sub, fraction=1)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(\n",
    "    sub,\n",
    "    color=[\"Supertype\"],\n",
    "    size=20,\n",
    "    palette=\"tab10\",\n",
    "    frameon=False,\n",
    "    legend_loc=\"on data\",\n",
    "    title=\"\",\n",
    "    save=\"_Oligodendrocyte_Supertype_tab10.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"umap_Oligodendrocyte_Supertype_tab10.pdf\"), os.path.join(pwd, \"output\", \"Figure 8a_umap_Oligodendrocyte_Supertype_tab10.pdf\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e4bac9-6e6c-4361-ae36-99b4ece1c178",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "309e4646-bca4-4f6f-a734-8a629ee78d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPC\n",
    "effect_sizes_OPC = effect_sizes_vs_all[:, (effect_sizes_vs_all.var[\"Subclass\"] == \"OPC\") & (effect_sizes_vs_all.var[\"Supertype\"] != \"OPC_2_1-SEAAD\")].copy()\n",
    "effect_sizes_OPC = pd.DataFrame(\n",
    "    effect_sizes_OPC.X,\n",
    "    index=effect_sizes_OPC.obs_names,\n",
    "    columns=effect_sizes_OPC.var_names\n",
    ")\n",
    "\n",
    "mean_expression_OPC = mean_expression[:, (mean_expression.var[\"Subclass\"] == \"OPC\") & (effect_sizes_vs_all.var[\"Supertype\"] != \"OPC_2_1-SEAAD\")].copy()\n",
    "mean_expression_OPC = pd.DataFrame(\n",
    "    mean_expression_OPC.X,\n",
    "    index=mean_expression_OPC.obs_names,\n",
    "    columns=mean_expression_OPC.var_names\n",
    ")\n",
    "\n",
    "z_score_OPC = mean_expression_OPC.T\n",
    "z_score_OPC = (z_score_OPC - z_score_OPC.mean(axis=0)) / z_score_OPC.std(axis=0)\n",
    "z_score_OPC = z_score_OPC.T\n",
    "\n",
    "\n",
    "genes = []\n",
    "for i in effect_sizes_OPC.columns:\n",
    "    gene_n = 3\n",
    "    if i == \"OPC_2\":\n",
    "        gene_n = 30\n",
    "    new_genes = effect_sizes_OPC.loc[(effect_sizes_OPC.max(axis=1) == effect_sizes_OPC.loc[:, i]) & (mean_expression_OPC.max(axis=1) == mean_expression_OPC.loc[:, i]), :].sort_values(by=i, ascending=False).iloc[:gene_n].index.to_list()\n",
    "    genes.extend(z_score_OPC.loc[new_genes, i].sort_values(ascending=False).index.to_list())\n",
    "    \n",
    "plt.rcParams[\"figure.figsize\"] = (12,3)\n",
    "sns.heatmap(\n",
    "    z_score_OPC.loc[genes, :].T,\n",
    "    cmap=\"RdBu_r\",\n",
    "    center=0,\n",
    "    xticklabels=True,\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8b_heatmap_OPC marker z scores.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# Oligodendrocyte\n",
    "effect_sizes_Oligodendrocyte = effect_sizes_vs_all[:, (effect_sizes_vs_all.var[\"Subclass\"] == \"Oligodendrocyte\") & (effect_sizes_vs_all.var[\"Supertype\"] != \"Oligo_2_1-SEAAD\")].copy()\n",
    "effect_sizes_Oligodendrocyte = pd.DataFrame(\n",
    "    effect_sizes_Oligodendrocyte.X,\n",
    "    index=effect_sizes_Oligodendrocyte.obs_names,\n",
    "    columns=effect_sizes_Oligodendrocyte.var_names\n",
    ")\n",
    "\n",
    "mean_expression_Oligodendrocyte = mean_expression[:, (mean_expression.var[\"Subclass\"] == \"Oligodendrocyte\") & (effect_sizes_vs_all.var[\"Supertype\"] != \"Oligo_2_1-SEAAD\")].copy()\n",
    "mean_expression_Oligodendrocyte = pd.DataFrame(\n",
    "    mean_expression_Oligodendrocyte.X,\n",
    "    index=mean_expression_Oligodendrocyte.obs_names,\n",
    "    columns=mean_expression_Oligodendrocyte.var_names\n",
    ")\n",
    "\n",
    "z_score_Oligodendrocyte = mean_expression_Oligodendrocyte.T\n",
    "z_score_Oligodendrocyte = (z_score_Oligodendrocyte - z_score_Oligodendrocyte.mean(axis=0)) / z_score_Oligodendrocyte.std(axis=0)\n",
    "z_score_Oligodendrocyte = z_score_Oligodendrocyte.T\n",
    "\n",
    "\n",
    "genes = []\n",
    "for i in effect_sizes_Oligodendrocyte.columns:\n",
    "    gene_n = 3\n",
    "    if i == \"Oligo_2\" or i == \"Oligo_4\":\n",
    "        gene_n = 30\n",
    "    new_genes = effect_sizes_Oligodendrocyte.loc[(effect_sizes_Oligodendrocyte.max(axis=1) == effect_sizes_Oligodendrocyte.loc[:, i]) & (mean_expression_Oligodendrocyte.max(axis=1) == mean_expression_Oligodendrocyte.loc[:, i]), :].sort_values(by=i, ascending=False).iloc[:gene_n].index.to_list()\n",
    "    genes.extend(z_score_Oligodendrocyte.loc[new_genes, i].sort_values(ascending=False).index.to_list())\n",
    "    \n",
    "plt.rcParams[\"figure.figsize\"] = (20,3)\n",
    "sns.heatmap(\n",
    "    z_score_Oligodendrocyte.loc[genes, :].T,\n",
    "    cmap=\"RdBu_r\",\n",
    "    center=0,\n",
    "    xticklabels=True,\n",
    ");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8b_heatmap_Oligodendrocyte marker z scores.pdf\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17db54cd-6f7e-4471-acfb-a28c1eb08cd2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bfe6db7-04b9-486c-8c12-28ec4f36023e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Green et al_2023\n",
    "df = green_mapping.merge(green_labels.loc[:, [\"cell\", \"state\"]], how=\"inner\", left_on=\"Original Cell ID\", right_on=\"cell\")\n",
    "df = df.loc[:, [\"Supertype\", \"state\"]].groupby(\"state\").value_counts(normalize=True).reset_index().pivot(index=\"Supertype\", columns=\"state\", values=\"proportion\")\n",
    "df = df.fillna(0)\n",
    "plt.rcParams[\"figure.figsize\"] = (5,3)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[\n",
    "        [\"Oligo_1\", \"Oligo_2\", \"Oligo_4\", \"Oligo_3\", \"OPC_1\", \"OPC_2\", \"OPC_2_2-SEAAD\"],\n",
    "        [\"Oli.5\", \"Oli.11\", \"Oli.6\", \"Oli.9\", \"Oli.3\", \"Oli.8\", \"Oli.7\", \"Oli.4\", \"Oli.10\", \"Oli.12\", \"Oli.2\", \"MFOL\", \"Oli.1\", \"COP\", \"OPC.1\", \"OPC.2\", \"OPC.3\"]\n",
    "    ],\n",
    "    cmap=\"YlGnBu\",\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    ");\n",
    "ax.set(xlabel=\"Green et al 2023\", ylabel=\"SEA-AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8c_heatmap_Green_OPC_Oligo_Comparison.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Mathys_2023\n",
    "df = mathys_mapping.merge(mathys_labels.loc[:, [\"barcode\", \"cell_type_high_resolution\"]], how=\"inner\", left_on=\"Original Cell ID Alternative\", right_on=\"barcode\")\n",
    "df = df.loc[:, [\"Supertype\", \"cell_type_high_resolution\"]].groupby(\"cell_type_high_resolution\").value_counts(normalize=True).reset_index().pivot(index=\"Supertype\", columns=\"cell_type_high_resolution\", values=\"proportion\")\n",
    "df = df.fillna(0)\n",
    "plt.rcParams[\"figure.figsize\"] = (1,3)\n",
    "ax = sns.heatmap(\n",
    "    df.loc[\n",
    "        [\"Oligo_1\", \"Oligo_2\", \"Oligo_4\", \"Oligo_3\", \"OPC_1\", \"OPC_2\", \"OPC_2_2-SEAAD\"],\n",
    "        [\"Oli\", \"OPC\"]\n",
    "    ],\n",
    "    cmap=\"YlGnBu\",\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    ");\n",
    "ax.set(xlabel=\"Mathys et al 2023\", ylabel=\"SEA-AD\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8c_heatmap_Mathys_OPC_Oligo_Comparison.pdf\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18edaac9-688f-4e0c-87fc-d905b40736d8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76b24ce7-db45-41b9-b8fe-daba49ca87d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "psen_genes = [\"PSEN1\", \"PSEN2\"]\n",
    "beta_secretase_genes = [\"BACE1\", \"BACE2\"]\n",
    "gamma_secretase_complex_genes = [\"PSENEN\", \"NCSTN\", \"APH1A\"]\n",
    "final_genes = [\"PSEN\", \"BACE\", \"PSENEN\", \"NCSTN\", \"APH1A\", \"APP\"]\n",
    "vars_to_get = [\"APP\"]\n",
    "vars_to_get.extend(psen_genes)\n",
    "vars_to_get.extend(beta_secretase_genes)\n",
    "vars_to_get.extend(gamma_secretase_complex_genes)\n",
    "vars_to_get.append(\"Subclass\")\n",
    "df = sc.get.obs_df(adata, vars_to_get)\n",
    "df[\"PSEN\"] = df.loc[:, psen_genes].sum(axis=1)\n",
    "df[\"BACE\"] = df.loc[:, beta_secretase_genes].sum(axis=1)\n",
    "df[\"AB Score\"] = df.loc[:, final_genes].mean(axis=1)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (6,4)\n",
    "ax = sns.boxplot(\n",
    "    data=df,\n",
    "    x=\"Subclass\",\n",
    "    y=\"AB Score\",\n",
    "    showfliers=False\n",
    ");\n",
    "plt.xticks(rotation=90);\n",
    "ax.set(xlabel=\"\", ylabel=\"Expression of beta- &\\n gamma-secretase and APP\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8d_boxplot_AB score_by_Subclass.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81200352-a427-4cbc-9962-fd09cd01fa48",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71adb579-a1e7-4686-9574-04a71e377e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.Series(effect_size_table[\"Gene\"].unique())\n",
    "# Define gene lists\n",
    "gene_lists = {}\n",
    "\n",
    "# Cholesterol biosynthesis\n",
    "# From https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-341\n",
    "# https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-3\n",
    "# and https://biocyc.org/HUMAN/NEW-IMAGE?type=PATHWAY&object=PWY66-4\n",
    "gene_lists[\"Cholesterol biosynthesis\"] = []\n",
    "gene_lists[\"Cholesterol biosynthesis\"].extend([\n",
    "    \"FDFT1\",\n",
    "    \"SQLE\",\n",
    "    \"LSS\",\n",
    "    \"CYP51A1\",\n",
    "    \"LBR\",\n",
    "    \"TM7SF2\",\n",
    "    \"MSMO1\",\n",
    "    \"NSDHL\",\n",
    "    \"HSD17B1\",\n",
    "    \"EBP\",\n",
    "    \"DHCR24\",\n",
    "    \"SC5D\",\n",
    "    \"DHCR7\",\n",
    "])\n",
    "\n",
    "# Based on UniProt Keywords\n",
    "gene_lists[\"Fatty Acid metabolism\"] = []\n",
    "tmp = pd.read_csv(os.path.join(pwd, \"input\", \"dbs\", \"uniprotkb_keyword_KW_0276_AND_reviewed_2024_02_06.tsv\"), sep=\"\\t\")\n",
    "tmp = [i.split(\" \")[0] for i in tmp[\"Gene Names\"]]\n",
    "gene_lists[\"Fatty Acid metabolism\"].extend(np.intersect1d(tmp, genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4ef3f5f-e08a-41ee-a4e3-63ab53a1b622",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Oligodendrocyte early\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Oligodendrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df2 = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Oligodendrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    effect_size_table[\"Gene\"].isin(gene_lists[\"Fatty Acid metabolism\"]),\n",
    "    :\n",
    "]\n",
    "df = pd.concat([df, df2], axis=0)\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axvline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axvline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = cp.copy(gene_lists[\"Fatty Acid metabolism\"])\n",
    "already_colored.extend([\"NCSTN\", \"CR1\", \"MYRF\", \"PLLP\"])\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"red\") for i in df.loc[(df[\"Gene\"].isin(gene_lists[\"Fatty Acid metabolism\"])) & (effect_size_table[\"Effect size across early pseudoprogression\"] > 1.5), :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"blue\") for i in df.loc[df[\"Gene\"].isin([\"MYRF\", \"PLLP\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"green\") for i in df.loc[df[\"Gene\"].isin([\"CR1\", \"NCSTN\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\", ascending=False).iloc[:10, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\").iloc[:10, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8e_scatterplot_Oligodendrocyte_Early.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# Oligodendrocyte late\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Oligodendrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df2 = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"Oligodendrocyte\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    effect_size_table[\"Gene\"].isin(gene_lists[\"Cholesterol biosynthesis\"]),\n",
    "    :\n",
    "]\n",
    "df = pd.concat([df, df2], axis=0)\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axhline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axhline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = cp.copy(gene_lists[\"Cholesterol biosynthesis\"])\n",
    "already_colored.extend([\"MOG\", \"MYRF\", \"OPALIN\", \"OMG\", \"MOBP\", \"PLLP\", \"CR1\"])\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"red\") for i in df.loc[df[\"Gene\"].isin(gene_lists[\"Cholesterol biosynthesis\"]), :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"blue\") for i in df.loc[df[\"Gene\"].isin([\"MOG\", \"MYRF\", \"OPALIN\", \"OMG\", \"MOBP\", \"PLLP\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"green\") for i in df.loc[df[\"Gene\"].isin([\"CR1\"]), :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across late pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across late pseudoprogression\", ascending=False).iloc[:15, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across late pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across late pseudoprogression\").iloc[:15, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8e_scatterplot_Oligodendrocyte_Late.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "# OPC early\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"OPC\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    (effect_size_table[\"Mean expression z-score\"] > 2),\n",
    "    :\n",
    "]\n",
    "df2 = effect_size_table.loc[\n",
    "    (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") &\n",
    "    (effect_size_table[\"Population\"] == \"OPC\") &\n",
    "    (effect_size_table[\"Mean expression (natural log UMIs per 10k plus 1)\"] > 0.05) &\n",
    "    effect_size_table[\"Gene\"].isin([\"DLL1\", \"DLL3\", \"CSPG4\", \"ASCL1\", \"OLIG1\", \"OLIG2\"]),\n",
    "    :\n",
    "]\n",
    "df = pd.concat([df, df2], axis=0)\n",
    "df = df.loc[~df.duplicated()].copy()\n",
    "ax = sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    y=\"Effect size across late pseudoprogression\",\n",
    "    alpha=0.1,\n",
    "    size=0.5,\n",
    "    color=\"grey\",\n",
    ")\n",
    "plt.axvline(1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.axvline(-1, linestyle=\"--\", color=\"lightgrey\");\n",
    "plt.legend('',frameon=False);\n",
    "\n",
    "already_colored = [\"DLL1\", \"DLL3\", \"CSPG4\", \"ASCL1\", \"OLIG1\", \"OLIG2\"]\n",
    "texts = [plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"blue\") for i in df.loc[df[\"Gene\"].isin([\"DLL1\", \"DLL3\", \"CSPG4\", \"ASCL1\", \"OLIG1\", \"OLIG2\"]), :].index]\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\", ascending=False).iloc[:10, :].index])\n",
    "texts.extend([plt.text(df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Effect size across late pseudoprogression\"], df.loc[i, \"Gene\"], size=8, color=\"grey\") for i in df.loc[(df[\"Effect size across early pseudoprogression\"].abs() > 1.5) & ~(df[\"Gene\"].isin(already_colored)), :].sort_values(by=\"Effect size across early pseudoprogression\").iloc[:10, :].index])\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\"), color=\"grey\");\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8e_scatterplot_OPC_Late.pdf\"), bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb566e4e-5b53-4ea9-a898-73f5687cc104",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48949afe-849b-47b4-a051-d5e2459113b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "# Scatterplot\n",
    "df = effect_size_table.loc[(effect_size_table[\"Gene\"].isin(TFs.loc[TFs[\"Direction\"] == \"+\", \"TF\"])) & (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") & (effect_size_table[\"Population\"].isin([\"OPC\"])), :]\n",
    "sns.scatterplot(\n",
    "    data=df,\n",
    "    x=\"Mean expression z-score\",\n",
    "    y=\"Effect size across early pseudoprogression\",\n",
    "    alpha=0.5\n",
    ")\n",
    "texts = [plt.text(df.loc[i, \"Mean expression z-score\"], df.loc[i, \"Effect size across early pseudoprogression\"], df.loc[i, \"Gene\"], size=8) for i in df.loc[(df[\"Mean expression z-score\"] > 2.5) & (df[\"Effect size across early pseudoprogression\"] > 1), :].index]\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color=\"gray\"));\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8f_scatterplot_OPC transcription factor regulators.pdf\"), bbox_inches=\"tight\")\n",
    "MR = TFs.loc[TFs[\"TF\"].isin(df.loc[(df[\"Mean expression z-score\"] > 2) & (df[\"Effect size across early pseudoprogression\"] > 1.1), \"Gene\"]) & (TFs[\"Direction\"] == \"+\"), \"Gene_signature_name\"].to_list()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c9f7e7-8a1a-497f-a8a0-96269d3fc675",
   "metadata": {},
   "outputs": [],
   "source": [
    "# KDEplot\n",
    "Downstream_Genes = eRegulons.loc[(eRegulons[\"Gene_signature_name\"].isin(MR)), [\"Gene\", \"TF\"]].drop_duplicates().loc[:, \"Gene\"].value_counts()\n",
    "Downstream_Genes = Downstream_Genes[Downstream_Genes > Downstream_Genes.median() + 1].index\n",
    "\n",
    "df = effect_size_table.loc[\n",
    "    (effect_size_table[\"Gene\"].isin(Downstream_Genes)) & (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") & (effect_size_table[\"Population\"].isin([\"OPC\"])),\n",
    "    :\n",
    "].merge(\n",
    "    eRegulons.loc[(eRegulons[\"Gene_signature_name\"].isin(MR)), [\"Gene\", \"TF2G_importance\"]].drop_duplicates().groupby(\"Gene\").mean(),\n",
    "    left_on=\"Gene\",\n",
    "    right_on=\"Gene\",\n",
    "    how=\"left\"\n",
    ")\n",
    "df_else = effect_size_table.loc[\n",
    "    ~(effect_size_table[\"Gene\"].isin(Downstream_Genes)) & (effect_size_table[\"Taxonomy Level\"] == \"Subclass\") & (effect_size_table[\"Population\"].isin([\"OPC\"])),\n",
    "    :\n",
    "]\n",
    "if df.shape[0] < df_else.shape[0]:\n",
    "    df_else = df_else.loc[np.random.choice(df_else.index, size=df.shape[0], replace=False), :].copy()\n",
    "\n",
    "\n",
    "sns.kdeplot(\n",
    "    data=df,\n",
    "    x=\"Effect size across early pseudoprogression\"\n",
    ");\n",
    "sns.kdeplot(\n",
    "    data=df_else,\n",
    "    x=\"Effect size across early pseudoprogression\",\n",
    "    color=\"orange\"\n",
    ");\n",
    "print(\"p-value early: \" + str(sp_stats.ttest_ind(df[\"Effect size across early pseudoprogression\"], df_else[\"Effect size across early pseudoprogression\"], equal_var=False)[1]))\n",
    "plt.savefig(os.path.join(pwd, \"output\", \"Figure 8f_kdeplot_OPC downstream targets.pdf\"), bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a34d4ad4-27d3-48a9-a4ee-2c5aba7eb0f6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Figure 8g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "717fc874-2664-46d4-a162-4d9acfa99710",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Deltaplots\n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "sub = adata[adata.obs[\"Neurotypical reference\"] == \"False\", :].copy()\n",
    "\n",
    "for i in [\"NCSTN\", \"OMG\", \"DHCR24\", \"OLIG2\"]:\n",
    "    ax = delta_plot(\n",
    "        adata=sub,\n",
    "        genes=[i],\n",
    "        groupby=\"Class\",\n",
    "        groupby_subset=[\"Non-neuronal and Non-neural\"],\n",
    "        plotby=\"Subclass\",\n",
    "        donor=\"Donor ID\",\n",
    "        across=\"Continuous Pseudo-progression Score\",\n",
    "        highlight=[],\n",
    "        colormap=subclass_colors,\n",
    "        title=\"Expression of \" + i,\n",
    "        legend=False,\n",
    "        save=os.path.join(pwd, \"output\", \"Figure 8g_deltaplot_{title}_by non-neuronal Subclass.pdf\")\n",
    "    )\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7854e20e-a6e3-4e08-b472-0211c972bf8e",
   "metadata": {},
   "source": [
    "### Figure 8h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d768499-b7a8-4155-9671-dd44be5da64a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.dotplot(\n",
    "    sub,\n",
    "    groupby=\"Subclass\",\n",
    "    var_names=[\"IGF1\", \"IGF2\", \"PDGFA\", \"PDGFB\", \"PDGFC\"],\n",
    "    cmap=\"YlGnBu\",\n",
    "    mean_only_expressed=True,\n",
    "    swap_axes=True,\n",
    "    smallest_dot=0.1,\n",
    "    figsize=(6,2),\n",
    "    save=\"Myelination factors by Subclass.pdf\"\n",
    ")\n",
    "os.rename(os.path.join(pwd, \"figures\", \"dotplot_Myelination factors by Subclass.pdf\"), os.path.join(pwd, \"output\", \"Figure 8h_dotplot_Myelination factors by Subclass.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6cb735c-1772-4d85-b9e7-35701a95ad2b",
   "metadata": {},
   "source": [
    "### Figure 8i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21fc0efa-bddd-4185-a904-e80296d88996",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deltaplots\n",
    "subclass_colors = color_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "subclass_colors.index = subclass_colors[\"subclass_label\"].copy()\n",
    "subclass_colors = subclass_colors[\"subclass_color\"].to_dict()\n",
    "\n",
    "sub = adata[adata.obs[\"Neurotypical reference\"] == \"False\", :].copy()\n",
    "\n",
    "for i in [\"IGF1\"]:\n",
    "    ax = delta_plot(\n",
    "        adata=sub,\n",
    "        genes=[i],\n",
    "        groupby=\"Class\",\n",
    "        groupby_subset=None,\n",
    "        plotby=\"Subclass\",\n",
    "        donor=\"Donor ID\",\n",
    "        across=\"Continuous Pseudo-progression Score\",\n",
    "        highlight=[],\n",
    "        colormap=subclass_colors,\n",
    "        title=\"Expression of \" + i,\n",
    "        legend=False,\n",
    "        save=os.path.join(pwd, \"output\", \"Figure 8i_deltaplot_{title}_groupby Class_plotby Subclass.pdf\")\n",
    "    )\n",
    "    plt.show();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "378890ea-6ed1-4ba4-9b8e-eabec6aa7c7f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ab69a5-fde2-4095-b38f-d1c00e27e4d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(os.path.join(pwd, \"figures\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
