{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4cb2aad0-38fa-45ce-b7ae-c7ff4cf2c4ca",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a17b3b0-f02e-41b4-a60a-d1b83b9c7713",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from datetime import datetime\n",
    "import warnings\n",
    "from helper_functions import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4d15727-cd96-4e22-a284-01154700ea12",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Build reference singleome and multiome adata from mtx files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e8e301-1989-4d4e-ae81-53cc78532d08",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dataset = \"reference\"\n",
    "technology = \"singleomeCR6\"\n",
    "region = \"MTG\"\n",
    "to_correct = [\n",
    "    \"estimated_number_of_cells\", \"mean_reads_per_cell\", \"median_genes_per_cell\",\n",
    "    \"number_of_reads\", \"sequencing_saturation\", \"q30_bases_in_Barcode\",\n",
    "    \"q30.bases_in_rna_read\", \"q30_bases_in_umi\",\n",
    "    \"reads_mapped_to_genome\", \"reads_mapped_confidently_to_genome\",\n",
    "    \"reads_mapped_confidently_to_intergenic_regions\", \"reads_mapped_confidently_to_intronic_regions\",\n",
    "    \"reads_mapped_confidently_to_exonic_regions\", \"reads_mapped_confidently_to_transcriptome\",\n",
    "    \"reads_mapped_antisense_to_gene\", \"fraction_reads_in_cells\", \"total_genes_detected\",\n",
    "    \"median_umi_counts_per_cell\"\n",
    "]\n",
    "\n",
    "build_anndata(\n",
    "    datadir=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \"_donor_mtx\"),\n",
    "    outfile=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"),\n",
    "    to_correct=to_correct,\n",
    "    metadata_file=os.path.join(pwd, \"input\", \"metadata\", \"Great_ApesMetadata_version101_20220321.csv\"),\n",
    "    metadata_name=\"Great Apes Metadata\",\n",
    "    adata_key=\"sample_id\",\n",
    "    external_key=\"sample_id\",\n",
    "    cell_ids=\"sample_id\",\n",
    "    remove_arid=True\n",
    ")\n",
    "\n",
    "dataset = \"reference\"\n",
    "technology = \"multiome\"\n",
    "region = \"MTG\"\n",
    "\n",
    "build_anndata(\n",
    "    datadir=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \"_donor_mtx\"),\n",
    "    outfile=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"),\n",
    "    metadata_file=None,\n",
    "    metadata_name=None\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58eab93e-3a49-4fe5-aa06-25275c5e803a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Build AD singleome and multiome adata from mtx files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbaaac6d-cd66-402f-9fbd-fb860e6c2842",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"AD\"\n",
    "technology = \"singleomeCR6\"\n",
    "region = \"MTG\"\n",
    "to_correct = [\n",
    "    \"estimated_number_of_cells\", \"mean_reads_per_cell\", \"median_genes_per_cell\",\n",
    "    \"number_of_reads\", \"sequencing_saturation\", \"q30_bases_in_Barcode\",\n",
    "    \"q30.bases_in_rna_read\", \"q30_bases_in_umi\",\n",
    "    \"reads_mapped_to_genome\", \"reads_mapped_confidently_to_genome\",\n",
    "    \"reads_mapped_confidently_to_intergenic_regions\", \"reads_mapped_confidently_to_intronic_regions\",\n",
    "    \"reads_mapped_confidently_to_exonic_regions\", \"reads_mapped_confidently_to_transcriptome\",\n",
    "    \"reads_mapped_antisense_to_gene\", \"fraction_reads_in_cells\", \"total_genes_detected\",\n",
    "    \"median_umi_counts_per_cell\"\n",
    "]\n",
    "\n",
    "\n",
    "build_anndata(\n",
    "    datadir=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \"_donor_mtx\"),\n",
    "    outfile=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"),\n",
    "    to_correct=to_correct,\n",
    "    metadata_file=os.path.join(pwd, \"input\", \"metadata\", \"NeuropathMetadata_version103_20220407.KT.sav\"),\n",
    "    metadata_name=\"UW Clinical Metadata\",\n",
    "    adata_key=\"external_donor_name\",\n",
    "    external_key=\"uwa\",\n",
    "    cell_ids=\"sample_id\",\n",
    "    remove_arid=True\n",
    ")\n",
    "\n",
    "dataset = \"AD\"\n",
    "technology = \"multiome\"\n",
    "region = \"MTG\"\n",
    "to_correct = []\n",
    "\n",
    "build_anndata(\n",
    "    datadir=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \"_donor_mtx\"),\n",
    "    outfile=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"),\n",
    "    to_correct=to_correct,\n",
    "    metadata_file=os.path.join(pwd, \"input\", \"metadata\", \"NeuropathMetadata_version103_20220407.KT.sav\"),\n",
    "    metadata_name=\"UW Clinical Metadata\",\n",
    "    adata_key=\"external_donor_name\",\n",
    "    external_key=\"uwa\",\n",
    "    cell_ids=\"sample_id\",\n",
    "    remove_arid=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d644591-3d55-47d3-82ec-5fcd8a844481",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Merge multiome and singleome datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "152fe2a1-ad2a-43ac-9633-eb43fe298ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"reference\"\n",
    "technology = \"singleomeCR6\"\n",
    "region = \"MTG\"\n",
    "date = \"2022-04-08\"\n",
    "\n",
    "adata_ref_singleome = sc.read_h5ad(filename=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + date + \".h5ad\"))\n",
    "\n",
    "dataset = \"reference\"\n",
    "technology = \"multiome\"\n",
    "region = \"MTG\"\n",
    "date = \"2022-09-27\"\n",
    "\n",
    "adata_ref_multiome = sc.read_h5ad(filename=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + date + \".h5ad\"))\n",
    "\n",
    "dataset = \"AD\"\n",
    "technology = \"singleomeCR6\"\n",
    "region = \"MTG\"\n",
    "date = \"2022-04-07\"\n",
    "\n",
    "adata_singleome = sc.read_h5ad(filename=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + date + \".h5ad\"))\n",
    "\n",
    "dataset = \"AD\"\n",
    "technology = \"multiome\"\n",
    "region = \"MTG\"\n",
    "date = \"2022-04-07\"\n",
    "\n",
    "adata_multiome = sc.read_h5ad(filename=os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + date + \".h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d162d998-f2d6-4aa8-b6b9-5867557fa9e5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "to_drop_singleome = [\"sex_y\"]\n",
    "to_drop_multiome = [\"sex_y\", \"file\"]\n",
    "\n",
    "to_divide_reference_singleome = [\n",
    "    \"sequencing_saturation\", \"q30_bases_in_Barcode\", \"q30.bases_in_rna_read\",\n",
    "    \"q30_bases_in_umi\", \"reads_mapped_to_genome\", \"reads_mapped_confidently_to_genome\",\n",
    "    \"reads_mapped_confidently_to_intergenic_regions\", \"reads_mapped_confidently_to_intronic_regions\",\n",
    "    \"reads_mapped_confidently_to_exonic_regions\", \"reads_mapped_confidently_to_transcriptome\",\n",
    "    \"reads_mapped_antisense_to_gene\", \"fraction_reads_in_cells\"\n",
    "]\n",
    "to_divide_singleome = [\n",
    "    \"sequencing_saturation\", \"q30_bases_in_Barcode\", \"q30.bases_in_rna_read\",\n",
    "    \"q30_bases_in_umi\", \"reads_mapped_to_genome\", \"reads_mapped_confidently_to_genome\",\n",
    "    \"reads_mapped_confidently_to_intergenic_regions\", \"reads_mapped_confidently_to_intronic_regions\",\n",
    "    \"reads_mapped_confidently_to_exonic_regions\", \"reads_mapped_confidently_to_transcriptome\",\n",
    "    \"reads_mapped_antisense_to_gene\", \"fraction_reads_in_cells\"\n",
    "]\n",
    "\n",
    "to_rename_reference_singleome = {\n",
    "    \"age\": \"age_at_death\",\n",
    "    \"estimated_number_of_cells\": \"Estimated_number_of_cells\",\n",
    "    \"fraction_reads_in_cells\": \"GEX_Fraction_of_transcriptomic_reads_in_cells\",\n",
    "    \"mean_reads_per_cell\": \"GEX_Mean_raw_reads_per_cell\",\n",
    "    \"median_genes_per_cell\": \"GEX_Median_genes_per_cell\",\n",
    "    \"median_umi_counts_per_cell\": \"GEX_Median_UMI_counts_per_cell\",\n",
    "    \"q30.bases_in_rna_read\": \"GEX_Q30_bases_in_read_2\",\n",
    "    \"q30_bases_in_Barcode\": \"GEX_Q30_bases_in_barcode\",\n",
    "    \"q30_bases_in_umi\": \"GEX_Q30_bases_in_UMI\",\n",
    "    \"reads_mapped_antisense_to_gene\": \"GEX_Reads_mapped_antisense_to_gene\",\n",
    "    \"reads_mapped_confidently_to_exonic_regions\": \"GEX_Reads_mapped_confidently_to_exonic_regions\",\n",
    "    \"reads_mapped_confidently_to_genome\": \"GEX_Reads_mapped_confidently_to_genome\",\n",
    "    \"reads_mapped_confidently_to_intergenic_regions\": \"GEX_Reads_mapped_confidently_to_intergenic_regions\",\n",
    "    \"reads_mapped_confidently_to_intronic_regions\": \"GEX_Reads_mapped_confidently_to_intronic_regions\",\n",
    "    \"reads_mapped_confidently_to_transcriptome\": \"GEX_Reads_mapped_confidently_to_transcriptome\",\n",
    "    \"reads_mapped_to_genome\": \"GEX_Reads_mapped_to_genome\",\n",
    "    \"total_genes_detected\": \"GEX_Total_genes_detected\",\n",
    "}\n",
    "to_rename_reference_multiome = {\n",
    "    \"age\": \"age_at_death\",\n",
    "    \"Pipeline_version\": \"alignment\",\n",
    "}\n",
    "to_rename_singleome = {\n",
    "    \"sex_x\": \"sex\",\n",
    "    \"estimated_number_of_cells\": \"Estimated_number_of_cells\",\n",
    "    \"fraction_reads_in_cells\": \"GEX_Fraction_of_transcriptomic_reads_in_cells\",\n",
    "    \"mean_reads_per_cell\": \"GEX_Mean_raw_reads_per_cell\",\n",
    "    \"median_genes_per_cell\": \"GEX_Median_genes_per_cell\",\n",
    "    \"median_umi_counts_per_cell\": \"GEX_Median_UMI_counts_per_cell\",\n",
    "    \"q30.bases_in_rna_read\": \"GEX_Q30_bases_in_read_2\",\n",
    "    \"q30_bases_in_Barcode\": \"GEX_Q30_bases_in_barcode\",\n",
    "    \"q30_bases_in_umi\": \"GEX_Q30_bases_in_UMI\",\n",
    "    \"reads_mapped_antisense_to_gene\": \"GEX_Reads_mapped_antisense_to_gene\",\n",
    "    \"reads_mapped_confidently_to_exonic_regions\": \"GEX_Reads_mapped_confidently_to_exonic_regions\",\n",
    "    \"reads_mapped_confidently_to_genome\": \"GEX_Reads_mapped_confidently_to_genome\",\n",
    "    \"reads_mapped_confidently_to_intergenic_regions\": \"GEX_Reads_mapped_confidently_to_intergenic_regions\",\n",
    "    \"reads_mapped_confidently_to_intronic_regions\": \"GEX_Reads_mapped_confidently_to_intronic_regions\",\n",
    "    \"reads_mapped_confidently_to_transcriptome\": \"GEX_Reads_mapped_confidently_to_transcriptome\",\n",
    "    \"reads_mapped_to_genome\": \"GEX_Reads_mapped_to_genome\",\n",
    "    \"total_genes_detected\": \"GEX_Total_genes_detected\",\n",
    "}\n",
    "to_rename_multiome = {\n",
    "    \"Pipeline_version\": \"alignment\",\n",
    "    \"sex_x\": \"sex\"\n",
    "}\n",
    "\n",
    "# Adding metadata\n",
    "adata_ref_singleome.obs[\"Genome\"] = \"GRCh38\"\n",
    "adata_ref_singleome.obs[\"gender\"] = adata_ref_singleome.obs[\"sex\"].copy()\n",
    "adata_ref_multiome.obs[\"gender\"] = adata_ref_multiome.obs[\"sex\"].copy()\n",
    "\n",
    "adata_singleome.obs[\"Genome\"] = \"GRCh38\"\n",
    "\n",
    "\n",
    "# Dropping metadata\n",
    "adata_singleome.obs.drop(to_drop_singleome, axis=1, inplace=True)\n",
    "adata_multiome.obs.drop(to_drop_multiome, axis=1, inplace=True)\n",
    "\n",
    "\n",
    "# Editing metadata\n",
    "for i in to_divide_reference_singleome:\n",
    "    adata_ref_singleome.obs[i] = adata_ref_singleome.obs[i] / 100\n",
    "    \n",
    "for i in to_divide_singleome:\n",
    "    adata_singleome.obs[i] = adata_singleome.obs[i] / 100\n",
    "\n",
    "\n",
    "# Renaming metadata\n",
    "adata_ref_singleome.obs.rename(columns=to_rename_reference_singleome, inplace=True)\n",
    "adata_ref_multiome.obs.rename(columns=to_rename_reference_multiome, inplace=True)\n",
    "\n",
    "adata_singleome.obs.rename(columns=to_rename_singleome, inplace=True)\n",
    "adata_multiome.obs.rename(columns=to_rename_multiome, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49dc6eab-488a-40a3-b7f4-19845c1c92e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final reference metadata cleanup\n",
    "adata_ref_singleome.obs[\"fraction_mito\"] = adata_ref_singleome[:, adata_ref_singleome.var_names.str.startswith(\"MT-\")].layers[\"UMIs\"].sum(axis=1) / adata_ref_singleome.layers[\"UMIs\"].sum(axis=1)\n",
    "adata_ref_singleome.obs[\"age_at_death\"] = adata_ref_singleome.obs[\"age_at_death\"].str.replace(\" yrs\", \"\").astype(\"float\")\n",
    "adata_ref_singleome.obs[\"reference_cell\"] = 1\n",
    "\n",
    "adata_ref_multiome.obs[\"fraction_mito\"] = adata_ref_multiome[:, adata_ref_multiome.var_names.str.startswith(\"MT-\")].layers[\"UMIs\"].sum(axis=1) / adata_ref_multiome.layers[\"UMIs\"].sum(axis=1)\n",
    "adata_ref_multiome.obs[\"age_at_death\"] = adata_ref_multiome.obs[\"age_at_death\"].str.replace(\" yrs\", \"\").astype(\"float\")\n",
    "adata_ref_multiome.obs[\"reference_cell\"] = 1\n",
    "\n",
    "# Combine reference object\n",
    "adata_ref_combined = adata_ref_singleome.concatenate(adata_ref_smartseq, index_unique=None)\n",
    "adata_ref_combined.uns = adata_ref_singleome.uns.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfada073-c128-46d4-90cb-43a4c3318df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine AD object\n",
    "adata_query = adata_singleome.concatenate(adata_multiome, index_unique=None)\n",
    "adata_query.uns = adata_singleome.uns.copy()\n",
    "del adata_singleome\n",
    "del adata_multiome\n",
    "\n",
    "# Final query metadata cleanup\n",
    "adata_query.obs[\"fraction_mito\"] = adata_query[:, adata_query.var_names.str.startswith(\"MT-\")].layers[\"UMIs\"].sum(axis=1) / adata_query.layers[\"UMIs\"].sum(axis=1)\n",
    "adata_query.obs[\"reference_cell\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5adef82d-271a-473e-8bd2-4e0c93d2ee9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"reference\"\n",
    "technology = \"singleomeCR6\"\n",
    "region = \"MTG\"\n",
    "\n",
    "adata_ref_singleome.write(os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"))\n",
    "\n",
    "dataset = \"reference\"\n",
    "technology = \"multiome\"\n",
    "region = \"MTG\"\n",
    "\n",
    "adata_ref_multiome.write(os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"))\n",
    "\n",
    "dataset = \"reference\"\n",
    "technology = \"combined\"\n",
    "region = \"MTG\"\n",
    "\n",
    "adata_ref_combined.write(os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"))\n",
    "\n",
    "dataset = \"AD\"\n",
    "technology = \"combined\"\n",
    "region = \"MTG\"\n",
    "\n",
    "adata_query.write(os.path.join(pwd, \"output\", region + \"_\" + dataset + \"_\" + technology + \".\" + str(datetime.date(datetime.now())) + \".h5ad\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
