{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "361b5a67-bb4a-4bab-ac28-c22bb1628790",
   "metadata": {},
   "source": [
    "### Load needed libraries and make needed directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74151e8b-8049-4260-b76e-a46a99edbdfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scenicplus\n",
    "scenicplus.__version__\n",
    "\n",
    "import pycisTopic\n",
    "pycisTopic.__version__\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "import dill\n",
    "import scanpy as sc\n",
    "import sys\n",
    "import warnings\n",
    "import pyranges as pr\n",
    "import seaborn as sns\n",
    "from scipy.io import mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "from pycisTopic.cistopic_class import *\n",
    "from pycisTopic.lda_models import *\n",
    "from pycisTopic.topic_binarization import *\n",
    "from pycisTopic.diff_features import *\n",
    "from pycistarget.utils import region_names_to_coordinates\n",
    "from scenicplus.wrappers.run_pycistarget import run_pycistarget\n",
    "from scenicplus.scenicplus_class import create_SCENICPLUS_object\n",
    "from scenicplus.wrappers.run_scenicplus import run_scenicplus\n",
    "from scenicplus.preprocessing.filtering import apply_std_filtering_to_eRegulons\n",
    "from scenicplus.eregulon_enrichment import score_eRegulons\n",
    "from scenicplus.cistromes import TF_cistrome_correlation, generate_pseudobulks\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')\n",
    "\n",
    "projDir = os.getcwd()\n",
    "\n",
    "## Define rankings, score and motif annotation database\n",
    "# From https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/screen/mc_v10_clust/region_based/\n",
    "# and https://resources.aertslab.org/cistarget/motif2tf/\n",
    "db_anno_fpath = os.path.join(projDir, \"input\", \"motif_db\")\n",
    "rankings_db = os.path.join(db_anno_fpath, 'hg38_screen_v10_clust.regions_vs_motifs.rankings.feather')\n",
    "scores_db =  os.path.join(db_anno_fpath, 'hg38_screen_v10_clust.regions_vs_motifs.scores.feather')\n",
    "motif_annotation = os.path.join(db_anno_fpath, 'motifs-v10nr_clust-nr.hgnc-m0.001-o0.0.tbl')\n",
    "\n",
    "# From http://humantfs.ccbr.utoronto.ca/download/v_1.01/\n",
    "tfdb = os.path.join(profDir, \"input\", \"TF_names_v_1.01.txt\")\n",
    "biomart_host = \"http://sep2019.archive.ensembl.org/\"\n",
    "\n",
    "# Output directories\n",
    "outDir = os.path.join(pwd, \"output\")\n",
    "if not os.path.exists(outDir):\n",
    "    os.makedirs(outDir)\n",
    "\n",
    "workDir = os.path.join(outDir, \"tmp\")\n",
    "if not os.path.exists(workDir):\n",
    "    os.makedirs(workDir)\n",
    "\n",
    "if not os.path.exists(os.path.join(workDir, 'models')):\n",
    "    os.makedirs(os.path.join(workDir, 'models'))\n",
    "\n",
    "if not os.path.exists(os.path.join(workDir, 'candidate_enhancers')):\n",
    "    os.makedirs(os.path.join(workDir, 'candidate_enhancers'))\n",
    "\n",
    "if not os.path.exists(os.path.join(workDir, 'motifs')):\n",
    "    os.makedirs(os.path.join(workDir, 'motifs'))\n",
    "\n",
    "if not os.path.exists(os.path.join(workDir, 'scenicplus')):\n",
    "    os.makedirs(os.path.join(workDir, 'scenicplus'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1276b94-915a-4d02-b017-fa7c3cd557f8",
   "metadata": {},
   "source": [
    "### Read in needed datasets for cisTopic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c837e089-3e24-4a25-8c95-8a3ebfb17164",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the Matrix Market format file\n",
    "data_folder = projDir\n",
    "sparse_matrix = mmread(os.path.join(outDir, \"sparse_matrix.mtx\"))\n",
    "\n",
    "# Convert to sparse.csr_matrix\n",
    "sparse_csr_matrix = csr_matrix(sparse_matrix)\n",
    "\n",
    "import pandas as pd\n",
    "df = pd.read_csv(os.path.join(outDir, \"/seqnames_ranges.csv\"))\n",
    "\n",
    "region_names = df.loc[:, 'Seqnames_Ranges'].tolist()\n",
    "df = pd.read_csv(os.path.join(outDir, \"/cell_names.csv\"))\n",
    "cell_names = df[\"cells\"].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5fcff0-fafe-4701-a35b-63ca6d7fbcb9",
   "metadata": {},
   "source": [
    "### Create cisTopic object and run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2fae94-885c-42fc-8a57-3fe02ac5d9cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "cistopic_obj = create_cistopic_object(\n",
    "    fragment_matrix=sparse_csr_matrix,\n",
    "    cell_names=cell_names,\n",
    "    region_names=region_names\n",
    ")\n",
    "\n",
    "## Checkpoint\n",
    "pickle.dump(cistopic_obj, open(os.path.join(workDir, 'cistopic_obj.pkl'), 'wb'))\n",
    "\n",
    "models = run_cgs_models(\n",
    "    cistopic_obj,\n",
    "    n_topics=[2,4,10,16,32,48],\n",
    "    n_cpu=32,\n",
    "    n_iter=500,\n",
    "    random_state=555,\n",
    "    alpha=50,\n",
    "    alpha_by_topic=True,\n",
    "    eta=0.1,\n",
    "    eta_by_topic=False,\n",
    "    save_path=None,\n",
    ")\n",
    "\n",
    "pickle.dump(models, open(os.path.join(workDir, \"models\", \"ATAC_models_500_iter_LDA.pkl\"), 'wb'))\n",
    "\n",
    "model = evaluate_models(\n",
    "    models,\n",
    "    select_model=16,\n",
    "    return_model=True,\n",
    "    metrics=['Arun_2010','Cao_Juan_2009', 'Minmo_2011', 'loglikelihood'],\n",
    "    plot_metrics=False\n",
    ")\n",
    "\n",
    "cistopic_obj.add_LDA_model(model)\n",
    "\n",
    "## Checkpoint\n",
    "pickle.dump(cistopic_obj, open(os.path.join(workDir, 'cistopic_obj.pkl'), 'wb'))\n",
    "\n",
    "region_bin_topics_otsu = binarize_topics(cistopic_obj, method='otsu')\n",
    "region_bin_topics_top3k = binarize_topics(cistopic_obj, method='ntop', ntop = 3000)\n",
    "\n",
    "cell_metadata = pd.read_csv(os.path.join(outDir, \"cell_metadata.csv\"), index_col=0)\n",
    "\n",
    "cell_metadata = cell_metadata.reset_index()\n",
    "cell_metadata[\"index\"] = cell_metadata[\"index\"] + \"___cisTopic\"\n",
    "\n",
    "tmp = cistopic_obj.cell_data.copy()\n",
    "tmp = tmp.reset_index()\n",
    "tmp = pd.merge(tmp, cell_metadata, how='left', on='index')\n",
    "\n",
    "tmp = tmp.set_index(\"index\")\n",
    "tmp.index.name = None\n",
    "\n",
    "cistopic_obj.cell_data = tmp\n",
    "\n",
    "imputed_acc_obj = impute_accessibility(\n",
    "    cistopic_obj,\n",
    "    selected_cells=None,\n",
    "    selected_regions=None,\n",
    "    scale_factor=10**6\n",
    ")\n",
    "\n",
    "normalized_imputed_acc_obj = normalize_scores(\n",
    "    imputed_acc_obj,\n",
    "    scale_factor=10**4\n",
    ")\n",
    "\n",
    "variable_regions = find_highly_variable_features(\n",
    "    normalized_imputed_acc_obj,\n",
    "    plot=False\n",
    ")\n",
    "\n",
    "markers_dict = find_diff_features(\n",
    "    cistopic_obj,\n",
    "    imputed_acc_obj,\n",
    "    variable='Subclass',\n",
    "    var_features=variable_regions,\n",
    "    split_pattern='-'\n",
    ")\n",
    "\n",
    "## Checkpoint\n",
    "pickle.dump(cistopic_obj, open(os.path.join(workDir, 'cistopic_obj.pkl'), 'wb'))\n",
    "pickle.dump(region_bin_topics_otsu, open(os.path.join(workDir, \"candidate_enhancers\", \"region_bin_topics_otsu.pkl\"), 'wb'))\n",
    "pickle.dump(region_bin_topics_top3k, open(os.path.join(workDir, \"candidate_enhancers\", \"region_bin_topics_top3k.pkl\"), 'wb'))\n",
    "pickle.dump(markers_dict, open(os.path.join(workDir, \"candidate_enhancers\", \"markers_dict.pkl\"), 'wb'))\n",
    "\n",
    "\n",
    "region_sets = {}\n",
    "region_sets['topics_otsu'] = {}\n",
    "region_sets['topics_top_3'] = {}\n",
    "region_sets['DARs'] = {}\n",
    "for topic in region_bin_topics_otsu.keys():\n",
    "    regions = region_bin_topics_otsu[topic].index[region_bin_topics_otsu[topic].index.str.startswith('chr')] #only keep regions on known chromosomes\n",
    "    region_sets['topics_otsu'][topic] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "for topic in region_bin_topics_top3k.keys():\n",
    "    regions = region_bin_topics_top3k[topic].index[region_bin_topics_top3k[topic].index.str.startswith('chr')] #only keep regions on known chromosomes\n",
    "    region_sets['topics_top_3'][topic] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "for DAR in markers_dict.keys():\n",
    "    regions = markers_dict[DAR].index[markers_dict[DAR].index.str.startswith('chr')] #only keep regions on known chromosomes\n",
    "    region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "\n",
    "\n",
    "run_pycistarget(\n",
    "    region_sets=region_sets,\n",
    "    species='homo_sapiens',\n",
    "    save_path=os.path.join(workDir, 'motifs'),\n",
    "    ctx_db_path=rankings_db,\n",
    "    dem_db_path=scores_db,\n",
    "    path_to_motif_annotations=motif_annotation,\n",
    "    run_without_promoters=True,\n",
    "    n_cpu=30,\n",
    "    annotation_version='v10nr_clust',\n",
    ")\n",
    "\n",
    "## Checkpoint\n",
    "menr = dill.load(open(os.path.join(workDir, 'motifs/menr.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bcf8037-e01f-4297-8a37-d1d922312d19",
   "metadata": {},
   "source": [
    "### Read in needed datasets for scenicplus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d74a2190-ed09-4d75-bc63-12b3ed12a51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(os.path.join(projDir, \"input\", \"SEAAD_MTG_RNAseq_final-nuclei.2024-02-13.h5ad\")\n",
    "adata = adata[(adata.obs[\"Class\"] == \"Non-neuronal and Non-neural\") & (adata.obs[\"Severely Affected Donor\"] == \"N\"), :].copy()\n",
    "adata.X = adata.layers['UMIs'].copy()\n",
    "del adata.layers['UMIs']\n",
    "\n",
    "cistopic_obj = dill.load(open(os.path.join(workDir, \"scATAC\", \"cistopic_obj.pkl\"), 'rb'))\n",
    "menr = dill.load(open(os.path.join(workDir, \"motifs\", \"menr.pkl\"), 'rb'))\n",
    "adata.obs[\"Subclass\"] = adata.obs['Subclass'].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0cff43d-0aca-4d24-a2e7-72361666731f",
   "metadata": {},
   "source": [
    "### Create scenicplus object and run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5be3d74c-b82e-4f78-a1c1-73c3ff044bfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj = create_SCENICPLUS_object(\n",
    "        GEX_anndata=adata,\n",
    "        cisTopic_obj=cistopic_obj,\n",
    "        menr=menr,\n",
    "        multi_ome_mode=False,\n",
    "        key_to_group_by=\"Subclass\",\n",
    "        nr_cells_per_metacells=5\n",
    ")\n",
    "\n",
    "# Checkpoint\n",
    "dill.dump(scplus_obj, open(os.path.join(workDir, 'scenicplus/scplus_obj.pkl'), 'wb'), protocol=-1)\n",
    "\n",
    "scplus_obj.dr_cell['eRegulons_UMAP'] = True\n",
    "\n",
    "\n",
    "sys.stderr = open(os.devnull, \"w\")  # silence stderr\n",
    "run_scenicplus(\n",
    "    scplus_obj=scplus_obj,\n",
    "    variable=['Subclass'],\n",
    "    species=\"hsapiens\",\n",
    "    assembly=\"hg38\",\n",
    "    tf_file=tfdb,\n",
    "    save_path=os.path.join(workDir, \"scenicplus\"),\n",
    "    biomart_host=biomart_host,\n",
    "    upstream=[1000, 150000],\n",
    "    downstream=[1000, 150000],\n",
    "    calculate_TF_eGRN_correlation=True,\n",
    "    calculate_DEGs_DARs=True,\n",
    "    export_to_loom_file=False,\n",
    "    export_to_UCSC_file=False,\n",
    "    n_cpu=30,\n",
    ")\n",
    "sys.stderr = sys.__stderr__  # unsilence stderr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d6d09dd-3a02-4e83-9493-80b9e9c98106",
   "metadata": {},
   "source": [
    "### Filter eRegulons to produce final output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69fe7a10-c8b6-4f8a-8fd2-01a9ad90b0db",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj = dill.load(open(os.path.join(workDir, 'scenicplus/scplus_obj.pkl'), 'rb'))\n",
    "region_ranking = dill.load(open(os.path.join(workDir, 'scenicplus/region_ranking.pkl'), 'rb')) #load ranking calculated using the wrapper function\n",
    "gene_ranking = dill.load(open(os.path.join(workDir, 'scenicplus/gene_ranking.pkl'), 'rb')) #load ranking calculated using the wrapper function\n",
    "score_eRegulons(\n",
    "    scplus_obj,\n",
    "    ranking=region_ranking,\n",
    "    eRegulon_signatures_key='eRegulon_signatures_filtered',\n",
    "    key_added='eRegulon_AUC_filtered',\n",
    "    enrichment_type='region',\n",
    "    auc_threshold=0.05,\n",
    "    normalize=False,\n",
    "    n_cpu=30\n",
    ")\n",
    "score_eRegulons(\n",
    "    scplus_obj,\n",
    "    gene_ranking,\n",
    "    eRegulon_signatures_key='eRegulon_signatures_filtered',\n",
    "    key_added='eRegulon_AUC_filtered',\n",
    "    enrichment_type='gene',\n",
    "    auc_threshold=0.05,\n",
    "    normalize=False,\n",
    "    n_cpu=30\n",
    ")\n",
    "\n",
    "generate_pseudobulks(\n",
    "    scplus_obj=scplus_obj,\n",
    "    variable='label_transfer',\n",
    "    auc_key='eRegulon_AUC_filtered',\n",
    "    signature_key='Gene_based')\n",
    "generate_pseudobulks(\n",
    "    scplus_obj=scplus_obj,\n",
    "    variable='label_transfer',\n",
    "    auc_key='eRegulon_AUC_filtered',\n",
    "    signature_key='Region_based'\n",
    ")\n",
    "\n",
    "TF_cistrome_correlation(\n",
    "    scplus_obj,\n",
    "    use_pseudobulk=True,\n",
    "    variable='label_transfer',\n",
    "    auc_key='eRegulon_AUC_filtered',\n",
    "    signature_key='Gene_based',\n",
    "    out_key='filtered_gene_based'\n",
    ")\n",
    "TF_cistrome_correlation(\n",
    "    scplus_obj,\n",
    "    use_pseudobulk=True,\n",
    "    variable='label_transfer',\n",
    "    auc_key='eRegulon_AUC_filtered',\n",
    "    signature_key='Region_based',\n",
    "    out_key='filtered_region_based'\n",
    ")\n",
    "\n",
    "n_targets = [int(x.split('(')[1].replace('r)', '')) for x in scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Cistrome']]\n",
    "rho = scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Rho'].to_list()\n",
    "adj_pval = scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Adjusted_p-value'].to_list()\n",
    "\n",
    "thresholds = {\n",
    "        'rho': [-0.75, 0.70],\n",
    "        'n_targets': 0\n",
    "}\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (10, 5))\n",
    "sc = ax.scatter(rho, n_targets, c = -np.log10(adj_pval), s = 5)\n",
    "ax.set_xlabel('Correlation coefficient')\n",
    "ax.set_ylabel('nr. target regions')\n",
    "#ax.hlines(y = thresholds['n_targets'], xmin = min(rho), xmax = max(rho), color = 'black', ls = 'dashed', lw = 1)\n",
    "ax.vlines(x = thresholds['rho'], ymin = 0, ymax = max(n_targets), color = 'black', ls = 'dashed', lw = 1)\n",
    "ax.text(x = thresholds['rho'][0], y = max(n_targets), s = str(thresholds['rho'][0]))\n",
    "ax.text(x = thresholds['rho'][1], y = max(n_targets), s = str(thresholds['rho'][1]))\n",
    "sns.despine(ax = ax)\n",
    "fig.colorbar(sc, label = '-log10(adjusted_pvalue)', ax = ax)\n",
    "plt.show()\n",
    "\n",
    "selected_cistromes = scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based'].loc[\n",
    "        np.logical_or(\n",
    "                scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Rho'] > thresholds['rho'][1],\n",
    "                scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Rho'] < thresholds['rho'][0]\n",
    "        )]['Cistrome'].to_list()\n",
    "selected_eRegulons = [x.split('_(')[0] for x in selected_cistromes]\n",
    "selected_eRegulons_gene_sig = [\n",
    "        x for x in scplus_obj.uns['eRegulon_signatures_filtered']['Gene_based'].keys()\n",
    "        if x.split('_(')[0] in selected_eRegulons]\n",
    "selected_eRegulons_region_sig = [\n",
    "        x for x in scplus_obj.uns['eRegulon_signatures_filtered']['Region_based'].keys()\n",
    "        if x.split('_(')[0] in selected_eRegulons]\n",
    "\n",
    "#save the results in the scenicplus object\n",
    "scplus_obj.uns['selected_eRegulon'] = {'Gene_based': selected_eRegulons_gene_sig, 'Region_based': selected_eRegulons_region_sig}\n",
    "print(f'selected: {len(selected_eRegulons_gene_sig)} eRegulons')\n",
    "\n",
    "# Checkpoint the selected results\n",
    "dill.dump(scplus_obj, open(os.path.join(workDir, 'scenicplus/scplus_obj.pkl'), 'wb'), protocol=-1)\n",
    "\n",
    "scplus_obj.uns['eRegulon_metadata_filtered'].to_csv(\"eRegulon_metadata_filtered.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
